{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Quickstart\" data-toc-modified-id=\"Quickstart-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Quickstart</a></span></li><li><span><a href=\"#Initialize-your-IcaData-object\" data-toc-modified-id=\"Initialize-your-IcaData-object-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Initialize your IcaData object</a></span><ul class=\"toc-item\"><li><span><a href=\"#Minimum-requirements\" data-toc-modified-id=\"Minimum-requirements-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Minimum requirements</a></span></li><li><span><a href=\"#Viewing-matrices\" data-toc-modified-id=\"Viewing-matrices-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Viewing matrices</a></span></li><li><span><a href=\"#Sample,-gene,-and-iModulon-names\" data-toc-modified-id=\"Sample,-gene,-and-iModulon-names-2.3\"><span class=\"toc-item-num\">2.3&nbsp;&nbsp;</span>Sample, gene, and iModulon names</a></span></li><li><span><a href=\"#Adding-the-X-matrix\" data-toc-modified-id=\"Adding-the-X-matrix-2.4\"><span class=\"toc-item-num\">2.4&nbsp;&nbsp;</span>Adding the X matrix</a></span></li></ul></li><li><span><a href=\"#Sample,-gene,-and-iModulon-tables\" data-toc-modified-id=\"Sample,-gene,-and-iModulon-tables-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Sample, gene, and iModulon tables</a></span><ul class=\"toc-item\"><li><span><a href=\"#Loading-tables\" data-toc-modified-id=\"Loading-tables-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Loading tables</a></span></li><li><span><a href=\"#Viewing-tables\" data-toc-modified-id=\"Viewing-tables-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Viewing tables</a></span></li></ul></li><li><span><a href=\"#TRN-Table\" data-toc-modified-id=\"TRN-Table-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>TRN Table</a></span><ul class=\"toc-item\"><li><span><a href=\"#Load-the-TRN\" data-toc-modified-id=\"Load-the-TRN-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Load the TRN</a></span></li></ul></li><li><span><a href=\"#iModulon-properties\" data-toc-modified-id=\"iModulon-properties-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>iModulon properties</a></span><ul class=\"toc-item\"><li><span><a href=\"#Inspecting-iModulons\" data-toc-modified-id=\"Inspecting-iModulons-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Inspecting iModulons</a></span></li><li><span><a href=\"#Renaming-iModulons\" data-toc-modified-id=\"Renaming-iModulons-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Renaming iModulons</a></span></li><li><span><a href=\"#Identifying-&quot;single-gene&quot;-iModulons\" data-toc-modified-id=\"Identifying-&quot;single-gene&quot;-iModulons-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>Identifying \"single-gene\" iModulons</a></span></li></ul></li><li><span><a href=\"#Gene-set-enrichment-analysis\" data-toc-modified-id=\"Gene-set-enrichment-analysis-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Gene set enrichment analysis</a></span><ul class=\"toc-item\"><li><span><a href=\"#General-functions\" data-toc-modified-id=\"General-functions-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>General functions</a></span></li><li><span><a href=\"#Using-the-IcaData-object\" data-toc-modified-id=\"Using-the-IcaData-object-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>Using the IcaData object</a></span><ul class=\"toc-item\"><li><span><a href=\"#Save-to-sample-table\" data-toc-modified-id=\"Save-to-sample-table-6.2.1\"><span class=\"toc-item-num\">6.2.1&nbsp;&nbsp;</span>Save to sample table</a></span></li></ul></li></ul></li><li><span><a href=\"#iModulon-Thresholds\" data-toc-modified-id=\"iModulon-Thresholds-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>iModulon Thresholds</a></span><ul class=\"toc-item\"><li><span><a href=\"#Fitting-the-threshold\" data-toc-modified-id=\"Fitting-the-threshold-7.1\"><span class=\"toc-item-num\">7.1&nbsp;&nbsp;</span>Fitting the threshold</a></span></li><li><span><a href=\"#Changing-the-D'agostino-statistic\" data-toc-modified-id=\"Changing-the-D'agostino-statistic-7.2\"><span class=\"toc-item-num\">7.2&nbsp;&nbsp;</span>Changing the D'agostino statistic</a></span></li><li><span><a href=\"#Manually-changing-thresholds\" data-toc-modified-id=\"Manually-changing-thresholds-7.3\"><span class=\"toc-item-num\">7.3&nbsp;&nbsp;</span>Manually changing thresholds</a></span></li></ul></li><li><span><a href=\"#Plotting\" data-toc-modified-id=\"Plotting-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Plotting</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.util import *\n",
    "from os.path import abspath, join\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = join('pymodulon', 'test', 'data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quickstart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = join(DATA_DIR, 'ecoli_M.csv')\n",
    "a = join(DATA_DIR, 'ecoli_A.csv')\n",
    "x = join(DATA_DIR, 'ecoli_X.csv')\n",
    "\n",
    "gene_file = join(DATA_DIR, 'ecoli_gene_table.csv')\n",
    "sample_file = join(DATA_DIR, 'ecoli_sample_table.csv')\n",
    "imodulon_file = join(DATA_DIR, 'ecoli_imodulon_table.csv')\n",
    "\n",
    "trn = pd.read_csv(join(DATA_DIR, 'ecoli_trn.csv'))\n",
    "\n",
    "ica_data = IcaData(m,a,x,\n",
    "                   gene_table=gene_file,\n",
    "                   sample_table=sample_file,\n",
    "                   imodulon_table=imodulon_file,\n",
    "                   trn=trn,\n",
    "                   dagostino_cutoff=550)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0306</th>\n",
       "      <td>0.237575</td>\n",
       "      <td>321607</td>\n",
       "      <td>322327</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgE</td>\n",
       "      <td>720</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>lrp,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0307</th>\n",
       "      <td>0.157379</td>\n",
       "      <td>322337</td>\n",
       "      <td>323765</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgF</td>\n",
       "      <td>1428</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>lrp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0308</th>\n",
       "      <td>0.157886</td>\n",
       "      <td>323757</td>\n",
       "      <td>324453</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgG</td>\n",
       "      <td>696</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>lrp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0447</th>\n",
       "      <td>0.098046</td>\n",
       "      <td>468382</td>\n",
       "      <td>468841</td>\n",
       "      <td>+</td>\n",
       "      <td>ybaO</td>\n",
       "      <td>459</td>\n",
       "      <td>decR</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>marA,rpoD,soxS,ybaO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1826</th>\n",
       "      <td>-0.124414</td>\n",
       "      <td>1908622</td>\n",
       "      <td>1908766</td>\n",
       "      <td>-</td>\n",
       "      <td>mgrB</td>\n",
       "      <td>144</td>\n",
       "      <td>mgrB</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>phoP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2141</th>\n",
       "      <td>0.264099</td>\n",
       "      <td>2230623</td>\n",
       "      <td>2231022</td>\n",
       "      <td>+</td>\n",
       "      <td>yohJ</td>\n",
       "      <td>399</td>\n",
       "      <td>yohJK</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2142</th>\n",
       "      <td>0.202429</td>\n",
       "      <td>2231018</td>\n",
       "      <td>2231714</td>\n",
       "      <td>+</td>\n",
       "      <td>yohK</td>\n",
       "      <td>696</td>\n",
       "      <td>yohJK</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2143</th>\n",
       "      <td>0.093723</td>\n",
       "      <td>2231843</td>\n",
       "      <td>2232728</td>\n",
       "      <td>+</td>\n",
       "      <td>cdd</td>\n",
       "      <td>885</td>\n",
       "      <td>cdd</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>crp,cytR,nsrR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3755</th>\n",
       "      <td>-0.348577</td>\n",
       "      <td>3940634</td>\n",
       "      <td>3941327</td>\n",
       "      <td>-</td>\n",
       "      <td>yieP</td>\n",
       "      <td>693</td>\n",
       "      <td>yieP-hsrA</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4116</th>\n",
       "      <td>0.137769</td>\n",
       "      <td>4337167</td>\n",
       "      <td>4337929</td>\n",
       "      <td>-</td>\n",
       "      <td>adiY</td>\n",
       "      <td>762</td>\n",
       "      <td>adiY</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>h-NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4354</th>\n",
       "      <td>0.136132</td>\n",
       "      <td>4589128</td>\n",
       "      <td>4591279</td>\n",
       "      <td>-</td>\n",
       "      <td>yjiY</td>\n",
       "      <td>2151</td>\n",
       "      <td>btsT</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>btsR,crp,rpoS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length     operon  \\\n",
       "b0306     0.237575   321607   322327      +      ykgE     720     ykgEFG   \n",
       "b0307     0.157379   322337   323765      +      ykgF    1428     ykgEFG   \n",
       "b0308     0.157886   323757   324453      +      ykgG     696     ykgEFG   \n",
       "b0447     0.098046   468382   468841      +      ybaO     459       decR   \n",
       "b1826    -0.124414  1908622  1908766      -      mgrB     144       mgrB   \n",
       "b2141     0.264099  2230623  2231022      +      yohJ     399      yohJK   \n",
       "b2142     0.202429  2231018  2231714      +      yohK     696      yohJK   \n",
       "b2143     0.093723  2231843  2232728      +       cdd     885        cdd   \n",
       "b3755    -0.348577  3940634  3941327      -      yieP     693  yieP-hsrA   \n",
       "b4116     0.137769  4337167  4337929      -      adiY     762       adiY   \n",
       "b4354     0.136132  4589128  4591279      -      yjiY    2151       btsT   \n",
       "\n",
       "                                          cog            regulator  \n",
       "b0306        Energy production and conversion             lrp,yieP  \n",
       "b0307        Energy production and conversion                  lrp  \n",
       "b0308                        Function unknown                  lrp  \n",
       "b0447                           Transcription  marA,rpoD,soxS,ybaO  \n",
       "b1826                        Function unknown                 phoP  \n",
       "b2141                        Function unknown                 yieP  \n",
       "b2142  Cell wall/membrane/envelope biogenesis                  NaN  \n",
       "b2143     Nucleotide transport and metabolism   crp,cytR,nsrR,rpoD  \n",
       "b3755                           Transcription                  NaN  \n",
       "b4116                           Transcription                 h-NS  \n",
       "b4354          Signal transduction mechanisms        btsR,crp,rpoS  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Initialize your IcaData object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Minimum requirements\n",
    "The IcaData object only requires two matrices, the M and the A matrices. However, many functions (such as TRN enrichment) will be limited without including additional data.  \n",
    "These matrices can be added as either a filename or a pandas DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = abspath(join(DATA_DIR, 'ecoli_M.csv'))\n",
    "a = abspath(join(DATA_DIR, 'ecoli_A.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = IcaData(m,a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viewing matrices\n",
    "You can directly view the M and A matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control__wt_glc__1</th>\n",
       "      <th>control__wt_glc__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_dpd__1</th>\n",
       "      <th>fur__delfur_dpd__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>...</th>\n",
       "      <th>efeU__menFentC_ale29__1</th>\n",
       "      <th>efeU__menFentC_ale29__2</th>\n",
       "      <th>efeU__menFentC_ale30__1</th>\n",
       "      <th>efeU__menFentC_ale30__2</th>\n",
       "      <th>efeU__menFentCubiC_ale36__1</th>\n",
       "      <th>efeU__menFentCubiC_ale36__2</th>\n",
       "      <th>efeU__menFentCubiC_ale37__1</th>\n",
       "      <th>efeU__menFentCubiC_ale37__2</th>\n",
       "      <th>efeU__menFentCubiC_ale38__1</th>\n",
       "      <th>efeU__menFentCubiC_ale38__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.076035</td>\n",
       "      <td>-0.076035</td>\n",
       "      <td>1.483992</td>\n",
       "      <td>1.719034</td>\n",
       "      <td>-1.184220</td>\n",
       "      <td>-1.350203</td>\n",
       "      <td>1.451882</td>\n",
       "      <td>1.329577</td>\n",
       "      <td>0.265718</td>\n",
       "      <td>0.024864</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.892372</td>\n",
       "      <td>-1.022146</td>\n",
       "      <td>0.253722</td>\n",
       "      <td>-0.880352</td>\n",
       "      <td>0.526509</td>\n",
       "      <td>1.261740</td>\n",
       "      <td>1.727782</td>\n",
       "      <td>1.512944</td>\n",
       "      <td>2.584012</td>\n",
       "      <td>2.787686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.408568</td>\n",
       "      <td>0.408568</td>\n",
       "      <td>1.760680</td>\n",
       "      <td>1.357274</td>\n",
       "      <td>-0.446482</td>\n",
       "      <td>-0.931725</td>\n",
       "      <td>0.517458</td>\n",
       "      <td>0.427993</td>\n",
       "      <td>-0.499529</td>\n",
       "      <td>-0.581335</td>\n",
       "      <td>...</td>\n",
       "      <td>1.781224</td>\n",
       "      <td>2.785099</td>\n",
       "      <td>0.439133</td>\n",
       "      <td>-0.015471</td>\n",
       "      <td>0.366544</td>\n",
       "      <td>0.369589</td>\n",
       "      <td>0.325470</td>\n",
       "      <td>1.027508</td>\n",
       "      <td>1.629269</td>\n",
       "      <td>2.154717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.103912</td>\n",
       "      <td>-0.103912</td>\n",
       "      <td>0.847276</td>\n",
       "      <td>1.530875</td>\n",
       "      <td>0.258492</td>\n",
       "      <td>0.188903</td>\n",
       "      <td>1.232846</td>\n",
       "      <td>0.883215</td>\n",
       "      <td>0.505356</td>\n",
       "      <td>0.132219</td>\n",
       "      <td>...</td>\n",
       "      <td>1.422466</td>\n",
       "      <td>2.396554</td>\n",
       "      <td>2.130935</td>\n",
       "      <td>1.656316</td>\n",
       "      <td>0.864338</td>\n",
       "      <td>0.093698</td>\n",
       "      <td>0.641705</td>\n",
       "      <td>1.306455</td>\n",
       "      <td>1.366912</td>\n",
       "      <td>2.452061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.456843</td>\n",
       "      <td>0.456843</td>\n",
       "      <td>5.890468</td>\n",
       "      <td>6.051415</td>\n",
       "      <td>-0.914670</td>\n",
       "      <td>-1.087245</td>\n",
       "      <td>5.210846</td>\n",
       "      <td>5.028641</td>\n",
       "      <td>3.820172</td>\n",
       "      <td>3.961980</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.343743</td>\n",
       "      <td>-1.991048</td>\n",
       "      <td>3.520794</td>\n",
       "      <td>3.105551</td>\n",
       "      <td>6.620301</td>\n",
       "      <td>5.400225</td>\n",
       "      <td>5.545880</td>\n",
       "      <td>5.555621</td>\n",
       "      <td>9.672308</td>\n",
       "      <td>10.109228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.269057</td>\n",
       "      <td>-0.269057</td>\n",
       "      <td>-0.254271</td>\n",
       "      <td>0.184537</td>\n",
       "      <td>0.651527</td>\n",
       "      <td>0.691914</td>\n",
       "      <td>0.016636</td>\n",
       "      <td>-0.465781</td>\n",
       "      <td>1.212097</td>\n",
       "      <td>1.122801</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.646464</td>\n",
       "      <td>-3.606055</td>\n",
       "      <td>-0.665810</td>\n",
       "      <td>-0.830586</td>\n",
       "      <td>-1.044574</td>\n",
       "      <td>0.119876</td>\n",
       "      <td>-0.045670</td>\n",
       "      <td>-0.826608</td>\n",
       "      <td>-1.393160</td>\n",
       "      <td>-1.587978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 278 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   control__wt_glc__1  control__wt_glc__2  fur__wt_dpd__1  fur__wt_dpd__2  \\\n",
       "0            0.076035           -0.076035        1.483992        1.719034   \n",
       "1           -0.408568            0.408568        1.760680        1.357274   \n",
       "2            0.103912           -0.103912        0.847276        1.530875   \n",
       "3           -0.456843            0.456843        5.890468        6.051415   \n",
       "4            0.269057           -0.269057       -0.254271        0.184537   \n",
       "\n",
       "   fur__wt_fe__1  fur__wt_fe__2  fur__delfur_dpd__1  fur__delfur_dpd__2  \\\n",
       "0      -1.184220      -1.350203            1.451882            1.329577   \n",
       "1      -0.446482      -0.931725            0.517458            0.427993   \n",
       "2       0.258492       0.188903            1.232846            0.883215   \n",
       "3      -0.914670      -1.087245            5.210846            5.028641   \n",
       "4       0.651527       0.691914            0.016636           -0.465781   \n",
       "\n",
       "   fur__delfur_fe2__1  fur__delfur_fe2__2  ...  efeU__menFentC_ale29__1  \\\n",
       "0            0.265718            0.024864  ...                -0.892372   \n",
       "1           -0.499529           -0.581335  ...                 1.781224   \n",
       "2            0.505356            0.132219  ...                 1.422466   \n",
       "3            3.820172            3.961980  ...                -2.343743   \n",
       "4            1.212097            1.122801  ...                -3.646464   \n",
       "\n",
       "   efeU__menFentC_ale29__2  efeU__menFentC_ale30__1  efeU__menFentC_ale30__2  \\\n",
       "0                -1.022146                 0.253722                -0.880352   \n",
       "1                 2.785099                 0.439133                -0.015471   \n",
       "2                 2.396554                 2.130935                 1.656316   \n",
       "3                -1.991048                 3.520794                 3.105551   \n",
       "4                -3.606055                -0.665810                -0.830586   \n",
       "\n",
       "   efeU__menFentCubiC_ale36__1  efeU__menFentCubiC_ale36__2  \\\n",
       "0                     0.526509                     1.261740   \n",
       "1                     0.366544                     0.369589   \n",
       "2                     0.864338                     0.093698   \n",
       "3                     6.620301                     5.400225   \n",
       "4                    -1.044574                     0.119876   \n",
       "\n",
       "   efeU__menFentCubiC_ale37__1  efeU__menFentCubiC_ale37__2  \\\n",
       "0                     1.727782                     1.512944   \n",
       "1                     0.325470                     1.027508   \n",
       "2                     0.641705                     1.306455   \n",
       "3                     5.545880                     5.555621   \n",
       "4                    -0.045670                    -0.826608   \n",
       "\n",
       "   efeU__menFentCubiC_ale38__1  efeU__menFentCubiC_ale38__2  \n",
       "0                     2.584012                     2.787686  \n",
       "1                     1.629269                     2.154717  \n",
       "2                     1.366912                     2.452061  \n",
       "3                     9.672308                    10.109228  \n",
       "4                    -1.393160                    -1.587978  \n",
       "\n",
       "[5 rows x 278 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.A.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample, gene, and iModulon names\n",
    "If the M and A matrices have sample/gene/iModulon names, these will be saved as the official sample/gene/iModulon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene names: ['b0002', 'b0003', 'b0004', 'b0005', 'b0006']\n",
      "Sample names: ['control__wt_glc__1', 'control__wt_glc__2', 'fur__wt_dpd__1', 'fur__wt_dpd__2', 'fur__wt_fe__1']\n",
      "iModulon names: [0, 1, 2, 3, 4]\n"
     ]
    }
   ],
   "source": [
    "print('Gene names:',ica_data.gene_names[:5])\n",
    "print('Sample names:',ica_data.sample_names[:5])\n",
    "print('iModulon names:',ica_data.imodulon_names[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding the X matrix\n",
    "The X matrix contains expression data and is mainly used for plotting functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = abspath(join(DATA_DIR, 'ecoli_X.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.X = x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample, gene, and iModulon tables\n",
    "You may load in additional data tables with information about your samples, genes, or iModulons. If you do not load any data, these tables are originally empty but can be altered like any pandas table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [b0002, b0003, b0004, b0005, b0006]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.gene_table['gene_name'] = 'example'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "b0002   example\n",
       "b0003   example\n",
       "b0004   example\n",
       "b0005   example\n",
       "b0006   example"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading tables\n",
    "Sample/gene/iModulon tables contain one sample/gene/iModulon per row, and information about the respective item in columns.  \n",
    "**The sample/gene/iModulon names must match the names in your M and A matrix.** However, the order of the names does not matter.  \n",
    "The tables can be passed in using either a filename or a pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene file: /home/sbrg/Desktop/pymodulon/pymodulon/test/data/ecoli_gene_table.csv\n"
     ]
    }
   ],
   "source": [
    "gene_file = abspath(join(DATA_DIR, 'ecoli_gene_table.csv'))\n",
    "sample_file = abspath(join(DATA_DIR, 'ecoli_sample_table.csv'))\n",
    "imodulon_file = abspath(join(DATA_DIR, 'ecoli_imodulon_table.csv'))\n",
    "print('Gene file:',gene_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.gene_table = gene_file\n",
    "ica_data.sample_table = sample_file\n",
    "ica_data.imodulon_table = imodulon_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viewing tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>project_id</th>\n",
       "      <th>condition_id</th>\n",
       "      <th>rep_id</th>\n",
       "      <th>contact</th>\n",
       "      <th>library creator</th>\n",
       "      <th>run_date</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>BAM</th>\n",
       "      <th>alignment</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "      <th>Public</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>control__wt_glc__1</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>04/17/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>control/MG1655-M9-glucose-1_S1_L001_R1_001.fas...</td>\n",
       "      <td>control/MG1655-M9-glucose-1_S1_L001_R2_001.fas...</td>\n",
       "      <td>control__wt_glc__1.bam</td>\n",
       "      <td>94.33</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control__wt_glc__2</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>2</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>04/17/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>control/MG1655-M9-glucose-2_S2_L001_R1_001.fas...</td>\n",
       "      <td>control/MG1655-M9-glucose-2_S2_L001_R2_001.fas...</td>\n",
       "      <td>control__wt_glc__2.bam</td>\n",
       "      <td>94.24</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>10/02/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WTDPD1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WTDPD1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_dpd__1.bam</td>\n",
       "      <td>98.04</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>2</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>10/02/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WTDPD2_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WTDPD2_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_dpd__2.bam</td>\n",
       "      <td>98.30</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>09/05/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WT-Fe2-1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WT-Fe2-1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_fe__1.bam</td>\n",
       "      <td>93.35</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      study project_id condition_id  rep_id      contact  \\\n",
       "control__wt_glc__1  Control    control       wt_glc       1  Sangwoo Seo   \n",
       "control__wt_glc__2  Control    control       wt_glc       2  Sangwoo Seo   \n",
       "fur__wt_dpd__1          Fur        fur       wt_dpd       1  Sangwoo Seo   \n",
       "fur__wt_dpd__2          Fur        fur       wt_dpd       2  Sangwoo Seo   \n",
       "fur__wt_fe__1           Fur        fur        wt_fe       1  Sangwoo Seo   \n",
       "\n",
       "                   library creator  run_date            Strain Description  \\\n",
       "control__wt_glc__1     Sangwoo Seo  04/17/13  Escherichia coli K-12 MG1655   \n",
       "control__wt_glc__2     Sangwoo Seo  04/17/13  Escherichia coli K-12 MG1655   \n",
       "fur__wt_dpd__1         Sangwoo Seo  10/02/13  Escherichia coli K-12 MG1655   \n",
       "fur__wt_dpd__2         Sangwoo Seo  10/02/13  Escherichia coli K-12 MG1655   \n",
       "fur__wt_fe__1          Sangwoo Seo  09/05/13  Escherichia coli K-12 MG1655   \n",
       "\n",
       "                    Strain Base Media  ... Sequencing Machine  \\\n",
       "control__wt_glc__1  MG1655         M9  ...              MiSeq   \n",
       "control__wt_glc__2  MG1655         M9  ...              MiSeq   \n",
       "fur__wt_dpd__1      MG1655         M9  ...              MiSeq   \n",
       "fur__wt_dpd__2      MG1655         M9  ...              MiSeq   \n",
       "fur__wt_fe__1       MG1655         M9  ...              MiSeq   \n",
       "\n",
       "                   Additional Details Biological Replicates  \\\n",
       "control__wt_glc__1                NaN                     2   \n",
       "control__wt_glc__2                NaN                     2   \n",
       "fur__wt_dpd__1                    NaN                     2   \n",
       "fur__wt_dpd__2                    NaN                     2   \n",
       "fur__wt_fe__1                     NaN                     2   \n",
       "\n",
       "                                                                   R1  \\\n",
       "control__wt_glc__1  control/MG1655-M9-glucose-1_S1_L001_R1_001.fas...   \n",
       "control__wt_glc__2  control/MG1655-M9-glucose-2_S2_L001_R1_001.fas...   \n",
       "fur__wt_dpd__1                     fur/WTDPD1_S1_L001_R1_001.fastq.gz   \n",
       "fur__wt_dpd__2                     fur/WTDPD2_S1_L001_R1_001.fastq.gz   \n",
       "fur__wt_fe__1                    fur/WT-Fe2-1_S1_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                                                   R2  \\\n",
       "control__wt_glc__1  control/MG1655-M9-glucose-1_S1_L001_R2_001.fas...   \n",
       "control__wt_glc__2  control/MG1655-M9-glucose-2_S2_L001_R2_001.fas...   \n",
       "fur__wt_dpd__1                     fur/WTDPD1_S1_L001_R2_001.fastq.gz   \n",
       "fur__wt_dpd__2                     fur/WTDPD2_S1_L001_R2_001.fastq.gz   \n",
       "fur__wt_fe__1                    fur/WT-Fe2-1_S1_L001_R2_001.fastq.gz   \n",
       "\n",
       "                                       BAM  alignment  \\\n",
       "control__wt_glc__1  control__wt_glc__1.bam      94.33   \n",
       "control__wt_glc__2  control__wt_glc__2.bam      94.24   \n",
       "fur__wt_dpd__1          fur__wt_dpd__1.bam      98.04   \n",
       "fur__wt_dpd__2          fur__wt_dpd__2.bam      98.30   \n",
       "fur__wt_fe__1            fur__wt_fe__1.bam      93.35   \n",
       "\n",
       "                                           DOI       GEO  Public  \n",
       "control__wt_glc__1      doi.org/10.1101/080929  GSE65643       1  \n",
       "control__wt_glc__2      doi.org/10.1101/080929  GSE65643       1  \n",
       "fur__wt_dpd__1      doi.org/10.1038/ncomms5910  GSE54900       1  \n",
       "fur__wt_dpd__2      doi.org/10.1038/ncomms5910  GSE54900       1  \n",
       "fur__wt_fe__1       doi.org/10.1038/ncomms5910  GSE54900       1  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.sample_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>336</td>\n",
       "      <td>2799</td>\n",
       "      <td>+</td>\n",
       "      <td>thrA</td>\n",
       "      <td>2463</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>2800</td>\n",
       "      <td>3733</td>\n",
       "      <td>+</td>\n",
       "      <td>thrB</td>\n",
       "      <td>933</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>3733</td>\n",
       "      <td>5020</td>\n",
       "      <td>+</td>\n",
       "      <td>thrC</td>\n",
       "      <td>1287</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>5233</td>\n",
       "      <td>5530</td>\n",
       "      <td>+</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>297</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>5682</td>\n",
       "      <td>6459</td>\n",
       "      <td>-</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>777</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start  stop strand gene_name  length   operon  \\\n",
       "b0002    336  2799      +      thrA    2463  thrLABC   \n",
       "b0003   2800  3733      +      thrB     933  thrLABC   \n",
       "b0004   3733  5020      +      thrC    1287  thrLABC   \n",
       "b0005   5233  5530      +      yaaX     297     yaaX   \n",
       "b0006   5682  6459      -      yaaA     777     yaaA   \n",
       "\n",
       "                                       cog  \n",
       "b0002  Amino acid transport and metabolism  \n",
       "b0003  Amino acid transport and metabolism  \n",
       "b0004  Amino acid transport and metabolism  \n",
       "b0005                     Function unknown  \n",
       "b0006                     Function unknown  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YieP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YgbI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>efeU-repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ydcI-KO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          name regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "0         YieP       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "1         GlpR      glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "2         YgbI       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "3  efeU-repair       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "4      ydcI-KO       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "   regulon_size  n_regs                                           Function  \\\n",
       "0            11       1                                            Unknown   \n",
       "1             9       1                                Glycerol catabolism   \n",
       "2             9       1                                   Unknown Function   \n",
       "3             8       0  Accounts for repair and expression of efeU operon   \n",
       "4             3       0                        Accounts for ydcI knock-out   \n",
       "\n",
       "                    Category  \n",
       "0        Regulator Discovery  \n",
       "1  Carbon Source Utilization  \n",
       "2        Regulator Discovery  \n",
       "3        Genomic Alterations  \n",
       "4        Genomic Alterations  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRN Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the TRN\n",
    "Again, this table can be passed in using either a filename or a pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "trn = pd.read_csv(abspath(join(DATA_DIR, 'ecoli_trn.csv')))\n",
    "ica_data.trn = trn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>regulator_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>effect</th>\n",
       "      <th>ev_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FMN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b3041</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-tryptophan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b3708</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-tryptophan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b3709</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TPP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b0066</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TPP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b0067</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      regulator  regulator_id  gene_name gene_id effect  ev_level\n",
       "0           FMN           NaN        NaN   b3041      -       NaN\n",
       "1  L-tryptophan           NaN        NaN   b3708      +       NaN\n",
       "2  L-tryptophan           NaN        NaN   b3709      +       NaN\n",
       "3           TPP           NaN        NaN   b0066      -       NaN\n",
       "4           TPP           NaN        NaN   b0067      -       NaN"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iModulon properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspecting iModulons\n",
    "The `view_imodulon` shows the information about each gene in the iModulon. Most information is stored in the `gene_table`, but the `regulator` column comes from the TRN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0306</th>\n",
       "      <td>0.237575</td>\n",
       "      <td>321607</td>\n",
       "      <td>322327</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgE</td>\n",
       "      <td>720</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>lrp,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0307</th>\n",
       "      <td>0.157379</td>\n",
       "      <td>322337</td>\n",
       "      <td>323765</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgF</td>\n",
       "      <td>1428</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>lrp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0308</th>\n",
       "      <td>0.157886</td>\n",
       "      <td>323757</td>\n",
       "      <td>324453</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgG</td>\n",
       "      <td>696</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>lrp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0447</th>\n",
       "      <td>0.098046</td>\n",
       "      <td>468382</td>\n",
       "      <td>468841</td>\n",
       "      <td>+</td>\n",
       "      <td>ybaO</td>\n",
       "      <td>459</td>\n",
       "      <td>decR</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>marA,rpoD,soxS,ybaO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1826</th>\n",
       "      <td>-0.124414</td>\n",
       "      <td>1908622</td>\n",
       "      <td>1908766</td>\n",
       "      <td>-</td>\n",
       "      <td>mgrB</td>\n",
       "      <td>144</td>\n",
       "      <td>mgrB</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>phoP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2141</th>\n",
       "      <td>0.264099</td>\n",
       "      <td>2230623</td>\n",
       "      <td>2231022</td>\n",
       "      <td>+</td>\n",
       "      <td>yohJ</td>\n",
       "      <td>399</td>\n",
       "      <td>yohJK</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2142</th>\n",
       "      <td>0.202429</td>\n",
       "      <td>2231018</td>\n",
       "      <td>2231714</td>\n",
       "      <td>+</td>\n",
       "      <td>yohK</td>\n",
       "      <td>696</td>\n",
       "      <td>yohJK</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2143</th>\n",
       "      <td>0.093723</td>\n",
       "      <td>2231843</td>\n",
       "      <td>2232728</td>\n",
       "      <td>+</td>\n",
       "      <td>cdd</td>\n",
       "      <td>885</td>\n",
       "      <td>cdd</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>crp,cytR,nsrR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3755</th>\n",
       "      <td>-0.348577</td>\n",
       "      <td>3940634</td>\n",
       "      <td>3941327</td>\n",
       "      <td>-</td>\n",
       "      <td>yieP</td>\n",
       "      <td>693</td>\n",
       "      <td>yieP-hsrA</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4116</th>\n",
       "      <td>0.137769</td>\n",
       "      <td>4337167</td>\n",
       "      <td>4337929</td>\n",
       "      <td>-</td>\n",
       "      <td>adiY</td>\n",
       "      <td>762</td>\n",
       "      <td>adiY</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>h-NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4354</th>\n",
       "      <td>0.136132</td>\n",
       "      <td>4589128</td>\n",
       "      <td>4591279</td>\n",
       "      <td>-</td>\n",
       "      <td>yjiY</td>\n",
       "      <td>2151</td>\n",
       "      <td>btsT</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>btsR,crp,rpoS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length     operon  \\\n",
       "b0306     0.237575   321607   322327      +      ykgE     720     ykgEFG   \n",
       "b0307     0.157379   322337   323765      +      ykgF    1428     ykgEFG   \n",
       "b0308     0.157886   323757   324453      +      ykgG     696     ykgEFG   \n",
       "b0447     0.098046   468382   468841      +      ybaO     459       decR   \n",
       "b1826    -0.124414  1908622  1908766      -      mgrB     144       mgrB   \n",
       "b2141     0.264099  2230623  2231022      +      yohJ     399      yohJK   \n",
       "b2142     0.202429  2231018  2231714      +      yohK     696      yohJK   \n",
       "b2143     0.093723  2231843  2232728      +       cdd     885        cdd   \n",
       "b3755    -0.348577  3940634  3941327      -      yieP     693  yieP-hsrA   \n",
       "b4116     0.137769  4337167  4337929      -      adiY     762       adiY   \n",
       "b4354     0.136132  4589128  4591279      -      yjiY    2151       btsT   \n",
       "\n",
       "                                          cog            regulator  \n",
       "b0306        Energy production and conversion             lrp,yieP  \n",
       "b0307        Energy production and conversion                  lrp  \n",
       "b0308                        Function unknown                  lrp  \n",
       "b0447                           Transcription  marA,rpoD,soxS,ybaO  \n",
       "b1826                        Function unknown                 phoP  \n",
       "b2141                        Function unknown                 yieP  \n",
       "b2142  Cell wall/membrane/envelope biogenesis                  NaN  \n",
       "b2143     Nucleotide transport and metabolism   crp,cytR,nsrR,rpoD  \n",
       "b3755                           Transcription                  NaN  \n",
       "b4116                           Transcription                 h-NS  \n",
       "b4354          Signal transduction mechanisms        btsR,crp,rpoS  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Renaming iModulons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>YieP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YgbI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>efeU-repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ydcI-KO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             name regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "YieP         YieP       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "1            GlpR      glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "2            YgbI       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "3     efeU-repair       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "4         ydcI-KO       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "      regulon_size  n_regs                                           Function  \\\n",
       "YieP            11       1                                            Unknown   \n",
       "1                9       1                                Glycerol catabolism   \n",
       "2                9       1                                   Unknown Function   \n",
       "3                8       0  Accounts for repair and expression of efeU operon   \n",
       "4                3       0                        Accounts for ydcI knock-out   \n",
       "\n",
       "                       Category  \n",
       "YieP        Regulator Discovery  \n",
       "1     Carbon Source Utilization  \n",
       "2           Regulator Discovery  \n",
       "3           Genomic Alterations  \n",
       "4           Genomic Alterations  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename iModulons using a dict\n",
    "ica_data.rename_imodulons({0:'YieP'})\n",
    "\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "name                                                                     \n",
       "YieP              NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "GlpR             glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "YgbI              NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "efeU-repair       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "ydcI-KO           NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "             regulon_size  n_regs  \\\n",
       "name                                \n",
       "YieP                   11       1   \n",
       "GlpR                    9       1   \n",
       "YgbI                    9       1   \n",
       "efeU-repair             8       0   \n",
       "ydcI-KO                 3       0   \n",
       "\n",
       "                                                      Function  \\\n",
       "name                                                             \n",
       "YieP                                                   Unknown   \n",
       "GlpR                                       Glycerol catabolism   \n",
       "YgbI                                          Unknown Function   \n",
       "efeU-repair  Accounts for repair and expression of efeU operon   \n",
       "ydcI-KO                            Accounts for ydcI knock-out   \n",
       "\n",
       "                              Category  \n",
       "name                                    \n",
       "YieP               Regulator Discovery  \n",
       "GlpR         Carbon Source Utilization  \n",
       "YgbI               Regulator Discovery  \n",
       "efeU-repair        Genomic Alterations  \n",
       "ydcI-KO            Genomic Alterations  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename iModulons using the iModulon table\n",
    "ica_data.rename_imodulons(column='name')\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>name</th>\n",
       "      <th>YieP</th>\n",
       "      <th>GlpR</th>\n",
       "      <th>YgbI</th>\n",
       "      <th>efeU-repair</th>\n",
       "      <th>ydcI-KO</th>\n",
       "      <th>BW25113</th>\n",
       "      <th>proVWX</th>\n",
       "      <th>DhaR/Mlc</th>\n",
       "      <th>ArgR</th>\n",
       "      <th>YneJ</th>\n",
       "      <th>...</th>\n",
       "      <th>RpoS</th>\n",
       "      <th>membrane</th>\n",
       "      <th>duplication-1</th>\n",
       "      <th>uncharacterized-3</th>\n",
       "      <th>ExuR/FucR</th>\n",
       "      <th>FecI</th>\n",
       "      <th>iron-related</th>\n",
       "      <th>RbsR</th>\n",
       "      <th>sgrT</th>\n",
       "      <th>insertion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>0.072693</td>\n",
       "      <td>-0.006320</td>\n",
       "      <td>-0.038374</td>\n",
       "      <td>-0.010045</td>\n",
       "      <td>-0.004345</td>\n",
       "      <td>-0.005599</td>\n",
       "      <td>0.009043</td>\n",
       "      <td>-0.005072</td>\n",
       "      <td>-0.012186</td>\n",
       "      <td>-0.013543</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.010152</td>\n",
       "      <td>-0.012124</td>\n",
       "      <td>-0.012027</td>\n",
       "      <td>-0.002123</td>\n",
       "      <td>-0.000036</td>\n",
       "      <td>-0.013667</td>\n",
       "      <td>-0.018883</td>\n",
       "      <td>0.004408</td>\n",
       "      <td>-0.006315</td>\n",
       "      <td>-0.001780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>0.015975</td>\n",
       "      <td>-0.009969</td>\n",
       "      <td>-0.024571</td>\n",
       "      <td>-0.006775</td>\n",
       "      <td>-0.003583</td>\n",
       "      <td>0.006680</td>\n",
       "      <td>-0.010841</td>\n",
       "      <td>-0.012620</td>\n",
       "      <td>-0.003685</td>\n",
       "      <td>0.003512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004702</td>\n",
       "      <td>0.004402</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>-0.004415</td>\n",
       "      <td>-0.001514</td>\n",
       "      <td>-0.005052</td>\n",
       "      <td>0.015502</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>-0.022438</td>\n",
       "      <td>-0.020087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>0.015319</td>\n",
       "      <td>-0.003808</td>\n",
       "      <td>-0.027018</td>\n",
       "      <td>-0.004455</td>\n",
       "      <td>-0.008418</td>\n",
       "      <td>0.003245</td>\n",
       "      <td>-0.016494</td>\n",
       "      <td>-0.013818</td>\n",
       "      <td>-0.008104</td>\n",
       "      <td>0.002144</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003507</td>\n",
       "      <td>0.005342</td>\n",
       "      <td>0.003759</td>\n",
       "      <td>0.000700</td>\n",
       "      <td>0.001859</td>\n",
       "      <td>-0.002344</td>\n",
       "      <td>0.016263</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>-0.016565</td>\n",
       "      <td>-0.014142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>0.008229</td>\n",
       "      <td>-0.000685</td>\n",
       "      <td>-0.005215</td>\n",
       "      <td>-0.002029</td>\n",
       "      <td>-0.028069</td>\n",
       "      <td>0.014748</td>\n",
       "      <td>-0.019240</td>\n",
       "      <td>-0.002403</td>\n",
       "      <td>-0.008507</td>\n",
       "      <td>-0.019092</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.006856</td>\n",
       "      <td>0.010676</td>\n",
       "      <td>0.008874</td>\n",
       "      <td>0.018497</td>\n",
       "      <td>0.000700</td>\n",
       "      <td>-0.013267</td>\n",
       "      <td>0.015265</td>\n",
       "      <td>0.016886</td>\n",
       "      <td>-0.015319</td>\n",
       "      <td>-0.020727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>-0.004941</td>\n",
       "      <td>-0.001849</td>\n",
       "      <td>-0.002326</td>\n",
       "      <td>0.006075</td>\n",
       "      <td>0.001607</td>\n",
       "      <td>0.009678</td>\n",
       "      <td>0.001661</td>\n",
       "      <td>0.003754</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>-0.001013</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014761</td>\n",
       "      <td>0.005071</td>\n",
       "      <td>0.001496</td>\n",
       "      <td>-0.002506</td>\n",
       "      <td>-0.002819</td>\n",
       "      <td>0.005800</td>\n",
       "      <td>0.003975</td>\n",
       "      <td>-0.010976</td>\n",
       "      <td>-0.008756</td>\n",
       "      <td>0.006169</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "name       YieP      GlpR      YgbI  efeU-repair   ydcI-KO   BW25113  \\\n",
       "b0002  0.072693 -0.006320 -0.038374    -0.010045 -0.004345 -0.005599   \n",
       "b0003  0.015975 -0.009969 -0.024571    -0.006775 -0.003583  0.006680   \n",
       "b0004  0.015319 -0.003808 -0.027018    -0.004455 -0.008418  0.003245   \n",
       "b0005  0.008229 -0.000685 -0.005215    -0.002029 -0.028069  0.014748   \n",
       "b0006 -0.004941 -0.001849 -0.002326     0.006075  0.001607  0.009678   \n",
       "\n",
       "name     proVWX  DhaR/Mlc      ArgR      YneJ  ...      RpoS  membrane  \\\n",
       "b0002  0.009043 -0.005072 -0.012186 -0.013543  ... -0.010152 -0.012124   \n",
       "b0003 -0.010841 -0.012620 -0.003685  0.003512  ...  0.004702  0.004402   \n",
       "b0004 -0.016494 -0.013818 -0.008104  0.002144  ...  0.003507  0.005342   \n",
       "b0005 -0.019240 -0.002403 -0.008507 -0.019092  ... -0.006856  0.010676   \n",
       "b0006  0.001661  0.003754 -0.000152 -0.001013  ... -0.014761  0.005071   \n",
       "\n",
       "name   duplication-1  uncharacterized-3  ExuR/FucR      FecI  iron-related  \\\n",
       "b0002      -0.012027          -0.002123  -0.000036 -0.013667     -0.018883   \n",
       "b0003       0.001182          -0.004415  -0.001514 -0.005052      0.015502   \n",
       "b0004       0.003759           0.000700   0.001859 -0.002344      0.016263   \n",
       "b0005       0.008874           0.018497   0.000700 -0.013267      0.015265   \n",
       "b0006       0.001496          -0.002506  -0.002819  0.005800      0.003975   \n",
       "\n",
       "name       RbsR      sgrT  insertion  \n",
       "b0002  0.004408 -0.006315  -0.001780  \n",
       "b0003  0.000860 -0.022438  -0.020087  \n",
       "b0004  0.002486 -0.016565  -0.014142  \n",
       "b0005  0.016886 -0.015319  -0.020727  \n",
       "b0006 -0.010976 -0.008756   0.006169  \n",
       "\n",
       "[5 rows x 92 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This also renames the M and A matrix\n",
    "ica_data.M.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying \"single-gene\" iModulons\n",
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "2. Artificial knock-out of single genes\n",
    "3. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The `find_single_gene_imodulons` function identifies iModulons that are likely dominated by a single gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ydcI-KO', 'fur-KO', 'Pyruvate', 'thrA-KO', 'sgrT']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.find_single_gene_imodulons(save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>fur-KO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for fur knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       regulator  pvalue  qvalue  precision  recall  f1score  TP  \\\n",
       "name                                                               \n",
       "fur-KO       NaN     NaN     NaN        NaN     NaN      NaN NaN   \n",
       "\n",
       "        regulon_size  n_regs                    Function             Category  \\\n",
       "name                                                                            \n",
       "fur-KO            13       0  Accounts for fur knock-out  Genomic Alterations   \n",
       "\n",
       "       single_gene  \n",
       "name                \n",
       "fur-KO        True  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.loc[['fur-KO']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene set enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.enrichment import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform a basic enrichment test between two samples, use the `compute_enrichment` function.\n",
    "Optional arguments:\n",
    "* `label`: Label for your target gene set (e.g. regulator name or GO term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set = ['b0002','b0003','b0004'] # e.g. an iModulon or other gene set\n",
    "target_genes = ['b0002','b0005','b0007'] # e.g. a regulon or genes in a COG category\n",
    "all_genes = ica_data.M.index # List of all genes in your expression dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pvalue             0.002293\n",
       "precision          0.333333\n",
       "recall             0.333333\n",
       "f1score            0.333333\n",
       "TP                 1.000000\n",
       "target_set_size    3.000000\n",
       "gene_set_size      3.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_enrichment(gene_set,target_genes,all_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform an enrichment test against a regulon in a TRN table, use the `compute_regulon` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sbrg/Desktop/pymodulon/pymodulon/enrichment.py:126: UserWarning: Some genes are in the regulon but not in all_genes. These genes are removed before enrichment analysis.\n",
      "  'analysis.', category=UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "pvalue             0.000131\n",
       "precision          0.015000\n",
       "recall             1.000000\n",
       "f1score            0.029557\n",
       "TP                 3.000000\n",
       "regulon_size     200.000000\n",
       "gene_set_size      3.000000\n",
       "n_regs             1.000000\n",
       "Name: lrp, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_regulon_enrichment(gene_set,'lrp',all_genes,trn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform an enrichment against all regulons in a TRN table, use the `compute_trn_enrichment` function. Optional arguments:\n",
    "* `max_regs`: Maximum number of regulators to include in complex regulon (default: 1) \n",
    "* `fdr`: False detection rate for Benjamini Hochberg correction\n",
    "* `method`: How to combine complex regulons. `'or'` computes enrichment against union of regulons, `'and'` computes enrichment against intersection of regulons, and `'both'` performs both tests (default: `'both'`) \n",
    "* `force`: Allows computation of >2 regulators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>gene_set_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>qvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thr-tRNA</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ile-tRNA</th>\n",
       "      <td>8.354256e-09</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pvalue  precision  recall  f1score   TP  regulon_size  \\\n",
       "thr-tRNA  0.000000e+00   1.000000     1.0      1.0  3.0           3.0   \n",
       "ile-tRNA  8.354256e-09   0.333333     1.0      0.5  3.0           9.0   \n",
       "\n",
       "          gene_set_size  n_regs    qvalue  \n",
       "thr-tRNA            3.0     1.0  0.000000  \n",
       "ile-tRNA            3.0     1.0  0.000002  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_trn_enrichment(gene_set,all_genes,trn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sbrg/Desktop/pymodulon/pymodulon/enrichment.py:153: RuntimeWarning: Using >2 maximum regulators may take time to compute. To perform analysis, use force=True\n",
      "  category=RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "compute_trn_enrichment(gene_set,all_genes,trn,max_regs=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>gene_set_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>qvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>arcA+ile-tRNA+rpoD</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arcA+lrp+thr-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arcA+rpoD+thr-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>crp+fnr+ile-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>crp+fnr+thr-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    pvalue  precision  recall  f1score   TP  regulon_size  \\\n",
       "arcA+ile-tRNA+rpoD     0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "arcA+lrp+thr-tRNA      0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "arcA+rpoD+thr-tRNA     0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "crp+fnr+ile-tRNA       0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "crp+fnr+thr-tRNA       0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "\n",
       "                    gene_set_size  n_regs  qvalue  \n",
       "arcA+ile-tRNA+rpoD            3.0     3.0     0.0  \n",
       "arcA+lrp+thr-tRNA             3.0     3.0     0.0  \n",
       "arcA+rpoD+thr-tRNA            3.0     3.0     0.0  \n",
       "crp+fnr+ile-tRNA              3.0     3.0     0.0  \n",
       "crp+fnr+thr-tRNA              3.0     3.0     0.0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_trn_enrichment(gene_set,all_genes,trn,max_regs=3,force=True).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the IcaData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pvalue           0\n",
       "precision        1\n",
       "recall           1\n",
       "f1score          1\n",
       "TP               9\n",
       "regulon_size     9\n",
       "imodulon_size    9\n",
       "n_regs           1\n",
       "Name: glpR, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compare a single iModulon to a simple regulon\n",
    "ica_data.compute_regulon_enrichment('GlpR',regulator='glpR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pvalue           5.471346e-20\n",
       "precision        4.210526e-01\n",
       "recall           1.000000e+00\n",
       "f1score          5.925926e-01\n",
       "TP               8.000000e+00\n",
       "regulon_size     1.900000e+01\n",
       "imodulon_size    8.000000e+00\n",
       "n_regs           2.000000e+00\n",
       "Name: cusR/cueR, dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compare a single iModulon to a complex regulon\n",
    "ica_data.compute_regulon_enrichment('Copper','cusR/cueR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>glpR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>crp</td>\n",
       "      <td>3.567813e-08</td>\n",
       "      <td>8.420040e-06</td>\n",
       "      <td>0.015332</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.030201</td>\n",
       "      <td>9.0</td>\n",
       "      <td>587.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>efeU-repair</td>\n",
       "      <td>fur</td>\n",
       "      <td>2.409711e-09</td>\n",
       "      <td>1.137384e-06</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.076503</td>\n",
       "      <td>7.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BW25113</td>\n",
       "      <td>mhpR</td>\n",
       "      <td>3.603815e-09</td>\n",
       "      <td>1.701001e-06</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.347826</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DhaR/Mlc</td>\n",
       "      <td>mlc</td>\n",
       "      <td>1.497841e-11</td>\n",
       "      <td>7.069810e-09</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>FecI</td>\n",
       "      <td>fecI</td>\n",
       "      <td>3.262855e-14</td>\n",
       "      <td>1.540067e-11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>FecI</td>\n",
       "      <td>fur</td>\n",
       "      <td>5.508595e-12</td>\n",
       "      <td>1.300028e-09</td>\n",
       "      <td>0.051429</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.097297</td>\n",
       "      <td>9.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>FecI</td>\n",
       "      <td>hypT</td>\n",
       "      <td>4.163677e-11</td>\n",
       "      <td>6.550851e-09</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>FecI</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>2.352987e-08</td>\n",
       "      <td>2.776524e-06</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.196078</td>\n",
       "      <td>5.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>RbsR</td>\n",
       "      <td>rbsR</td>\n",
       "      <td>9.780236e-14</td>\n",
       "      <td>4.616271e-11</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0           GlpR      glpR  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "1           GlpR       crp  3.567813e-08  8.420040e-06   0.015332  1.000000   \n",
       "2    efeU-repair       fur  2.409711e-09  1.137384e-06   0.040000  0.875000   \n",
       "3        BW25113      mhpR  3.603815e-09  1.701001e-06   0.666667  0.235294   \n",
       "4       DhaR/Mlc       mlc  1.497841e-11  7.069810e-09   0.500000  0.454545   \n",
       "..           ...       ...           ...           ...        ...       ...   \n",
       "145         FecI      fecI  3.262855e-14  1.540067e-11   1.000000  0.500000   \n",
       "146         FecI       fur  5.508595e-12  1.300028e-09   0.051429  0.900000   \n",
       "147         FecI      hypT  4.163677e-11  6.550851e-09   0.384615  0.500000   \n",
       "148         FecI      pdhR  2.352987e-08  2.776524e-06   0.121951  0.500000   \n",
       "149         RbsR      rbsR  9.780236e-14  4.616271e-11   0.555556  0.833333   \n",
       "\n",
       "      f1score   TP  regulon_size  imodulon_size  n_regs  \n",
       "0    1.000000  9.0           9.0            9.0     1.0  \n",
       "1    0.030201  9.0         587.0            9.0     1.0  \n",
       "2    0.076503  7.0         175.0            8.0     1.0  \n",
       "3    0.347826  4.0           6.0           17.0     1.0  \n",
       "4    0.476190  5.0          10.0           11.0     1.0  \n",
       "..        ...  ...           ...            ...     ...  \n",
       "145  0.666667  5.0           5.0           10.0     1.0  \n",
       "146  0.097297  9.0         175.0           10.0     1.0  \n",
       "147  0.434783  5.0          13.0           10.0     1.0  \n",
       "148  0.196078  5.0          41.0           10.0     1.0  \n",
       "149  0.666667  5.0           9.0            6.0     1.0  \n",
       "\n",
       "[150 rows x 11 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all significant enrichments for all iModulons. \n",
    "# Optional parameters are same as the global compute_trn_enrichment function\n",
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save to sample table\n",
    "You can directly save iModulon enrichments to the sample table using either the `ica_data.compute_regulon_enrichment` function or `ica_data.compute_trn_enrichment` function. Columns unrelated to regulon enrichments (e.g. `Function` and `Category`) are left alone."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator  pvalue  qvalue  precision  recall  f1score   TP  regulon_size  \\\n",
       "name                                                                            \n",
       "GlpR      glpR     0.0     NaN        1.0     1.0      4.0  9.0             9   \n",
       "\n",
       "      n_regs             Function                   Category single_gene  \n",
       "name                                                                      \n",
       "GlpR       1  Glycerol catabolism  Carbon Source Utilization         NaN  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.loc[['GlpR']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>crp</td>\n",
       "      <td>3.567813e-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015332</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.030201</td>\n",
       "      <td>9.0</td>\n",
       "      <td>587.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator        pvalue  qvalue  precision  recall   f1score   TP  \\\n",
       "name                                                                     \n",
       "GlpR       crp  3.567813e-08     NaN   0.015332     1.0  0.030201  9.0   \n",
       "\n",
       "      regulon_size  n_regs             Function                   Category  \\\n",
       "name                                                                         \n",
       "GlpR         587.0     1.0  Glycerol catabolism  Carbon Source Utilization   \n",
       "\n",
       "     single_gene  imodulon_size  \n",
       "name                             \n",
       "GlpR         NaN            9.0  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.compute_regulon_enrichment('GlpR',regulator='crp',save=True)\n",
    "ica_data.imodulon_table.loc[['GlpR']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>fur</td>\n",
       "      <td>2.409711e-09</td>\n",
       "      <td>1.137384e-06</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.076503</td>\n",
       "      <td>7.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FecI</th>\n",
       "      <td>fecI</td>\n",
       "      <td>3.262855e-14</td>\n",
       "      <td>1.540067e-11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ferric citrate transport</td>\n",
       "      <td>Metal Homeostasis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iron-related</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Related to iron metabolism</td>\n",
       "      <td>Biological Enrichment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RbsR</th>\n",
       "      <td>rbsR</td>\n",
       "      <td>9.780236e-14</td>\n",
       "      <td>4.616271e-11</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>D-ribose catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgrT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Contains single dominating gene: sgrT</td>\n",
       "      <td>Biological Enrichment</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insertion</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>IS2 insertion element after laboratory evolution</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             regulator        pvalue        qvalue  precision    recall  \\\n",
       "YieP               NaN           NaN           NaN        NaN       NaN   \n",
       "GlpR              glpR  0.000000e+00  0.000000e+00   1.000000  1.000000   \n",
       "YgbI               NaN           NaN           NaN        NaN       NaN   \n",
       "efeU-repair        fur  2.409711e-09  1.137384e-06   0.040000  0.875000   \n",
       "ydcI-KO            NaN           NaN           NaN        NaN       NaN   \n",
       "...                ...           ...           ...        ...       ...   \n",
       "FecI              fecI  3.262855e-14  1.540067e-11   1.000000  0.500000   \n",
       "iron-related       NaN           NaN           NaN        NaN       NaN   \n",
       "RbsR              rbsR  9.780236e-14  4.616271e-11   0.555556  0.833333   \n",
       "sgrT               NaN           NaN           NaN        NaN       NaN   \n",
       "insertion          NaN           NaN           NaN        NaN       NaN   \n",
       "\n",
       "               f1score   TP  regulon_size  imodulon_size  n_regs  \\\n",
       "YieP               NaN  NaN          11.0            NaN     1.0   \n",
       "GlpR          1.000000  9.0           9.0            9.0     1.0   \n",
       "YgbI               NaN  NaN           9.0            NaN     1.0   \n",
       "efeU-repair   0.076503  7.0         175.0            8.0     1.0   \n",
       "ydcI-KO            NaN  NaN           3.0            NaN     0.0   \n",
       "...                ...  ...           ...            ...     ...   \n",
       "FecI          0.666667  5.0           5.0           10.0     1.0   \n",
       "iron-related       NaN  NaN          20.0            NaN     0.0   \n",
       "RbsR          0.666667  5.0           9.0            6.0     1.0   \n",
       "sgrT               NaN  NaN           4.0            NaN     0.0   \n",
       "insertion          NaN  NaN           9.0            NaN     0.0   \n",
       "\n",
       "                                                       Function  \\\n",
       "YieP                                                    Unknown   \n",
       "GlpR                                        Glycerol catabolism   \n",
       "YgbI                                           Unknown Function   \n",
       "efeU-repair   Accounts for repair and expression of efeU operon   \n",
       "ydcI-KO                             Accounts for ydcI knock-out   \n",
       "...                                                         ...   \n",
       "FecI                                   Ferric citrate transport   \n",
       "iron-related                         Related to iron metabolism   \n",
       "RbsR                                        D-ribose catabolism   \n",
       "sgrT                      Contains single dominating gene: sgrT   \n",
       "insertion      IS2 insertion element after laboratory evolution   \n",
       "\n",
       "                               Category single_gene  \n",
       "YieP                Regulator Discovery         NaN  \n",
       "GlpR          Carbon Source Utilization         NaN  \n",
       "YgbI                Regulator Discovery         NaN  \n",
       "efeU-repair         Genomic Alterations         NaN  \n",
       "ydcI-KO             Genomic Alterations        True  \n",
       "...                                 ...         ...  \n",
       "FecI                  Metal Homeostasis         NaN  \n",
       "iron-related      Biological Enrichment         NaN  \n",
       "RbsR          Carbon Source Utilization         NaN  \n",
       "sgrT              Biological Enrichment        True  \n",
       "insertion           Genomic Alterations         NaN  \n",
       "\n",
       "[92 rows x 13 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The enrichment with the lowest qvalue is saved as the official enrichment\n",
    "ica_data.compute_trn_enrichment(save=True)\n",
    "ica_data.imodulon_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iModulon Thresholds\n",
    "iModulon thresholds are computed using an outlier test (See Methods section \"*Determination of the gene coefficient threshold*\" in [Sastry et al 2019](https://www.nature.com/articles/s41467-019-13483-w#Sec11)). This outlier test relies on the D'agostino test statistic.\n",
    "* If you have a relatively complete TRN, you may fit this test statistic using the `reoptimize_thresholds` function.\n",
    "* To manually adjust this statistic, use the `recompute_thresholds` function.\n",
    "* To change a single iModulon threshold, use the `change_threshold` function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fitting the threshold\n",
    "The `reoptimize_threshold` function may take a few minutes to complete. Fortunately, this only needs to be performed once. You can turn off the progress bar or plot using the `progress` and `plot` arguments, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f7175be88f0944cc971e0c01e8ab508f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=34.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASEAAAEKCAYAAABdduPuAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nO3deXxU1fn48c+TPWRhywKEJSEgEBBQAyIorrjvS10QxaVoXbtYtbXf2ta239ra1lq1fvm6K5UqLvVn/Sq44C6yCAiEQFiEsCQBJAsh+/P7497oEJIwWSZ3MvO8X695ZebOufc+dyZ5cu4595wrqooxxnglwusAjDHhzZKQMcZTloSMMZ6yJGSM8ZQlIWOMp6K8DqAzpaSkaGZmptdhGBN2li5duktVU9uzbkgloczMTJYsWeJ1GMaEHRH5ur3r2umYMcZTloSMMZ6yJGSM8ZQlIWOMpywJGWM8ZUnIGOMpS0LGGE9ZEjLGeMqSkNfmzIHMTIiIcH7OmeN1RMZ0qZC6YrrbmTMHZs2Cykrn9ddfO68Bpk/3Li5jupDVhDxUfu99vNcvh79NvoySHr2chZWVcM893gZmTBeymlAX2l9Tz9Kvv+HTDbv4bONuVl74R+ojIgFYlzKYR17/o1NwyxYPozSma1kSCrC6+gY+Wr+LeUsLWZBXRE1dA1ERwrhBvfhB3nwmr/yIT4eM5eHJlzF9+VtM3rISBg9u1772VddRUV1HenJcJx+FMYFjSShACooreGnpVl5dto3i8mr6JMRwxcTBHD8ilQmZfUiMjYLkTfDeUxy5fS3/zjmBe6fdwJv/upvo3/2uTftSVV5bvo3fvpHH7n01DOwdz8SsPhyd1YeJWX3J7NsDEQnQkRrTMRJKd9vIzc1Vr6fy+KqwlF++voovt+wlMkI4cUQqFx81iJNGphET1UwT3Jw5cM89vBOdzvUX/ZJf9N/P9bdf7Pf+Nu3axy9e+4pPCnYzflAvzjq8P8u2fMMXm/awe18NAKlJsZw9tj//dVYOERGWjEznE5GlqprbnnWtJtSJ1heVM+PJRcRFRXLPmaM4/4gMUpNiW19p+nSYPp2TVTnx6cU8uPkbzi2vIi2p9VOqmroGZn+4gYfeKyA2MoL7zh/DFRMHE+kmGVVlQ8k+vti0hw/WFfPUJ5sZ0qcHM6dkddbhGtMprCbUSbbuqeSSxz6jXpWXb5zM4L492ryNTbv2cdpfP+Tscf35y/fGt1huyeY93P3KVxQUV3DW2P7ce3YOaa20A6kq1zy9mM837ubN245jaGpim2MzpjUdqQlZF30nKCmvZsYTi6isqeO56ya2KwEBZKUk8P2pWbyybBtLNu856P2q2nrue2MNl/zPZ+yvqeepmRN45IojW01AACLC/ReNJTYqkp+8tIL6htD5x2O6P0tCHVS6v5arn/yCorJqnrpmAiP7JXdoezefOIz+PeP45b9XH5AsvtzyDWc+9BFPfLyJK48ewvwfTeXEkWl+bzc9OY7fnDeaL7fsZfaHGzsUozGdyZJQB+yvqef6Zxazvricx2YcxVFD+nR4mz1iovjFWTms2VHGP7/YQnVdPfe/tZaL/vEp1bUNzLn+aO47fwwJsW1vzjt33ADOPLwff12wjrU7yzocqzGdwdqE2qm2voEbnlvK+/nFPHTZEZwzbkCnbVtVmf74IlZvL6Nfchz5ReVcNmEQ95w1iqS46A5te3dFNac9+CHpyXG8etOU5nvsjGkjaxPywN/fXc97a4v57fljOjUBgdOG8+tzR7Ovuo69+2t46poJ/OGisR1OQAB9E2P53QWHs3p7GQ+/t74TojWmY6yLvh2ciwO3c/xhqUw/ekhA9jE8PYm3fjiVtORYkjsh+fg6bXQ/Ljwyg0cWbuDkUemMG9SrU7dvTFtYTagd8ovK2bKnktNG9wvofoalJXZ6Amp07zmjSUuK5ccvLqeiui4g+zDGH5aE2mHB6iJE4JQc/3ungk3P+GgeuGQcm3dXMuvZJVTX1XsdkglTloTaYf6aIsYP6nXIq5qD3ZRhKfzxorF8umE3t7+w3K4fMp6wJNRG2/fu56ttpZyaE9hTsa5y0VED+a+zc3hr9U7uefUrQqm31HQP1jDdRgvWFAFw6uh0jyPpPNcdm8U3+2p4+P0CeifEcNfpI70OyYQRS0JttGBNEUNTE8gOsfFXPzn1MPZU1vCPhRvo3SOaWVOzvQ7JhAlLQm1QWlnL5xt3c/1xQ70OpdOJCPedN4bSylp+/+ZaevWI4Xu5g7wOy4QBS0Jt8H5+MXUNGlKnYr4iI4S/XDqOsqpa7n55JQN7xTN5WIrXYZkQZw3TbbBgTRGpSbGMHxi6F/fFRkXy2JVHkZmSwI9eXM4ed2I0YwLFkpCfqmrrWZhfzCmj0kN+dsKE2CgeuuwIvtlXy53zVlqPmQkoS0J++mzDbvbV1IfsqVhTYzJ6cufpI3gnr4jnF9ndP0zgWBLy0/w1O0mIiWRydl+vQ+ky107J4vjDUvntG2vI31nudTgmRFkS8kNDg7JgTTEnjEgjNirS63C6TESE8MAl40iKi+K2F76kqtaGdpjOF9AkJCKni0i+iBSIyN2tlJsgIvUicrHPsh+JyGoRWSUiL4iIZ2Mkvty6l10V1WFzKuYrNSmWBy4ZR35ROf/9Zp7X4ZgQFLAkJCKRwCPAGUAOcLmI5LRQ7n7gbZ9lGcBtQK6qjgEigcsCFeuhzF+zk6gI4YQR3XfAakecMCKN647N4pnPvuYd94pxYzpLIGtCE4ECVd2oqjXAXOC8ZsrdCrwMFDdZHgXEi0gU0APYHsBYW7VgTRGThvalZ3xgptXoDu48fQSjByTz03krKCqr8jocE0ICmYQygK0+rwvdZd9yazwXAI/5LlfVbcADwBZgB1CqqvMDGGuLCoor2FiyLyxPxXzFRkXy0OVHUFFdxz8WbvA6HBNCApmEmruYpukFJw8Cd6nqAS2eItIbp9aUBQwAEkTkymZ3IjJLRJaIyJKSkpJOCPtA89fsBOCUUeGdhACyUxM5Z+wAXlqylbKqWq/DMSEikEmoEPAdfDSQg0+pcoG5IrIZuBh4VETOB04BNqlqiarWAq8Ak5vbiarOVtVcVc1NTU3t7GNgYX4JYzKSGdArvtO33R1dMyWLfTX1vLh466ELG+OHQCahxcBwEckSkRichuXXfQuoapaqZqpqJjAPuElVX8M5DZskIj1ERICTgS7vmqlvUFZtKyW3E27lEyoOH9iTCZm9efrTzTYJmukUAUtCqloH3ILT65UHvKiqq0XkRhG58RDrLsJJSsuAr9w4Zwcq1pYUFFdQWVPP2IE9u3rXQe3aKVkUfrOfd/Ksp8x0XEBH0avqm8CbTZY91kLZmU1e3wvcG7Dg/LCicC8AY0N4wGp7TMtJJ6NXPE9+vCngk/2b0GdXTLdiZeFekmKjGJqS4HUoQSUqMoKrJw9h0aY9rN5e6nU4ppuzJNSKlYWljMnoGfKj5tvj0tzB9IiJ5KlPNnsdiunmLAm1oLqunrwdZYwdZO1BzenZI5qLjhzI68u3s6ui2utwTDdmSagFa3eUU1uvjLP2oBbNnJJJTX0Dcz63qT5M+1kSasHKbxulrSbUkuzURE4Ykcrzi762myeadrMk1IIVhaX0TYghwy5SbNW1U7IoKa/mPyt3eB2K6aYsCbVgZeFexg3qhXOtpGnJccNTGJaWyJOfbDpgGtiK6jqWfr2HOYu+th400yq720YzKqrrWF9cwZmH9/c6lKAnIlwzJZN7Xl3FL/+9mqKyKtbuLGfLnspvy4wekMx/bjvOwyhNMLOaUDNWbStFFWuU9tOFRwwkJTGWOYu+ZkNJBYcP7Mkdpx7G41flcuPx2azeXsa2vfu9DtMEKasJNcMapdsmPiaShT89gUgR4mMOnP42KzWBxz7YwDtrirh6cqY3AZqgZjWhZqwoLCWjVzx9E2O9DqXbSIyNOigBgdODNjQ1gQU2I6NpgSWhZjiN0lYL6izTctL5fONuSvfbHETmYJaEmtizr4ate/bboNVOdGpOOnUNysL8pjP4GmNJ6CDWHtT5xg/qTUpijJ2SmWZZEmpiZWEpInB4hiWhzhIZIZw8Mp0P8kuoqWvwOhwTZCwJNbGycC9DUxJIigvfO2sEwrScdMqr6/h8426vQzFBxpKQD1VlRWEp4wZZe1BnmzIshbjoCDslMwexJORjZ1kVJeXVdpFiAMTHRHLc8FTeySs6YHiHMZaEfKzY6oxxskbpwJiWk86O0ipWbSvzOhQTRCwJ+VhRuJeoCGFU/2SvQwlJJ49MI0JggXsvN2PAktABVhbuZWT/JOKiD77y13Rc38RYjhrSmwV5dr2Q+Y4lIVdDg7KysNQuUgywaTnp5O0oY6vPKHsT3iwJuTbv3kd5VR3jrD0ooKblOLcIsnuWmUaWhFwrCxsbpa0mFEhZKQkMS0u0rnrzLUtCrhWFe4mLjmB4WqLXoYS8aTnpLNq0h9JKG9BqLAl9a2VhKWMG9CQq0j6SQJuWk059g/K+DWg1WBICnEbpvB1ljLHxYl1i/MBepCbF8uZXNjm+aUMSEpGQvRfytr37qaypZ2S/JK9DCQsREcKFR2Ywf00Rv/z3KurqbVBrODtkEhKRySKyBshzX48TkUcDHlkXWruzHIDDLAl1mTtPG8msqUN59rOvue6ZJZRXWftQuPKnJvRX4DRgN4CqrgCmBjKorrauyE1C6ZaEukpkhPDzM0fx+wsO5+OCXVz8j88o/MauHQpHfp2OqerWJotC6nab+TvLGdg7nsRYm/e/q11x9GCeuWYi20v3c/4jn/Lllm+8Dsl0MX+S0FYRmQyoiMSIyB24p2ahYl1ROSOsFuSZY4en8OpNk4mPieCy2Z8zb2mhTX4WRvxJQjcCNwMZQCEw3n0dEmrrG9hQUmHtQR4blpbEazdN4fCMntzx0gqOum8BN89ZxivLCvlmX43X4ZkAavX8Q0QigRmqOr2L4ulym3bto7ZerSYUBPomxvLCrEkszC/h3bwi3l1bzH++2kGEwFFDenPyqHSOG57CqH7JRETY7blDRatJSFXrReQ8nMbpkJS/0xqlg0l0ZATTctKZlpNOQ4Py1bZS3s0r4p28Yv7wf2v5w/9B34QYJg9L4bhhKRw7PIUBveK/Xb+hQdlXU0dZVR1VtfUMTUlAxBJWMPOnJfYTEXkY+Bewr3Ghqi471IoicjrwNyASeFxV/9BCuQnA58ClqjrPXdYLeBwYAyhwrap+5ke8bbKuqJzICCE7LWQvg+q2IiKEcYN6MW5QL3586giKyqr4eP0uPinYxUcFu/h/K7YDMLC3k4TKq+oor6qlwWfixnPGDeDBS8cTaTWnoOVPEprs/vyNzzIFTmptJfdU7hFgGk5b0mIReV1V1zRT7n7g7Sab+BvwlqpeLCIxQA8/Ym2z/J3lZKUkEBtlcwgFu/TkOC46aiAXHTUQVWVdUQUfF+xi2ZZviImMIDkuiuT4aJLjokmOj2JDyT5mf7iRXvHR/Oa80VYjClKHTEKqemI7tz0RKFDVjQAiMhc4D1jTpNytwMvAhMYFIpKMcy3STDeGGiAgrZP5ReWMGWDDNbobEWFEvyRG9EviOrJaLgf8z4cb6d0jmh+fOqLrAjR+8+eK6Z4i8hcRWeI+/iwi/vzVZgC+1xcVust8t50BXAA81mTdoUAJ8JSIfCkij7c0bEREZjXGVlJS4kdY36msqWPLnkprDwphd58xkktzB/HQewU8+fEmr8MxzfCni/5JoBz4nvsoA57yY73m6r5Nb7PwIHCXqja9+DEKOBL4h6oegdMWdXdzO1HV2aqaq6q5qampfoT1nYLiClRhRD+bviNUiQi/u2AMp41O5zdvrOHVLwu9Dsk04U+bULaqXuTz+tcistyP9QqBQT6vBwLbm5TJBea65+opwJkiUofTSF2oqovccvNoIQl1RGPP2Ih+NrF9KIuKjOBvlx3BtU8v5o6XVpIcF83Jo9K9Dsu4/KkJ7ReRYxtfiMgUYL8f6y0GhotIltuwfBnwum8BVc1S1UxVzcRJNDep6muquhPnSu3Gk/iTObgtqcPWFZUTGxXB4D4BafM2QSQuOpLZV+UyekAyN81ZxqcFu7wOybj8SUI/AB4Rkc0ishl4GOcq6lapah1wC06vVx7woqquFpEbReSQ6+M0WM8RkZU4V2n/3o912mTtznKGpyda922YSIyN4ulrJjK4Tw+uevIL5iz62uuQDP71ji0Hxrk9Vqiq33euU9U3gTebLGvaCN24fGYz+831d1/tsa6onCnDUgK5CxNk+iTEMO8Hk7nthS+559VV5O0o495zRhPdwoyaq7eX8vs38yjdX8vz1x1Nrx4xXRxx6POnd+z3ItJLVctUtUxEeovIb7siuEDaW1lDUVm1DdcIQz3jo3ly5gRumDqU5z/fwvTHF7G7ovqAMsVlVdw5bwVn//1j1mwvY93OCmY9t5TqupCaQCIo+HM6doaq7m18oarfAGcGLqSusa6oAoARNnA1LEVGCD87cxQPXjqeFVv3cu7Dn7B6eyn7a+p56N31nPDAQl79chvfP24oC396In+6ZCxfbNrDnfNW0tDQtJPXdIQ/vWORIhKrqtUAIhIPxAY2rMDLL2rsGbMkFM7OPyKDoakJzHp2KRf/4zN69YhmR2kVZ4zpx91njGRIX+fytPPGZ1D4zX7+9HY+g3r34I7T7MLHzuJPEnoeeFdEnsIdwwU8E9CoukD+zjKS4qLolxzndSjGY2MH9uL1W6dw6z+/pKqugQcvHc/RQ/seVO6mE7LZuqeSh98vYFCfeC6dMNiDaEOPPw3Tf3R7qE7BuQDxPlVtOs6r21m3s4IR6Uk2nsgAkJYUx79uOKbVMiLCfeePYXtpFT9/dRX9e8Yz9bC2XSBrDuZPw3QCMF9V7wBmA7EiEh3wyAJIVckvKreJzEybRUdG8MgVRzA8LZGb5iwjb4ffncWmBf40TH8IxLnjvN4BrgGeDmRQgVZcXk3p/lq7xY9pl6S4aJ66ZgKJsVFc89RiNpZUeB1St+ZPEhJVrQQuBP6uqhcAOYENK7BsIjPTUf17xvP0tROorW/gksc+Y9W2Uq9D6rb8SkIicgwwHfiPu6xb35bCkpDpDCP7JfPSjccQGxXB5bM/Z/HmPV6H1C35k4RuB34GvOoOuxgKvB/YsAIrv6ic1KRY+iTY1a+mY4amJjLvB5NJTY5lxhOLeD+/2OuQup1DJiFV/VBVz1XV+93XG1X1tsCHFjh2ix/TmQb0iufFG44hOzWR7z+z5NtpZ41//L4XfahoaFAnCVmjtOlEKe6dQo4Y3Ivb5n7JPxdt8TqkbiPsktDWbyqpqm2wmpDpdMlx0Tx77dEcf1gqP3/1K65/ZjEFxdZzdihhl4TWNjZKW03IBEB8TCSzZ+Ry5+kjWLRxD6c9+CG/eO0rdjUZIGu+064kJCK/7OxAuso6NwkNT7MpXU1gxERFcNMJw1j40xO48ujBvPDFVk7400Ieeb+Aqlobhd9Ue2tC13dqFF0ov6icQX3iSYjt1lcZmG6gb2Isvz5vDPN/NJXJ2X3509v5nPjAQjbv2nfolcNIi0lIRMpaeJQDA7owxk7l9IzZnNKm62SnJjL7qlzmzppESXk1cxdvPfRKYaS1mtBeYLiqJjd5JAE7uii+TlVT18DGkn12dw3jiUlD+zJpaF/mr9npdShBpbUk9CwwpIX3/hmAWAKurqGBO08fwUkj7U4Lxhunjk5nY8k+6zXz0WISUtVfqOoXLbx3V+BCCpweMVHMmprNUUN6ex2KCVOnuLcaWrCmyONIgkdrbUK3+Dwf3TXhGBPaBvSK5/CMnnZK5qO107FrfZ4/F+hAjAkX03LSWb51L8VlVV6HEhT87aK36QeN6SSnjk5HFd7Js8Gu0PqUHL1E5AKcRJUsIhf6vqmqrwQ0MmNC1Ij0JAb36cGCNTu54mibp7q1JPQBcK77/EPgHJ/3FLAkZEw7iAjTctJ57rOvqaiuIzHML5xt8ehV9ZquDMSYcHJqTjpPfLyJD/JLOGtsf6/D8VTYDWA1JhgcNaQ3vXtEs8B6ySwJGeOFqMgITh6Vzntri6mtb/A6HE9ZEjLGI9Ny0imrquOLTeE9N7VfLWIiMhnI9C2vqs8GKCZjwsLU4anERUcwf/VOpgxL8Tocz/hz88PngAeAY4EJ7iM3wHEZE/LiYyI5bngqC9YUoapeh+MZf2pCuUCOhvOnZEyATMtJZ8GaIlZvL2NMRk+vw/GEP21Cq4B+gQ7EmHB08sg0IgTmrw7fXjJ/klAKsEZE3haR1xsfgQ7MmHDQNzGW3CF9mB/Go+r9OR37VaCDMCacnTo6nd/+J4+teyoZ1KeH1+F0OX9ufvhBcw9/Ni4ip4tIvogUiMjdrZSbICL1InJxk+WRIvKliLzhz/6M6Y6m5ThzDD307nrqG8Kv6dWf3rFJIrJYRCpEpMZNFmV+rBcJPAKcAeQAl4tITgvl7gfebmYztwN5h9qXMd3ZkL4J3HD8UF5aWsiNzy+lsqbO65C6lD9tQg8DlwPrgXicO2087Md6E4EC97bRNcBc4Lxmyt0KvAwcMK+BiAwEzgIe92NfxnRrPztjFL86J4d384q4fPbnlJSHz33K/LpiWlULgEhVrVfVp4AT/FgtA/C9rUChu+xbIpIBXAA81sz6DwJ3Aq1e0y4is0RkiYgsKSkp8SMsY4LTzClZ/M+MXNYVVXDBo59QUFzudUhdwp8kVCkiMcByEfmjiPwISPBjveYmQmt6wvsgcJeqHnBHOBE5GyhW1aWH2omqzlbVXFXNTU1N9SMsY4LXtJx0/nXDJKpqG7jw0U/5bMNur0MKOH+S0Ay33C3APmAQcJEf6xW6ZRsNBLY3KZMLzBWRzcDFwKMicj4wBTjXXT4XOElEnvdjn8Z0e2MH9uLVmyaTlhzHVU8u4uH31lNcHrpTwYo/F0KLSDwwWFXz/d6wSBSwDjgZ2AYsBq5Q1dUtlH8aeENV5zVZfgJwh6qefah95ubm6pIlS/wN0ZigVlpZy49fXM67a4uJjBBOHJHG93IHcuLINKIjg2vsuYgsVdV2Dec65HVCInIOztixGCBLRMYDv1HVc1tbT1Xr3Dt2vA1EAk+q6moRudF9v7l2IGOMq2ePaJ6YOYGNJRW8uKSQl5cV8k5eESmJsVx0ZAZXTc4ko1e812F22CFrQiKyFDgJWKiqR7jLVqrq2C6Ir02sJmRCWV19AwvzS3hxyVbeW1vM4D49eOfHxxMR4f19KDpSE/KnTlenqqXt2bgxpvNERUZwSk46s6/K5S+Xjmfjrn0sXNf979jh1wBWEbkCiBSR4SLyd+DTAMdljGnFGWP60S85jqc+2ex1KB3mTxK6FRgNVAMvAGXADwMZlDGmddGREcw4Zggfrd/F+qLufT2RP2PHKlX1HlWd4F6Pc4+qhm5/oTHdxOUTBxMTFcHTn272OpQOabF37FDTdRyqd8wYE1h9EmI4f/wAXlm2jTtPG0nPHtFeh9QurXXRH4Mz7OIFYBF2K2hjgs7MyVm8uKSQfy3Zwqyp2V6H0y6tnY71A34OjAH+BkwDdrVlKg9jTGDlDEjm6Kw+PPPp19R101sHtZiE3MGqb6nq1cAkoABYKCK3dll0xphDumZKFtv27uedvO45O2OrDdMiEisiFwLPAzcDD2H3oDcmqEzLSSejV3y37a5vMQmJyDM41wMdCfza7R27T1W3dVl0xphDiowQrp48hEWb9rB6e/e7rri1mtAM4DCc2Q0/FZEy91Huz8yKxpiuc2nuYOKjI3m6G9aGWmsTilDVJPeR7PNIUtXkrgzSGNO6nj2iufDIDP69Yju7K7rXrIzBNR+AMabdZk7OpKaugRe+2OJ1KG1iSciYEDE8PYnjhqfw7GdfU11Xf+gVgoQlIWNCyKypQykur+blpd2n/8iSkDEh5NhhKYwd2JPHPtjQbS5etCRkTAgREW4+cRhb9lTyxsodXofjF0tCxoSYaaPSOSw9kUfeL6ChG9zR1ZKQMSEmIsKpDa0vrmD+muAfymFJyJgQdNbh/RnStwePvF+AP3fU8ZIlIWNCUFRkBD84PpuvtpXy0fpdXofTKktCxoSoC48cSP+ecTz8foHXobTKkpAxISomKoLvHzeULzbtYfHmPV6H0yJLQsaEsMsnDqZvQgwPvxe8tSFLQsaEsPiYSK49NosP1pXwVWFwTvNhSciYEDfjmCEkxUXxSJC2DVkSMibEJcdFM2PSEN5avTMop/mwJGRMGJg0tC8A64srPI7kYJaEjAkD2WmJAGwosSRkjPFA/+Q44qMj2VC8z+tQDmJJyJgwEBEhDE1NsJqQMcY72amJloSMMd7JTk1k29797K8JrqlfLQkZEyaGpSWiCht3BVdtKKBJSEROF5F8ESkQkbtbKTdBROpF5GL39SAReV9E8kRktYjcHsg4jQkH2WkJAGwoCa7G6YAlIRGJBB4BzgBygMtFJKeFcvcDb/ssrgN+oqqjgEnAzc2ta4zxX2bfBERgQ5BdKxTImtBEoEBVN6pqDTAXOK+ZcrcCLwPFjQtUdYeqLnOflwN5QEYAYzUm5MVFRzKod4+ga5wOZBLKALb6vC6kSSIRkQzgAuCxljYiIpnAEcCiTo/QmDCTnZoQPqdjgDSzrOk8kw8Cd6lqs831IpKIU0v6oaqWtVBmlogsEZElJSUlHQrYmFCXnZrIxpKKoJoAP5BJqBAY5PN6ILC9SZlcYK6IbAYuBh4VkfMBRCQaJwHNUdVXWtqJqs5W1VxVzU1NTe3M+I0JOdlpiVTXNbBt736vQ/lWIJPQYmC4iGSJSAxwGfC6bwFVzVLVTFXNBOYBN6nqayIiwBNAnqr+JYAxGhNWhrljyAqCqF0oYElIVeuAW3B6vfKAF1V1tYjcKCI3HmL1KcAM4CQRWe4+zgxUrMaEi+xUdyBrEPWQRQVy46r6JvBmk2XNNkKr6kyf5x/TfJuSMaYD+iTE0LtHdFA1TtsV08aEmWAbQ2ZJyJgw09hDFiwsCRkTZrLTEthVUcPeyhqvQwEsCRkTdr5tnA6SdiFLQsaEmcZu+mDpIbMkZEyYGdi7BzGREUHTONqAJ0YAAA0fSURBVG1JyJgwExkhZKUEz1SvloSMCUPZacEzkNWSkDFhKDs1kS17Kqmu836qV0tCxoSh7NRE6huULbsrvQ7FkpAx4ei7bnrv24UsCRkThoamOvNNFwRBN70lIWPCUEJsFAN6xgVF47QlIWPCVHZacAxktSRkTJjKTk1kQ3EFqt5O9WpJyJgwlZ2awL6aeorKqj2Nw5KQMWEqWHrILAkZE6a+HchqScgY44XUpFiSYqM876a3JGRMmBIRhgZBD5klIWPCWHZqAhuKvb1WyJKQMWEsOzWRnWVVVFTXeRaDJSFjwlhj4/TC/GLPYrAkZEwYO/6wVMYO7Mmd81ayenupJzFYEjImjMVFR/L4Vbn0jI/muqeXsLO0qstjsCRkTJhLS47jyZkTKK+q5bpnFrOvi9uHLAkZYxjVP5mHrziSvB1l3D53OfUNXTeezJKQMQaAE0emce85o3knr4j/fjOvy/Yb1WV7MsYEvasnZ7Jp1z4e/3gTmSkJXDlpSMD3aUnIGHOA/zo7hy17Krn39dUM6tOD4w9LDej+7HTMGHOAyAjhocuP4LjhKfTuER3w/VlNyBhzkMTYKJ6+ZmKX7MtqQsYYT1kSMsZ4ypKQMcZTAU1CInK6iOSLSIGI3N1KuQkiUi8iF7d1XWNM9xawJCQikcAjwBlADnC5iOS0UO5+4O22rmuM6f4CWROaCBSo6kZVrQHmAuc1U+5W4GWguB3rGmO6uUAmoQxgq8/rQnfZt0QkA7gAeKyt6/psY5aILBGRJSUlJR0O2hjTtQKZhKSZZU1HxT0I3KWq9e1Y11moOltVc1U1NzU1sFd2GmM6XyAvViwEBvm8Hghsb1ImF5grIgApwJkiUufnugdZunTpLhH5uiNBd5IUYJfXQXggHI87HI8ZDj7udg8yk0DdAlZEooB1wMnANmAxcIWqrm6h/NPAG6o6r63rBhsRWaKquV7H0dXC8bjD8Zihc487YDUhVa0TkVtwer0igSdVdbWI3Oi+37Qd6JDrBipWY4x3AlYTCmf23zF8hOMxQ+cet10xHRizvQ7AI+F43OF4zNCJx201IWOMp6wmZIzxlCUhY4ynLAm1k4hsFpGvRGS5iCxxl/URkQUist792dun/M/cwbj5InKad5H7T0SeFJFiEVnls6zNxygiR7mfVYGIPCTuhWHBqoXj/pWIbHO/7+UicqbPe93+uEVkkIi8LyJ5IrJaRG53lwf++1ZVe7TjAWwGUpos+yNwt/v8buB+93kOsAKIBbKADUCk18fgxzFOBY4EVnXkGIEvgGNwroT/P+AMr4+tHcf9K+COZsqGxHED/YEj3edJONfp5XTF9201oc51HvCM+/wZ4Hyf5XNVtVpVNwEFOIN0g5qqfgjsabK4TccoIv2BZFX9TJ3f0Gd91glKLRx3S0LiuFV1h6ouc5+XA3k44zUD/n1bEmo/BeaLyFIRmeUuS1fVHeB8qUCau9zvAbndQFuPMcN93nR5d3SLiKx0T9caT0tC7rhFJBM4AlhEF3zfloTab4qqHokz59HNIjK1lbJ+D8jtxlo6xlA59n8A2cB4YAfwZ3d5SB23iCTiTK3zQ1Uta61oM8vaddyWhNpJVbe7P4uBV3FOr4rc6ijuz8Y5kto1IDdItfUYC93nTZd3K6papKr1qtoA/C/fnU6HzHGLSDROApqjqq+4iwP+fVsSagcRSRCRpMbnwKnAKuB14Gq32NXAv93nrwOXiUisiGQBw3Ea77qjNh2jW4UvF5FJbi/JVT7rdBuNf4iuC3C+bwiR43ZjfALIU9W/+LwV+O/b61b57vgAhuL0DKwAVgP3uMv7Au8C692ffXzWuQenByGfIO4laXKcL+CcetTi/Ie7rj3HiDNlyyr3vYdxr9QP1kcLx/0c8BWw0v0D7B9Kxw0ci3PatBJY7j7O7Irv24ZtGGM8ZadjxhhPWRIyxnjKkpAxxlOWhIwxnrIkZIzxlCUhcwARUfG5HbfpXCIS537GZ3dgGyPdbYzpzNi8YkmonUTkafcXQUWk1p364X0Rudm98tS37EIRmelRqM1y43+jmbf6A/8vgPv9lc/n1tIjsxP2M1dE5vlRLklE/igiG0SkSkRKROQj30QsIp+LyAOdEYOqVuF8xgv83EZz+17vbmNtW2MKRoG871g4eAeYgXNHkFTgJODXwAwROVlV93kZXHuo6s4A7+IBDrzj7kLgDXd5o668le4TwDic25Gvwbk4b5L7MyA6+hmrc7PQQH9PXcfrKzW76wN4Guc+aU2XjwFqgF/7LFsIzPR5/WOcK1P34dxX7XGgV5PtXAtsASpxaiY3OV/XAWVuwJlCocb9+f1m3l8HVOH8Yb+N84/nVzhXx/o+TnDXUeBi93mm+/oinP/clTh/qNOa7GcqzojrKqAI+CsQ4+fnuAr4VQvvzcL5b1/l/rwFn6tvcRJHAVDtHt+b7vI/NHN8k5rZvrjHdGUr8c1tZlv9cObReQpnXqn97uf8o8b4WooBiHOfn+0Tw33ud12NM87q8UPse6T7fEyT37v/AGVAOfAJMNLrvxO/fge8DqC7PmghCbnvvc6BE2It5MAk9EOcWlMmcDxOQnrO5/1jgAbgLuAw4PvuH5n6lLkAZ1jBLW6ZW93X57jv5wJ1wHScu2OOc/9IooBE4F84iaWf+4hx12suCa0FzsEZH/QMsBtIdMtk4CTTx4BRwNk4/6X/7Ofn2GwSco+n0D3OLJw5aUqA6933p+Ak30vd4xsP/MR9LxF4DSd5Nx5fdAv73wz8E0hq4f2ewFLgUZ9tRQAJwL3u55wFXO7+8U9vLQYOTkLTceYuOg0YDEwAfnCIfR+QhNzj/waY58ZzGM44rzH+fAdePzwPoLs+aD0J/QGobMO2Tsf5Lxjhvn4BeKtJmdkcmIQ+wbkpZNOYPnafXwiUtvLH1Wz8NJ+EbvB5P8Nddqz7+nc4tZEInzIz3ePp4cexH5SEcGoHO4FLmiy/G1jmPr8C5zbEze4DpxYxz4/9n4xT+6gFlgAPASc2KfM58IAf23rQ9zNtLoZmktDPccakNTvTZnP7biYJ/dn9DqK8/rtoz8MapgNDaGUOFRE5yZ2vt1BEyoFXgBic/3Tg/JI1HWW/qMnrUTiJyNfHONNuglPL+RrYJCJzROTqxpH/7bDS53njtAyNk1uNAj5TZ4oL3zhigGHt3N9AIB14RkQqGh84p5HZbpk3caaV2Cwiz4nIle6MBm2iqu/i1CROwZnGYjTwnoj87VDrisit7qR2JW58P8CpzbTFXKAPzvf0vyJyUdOODT8cAXygqnVtXC8oWBIKjBxgY3NviMgQnHP3POAS4Cic9h9w/nDhEEnMR3NlnOqMM0XnkcD3cNobfgasFZEB/h3CAWq/3bj7r5fvfndai7W9o6Mbt30NzmlW42MMzh8cqroX5xRzOk5i/CWwRkTSDtraIahqrap+oKr/raon49TubmsyfccBRORq4H6cuYVOdeN7nO++Q3/3vREnWd+C0z71ELBIROLasJmgnUDfH5aEOpl77cbpOOfnzcnF+UX9kTrz8K4DmiaGPA6eg7rp6zyc6Rd8HYvTcAyAqtap6nuq+jNgLE47RuP1KTU4vXodtQY4RkR8f5eOdbe/oZ3b3IpzqjVUVQuaPL5N7m7yWKCqd+EkpFSczx46dnyNn2Fjzaq5bR0LfKSqj6nql6pagNNm5suvGFR1v6q+rqq3A5NxEm3j9+3PNpYBU0WkW/Z2d8ugg0isiDQ2FqbitC/8HKcxsaXrSta75X8oIq/g9Jj8sEmZh4CPReSnOI2bU3EaaH39CXhJRJYC83H++KbjtAXhXgyXDTRO2n4izl0U8tz1NwNniMgInIbmUlWtpe0edeN/1D2FGYrTJvawqla2Y3uoaoOI/Bq43z3NeRunN+ooIFVV/yQiF+Ik749xGmVPxWlv8T2+K0VkuPv+3uZOV0TkY5z2saVuuTHAb3DaaTb4bGuSW4vdh/N5rQMuEZFpOKe9M4CjceYhwme9A2JoZv/fx+lAWOxuewYHJvDm9t3U34HrgRdE5A84PWQTgRWquqqZ8sHF60ap7vrA+cVt7Datw/nPvRCnV6fV7mngNpyu+f04E0V9z91Opk+Za3FqBPtxelh+Auxvsp0bcRoka2nSRY/zn/p9nF/a/TgNwNf4vJ+Kk7zKOXQXfW6T/X5bxn3d2EVfzXdd9LF+fo6tddFfhTO5VjVOIv0QuMh970TgA3d5JU671ZU+6/Z3P9sKWuiid8v9EvjU/ZyqgE04PX0DfMrk4LTRVfJdN3kcTk/hXpwE8z/Ab4G1rcXAwQ3Tl7ifXalbbhFw+iH23VwX/TicZL0PJwl9BIzw+u/En4dNatZNiMhfgVNU9XCvYzGmM9npWJByT8UW4Px3PAWn1vNzT4MyJgCsJhSkRORfwAk4F6xtwqnu/03tCzMhxpKQMcZT1kVvjPGUJSFjjKcsCRljPGVJyBjjKUtCxhhP/X/leSODxeADogAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 288x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ica_data.reoptimize_thresholds()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Changing the D'agostino statistic\n",
    "To test out different D'agostino test statistics, use the `recompute_thresholds` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2239</th>\n",
       "      <td>0.211384</td>\n",
       "      <td>2349934</td>\n",
       "      <td>2351011</td>\n",
       "      <td>-</td>\n",
       "      <td>glpQ</td>\n",
       "      <td>1077</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2240</th>\n",
       "      <td>0.306134</td>\n",
       "      <td>2351015</td>\n",
       "      <td>2352374</td>\n",
       "      <td>-</td>\n",
       "      <td>glpT</td>\n",
       "      <td>1359</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2241</th>\n",
       "      <td>0.375662</td>\n",
       "      <td>2352646</td>\n",
       "      <td>2354275</td>\n",
       "      <td>+</td>\n",
       "      <td>glpA</td>\n",
       "      <td>1629</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2242</th>\n",
       "      <td>0.328961</td>\n",
       "      <td>2354264</td>\n",
       "      <td>2355524</td>\n",
       "      <td>+</td>\n",
       "      <td>glpB</td>\n",
       "      <td>1260</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2243</th>\n",
       "      <td>0.315752</td>\n",
       "      <td>2355520</td>\n",
       "      <td>2356711</td>\n",
       "      <td>+</td>\n",
       "      <td>glpC</td>\n",
       "      <td>1191</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3426</th>\n",
       "      <td>0.350034</td>\n",
       "      <td>3562012</td>\n",
       "      <td>3563518</td>\n",
       "      <td>+</td>\n",
       "      <td>glpD</td>\n",
       "      <td>1506</td>\n",
       "      <td>glpD</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,glpR,rpoD,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3925</th>\n",
       "      <td>0.084278</td>\n",
       "      <td>4114568</td>\n",
       "      <td>4115579</td>\n",
       "      <td>-</td>\n",
       "      <td>glpX</td>\n",
       "      <td>1011</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3926</th>\n",
       "      <td>0.290235</td>\n",
       "      <td>4115713</td>\n",
       "      <td>4117222</td>\n",
       "      <td>-</td>\n",
       "      <td>glpK</td>\n",
       "      <td>1509</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3927</th>\n",
       "      <td>0.312307</td>\n",
       "      <td>4117244</td>\n",
       "      <td>4118090</td>\n",
       "      <td>-</td>\n",
       "      <td>glpF</td>\n",
       "      <td>846</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length  operon  \\\n",
       "b2239     0.211384  2349934  2351011      -      glpQ    1077   glpTQ   \n",
       "b2240     0.306134  2351015  2352374      -      glpT    1359   glpTQ   \n",
       "b2241     0.375662  2352646  2354275      +      glpA    1629  glpABC   \n",
       "b2242     0.328961  2354264  2355524      +      glpB    1260  glpABC   \n",
       "b2243     0.315752  2355520  2356711      +      glpC    1191  glpABC   \n",
       "b3426     0.350034  3562012  3563518      +      glpD    1506    glpD   \n",
       "b3925     0.084278  4114568  4115579      -      glpX    1011  glpFKX   \n",
       "b3926     0.290235  4115713  4117222      -      glpK    1509  glpFKX   \n",
       "b3927     0.312307  4117244  4118090      -      glpF     846  glpFKX   \n",
       "\n",
       "                                         cog  \\\n",
       "b2239       Energy production and conversion   \n",
       "b2240  Carbohydrate transport and metabolism   \n",
       "b2241       Energy production and conversion   \n",
       "b2242    Amino acid transport and metabolism   \n",
       "b2243       Energy production and conversion   \n",
       "b3426       Energy production and conversion   \n",
       "b3925  Carbohydrate transport and metabolism   \n",
       "b3926       Energy production and conversion   \n",
       "b3927  Carbohydrate transport and metabolism   \n",
       "\n",
       "                                  regulator  \n",
       "b2239         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2240         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2241  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2242  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2243  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b3426               arcA,crp,glpR,rpoD,yieP  \n",
       "b3925                         crp,glpR,rpoD  \n",
       "b3926                         crp,glpR,rpoD  \n",
       "b3927                         crp,glpR,rpoD  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('GlpR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.recompute_thresholds(300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b1199</th>\n",
       "      <td>0.045686</td>\n",
       "      <td>1249124</td>\n",
       "      <td>1249757</td>\n",
       "      <td>-</td>\n",
       "      <td>dhaL</td>\n",
       "      <td>633</td>\n",
       "      <td>dhaKLM</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>arcA,dhaR,fnr,lrp,rpoS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1200</th>\n",
       "      <td>0.045491</td>\n",
       "      <td>1249767</td>\n",
       "      <td>1250838</td>\n",
       "      <td>-</td>\n",
       "      <td>dhaK</td>\n",
       "      <td>1071</td>\n",
       "      <td>dhaKLM</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>arcA,dhaR,fnr,lrp,rpoS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2239</th>\n",
       "      <td>0.211384</td>\n",
       "      <td>2349934</td>\n",
       "      <td>2351011</td>\n",
       "      <td>-</td>\n",
       "      <td>glpQ</td>\n",
       "      <td>1077</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2240</th>\n",
       "      <td>0.306134</td>\n",
       "      <td>2351015</td>\n",
       "      <td>2352374</td>\n",
       "      <td>-</td>\n",
       "      <td>glpT</td>\n",
       "      <td>1359</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2241</th>\n",
       "      <td>0.375662</td>\n",
       "      <td>2352646</td>\n",
       "      <td>2354275</td>\n",
       "      <td>+</td>\n",
       "      <td>glpA</td>\n",
       "      <td>1629</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2242</th>\n",
       "      <td>0.328961</td>\n",
       "      <td>2354264</td>\n",
       "      <td>2355524</td>\n",
       "      <td>+</td>\n",
       "      <td>glpB</td>\n",
       "      <td>1260</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2243</th>\n",
       "      <td>0.315752</td>\n",
       "      <td>2355520</td>\n",
       "      <td>2356711</td>\n",
       "      <td>+</td>\n",
       "      <td>glpC</td>\n",
       "      <td>1191</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3011</th>\n",
       "      <td>0.049324</td>\n",
       "      <td>3155354</td>\n",
       "      <td>3156518</td>\n",
       "      <td>+</td>\n",
       "      <td>yqhD</td>\n",
       "      <td>1164</td>\n",
       "      <td>yqhD-dkgA</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>rpoD,yqhC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3426</th>\n",
       "      <td>0.350034</td>\n",
       "      <td>3562012</td>\n",
       "      <td>3563518</td>\n",
       "      <td>+</td>\n",
       "      <td>glpD</td>\n",
       "      <td>1506</td>\n",
       "      <td>glpD</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,glpR,rpoD,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3925</th>\n",
       "      <td>0.084278</td>\n",
       "      <td>4114568</td>\n",
       "      <td>4115579</td>\n",
       "      <td>-</td>\n",
       "      <td>glpX</td>\n",
       "      <td>1011</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3926</th>\n",
       "      <td>0.290235</td>\n",
       "      <td>4115713</td>\n",
       "      <td>4117222</td>\n",
       "      <td>-</td>\n",
       "      <td>glpK</td>\n",
       "      <td>1509</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3927</th>\n",
       "      <td>0.312307</td>\n",
       "      <td>4117244</td>\n",
       "      <td>4118090</td>\n",
       "      <td>-</td>\n",
       "      <td>glpF</td>\n",
       "      <td>846</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length     operon  \\\n",
       "b1199     0.045686  1249124  1249757      -      dhaL     633     dhaKLM   \n",
       "b1200     0.045491  1249767  1250838      -      dhaK    1071     dhaKLM   \n",
       "b2239     0.211384  2349934  2351011      -      glpQ    1077      glpTQ   \n",
       "b2240     0.306134  2351015  2352374      -      glpT    1359      glpTQ   \n",
       "b2241     0.375662  2352646  2354275      +      glpA    1629     glpABC   \n",
       "b2242     0.328961  2354264  2355524      +      glpB    1260     glpABC   \n",
       "b2243     0.315752  2355520  2356711      +      glpC    1191     glpABC   \n",
       "b3011     0.049324  3155354  3156518      +      yqhD    1164  yqhD-dkgA   \n",
       "b3426     0.350034  3562012  3563518      +      glpD    1506       glpD   \n",
       "b3925     0.084278  4114568  4115579      -      glpX    1011     glpFKX   \n",
       "b3926     0.290235  4115713  4117222      -      glpK    1509     glpFKX   \n",
       "b3927     0.312307  4117244  4118090      -      glpF     846     glpFKX   \n",
       "\n",
       "                                         cog  \\\n",
       "b1199  Carbohydrate transport and metabolism   \n",
       "b1200  Carbohydrate transport and metabolism   \n",
       "b2239       Energy production and conversion   \n",
       "b2240  Carbohydrate transport and metabolism   \n",
       "b2241       Energy production and conversion   \n",
       "b2242    Amino acid transport and metabolism   \n",
       "b2243       Energy production and conversion   \n",
       "b3011       Energy production and conversion   \n",
       "b3426       Energy production and conversion   \n",
       "b3925  Carbohydrate transport and metabolism   \n",
       "b3926       Energy production and conversion   \n",
       "b3927  Carbohydrate transport and metabolism   \n",
       "\n",
       "                                  regulator  \n",
       "b1199                arcA,dhaR,fnr,lrp,rpoS  \n",
       "b1200                arcA,dhaR,fnr,lrp,rpoS  \n",
       "b2239         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2240         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2241  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2242  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2243  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b3011                             rpoD,yqhC  \n",
       "b3426               arcA,crp,glpR,rpoD,yieP  \n",
       "b3925                         crp,glpR,rpoD  \n",
       "b3926                         crp,glpR,rpoD  \n",
       "b3927                         crp,glpR,rpoD  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('GlpR')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manually changing thresholds\n",
    "To manually change a threshold, use the `change_threshold` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.change_threshold('GlpR',0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2240</th>\n",
       "      <td>0.306134</td>\n",
       "      <td>2351015</td>\n",
       "      <td>2352374</td>\n",
       "      <td>-</td>\n",
       "      <td>glpT</td>\n",
       "      <td>1359</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2241</th>\n",
       "      <td>0.375662</td>\n",
       "      <td>2352646</td>\n",
       "      <td>2354275</td>\n",
       "      <td>+</td>\n",
       "      <td>glpA</td>\n",
       "      <td>1629</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2242</th>\n",
       "      <td>0.328961</td>\n",
       "      <td>2354264</td>\n",
       "      <td>2355524</td>\n",
       "      <td>+</td>\n",
       "      <td>glpB</td>\n",
       "      <td>1260</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2243</th>\n",
       "      <td>0.315752</td>\n",
       "      <td>2355520</td>\n",
       "      <td>2356711</td>\n",
       "      <td>+</td>\n",
       "      <td>glpC</td>\n",
       "      <td>1191</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3426</th>\n",
       "      <td>0.350034</td>\n",
       "      <td>3562012</td>\n",
       "      <td>3563518</td>\n",
       "      <td>+</td>\n",
       "      <td>glpD</td>\n",
       "      <td>1506</td>\n",
       "      <td>glpD</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,glpR,rpoD,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3927</th>\n",
       "      <td>0.312307</td>\n",
       "      <td>4117244</td>\n",
       "      <td>4118090</td>\n",
       "      <td>-</td>\n",
       "      <td>glpF</td>\n",
       "      <td>846</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length  operon  \\\n",
       "b2240     0.306134  2351015  2352374      -      glpT    1359   glpTQ   \n",
       "b2241     0.375662  2352646  2354275      +      glpA    1629  glpABC   \n",
       "b2242     0.328961  2354264  2355524      +      glpB    1260  glpABC   \n",
       "b2243     0.315752  2355520  2356711      +      glpC    1191  glpABC   \n",
       "b3426     0.350034  3562012  3563518      +      glpD    1506    glpD   \n",
       "b3927     0.312307  4117244  4118090      -      glpF     846  glpFKX   \n",
       "\n",
       "                                         cog  \\\n",
       "b2240  Carbohydrate transport and metabolism   \n",
       "b2241       Energy production and conversion   \n",
       "b2242    Amino acid transport and metabolism   \n",
       "b2243       Energy production and conversion   \n",
       "b3426       Energy production and conversion   \n",
       "b3927  Carbohydrate transport and metabolism   \n",
       "\n",
       "                                  regulator  \n",
       "b2240         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2241  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2242  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2243  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b3426               arcA,crp,glpR,rpoD,yieP  \n",
       "b3927                         crp,glpR,rpoD  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('GlpR')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organism Comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The compare_ica function is a function that allows for the comparison of different S matricies, even between different species. The function requires two S matricies as inputs in the form of Pandas Dataframes and a string path to the ortho_dir, as well as several different optional inputs. Ortho_dir is a string that describes the local path to the 'ortholog' directory under the Modulome repository. Metric can either be 'pearson' or 'spearman', which determines which correlation test is used for comparison between two components. Cutoff is a decimal value for of what minimum correlation value must occur between two components to be considered a match.\n",
    "\n",
    "The function returns two outputs, a digraph that describes which components from organism two are related to the ones in organsim one, and a name links dictionary which contains pairwise entries between related components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function compare_ica in module pymodulon.util:\n",
      "\n",
      "compare_ica(S1: pandas.core.frame.DataFrame, S2: pandas.core.frame.DataFrame, ortho_dir, metric='pearson', cutoff=0.2, show_all=False)\n",
      "    Compares two S matrices between a single organism or across organisms and returns the connected ICA components\n",
      "    :param S1: Pandas Dataframe of S matrix 1\n",
      "    :param S2: Pandas Dataframe of S Matrix 2\n",
      "    :param ortho_dir: String of the location where organism data can be found (can be found under modulome/data)\n",
      "    :param metric: A string of what statistical test to use (standard is 'pearson')\n",
      "    :param cutoff: Float cut off value for pearson statistical test\n",
      "    :param show_all: True will show all nodes of the digraph matrix\n",
      "    :return: Dot graph and name links of connected ICA components between the two runs or organisms.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(compare_ica)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3923, 92)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load in test data from precise\n",
    "mtb_S = pd.read_csv(\"pymodulon/test/data/mtb_S.csv\",index_col=0)\n",
    "ecoli_S = pd.read_csv(\"pymodulon/test/data/ecoli_S.csv\",index_col=0)\n",
    "ecoli_S_mod = pd.read_csv(\"pymodulon/test/data/ecoli_S_mod.csv\",index_col=0)\n",
    "ecoli_S.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lable the data for simpler viewing\n",
    "mtb_col = list(mtb_S.columns)\n",
    "for i in range(0,len(mtb_col)):\n",
    "    mtb_col[i] = \"mtb_\"+str(mtb_col[i])\n",
    "mtb_S.columns = mtb_col\n",
    "\n",
    "ecoli_col = list(ecoli_S.columns)\n",
    "for i in range(0,len(ecoli_col)):\n",
    "    ecoli_col[i] = \"ecoli_\"+str(ecoli_col[i])\n",
    "ecoli_S.columns = ecoli_col\n",
    "\n",
    "ecoli_col = list(ecoli_S_mod.columns)\n",
    "for i in range(0,len(ecoli_col)):\n",
    "    ecoli_col[i] = \"ecoli_mod_\"+str(ecoli_col[i])\n",
    "ecoli_S_mod.columns = ecoli_col\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 92/92 [00:01<00:00, 56.37it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"720pt\" height=\"562pt\"\n",
       " viewBox=\"0.00 0.00 720.00 562.33\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.551302 0.551302) rotate(0) translate(4 1016)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-1016 1302,-1016 1302,4 -4,4\"/>\n",
       "<!-- data2_ecoli_mod_0 -->\n",
       "<g id=\"node1\" class=\"node\"><title>data2_ecoli_mod_0</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_0</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_1 -->\n",
       "<g id=\"node2\" class=\"node\"><title>data2_ecoli_mod_1</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_1</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_10 -->\n",
       "<g id=\"node3\" class=\"node\"><title>data2_ecoli_mod_10</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_10</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_11 -->\n",
       "<g id=\"node4\" class=\"node\"><title>data2_ecoli_mod_11</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_11</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_12 -->\n",
       "<g id=\"node5\" class=\"node\"><title>data2_ecoli_mod_12</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_12</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_13 -->\n",
       "<g id=\"node6\" class=\"node\"><title>data2_ecoli_mod_13</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_13</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_14 -->\n",
       "<g id=\"node7\" class=\"node\"><title>data2_ecoli_mod_14</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_14</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_15 -->\n",
       "<g id=\"node8\" class=\"node\"><title>data2_ecoli_mod_15</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_15</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_16 -->\n",
       "<g id=\"node9\" class=\"node\"><title>data2_ecoli_mod_16</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_16</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_17 -->\n",
       "<g id=\"node10\" class=\"node\"><title>data2_ecoli_mod_17</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-932.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_17</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_18 -->\n",
       "<g id=\"node11\" class=\"node\"><title>data2_ecoli_mod_18</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_18</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_19 -->\n",
       "<g id=\"node12\" class=\"node\"><title>data2_ecoli_mod_19</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_19</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_2 -->\n",
       "<g id=\"node13\" class=\"node\"><title>data2_ecoli_mod_2</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_2</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_20 -->\n",
       "<g id=\"node14\" class=\"node\"><title>data2_ecoli_mod_20</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_20</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_21 -->\n",
       "<g id=\"node15\" class=\"node\"><title>data2_ecoli_mod_21</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_21</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_22 -->\n",
       "<g id=\"node16\" class=\"node\"><title>data2_ecoli_mod_22</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_22</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_23 -->\n",
       "<g id=\"node17\" class=\"node\"><title>data2_ecoli_mod_23</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_23</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_24 -->\n",
       "<g id=\"node18\" class=\"node\"><title>data2_ecoli_mod_24</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_24</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_25 -->\n",
       "<g id=\"node19\" class=\"node\"><title>data2_ecoli_mod_25</title>\n",
       "<text text-anchor=\"middle\" x=\"1030.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_25</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_26 -->\n",
       "<g id=\"node20\" class=\"node\"><title>data2_ecoli_mod_26</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_26</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_27 -->\n",
       "<g id=\"node21\" class=\"node\"><title>data2_ecoli_mod_27</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_27</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_28 -->\n",
       "<g id=\"node22\" class=\"node\"><title>data2_ecoli_mod_28</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_28</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_29 -->\n",
       "<g id=\"node23\" class=\"node\"><title>data2_ecoli_mod_29</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_29</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_3 -->\n",
       "<g id=\"node24\" class=\"node\"><title>data2_ecoli_mod_3</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_3</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_30 -->\n",
       "<g id=\"node25\" class=\"node\"><title>data2_ecoli_mod_30</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_30</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_31 -->\n",
       "<g id=\"node26\" class=\"node\"><title>data2_ecoli_mod_31</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_31</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_32 -->\n",
       "<g id=\"node27\" class=\"node\"><title>data2_ecoli_mod_32</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_32</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_33 -->\n",
       "<g id=\"node28\" class=\"node\"><title>data2_ecoli_mod_33</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_33</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_34 -->\n",
       "<g id=\"node29\" class=\"node\"><title>data2_ecoli_mod_34</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_34</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_35 -->\n",
       "<g id=\"node30\" class=\"node\"><title>data2_ecoli_mod_35</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-728.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_35</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_36 -->\n",
       "<g id=\"node31\" class=\"node\"><title>data2_ecoli_mod_36</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_36</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_37 -->\n",
       "<g id=\"node32\" class=\"node\"><title>data2_ecoli_mod_37</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_37</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_38 -->\n",
       "<g id=\"node33\" class=\"node\"><title>data2_ecoli_mod_38</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_38</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_39 -->\n",
       "<g id=\"node34\" class=\"node\"><title>data2_ecoli_mod_39</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_39</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_4 -->\n",
       "<g id=\"node35\" class=\"node\"><title>data2_ecoli_mod_4</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_4</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_40 -->\n",
       "<g id=\"node36\" class=\"node\"><title>data2_ecoli_mod_40</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_40</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_41 -->\n",
       "<g id=\"node37\" class=\"node\"><title>data2_ecoli_mod_41</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_41</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_42 -->\n",
       "<g id=\"node38\" class=\"node\"><title>data2_ecoli_mod_42</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_42</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_43 -->\n",
       "<g id=\"node39\" class=\"node\"><title>data2_ecoli_mod_43</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_43</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_44 -->\n",
       "<g id=\"node40\" class=\"node\"><title>data2_ecoli_mod_44</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-626.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_44</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_45 -->\n",
       "<g id=\"node41\" class=\"node\"><title>data2_ecoli_mod_45</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_45</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_46 -->\n",
       "<g id=\"node42\" class=\"node\"><title>data2_ecoli_mod_46</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_46</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_47 -->\n",
       "<g id=\"node43\" class=\"node\"><title>data2_ecoli_mod_47</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_47</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_48 -->\n",
       "<g id=\"node44\" class=\"node\"><title>data2_ecoli_mod_48</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_48</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_49 -->\n",
       "<g id=\"node45\" class=\"node\"><title>data2_ecoli_mod_49</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_49</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_5 -->\n",
       "<g id=\"node46\" class=\"node\"><title>data2_ecoli_mod_5</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_5</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_50 -->\n",
       "<g id=\"node47\" class=\"node\"><title>data2_ecoli_mod_50</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_50</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_51 -->\n",
       "<g id=\"node48\" class=\"node\"><title>data2_ecoli_mod_51</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_51</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_52 -->\n",
       "<g id=\"node49\" class=\"node\"><title>data2_ecoli_mod_52</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_52</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_53 -->\n",
       "<g id=\"node50\" class=\"node\"><title>data2_ecoli_mod_53</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-524.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_53</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_54 -->\n",
       "<g id=\"node51\" class=\"node\"><title>data2_ecoli_mod_54</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_54</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_55 -->\n",
       "<g id=\"node52\" class=\"node\"><title>data2_ecoli_mod_55</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_55</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_56 -->\n",
       "<g id=\"node53\" class=\"node\"><title>data2_ecoli_mod_56</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_56</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_57 -->\n",
       "<g id=\"node54\" class=\"node\"><title>data2_ecoli_mod_57</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_57</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_58 -->\n",
       "<g id=\"node55\" class=\"node\"><title>data2_ecoli_mod_58</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_58</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_59 -->\n",
       "<g id=\"node56\" class=\"node\"><title>data2_ecoli_mod_59</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_59</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_6 -->\n",
       "<g id=\"node57\" class=\"node\"><title>data2_ecoli_mod_6</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_6</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_60 -->\n",
       "<g id=\"node58\" class=\"node\"><title>data2_ecoli_mod_60</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_60</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_61 -->\n",
       "<g id=\"node59\" class=\"node\"><title>data2_ecoli_mod_61</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_61</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_62 -->\n",
       "<g id=\"node60\" class=\"node\"><title>data2_ecoli_mod_62</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_62</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_63 -->\n",
       "<g id=\"node61\" class=\"node\"><title>data2_ecoli_mod_63</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_63</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_64 -->\n",
       "<g id=\"node62\" class=\"node\"><title>data2_ecoli_mod_64</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_64</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_65 -->\n",
       "<g id=\"node63\" class=\"node\"><title>data2_ecoli_mod_65</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_65</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_66 -->\n",
       "<g id=\"node64\" class=\"node\"><title>data2_ecoli_mod_66</title>\n",
       "<text text-anchor=\"middle\" x=\"383.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_66</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_67 -->\n",
       "<g id=\"node65\" class=\"node\"><title>data2_ecoli_mod_67</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_67</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_68 -->\n",
       "<g id=\"node66\" class=\"node\"><title>data2_ecoli_mod_68</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_68</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_69 -->\n",
       "<g id=\"node67\" class=\"node\"><title>data2_ecoli_mod_69</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_69</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_7 -->\n",
       "<g id=\"node68\" class=\"node\"><title>data2_ecoli_mod_7</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_7</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_70 -->\n",
       "<g id=\"node69\" class=\"node\"><title>data2_ecoli_mod_70</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_70</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_71 -->\n",
       "<g id=\"node70\" class=\"node\"><title>data2_ecoli_mod_71</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-349.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">ecoli_mod_71</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_72 -->\n",
       "<g id=\"node71\" class=\"node\"><title>data2_ecoli_mod_72</title>\n",
       "<text text-anchor=\"middle\" x=\"1132.5\" y=\"-830.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_72</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_73 -->\n",
       "<g id=\"node72\" class=\"node\"><title>data2_ecoli_mod_73</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_73</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_74 -->\n",
       "<g id=\"node73\" class=\"node\"><title>data2_ecoli_mod_74</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_74</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_75 -->\n",
       "<g id=\"node74\" class=\"node\"><title>data2_ecoli_mod_75</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_75</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_76 -->\n",
       "<g id=\"node75\" class=\"node\"><title>data2_ecoli_mod_76</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_76</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_77 -->\n",
       "<g id=\"node76\" class=\"node\"><title>data2_ecoli_mod_77</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_77</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_78 -->\n",
       "<g id=\"node77\" class=\"node\"><title>data2_ecoli_mod_78</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_78</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_79 -->\n",
       "<g id=\"node78\" class=\"node\"><title>data2_ecoli_mod_79</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_79</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_8 -->\n",
       "<g id=\"node79\" class=\"node\"><title>data2_ecoli_mod_8</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-247.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">ecoli_mod_8</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_80 -->\n",
       "<g id=\"node80\" class=\"node\"><title>data2_ecoli_mod_80</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_80</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_81 -->\n",
       "<g id=\"node81\" class=\"node\"><title>data2_ecoli_mod_81</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_81</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_82 -->\n",
       "<g id=\"node82\" class=\"node\"><title>data2_ecoli_mod_82</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_82</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_83 -->\n",
       "<g id=\"node83\" class=\"node\"><title>data2_ecoli_mod_83</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_83</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_84 -->\n",
       "<g id=\"node84\" class=\"node\"><title>data2_ecoli_mod_84</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_84</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_85 -->\n",
       "<g id=\"node85\" class=\"node\"><title>data2_ecoli_mod_85</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_85</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_86 -->\n",
       "<g id=\"node86\" class=\"node\"><title>data2_ecoli_mod_86</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_86</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_87 -->\n",
       "<g id=\"node87\" class=\"node\"><title>data2_ecoli_mod_87</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_87</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_88 -->\n",
       "<g id=\"node88\" class=\"node\"><title>data2_ecoli_mod_88</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_88</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_89 -->\n",
       "<g id=\"node89\" class=\"node\"><title>data2_ecoli_mod_89</title>\n",
       "<text text-anchor=\"middle\" x=\"485.5\" y=\"-320.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_89</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_9 -->\n",
       "<g id=\"node90\" class=\"node\"><title>data2_ecoli_mod_9</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_9</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_90 -->\n",
       "<g id=\"node91\" class=\"node\"><title>data2_ecoli_mod_90</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_90</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_91 -->\n",
       "<g id=\"node92\" class=\"node\"><title>data2_ecoli_mod_91</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_91</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_92 -->\n",
       "<g id=\"node93\" class=\"node\"><title>data2_ecoli_mod_92</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-43.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">ecoli_mod_92</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_mod_93 -->\n",
       "<g id=\"node94\" class=\"node\"><title>data2_ecoli_mod_93</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_mod_93</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_0 -->\n",
       "<g id=\"node95\" class=\"node\"><title>data1_ecoli_0</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_0</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_0&#45;&gt;data2_ecoli_mod_44 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>data1_ecoli_0&#45;&gt;data2_ecoli_mod_44</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.88\" d=\"M1244.5,-669.894C1244.5,-664.575 1244.5,-658.634 1244.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.88\" points=\"1246.64,-653.001 1244.5,-648.001 1242.37,-653.001 1246.64,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_1 -->\n",
       "<g id=\"node96\" class=\"node\"><title>data1_ecoli_1</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_1</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_1&#45;&gt;data2_ecoli_mod_24 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>data1_ecoli_1&#45;&gt;data2_ecoli_mod_24</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M921.5,-873.894C921.5,-868.575 921.5,-862.634 921.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"923.674,-857.001 921.5,-852.001 919.326,-857.001 923.674,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_2 -->\n",
       "<g id=\"node97\" class=\"node\"><title>data1_ecoli_2</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_2</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_2&#45;&gt;data2_ecoli_mod_78 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>data1_ecoli_2&#45;&gt;data2_ecoli_mod_78</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.78\" d=\"M703.5,-261.894C703.5,-256.575 703.5,-250.634 703.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.78\" points=\"705.591,-245.001 703.5,-240.001 701.409,-245.001 705.591,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_3 -->\n",
       "<g id=\"node98\" class=\"node\"><title>data1_ecoli_3</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_3</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_3&#45;&gt;data2_ecoli_mod_2 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>data1_ecoli_3&#45;&gt;data2_ecoli_mod_2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.9\" d=\"M274.5,-873.894C274.5,-868.575 274.5,-862.634 274.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.9\" points=\"276.644,-857.001 274.5,-852.001 272.356,-857.001 276.644,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_4 -->\n",
       "<g id=\"node99\" class=\"node\"><title>data1_ecoli_4</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_4</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_4&#45;&gt;data2_ecoli_mod_79 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>data1_ecoli_4&#45;&gt;data2_ecoli_mod_79</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.88\" d=\"M812.5,-261.894C812.5,-256.575 812.5,-250.634 812.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.88\" points=\"814.635,-245.001 812.5,-240.001 810.365,-245.001 814.635,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_5 -->\n",
       "<g id=\"node100\" class=\"node\"><title>data1_ecoli_5</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_5</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_5&#45;&gt;data2_ecoli_mod_7 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>data1_ecoli_5&#45;&gt;data2_ecoli_mod_7</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.84\" d=\"M921.5,-363.894C921.5,-358.575 921.5,-352.634 921.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.84\" points=\"923.618,-347.001 921.5,-342.001 919.383,-347.001 923.618,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_6 -->\n",
       "<g id=\"node101\" class=\"node\"><title>data1_ecoli_6</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_6</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_6&#45;&gt;data2_ecoli_mod_11 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>data1_ecoli_6&#45;&gt;data2_ecoli_mod_11</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.68\" d=\"M434.5,-975.894C434.5,-970.575 434.5,-964.634 434.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.68\" points=\"436.548,-959.001 434.5,-954.001 432.453,-959.001 436.548,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_7 -->\n",
       "<g id=\"node102\" class=\"node\"><title>data1_ecoli_7</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_7</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_7&#45;&gt;data2_ecoli_mod_32 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>data1_ecoli_7&#45;&gt;data2_ecoli_mod_32</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.84\" d=\"M812.5,-771.894C812.5,-766.575 812.5,-760.634 812.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.84\" points=\"814.618,-755.001 812.5,-750.001 810.383,-755.001 814.618,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_8 -->\n",
       "<g id=\"node103\" class=\"node\"><title>data1_ecoli_8</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_8</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_8&#45;&gt;data2_ecoli_mod_9 -->\n",
       "<g id=\"edge9\" class=\"edge\"><title>data1_ecoli_8&#45;&gt;data2_ecoli_mod_9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M921.5,-159.894C921.5,-154.575 921.5,-148.634 921.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"923.666,-143.001 921.5,-138.001 919.334,-143.001 923.666,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_9 -->\n",
       "<g id=\"node104\" class=\"node\"><title>data1_ecoli_9</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_9</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_9&#45;&gt;data2_ecoli_mod_82 -->\n",
       "<g id=\"edge10\" class=\"edge\"><title>data1_ecoli_9&#45;&gt;data2_ecoli_mod_82</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.84\" d=\"M53.5,-159.894C53.5,-154.575 53.5,-148.634 53.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.84\" points=\"55.6176,-143.001 53.5,-138.001 51.3826,-143.001 55.6176,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_10 -->\n",
       "<g id=\"node105\" class=\"node\"><title>data1_ecoli_10</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_10</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_10&#45;&gt;data2_ecoli_mod_83 -->\n",
       "<g id=\"edge11\" class=\"edge\"><title>data1_ecoli_10&#45;&gt;data2_ecoli_mod_83</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.92\" d=\"M165.5,-159.894C165.5,-154.575 165.5,-148.634 165.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.92\" points=\"167.653,-143.001 165.5,-138.001 163.348,-143.001 167.653,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_11 -->\n",
       "<g id=\"node106\" class=\"node\"><title>data1_ecoli_11</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_11</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_11&#45;&gt;data2_ecoli_mod_61 -->\n",
       "<g id=\"edge12\" class=\"edge\"><title>data1_ecoli_11&#45;&gt;data2_ecoli_mod_61</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.92\" d=\"M1081.5,-465.894C1081.5,-460.575 1081.5,-454.634 1081.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.92\" points=\"1083.65,-449.001 1081.5,-444.001 1079.35,-449.001 1083.65,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_12 -->\n",
       "<g id=\"node107\" class=\"node\"><title>data1_ecoli_12</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_12</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_12&#45;&gt;data2_ecoli_mod_5 -->\n",
       "<g id=\"edge13\" class=\"edge\"><title>data1_ecoli_12&#45;&gt;data2_ecoli_mod_5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.94\" d=\"M703.5,-567.894C703.5,-562.575 703.5,-556.634 703.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.94\" points=\"705.661,-551.001 703.5,-546.001 701.339,-551.001 705.661,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_13 -->\n",
       "<g id=\"node108\" class=\"node\"><title>data1_ecoli_13</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_13</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_13&#45;&gt;data2_ecoli_mod_0 -->\n",
       "<g id=\"edge14\" class=\"edge\"><title>data1_ecoli_13&#45;&gt;data2_ecoli_mod_0</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.94\" d=\"M53.5,-975.894C53.5,-970.575 53.5,-964.634 53.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.94\" points=\"55.6613,-959.001 53.5,-954.001 51.3388,-959.001 55.6613,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_14 -->\n",
       "<g id=\"node109\" class=\"node\"><title>data1_ecoli_14</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_14</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_14&#45;&gt;data2_ecoli_mod_37 -->\n",
       "<g id=\"edge15\" class=\"edge\"><title>data1_ecoli_14&#45;&gt;data2_ecoli_mod_37</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.96\" d=\"M165.5,-669.894C165.5,-664.575 165.5,-658.634 165.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.96\" points=\"167.67,-653.001 165.5,-648.001 163.33,-653.001 167.67,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_15 -->\n",
       "<g id=\"node110\" class=\"node\"><title>data1_ecoli_15</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_15</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_15&#45;&gt;data2_ecoli_mod_6 -->\n",
       "<g id=\"edge16\" class=\"edge\"><title>data1_ecoli_15&#45;&gt;data2_ecoli_mod_6</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.81\" d=\"M812.5,-465.894C812.5,-460.575 812.5,-454.634 812.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.81\" points=\"814.604,-449.001 812.5,-444.001 810.396,-449.001 814.604,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_16 -->\n",
       "<g id=\"node111\" class=\"node\"><title>data1_ecoli_16</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_16</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_16&#45;&gt;data2_ecoli_mod_64 -->\n",
       "<g id=\"edge17\" class=\"edge\"><title>data1_ecoli_16&#45;&gt;data2_ecoli_mod_64</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.94\" d=\"M165.5,-363.894C165.5,-358.575 165.5,-352.634 165.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.94\" points=\"167.661,-347.001 165.5,-342.001 163.339,-347.001 167.661,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_17 -->\n",
       "<g id=\"node112\" class=\"node\"><title>data1_ecoli_17</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_17</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_17&#45;&gt;data2_ecoli_mod_46 -->\n",
       "<g id=\"edge18\" class=\"edge\"><title>data1_ecoli_17&#45;&gt;data2_ecoli_mod_46</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M165.5,-567.894C165.5,-562.575 165.5,-556.634 165.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"167.631,-551.001 165.5,-546.001 163.369,-551.001 167.631,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_18 -->\n",
       "<g id=\"node113\" class=\"node\"><title>data1_ecoli_18</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_18</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_18&#45;&gt;data2_ecoli_mod_14 -->\n",
       "<g id=\"edge19\" class=\"edge\"><title>data1_ecoli_18&#45;&gt;data2_ecoli_mod_14</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M812.5,-975.894C812.5,-970.575 812.5,-964.634 812.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"814.666,-959.001 812.5,-954.001 810.334,-959.001 814.666,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_19 -->\n",
       "<g id=\"node114\" class=\"node\"><title>data1_ecoli_19</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_19</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_19&#45;&gt;data2_ecoli_mod_1 -->\n",
       "<g id=\"edge20\" class=\"edge\"><title>data1_ecoli_19&#45;&gt;data2_ecoli_mod_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M165.5,-975.894C165.5,-970.575 165.5,-964.634 165.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"167.666,-959.001 165.5,-954.001 163.334,-959.001 167.666,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_20 -->\n",
       "<g id=\"node115\" class=\"node\"><title>data1_ecoli_20</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_20</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_20&#45;&gt;data2_ecoli_mod_15 -->\n",
       "<g id=\"edge21\" class=\"edge\"><title>data1_ecoli_20&#45;&gt;data2_ecoli_mod_15</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.93\" d=\"M921.5,-975.894C921.5,-970.575 921.5,-964.634 921.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.93\" points=\"923.657,-959.001 921.5,-954.001 919.343,-959.001 923.657,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_21 -->\n",
       "<g id=\"node116\" class=\"node\"><title>data1_ecoli_21</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_21</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_21&#45;&gt;data2_ecoli_mod_86 -->\n",
       "<g id=\"edge22\" class=\"edge\"><title>data1_ecoli_21&#45;&gt;data2_ecoli_mod_86</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.85\" d=\"M594.5,-159.894C594.5,-154.575 594.5,-148.634 594.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.85\" points=\"596.622,-143.001 594.5,-138.001 592.378,-143.001 596.622,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_22 -->\n",
       "<g id=\"node117\" class=\"node\"><title>data1_ecoli_22</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_22</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_22&#45;&gt;data2_ecoli_mod_81 -->\n",
       "<g id=\"edge23\" class=\"edge\"><title>data1_ecoli_22&#45;&gt;data2_ecoli_mod_81</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.94\" d=\"M1244.5,-261.894C1244.5,-256.575 1244.5,-250.634 1244.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.94\" points=\"1246.66,-245.001 1244.5,-240.001 1242.34,-245.001 1246.66,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_23 -->\n",
       "<g id=\"node118\" class=\"node\"><title>data1_ecoli_23</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_23</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_23&#45;&gt;data2_ecoli_mod_28 -->\n",
       "<g id=\"edge24\" class=\"edge\"><title>data1_ecoli_23&#45;&gt;data2_ecoli_mod_28</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.9\" d=\"M165.5,-771.894C165.5,-766.575 165.5,-760.634 165.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.9\" points=\"167.644,-755.001 165.5,-750.001 163.356,-755.001 167.644,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_24 -->\n",
       "<g id=\"node119\" class=\"node\"><title>data1_ecoli_24</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_24</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_24&#45;&gt;data2_ecoli_mod_54 -->\n",
       "<g id=\"edge25\" class=\"edge\"><title>data1_ecoli_24&#45;&gt;data2_ecoli_mod_54</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.89\" d=\"M53.5,-465.894C53.5,-460.575 53.5,-454.634 53.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.89\" points=\"55.6395,-449.001 53.5,-444.001 51.3607,-449.001 55.6395,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_25 -->\n",
       "<g id=\"node120\" class=\"node\"><title>data1_ecoli_25</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_25</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_25&#45;&gt;data2_ecoli_mod_93 -->\n",
       "<g id=\"edge26\" class=\"edge\"><title>data1_ecoli_25&#45;&gt;data2_ecoli_mod_93</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M165.5,-57.8939C165.5,-52.5746 165.5,-46.6338 165.5,-41.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"167.631,-41.0013 165.5,-36.0014 163.369,-41.0014 167.631,-41.0013\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_26 -->\n",
       "<g id=\"node121\" class=\"node\"><title>data1_ecoli_26</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_26</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_26&#45;&gt;data2_ecoli_mod_55 -->\n",
       "<g id=\"edge27\" class=\"edge\"><title>data1_ecoli_26&#45;&gt;data2_ecoli_mod_55</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M165.5,-465.894C165.5,-460.575 165.5,-454.634 165.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"167.674,-449.001 165.5,-444.001 163.326,-449.001 167.674,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_27 -->\n",
       "<g id=\"node122\" class=\"node\"><title>data1_ecoli_27</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_27</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_27&#45;&gt;data2_ecoli_mod_20 -->\n",
       "<g id=\"edge28\" class=\"edge\"><title>data1_ecoli_27&#45;&gt;data2_ecoli_mod_20</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.9\" d=\"M434.5,-873.894C434.5,-868.575 434.5,-862.634 434.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.9\" points=\"436.644,-857.001 434.5,-852.001 432.356,-857.001 436.644,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_28 -->\n",
       "<g id=\"node123\" class=\"node\"><title>data1_ecoli_28</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_28</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_28&#45;&gt;data2_ecoli_mod_40 -->\n",
       "<g id=\"edge29\" class=\"edge\"><title>data1_ecoli_28&#45;&gt;data2_ecoli_mod_40</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M703.5,-669.894C703.5,-664.575 703.5,-658.634 703.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"705.666,-653.001 703.5,-648.001 701.334,-653.001 705.666,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_29 -->\n",
       "<g id=\"node124\" class=\"node\"><title>data1_ecoli_29</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_29</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_29&#45;&gt;data2_ecoli_mod_74 -->\n",
       "<g id=\"edge30\" class=\"edge\"><title>data1_ecoli_29&#45;&gt;data2_ecoli_mod_74</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M165.5,-261.894C165.5,-256.575 165.5,-250.634 165.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"167.631,-245.001 165.5,-240.001 163.369,-245.001 167.631,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_30 -->\n",
       "<g id=\"node125\" class=\"node\"><title>data1_ecoli_30</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_30</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_30&#45;&gt;data2_ecoli_mod_39 -->\n",
       "<g id=\"edge31\" class=\"edge\"><title>data1_ecoli_30&#45;&gt;data2_ecoli_mod_39</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M434.5,-669.894C434.5,-664.575 434.5,-658.634 434.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"436.631,-653.001 434.5,-648.001 432.369,-653.001 436.631,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_31 -->\n",
       "<g id=\"node126\" class=\"node\"><title>data1_ecoli_31</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_31</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_31&#45;&gt;data2_ecoli_mod_87 -->\n",
       "<g id=\"edge32\" class=\"edge\"><title>data1_ecoli_31&#45;&gt;data2_ecoli_mod_87</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.98\" d=\"M703.5,-159.894C703.5,-154.575 703.5,-148.634 703.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.98\" points=\"705.679,-143.001 703.5,-138.001 701.321,-143.001 705.679,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_32 -->\n",
       "<g id=\"node127\" class=\"node\"><title>data1_ecoli_32</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_32</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_32&#45;&gt;data2_ecoli_mod_26 -->\n",
       "<g id=\"edge33\" class=\"edge\"><title>data1_ecoli_32&#45;&gt;data2_ecoli_mod_26</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.82\" d=\"M1244.5,-873.894C1244.5,-868.575 1244.5,-862.634 1244.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.82\" points=\"1246.61,-857.001 1244.5,-852.001 1242.39,-857.001 1246.61,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_33 -->\n",
       "<g id=\"node128\" class=\"node\"><title>data1_ecoli_33</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_33</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_33&#45;&gt;data2_ecoli_mod_84 -->\n",
       "<g id=\"edge34\" class=\"edge\"><title>data1_ecoli_33&#45;&gt;data2_ecoli_mod_84</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.89\" d=\"M274.5,-159.894C274.5,-154.575 274.5,-148.634 274.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.89\" points=\"276.639,-143.001 274.5,-138.001 272.361,-143.001 276.639,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_34 -->\n",
       "<g id=\"node129\" class=\"node\"><title>data1_ecoli_34</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_34</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_34&#45;&gt;data2_ecoli_mod_62 -->\n",
       "<g id=\"edge35\" class=\"edge\"><title>data1_ecoli_34&#45;&gt;data2_ecoli_mod_62</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.81\" d=\"M1244.5,-465.894C1244.5,-460.575 1244.5,-454.634 1244.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.81\" points=\"1246.6,-449.001 1244.5,-444.001 1242.4,-449.001 1246.6,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_35 -->\n",
       "<g id=\"node130\" class=\"node\"><title>data1_ecoli_35</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_35</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_35&#45;&gt;data2_ecoli_mod_52 -->\n",
       "<g id=\"edge36\" class=\"edge\"><title>data1_ecoli_35&#45;&gt;data2_ecoli_mod_52</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.93\" d=\"M1081.5,-567.894C1081.5,-562.575 1081.5,-556.634 1081.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.93\" points=\"1083.66,-551.001 1081.5,-546.001 1079.34,-551.001 1083.66,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_36 -->\n",
       "<g id=\"node131\" class=\"node\"><title>data1_ecoli_36</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_36</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_36&#45;&gt;data2_ecoli_mod_65 -->\n",
       "<g id=\"edge37\" class=\"edge\"><title>data1_ecoli_36&#45;&gt;data2_ecoli_mod_65</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.86\" d=\"M274.5,-363.894C274.5,-358.575 274.5,-352.634 274.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.86\" points=\"276.626,-347.001 274.5,-342.001 272.374,-347.001 276.626,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_37 -->\n",
       "<g id=\"node132\" class=\"node\"><title>data1_ecoli_37</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_37</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_37&#45;&gt;data2_ecoli_mod_30 -->\n",
       "<g id=\"edge38\" class=\"edge\"><title>data1_ecoli_37&#45;&gt;data2_ecoli_mod_30</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.8\" d=\"M594.5,-771.894C594.5,-766.575 594.5,-760.634 594.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.8\" points=\"596.6,-755.001 594.5,-750.001 592.4,-755.001 596.6,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_38 -->\n",
       "<g id=\"node133\" class=\"node\"><title>data1_ecoli_38</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_38</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_38&#45;&gt;data2_ecoli_mod_49 -->\n",
       "<g id=\"edge39\" class=\"edge\"><title>data1_ecoli_38&#45;&gt;data2_ecoli_mod_49</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M594.5,-567.894C594.5,-562.575 594.5,-556.634 594.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"596.631,-551.001 594.5,-546.001 592.369,-551.001 596.631,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_39 -->\n",
       "<g id=\"node134\" class=\"node\"><title>data1_ecoli_39</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_39</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_39&#45;&gt;data2_ecoli_mod_50 -->\n",
       "<g id=\"edge40\" class=\"edge\"><title>data1_ecoli_39&#45;&gt;data2_ecoli_mod_50</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.89\" d=\"M812.5,-567.894C812.5,-562.575 812.5,-556.634 812.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.89\" points=\"814.639,-551.001 812.5,-546.001 810.361,-551.001 814.639,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_40 -->\n",
       "<g id=\"node135\" class=\"node\"><title>data1_ecoli_40</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_40</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_40&#45;&gt;data2_ecoli_mod_17 -->\n",
       "<g id=\"edge41\" class=\"edge\"><title>data1_ecoli_40&#45;&gt;data2_ecoli_mod_17</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.92\" d=\"M1244.5,-975.894C1244.5,-970.575 1244.5,-964.634 1244.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.92\" points=\"1246.65,-959.001 1244.5,-954.001 1242.35,-959.001 1246.65,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_41 -->\n",
       "<g id=\"node136\" class=\"node\"><title>data1_ecoli_41</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_41</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_41&#45;&gt;data2_ecoli_mod_38 -->\n",
       "<g id=\"edge42\" class=\"edge\"><title>data1_ecoli_41&#45;&gt;data2_ecoli_mod_38</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.91\" d=\"M274.5,-669.894C274.5,-664.575 274.5,-658.634 274.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.91\" points=\"276.648,-653.001 274.5,-648.001 272.352,-653.001 276.648,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_42 -->\n",
       "<g id=\"node137\" class=\"node\"><title>data1_ecoli_42</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_42</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_42&#45;&gt;data2_ecoli_mod_67 -->\n",
       "<g id=\"edge43\" class=\"edge\"><title>data1_ecoli_42&#45;&gt;data2_ecoli_mod_67</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.88\" d=\"M594.5,-363.894C594.5,-358.575 594.5,-352.634 594.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.88\" points=\"596.635,-347.001 594.5,-342.001 592.365,-347.001 596.635,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_43 -->\n",
       "<g id=\"node138\" class=\"node\"><title>data1_ecoli_43</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_43</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_43&#45;&gt;data2_ecoli_mod_76 -->\n",
       "<g id=\"edge44\" class=\"edge\"><title>data1_ecoli_43&#45;&gt;data2_ecoli_mod_76</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.8\" d=\"M434.5,-261.894C434.5,-256.575 434.5,-250.634 434.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.8\" points=\"436.6,-245.001 434.5,-240.001 432.4,-245.001 436.6,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_44 -->\n",
       "<g id=\"node139\" class=\"node\"><title>data1_ecoli_44</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_44</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_44&#45;&gt;data2_ecoli_mod_10 -->\n",
       "<g id=\"edge45\" class=\"edge\"><title>data1_ecoli_44&#45;&gt;data2_ecoli_mod_10</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M274.5,-975.894C274.5,-970.575 274.5,-964.634 274.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"276.666,-959.001 274.5,-954.001 272.334,-959.001 276.666,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_45 -->\n",
       "<g id=\"node140\" class=\"node\"><title>data1_ecoli_45</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_45</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_45&#45;&gt;data2_ecoli_mod_23 -->\n",
       "<g id=\"edge46\" class=\"edge\"><title>data1_ecoli_45&#45;&gt;data2_ecoli_mod_23</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.81\" d=\"M812.5,-873.894C812.5,-868.575 812.5,-862.634 812.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.81\" points=\"814.604,-857.001 812.5,-852.001 810.396,-857.001 814.604,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_46 -->\n",
       "<g id=\"node141\" class=\"node\"><title>data1_ecoli_46</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_46</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_46&#45;&gt;data2_ecoli_mod_36 -->\n",
       "<g id=\"edge47\" class=\"edge\"><title>data1_ecoli_46&#45;&gt;data2_ecoli_mod_36</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.64\" d=\"M53.5,-669.894C53.5,-664.575 53.5,-658.634 53.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.64\" points=\"55.5301,-653.001 53.5,-648.001 51.4701,-653.001 55.5301,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_47 -->\n",
       "<g id=\"node142\" class=\"node\"><title>data1_ecoli_47</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_47</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_47&#45;&gt;data2_ecoli_mod_48 -->\n",
       "<g id=\"edge48\" class=\"edge\"><title>data1_ecoli_47&#45;&gt;data2_ecoli_mod_48</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.88\" d=\"M434.5,-567.894C434.5,-562.575 434.5,-556.634 434.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.88\" points=\"436.635,-551.001 434.5,-546.001 432.365,-551.001 436.635,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_48 -->\n",
       "<g id=\"node143\" class=\"node\"><title>data1_ecoli_48</title>\n",
       "<text text-anchor=\"middle\" x=\"165.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_48</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_48&#45;&gt;data2_ecoli_mod_19 -->\n",
       "<g id=\"edge49\" class=\"edge\"><title>data1_ecoli_48&#45;&gt;data2_ecoli_mod_19</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M165.5,-873.894C165.5,-868.575 165.5,-862.634 165.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"167.666,-857.001 165.5,-852.001 163.334,-857.001 167.666,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_49 -->\n",
       "<g id=\"node144\" class=\"node\"><title>data1_ecoli_49</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_49</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_49&#45;&gt;data2_ecoli_mod_66 -->\n",
       "<g id=\"edge50\" class=\"edge\"><title>data1_ecoli_49&#45;&gt;data2_ecoli_mod_66</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.5\" d=\"M418.921,-363.894C413.78,-358.249 408.002,-351.904 402.627,-346.002\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.5\" points=\"403.806,-344.372 398.983,-342.001 400.894,-347.024 403.806,-344.372\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_49&#45;&gt;data2_ecoli_mod_89 -->\n",
       "<g id=\"edge51\" class=\"edge\"><title>data1_ecoli_49&#45;&gt;data2_ecoli_mod_89</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.56\" d=\"M450.079,-363.894C455.22,-358.249 460.998,-351.904 466.373,-346.002\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.56\" points=\"467.944,-346.876 470.017,-342.001 465.356,-344.52 467.944,-346.876\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_50 -->\n",
       "<g id=\"node145\" class=\"node\"><title>data1_ecoli_50</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_50</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_50&#45;&gt;data2_ecoli_mod_29 -->\n",
       "<g id=\"edge52\" class=\"edge\"><title>data1_ecoli_50&#45;&gt;data2_ecoli_mod_29</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M274.5,-771.894C274.5,-766.575 274.5,-760.634 274.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"276.674,-755.001 274.5,-750.001 272.326,-755.001 276.674,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_51 -->\n",
       "<g id=\"node146\" class=\"node\"><title>data1_ecoli_51</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_51</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_51&#45;&gt;data2_ecoli_mod_34 -->\n",
       "<g id=\"edge53\" class=\"edge\"><title>data1_ecoli_51&#45;&gt;data2_ecoli_mod_34</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M1081.5,-771.894C1081.5,-766.575 1081.5,-760.634 1081.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"1083.67,-755.001 1081.5,-750.001 1079.33,-755.001 1083.67,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_52 -->\n",
       "<g id=\"node147\" class=\"node\"><title>data1_ecoli_52</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_52</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_52&#45;&gt;data2_ecoli_mod_73 -->\n",
       "<g id=\"edge54\" class=\"edge\"><title>data1_ecoli_52&#45;&gt;data2_ecoli_mod_73</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.82\" d=\"M53.5,-261.894C53.5,-256.575 53.5,-250.634 53.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.82\" points=\"55.6088,-245.001 53.5,-240.001 51.3913,-245.001 55.6088,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_53 -->\n",
       "<g id=\"node148\" class=\"node\"><title>data1_ecoli_53</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_53</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_53&#45;&gt;data2_ecoli_mod_80 -->\n",
       "<g id=\"edge55\" class=\"edge\"><title>data1_ecoli_53&#45;&gt;data2_ecoli_mod_80</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.92\" d=\"M1081.5,-261.894C1081.5,-256.575 1081.5,-250.634 1081.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.92\" points=\"1083.65,-245.001 1081.5,-240.001 1079.35,-245.001 1083.65,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_54 -->\n",
       "<g id=\"node149\" class=\"node\"><title>data1_ecoli_54</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_54</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_54&#45;&gt;data2_ecoli_mod_47 -->\n",
       "<g id=\"edge56\" class=\"edge\"><title>data1_ecoli_54&#45;&gt;data2_ecoli_mod_47</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M274.5,-567.894C274.5,-562.575 274.5,-556.634 274.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"276.674,-551.001 274.5,-546.001 272.326,-551.001 276.674,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_55 -->\n",
       "<g id=\"node150\" class=\"node\"><title>data1_ecoli_55</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_55</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_55&#45;&gt;data2_ecoli_mod_21 -->\n",
       "<g id=\"edge57\" class=\"edge\"><title>data1_ecoli_55&#45;&gt;data2_ecoli_mod_21</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M594.5,-873.894C594.5,-868.575 594.5,-862.634 594.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"596.674,-857.001 594.5,-852.001 592.326,-857.001 596.674,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_56 -->\n",
       "<g id=\"node151\" class=\"node\"><title>data1_ecoli_56</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_56</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_56&#45;&gt;data2_ecoli_mod_63 -->\n",
       "<g id=\"edge58\" class=\"edge\"><title>data1_ecoli_56&#45;&gt;data2_ecoli_mod_63</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.88\" d=\"M53.5,-363.894C53.5,-358.575 53.5,-352.634 53.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.88\" points=\"55.6351,-347.001 53.5,-342.001 51.3651,-347.001 55.6351,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_57 -->\n",
       "<g id=\"node152\" class=\"node\"><title>data1_ecoli_57</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_57</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_57&#45;&gt;data2_ecoli_mod_59 -->\n",
       "<g id=\"edge59\" class=\"edge\"><title>data1_ecoli_57&#45;&gt;data2_ecoli_mod_59</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M703.5,-465.894C703.5,-460.575 703.5,-454.634 703.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"705.631,-449.001 703.5,-444.001 701.369,-449.001 705.631,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_58 -->\n",
       "<g id=\"node153\" class=\"node\"><title>data1_ecoli_58</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_58</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_58&#45;&gt;data2_ecoli_mod_18 -->\n",
       "<g id=\"edge60\" class=\"edge\"><title>data1_ecoli_58&#45;&gt;data2_ecoli_mod_18</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.98\" d=\"M53.5,-873.894C53.5,-868.575 53.5,-862.634 53.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.98\" points=\"55.6788,-857.001 53.5,-852.001 51.3213,-857.001 55.6788,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_59 -->\n",
       "<g id=\"node154\" class=\"node\"><title>data1_ecoli_59</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_59</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_59&#45;&gt;data2_ecoli_mod_85 -->\n",
       "<g id=\"edge61\" class=\"edge\"><title>data1_ecoli_59&#45;&gt;data2_ecoli_mod_85</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M434.5,-159.894C434.5,-154.575 434.5,-148.634 434.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"436.674,-143.001 434.5,-138.001 432.326,-143.001 436.674,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_60 -->\n",
       "<g id=\"node155\" class=\"node\"><title>data1_ecoli_60</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_60</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_60&#45;&gt;data2_ecoli_mod_75 -->\n",
       "<g id=\"edge62\" class=\"edge\"><title>data1_ecoli_60&#45;&gt;data2_ecoli_mod_75</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.91\" d=\"M274.5,-261.894C274.5,-256.575 274.5,-250.634 274.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.91\" points=\"276.648,-245.001 274.5,-240.001 272.352,-245.001 276.648,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_61 -->\n",
       "<g id=\"node156\" class=\"node\"><title>data1_ecoli_61</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_61</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_61&#45;&gt;data2_ecoli_mod_53 -->\n",
       "<g id=\"edge63\" class=\"edge\"><title>data1_ecoli_61&#45;&gt;data2_ecoli_mod_53</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.86\" d=\"M1244.5,-567.894C1244.5,-562.575 1244.5,-556.634 1244.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.86\" points=\"1246.63,-551.001 1244.5,-546.001 1242.37,-551.001 1246.63,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_62 -->\n",
       "<g id=\"node157\" class=\"node\"><title>data1_ecoli_62</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_62</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_62&#45;&gt;data2_ecoli_mod_42 -->\n",
       "<g id=\"edge64\" class=\"edge\"><title>data1_ecoli_62&#45;&gt;data2_ecoli_mod_42</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M921.5,-669.894C921.5,-664.575 921.5,-658.634 921.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"923.674,-653.001 921.5,-648.001 919.326,-653.001 923.674,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_63 -->\n",
       "<g id=\"node158\" class=\"node\"><title>data1_ecoli_63</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_63</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_63&#45;&gt;data2_ecoli_mod_25 -->\n",
       "<g id=\"edge65\" class=\"edge\"><title>data1_ecoli_63&#45;&gt;data2_ecoli_mod_25</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.32\" d=\"M1065.92,-873.894C1060.78,-868.249 1055,-861.904 1049.63,-856.002\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.32\" points=\"1050.64,-854.52 1045.98,-852.001 1048.06,-856.876 1050.64,-854.52\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_63&#45;&gt;data2_ecoli_mod_72 -->\n",
       "<g id=\"edge66\" class=\"edge\"><title>data1_ecoli_63&#45;&gt;data2_ecoli_mod_72</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.23\" d=\"M1097.08,-873.894C1102.22,-868.249 1108,-861.904 1113.37,-856.002\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.23\" points=\"1115.02,-856.944 1117.02,-852.001 1112.28,-854.452 1115.02,-856.944\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_64 -->\n",
       "<g id=\"node159\" class=\"node\"><title>data1_ecoli_64</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_64</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_64&#45;&gt;data2_ecoli_mod_4 -->\n",
       "<g id=\"edge67\" class=\"edge\"><title>data1_ecoli_64&#45;&gt;data2_ecoli_mod_4</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.96\" d=\"M594.5,-669.894C594.5,-664.575 594.5,-658.634 594.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.96\" points=\"596.67,-653.001 594.5,-648.001 592.33,-653.001 596.67,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_65 -->\n",
       "<g id=\"node160\" class=\"node\"><title>data1_ecoli_65</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_65</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_65&#45;&gt;data2_ecoli_mod_3 -->\n",
       "<g id=\"edge68\" class=\"edge\"><title>data1_ecoli_65&#45;&gt;data2_ecoli_mod_3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.82\" d=\"M434.5,-771.894C434.5,-766.575 434.5,-760.634 434.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.82\" points=\"436.609,-755.001 434.5,-750.001 432.391,-755.001 436.609,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_66 -->\n",
       "<g id=\"node161\" class=\"node\"><title>data1_ecoli_66</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_66</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_66&#45;&gt;data2_ecoli_mod_33 -->\n",
       "<g id=\"edge69\" class=\"edge\"><title>data1_ecoli_66&#45;&gt;data2_ecoli_mod_33</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M921.5,-771.894C921.5,-766.575 921.5,-760.634 921.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"923.674,-755.001 921.5,-750.001 919.326,-755.001 923.674,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_67 -->\n",
       "<g id=\"node162\" class=\"node\"><title>data1_ecoli_67</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_67</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_67&#45;&gt;data2_ecoli_mod_12 -->\n",
       "<g id=\"edge70\" class=\"edge\"><title>data1_ecoli_67&#45;&gt;data2_ecoli_mod_12</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.87\" d=\"M594.5,-975.894C594.5,-970.575 594.5,-964.634 594.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.87\" points=\"596.631,-959.001 594.5,-954.001 592.369,-959.001 596.631,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_68 -->\n",
       "<g id=\"node163\" class=\"node\"><title>data1_ecoli_68</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_68</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_68&#45;&gt;data2_ecoli_mod_43 -->\n",
       "<g id=\"edge71\" class=\"edge\"><title>data1_ecoli_68&#45;&gt;data2_ecoli_mod_43</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.99\" d=\"M1081.5,-669.894C1081.5,-664.575 1081.5,-658.634 1081.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.99\" points=\"1083.68,-653.001 1081.5,-648.001 1079.32,-653.001 1083.68,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_69 -->\n",
       "<g id=\"node164\" class=\"node\"><title>data1_ecoli_69</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_69</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_69&#45;&gt;data2_ecoli_mod_22 -->\n",
       "<g id=\"edge72\" class=\"edge\"><title>data1_ecoli_69&#45;&gt;data2_ecoli_mod_22</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M703.5,-873.894C703.5,-868.575 703.5,-862.634 703.5,-857.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"705.674,-857.001 703.5,-852.001 701.326,-857.001 705.674,-857.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_70 -->\n",
       "<g id=\"node165\" class=\"node\"><title>data1_ecoli_70</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_70</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_70&#45;&gt;data2_ecoli_mod_31 -->\n",
       "<g id=\"edge73\" class=\"edge\"><title>data1_ecoli_70&#45;&gt;data2_ecoli_mod_31</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.9\" d=\"M703.5,-771.894C703.5,-766.575 703.5,-760.634 703.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.9\" points=\"705.644,-755.001 703.5,-750.001 701.356,-755.001 705.644,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_71 -->\n",
       "<g id=\"node166\" class=\"node\"><title>data1_ecoli_71</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-43.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">ecoli_71</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_72 -->\n",
       "<g id=\"node167\" class=\"node\"><title>data1_ecoli_72</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_72</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_72&#45;&gt;data2_ecoli_mod_88 -->\n",
       "<g id=\"edge74\" class=\"edge\"><title>data1_ecoli_72&#45;&gt;data2_ecoli_mod_88</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.85\" d=\"M812.5,-159.894C812.5,-154.575 812.5,-148.634 812.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.85\" points=\"814.622,-143.001 812.5,-138.001 810.378,-143.001 814.622,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_73 -->\n",
       "<g id=\"node168\" class=\"node\"><title>data1_ecoli_73</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_73</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_73&#45;&gt;data2_ecoli_mod_45 -->\n",
       "<g id=\"edge75\" class=\"edge\"><title>data1_ecoli_73&#45;&gt;data2_ecoli_mod_45</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.91\" d=\"M53.5,-567.894C53.5,-562.575 53.5,-556.634 53.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.91\" points=\"55.6482,-551.001 53.5,-546.001 51.352,-551.001 55.6482,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_74 -->\n",
       "<g id=\"node169\" class=\"node\"><title>data1_ecoli_74</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_74</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_74&#45;&gt;data2_ecoli_mod_77 -->\n",
       "<g id=\"edge76\" class=\"edge\"><title>data1_ecoli_74&#45;&gt;data2_ecoli_mod_77</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.94\" d=\"M594.5,-261.894C594.5,-256.575 594.5,-250.634 594.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.94\" points=\"596.661,-245.001 594.5,-240.001 592.339,-245.001 596.661,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_75 -->\n",
       "<g id=\"node170\" class=\"node\"><title>data1_ecoli_75</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_75</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_75&#45;&gt;data2_ecoli_mod_57 -->\n",
       "<g id=\"edge77\" class=\"edge\"><title>data1_ecoli_75&#45;&gt;data2_ecoli_mod_57</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.85\" d=\"M434.5,-465.894C434.5,-460.575 434.5,-454.634 434.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.85\" points=\"436.622,-449.001 434.5,-444.001 432.378,-449.001 436.622,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_76 -->\n",
       "<g id=\"node171\" class=\"node\"><title>data1_ecoli_76</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_76</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_76&#45;&gt;data2_ecoli_mod_58 -->\n",
       "<g id=\"edge78\" class=\"edge\"><title>data1_ecoli_76&#45;&gt;data2_ecoli_mod_58</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.91\" d=\"M594.5,-465.894C594.5,-460.575 594.5,-454.634 594.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.91\" points=\"596.648,-449.001 594.5,-444.001 592.352,-449.001 596.648,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_77 -->\n",
       "<g id=\"node172\" class=\"node\"><title>data1_ecoli_77</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_77</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_77&#45;&gt;data2_ecoli_mod_60 -->\n",
       "<g id=\"edge79\" class=\"edge\"><title>data1_ecoli_77&#45;&gt;data2_ecoli_mod_60</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.86\" d=\"M921.5,-465.894C921.5,-460.575 921.5,-454.634 921.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.86\" points=\"923.626,-449.001 921.5,-444.001 919.374,-449.001 923.626,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_78 -->\n",
       "<g id=\"node173\" class=\"node\"><title>data1_ecoli_78</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_78</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_78&#45;&gt;data2_ecoli_mod_70 -->\n",
       "<g id=\"edge80\" class=\"edge\"><title>data1_ecoli_78&#45;&gt;data2_ecoli_mod_70</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.98\" d=\"M1081.5,-363.894C1081.5,-358.575 1081.5,-352.634 1081.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.98\" points=\"1083.68,-347.001 1081.5,-342.001 1079.32,-347.001 1083.68,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_79 -->\n",
       "<g id=\"node174\" class=\"node\"><title>data1_ecoli_79</title>\n",
       "<text text-anchor=\"middle\" x=\"921.5\" y=\"-582.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_79</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_79&#45;&gt;data2_ecoli_mod_51 -->\n",
       "<g id=\"edge81\" class=\"edge\"><title>data1_ecoli_79&#45;&gt;data2_ecoli_mod_51</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.89\" d=\"M921.5,-567.894C921.5,-562.575 921.5,-556.634 921.5,-551.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.89\" points=\"923.639,-551.001 921.5,-546.001 919.361,-551.001 923.639,-551.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_80 -->\n",
       "<g id=\"node175\" class=\"node\"><title>data1_ecoli_80</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_80</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_80&#45;&gt;data2_ecoli_mod_16 -->\n",
       "<g id=\"edge82\" class=\"edge\"><title>data1_ecoli_80&#45;&gt;data2_ecoli_mod_16</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.95\" d=\"M1081.5,-975.894C1081.5,-970.575 1081.5,-964.634 1081.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.95\" points=\"1083.67,-959.001 1081.5,-954.001 1079.33,-959.001 1083.67,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_81 -->\n",
       "<g id=\"node176\" class=\"node\"><title>data1_ecoli_81</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-684.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_81</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_81&#45;&gt;data2_ecoli_mod_41 -->\n",
       "<g id=\"edge83\" class=\"edge\"><title>data1_ecoli_81&#45;&gt;data2_ecoli_mod_41</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.78\" d=\"M812.5,-669.894C812.5,-664.575 812.5,-658.634 812.5,-653.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.78\" points=\"814.591,-653.001 812.5,-648.001 810.409,-653.001 814.591,-653.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_82 -->\n",
       "<g id=\"node177\" class=\"node\"><title>data1_ecoli_82</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-990.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_82</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_82&#45;&gt;data2_ecoli_mod_13 -->\n",
       "<g id=\"edge84\" class=\"edge\"><title>data1_ecoli_82&#45;&gt;data2_ecoli_mod_13</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.97\" d=\"M703.5,-975.894C703.5,-970.575 703.5,-964.634 703.5,-959.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.97\" points=\"705.674,-959.001 703.5,-954.001 701.326,-959.001 705.674,-959.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_83 -->\n",
       "<g id=\"node178\" class=\"node\"><title>data1_ecoli_83</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_83</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_83&#45;&gt;data2_ecoli_mod_35 -->\n",
       "<g id=\"edge85\" class=\"edge\"><title>data1_ecoli_83&#45;&gt;data2_ecoli_mod_35</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.92\" d=\"M1244.5,-771.894C1244.5,-766.575 1244.5,-760.634 1244.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.92\" points=\"1246.65,-755.001 1244.5,-750.001 1242.35,-755.001 1246.65,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_84 -->\n",
       "<g id=\"node179\" class=\"node\"><title>data1_ecoli_84</title>\n",
       "<text text-anchor=\"middle\" x=\"274.5\" y=\"-480.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_84</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_84&#45;&gt;data2_ecoli_mod_56 -->\n",
       "<g id=\"edge86\" class=\"edge\"><title>data1_ecoli_84&#45;&gt;data2_ecoli_mod_56</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.96\" d=\"M274.5,-465.894C274.5,-460.575 274.5,-454.634 274.5,-449.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.96\" points=\"276.67,-449.001 274.5,-444.001 272.33,-449.001 276.67,-449.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_85 -->\n",
       "<g id=\"node180\" class=\"node\"><title>data1_ecoli_85</title>\n",
       "<text text-anchor=\"middle\" x=\"1244.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_85</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_85&#45;&gt;data2_ecoli_mod_91 -->\n",
       "<g id=\"edge87\" class=\"edge\"><title>data1_ecoli_85&#45;&gt;data2_ecoli_mod_91</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.82\" d=\"M1244.5,-159.894C1244.5,-154.575 1244.5,-148.634 1244.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.82\" points=\"1246.61,-143.001 1244.5,-138.001 1242.39,-143.001 1246.61,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_86 -->\n",
       "<g id=\"node181\" class=\"node\"><title>data1_ecoli_86</title>\n",
       "<text text-anchor=\"middle\" x=\"703.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_86</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_86&#45;&gt;data2_ecoli_mod_68 -->\n",
       "<g id=\"edge88\" class=\"edge\"><title>data1_ecoli_86&#45;&gt;data2_ecoli_mod_68</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.75\" d=\"M703.5,-363.894C703.5,-358.575 703.5,-352.634 703.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.75\" points=\"705.578,-347.001 703.5,-342.001 701.422,-347.001 705.578,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_87 -->\n",
       "<g id=\"node182\" class=\"node\"><title>data1_ecoli_87</title>\n",
       "<text text-anchor=\"middle\" x=\"812.5\" y=\"-378.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_87</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_87&#45;&gt;data2_ecoli_mod_69 -->\n",
       "<g id=\"edge89\" class=\"edge\"><title>data1_ecoli_87&#45;&gt;data2_ecoli_mod_69</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.88\" d=\"M812.5,-363.894C812.5,-358.575 812.5,-352.634 812.5,-347.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.88\" points=\"814.635,-347.001 812.5,-342.001 810.365,-347.001 814.635,-347.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_88 -->\n",
       "<g id=\"node183\" class=\"node\"><title>data1_ecoli_88</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-43.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">ecoli_88</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_89 -->\n",
       "<g id=\"node184\" class=\"node\"><title>data1_ecoli_89</title>\n",
       "<text text-anchor=\"middle\" x=\"1081.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_89</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_89&#45;&gt;data2_ecoli_mod_90 -->\n",
       "<g id=\"edge90\" class=\"edge\"><title>data1_ecoli_89&#45;&gt;data2_ecoli_mod_90</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.63\" d=\"M1081.5,-159.894C1081.5,-154.575 1081.5,-148.634 1081.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.63\" points=\"1083.53,-143.001 1081.5,-138.001 1079.47,-143.001 1083.53,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_90 -->\n",
       "<g id=\"node185\" class=\"node\"><title>data1_ecoli_90</title>\n",
       "<text text-anchor=\"middle\" x=\"53.5\" y=\"-786.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_90</text>\n",
       "</g>\n",
       "<!-- data1_ecoli_90&#45;&gt;data2_ecoli_mod_27 -->\n",
       "<g id=\"edge91\" class=\"edge\"><title>data1_ecoli_90&#45;&gt;data2_ecoli_mod_27</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.88\" d=\"M53.5,-771.894C53.5,-766.575 53.5,-760.634 53.5,-755.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.88\" points=\"55.2501,-755.001 53.5,-750.001 51.7501,-755.001 55.2501,-755.001\"/>\n",
       "</g>\n",
       "<!-- data1_ecoli_91 -->\n",
       "<g id=\"node186\" class=\"node\"><title>data1_ecoli_91</title>\n",
       "<text text-anchor=\"middle\" x=\"594.5\" y=\"-43.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">ecoli_91</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f000c6b6750>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# You can run compare_ica between different runs for the same organism by setting \"ortho_dir\" to None. \n",
    "# Cutoff is optional, and the default for show_all is False\n",
    "dots, name_links = compare_ica(ecoli_S,ecoli_S_mod,ortho_dir=None, cutoff=.3,show_all=True)\n",
    "dots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 65/65 [00:00<00:00, 65.61it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.38.0 (20140413.2041)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"285pt\" height=\"306pt\"\n",
       " viewBox=\"0.00 0.00 285.00 306.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 302)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-302 281,-302 281,4 -4,4\"/>\n",
       "<!-- data2_ecoli_3 -->\n",
       "<g id=\"node1\" class=\"node\"><title>data2_ecoli_3</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_3</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_30 -->\n",
       "<g id=\"node2\" class=\"node\"><title>data2_ecoli_30</title>\n",
       "<text text-anchor=\"middle\" x=\"168.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_30</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_32 -->\n",
       "<g id=\"node3\" class=\"node\"><title>data2_ecoli_32</title>\n",
       "<text text-anchor=\"middle\" x=\"243.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_32</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_33 -->\n",
       "<g id=\"node4\" class=\"node\"><title>data2_ecoli_33</title>\n",
       "<text text-anchor=\"middle\" x=\"63.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_33</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_49 -->\n",
       "<g id=\"node5\" class=\"node\"><title>data2_ecoli_49</title>\n",
       "<text text-anchor=\"middle\" x=\"168.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_49</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_5 -->\n",
       "<g id=\"node6\" class=\"node\"><title>data2_ecoli_5</title>\n",
       "<text text-anchor=\"middle\" x=\"243.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_5</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_51 -->\n",
       "<g id=\"node7\" class=\"node\"><title>data2_ecoli_51</title>\n",
       "<text text-anchor=\"middle\" x=\"93.5\" y=\"-218.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_51</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_54 -->\n",
       "<g id=\"node8\" class=\"node\"><title>data2_ecoli_54</title>\n",
       "<text text-anchor=\"middle\" x=\"63.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_54</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_60 -->\n",
       "<g id=\"node9\" class=\"node\"><title>data2_ecoli_60</title>\n",
       "<text text-anchor=\"middle\" x=\"168.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_60</text>\n",
       "</g>\n",
       "<!-- data2_ecoli_8 -->\n",
       "<g id=\"node10\" class=\"node\"><title>data2_ecoli_8</title>\n",
       "<text text-anchor=\"middle\" x=\"243.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ecoli_8</text>\n",
       "</g>\n",
       "<!-- data1_mtb_0 -->\n",
       "<g id=\"node11\" class=\"node\"><title>data1_mtb_0</title>\n",
       "<text text-anchor=\"middle\" x=\"243.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_0</text>\n",
       "</g>\n",
       "<!-- data1_mtb_0&#45;&gt;data2_ecoli_5 -->\n",
       "<g id=\"edge1\" class=\"edge\"><title>data1_mtb_0&#45;&gt;data2_ecoli_5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.93\" d=\"M243.5,-159.894C243.5,-154.575 243.5,-148.634 243.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.93\" points=\"245.25,-143.001 243.5,-138.001 241.75,-143.001 245.25,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_12 -->\n",
       "<g id=\"node12\" class=\"node\"><title>data1_mtb_12</title>\n",
       "<text text-anchor=\"middle\" x=\"63.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_12</text>\n",
       "</g>\n",
       "<!-- data1_mtb_12&#45;&gt;data2_ecoli_33 -->\n",
       "<g id=\"edge2\" class=\"edge\"><title>data1_mtb_12&#45;&gt;data2_ecoli_33</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.18\" d=\"M63.5,-159.894C63.5,-154.575 63.5,-148.634 63.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.18\" points=\"65.3288,-143.001 63.5,-138.001 61.6713,-143.001 65.3288,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_15 -->\n",
       "<g id=\"node13\" class=\"node\"><title>data1_mtb_15</title>\n",
       "<text text-anchor=\"middle\" x=\"168.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_15</text>\n",
       "</g>\n",
       "<!-- data1_mtb_15&#45;&gt;data2_ecoli_49 -->\n",
       "<g id=\"edge3\" class=\"edge\"><title>data1_mtb_15&#45;&gt;data2_ecoli_49</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.47\" d=\"M168.5,-159.894C168.5,-154.575 168.5,-148.634 168.5,-143.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.47\" points=\"170.25,-143.001 168.5,-138.001 166.75,-143.001 170.25,-143.001\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_18 -->\n",
       "<g id=\"node14\" class=\"node\"><title>data1_mtb_18</title>\n",
       "<text text-anchor=\"middle\" x=\"243.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_18</text>\n",
       "</g>\n",
       "<!-- data1_mtb_18&#45;&gt;data2_ecoli_32 -->\n",
       "<g id=\"edge4\" class=\"edge\"><title>data1_mtb_18&#45;&gt;data2_ecoli_32</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.61\" d=\"M243.5,-261.894C243.5,-256.575 243.5,-250.634 243.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.61\" points=\"245.25,-245.001 243.5,-240.001 241.75,-245.001 245.25,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_20 -->\n",
       "<g id=\"node15\" class=\"node\"><title>data1_mtb_20</title>\n",
       "<text text-anchor=\"middle\" x=\"168.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_20</text>\n",
       "</g>\n",
       "<!-- data1_mtb_20&#45;&gt;data2_ecoli_30 -->\n",
       "<g id=\"edge5\" class=\"edge\"><title>data1_mtb_20&#45;&gt;data2_ecoli_30</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.7\" d=\"M168.5,-261.894C168.5,-256.575 168.5,-250.634 168.5,-245.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.7\" points=\"170.25,-245.001 168.5,-240.001 166.75,-245.001 170.25,-245.001\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_26 -->\n",
       "<g id=\"node16\" class=\"node\"><title>data1_mtb_26</title>\n",
       "<text text-anchor=\"middle\" x=\"61.5\" y=\"-276.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_26</text>\n",
       "</g>\n",
       "<!-- data1_mtb_26&#45;&gt;data2_ecoli_3 -->\n",
       "<g id=\"edge6\" class=\"edge\"><title>data1_mtb_26&#45;&gt;data2_ecoli_3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.41\" d=\"M51.7251,-261.894C48.5614,-256.357 45.013,-250.148 41.6962,-244.343\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.41\" points=\"43.2152,-243.474 39.2151,-240.001 40.1764,-245.211 43.2152,-243.474\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_26&#45;&gt;data2_ecoli_51 -->\n",
       "<g id=\"edge7\" class=\"edge\"><title>data1_mtb_26&#45;&gt;data2_ecoli_51</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.08\" d=\"M71.2749,-261.894C74.4386,-256.357 77.987,-250.148 81.3038,-244.343\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.08\" points=\"82.8236,-245.211 83.7849,-240.001 79.7848,-243.474 82.8236,-245.211\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_45 -->\n",
       "<g id=\"node17\" class=\"node\"><title>data1_mtb_45</title>\n",
       "<text text-anchor=\"middle\" x=\"243.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_45</text>\n",
       "</g>\n",
       "<!-- data1_mtb_45&#45;&gt;data2_ecoli_8 -->\n",
       "<g id=\"edge8\" class=\"edge\"><title>data1_mtb_45&#45;&gt;data2_ecoli_8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.43\" d=\"M243.5,-57.8939C243.5,-52.5746 243.5,-46.6338 243.5,-41.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.43\" points=\"245.25,-41.0013 243.5,-36.0014 241.75,-41.0014 245.25,-41.0013\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_55 -->\n",
       "<g id=\"node18\" class=\"node\"><title>data1_mtb_55</title>\n",
       "<text text-anchor=\"middle\" x=\"168.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_55</text>\n",
       "</g>\n",
       "<!-- data1_mtb_55&#45;&gt;data2_ecoli_60 -->\n",
       "<g id=\"edge9\" class=\"edge\"><title>data1_mtb_55&#45;&gt;data2_ecoli_60</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.21\" d=\"M168.5,-57.8939C168.5,-52.5746 168.5,-46.6338 168.5,-41.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.21\" points=\"170.25,-41.0013 168.5,-36.0014 166.75,-41.0014 170.25,-41.0013\"/>\n",
       "</g>\n",
       "<!-- data1_mtb_61 -->\n",
       "<g id=\"node19\" class=\"node\"><title>data1_mtb_61</title>\n",
       "<text text-anchor=\"middle\" x=\"63.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">mtb_61</text>\n",
       "</g>\n",
       "<!-- data1_mtb_61&#45;&gt;data2_ecoli_54 -->\n",
       "<g id=\"edge10\" class=\"edge\"><title>data1_mtb_61&#45;&gt;data2_ecoli_54</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.42\" d=\"M63.5,-57.8939C63.5,-52.5746 63.5,-46.6338 63.5,-41.028\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.42\" points=\"65.2501,-41.0013 63.5,-36.0014 61.7501,-41.0014 65.2501,-41.0013\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f003cc3fb90>"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If you want to compare ICA runs between different organism, provide a string to the path of the ortholog\n",
    "# folder in the modulome github repository\n",
    "dots, name_links = compare_ica(mtb_S,ecoli_S,ortho_dir=\"/home/sbrg/Desktop/modulome/data/ortholog/\",\n",
    "                               metric='pearson')\n",
    "dots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('mtb_0', 'ecoli_5'),\n",
       " ('mtb_12', 'ecoli_33'),\n",
       " ('mtb_15', 'ecoli_49'),\n",
       " ('mtb_18', 'ecoli_32'),\n",
       " ('mtb_20', 'ecoli_30'),\n",
       " ('mtb_26', 'ecoli_3'),\n",
       " ('mtb_26', 'ecoli_51'),\n",
       " ('mtb_45', 'ecoli_8'),\n",
       " ('mtb_55', 'ecoli_60'),\n",
       " ('mtb_61', 'ecoli_54')]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
