{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparing iModulons\n",
    "\n",
    "`compare_ica` is a function specifically built to compare the components between different ICA runs. The function will return a Digraph plot which shows connections between ICA components of different runs, as well as Dictionary List of the one-to-one connections.\n",
    "\n",
    "The function requires a minimum of two inputs if running a comparison beween two ICA runs for the same organism:\n",
    "\n",
    "* S1: Pandas Dataframe of M matrix 1\n",
    "* S2: Pandas Dataframe of M Matrix 2\n",
    "\n",
    "`compare_ica` can be found in **pymodulon.compare**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon import example_data\n",
    "from pymodulon.compare import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparsion within the same organism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_data = example_data.load_ecoli_data()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will create a copy of the *E. coli* M matrix that has slightly different iModulon names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_M = ecoli_data.M.copy()\n",
    "new_M.columns = ['run2_'+str(i) for i in new_M.columns]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this example, we will only compare the first 5 iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "links,dots = compare_ica(ecoli_data.M.iloc[:,:5],new_M.iloc[:,:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The comparison data are stored in ``links``. The first two values in each element of ``links`` are the iModulon name from the first and second datasets respectively. The final value is the Pearson R correlation between the Independent Component gene weights."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('AllR/AraC/FucR', 'run2_AllR/AraC/FucR', 1.0),\n",
       " ('ArcA-1', 'run2_ArcA-1', 1.0),\n",
       " ('ArcA-2', 'run2_ArcA-2', 1.0),\n",
       " ('ArgR', 'run2_ArgR', 1.0),\n",
       " ('AtoC', 'run2_AtoC', 1.0)]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The graphical comparison is saved as ``dots``. The thickness of the arrows represents how similar the iModulons in the first dataset are to the dataset in the second dataset. Since we are comparing two copies of the same dataset, all arrows are at maximum thickness."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"251pt\" height=\"208pt\"\n",
       " viewBox=\"0.00 0.00 250.50 208.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 204)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-204 246.5,-204 246.5,4 -4,4\"/>\n",
       "<!-- data2_run2_AllR/AraC/FucR -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_run2_AllR/AraC/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-124.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_AllR&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-109.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">/AraC/FucR</text>\n",
       "</g>\n",
       "<!-- data2_run2_ArcA&#45;1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_run2_ArcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"135.5\" y=\"-125.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"135.5\" y=\"-110.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_run2_ArcA&#45;2 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_run2_ArcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-125.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-110.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_run2_ArgR -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_run2_ArgR</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_ArgR</text>\n",
       "</g>\n",
       "<!-- data2_run2_AtoC -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_run2_AtoC</title>\n",
       "<text text-anchor=\"middle\" x=\"135.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_AtoC</text>\n",
       "</g>\n",
       "<!-- data1_AllR/AraC/FucR -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data1_AllR/AraC/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-184.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AllR/Ar&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-169.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">aC/FucR</text>\n",
       "</g>\n",
       "<!-- data1_AllR/AraC/FucR&#45;&gt;data2_run2_AllR/AraC/FucR -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_AllR/AraC/FucR&#45;&gt;data2_run2_AllR/AraC/FucR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"5\" d=\"M43.5,-161.6854C43.5,-156.5467 43.5,-150.9269 43.5,-145.549\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"5\" points=\"45.6876,-145.2784 43.5,-140.2785 41.3126,-145.2785 45.6876,-145.2784\"/>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;1 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data1_ArcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"135.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;1&#45;&gt;data2_run2_ArcA&#45;1 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_ArcA&#45;1&#45;&gt;data2_run2_ArcA&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"5\" d=\"M135.5,-162.9775C135.5,-157.6845 135.5,-151.8076 135.5,-146.1955\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"5\" points=\"137.6876,-146.1403 135.5,-141.1403 133.3126,-146.1404 137.6876,-146.1403\"/>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;2 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data1_ArcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;2&#45;&gt;data2_run2_ArcA&#45;2 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_ArcA&#45;2&#45;&gt;data2_run2_ArcA&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"5\" d=\"M213.5,-162.9775C213.5,-157.6845 213.5,-151.8076 213.5,-146.1955\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"5\" points=\"215.6876,-146.1403 213.5,-141.1403 211.3126,-146.1404 215.6876,-146.1403\"/>\n",
       "</g>\n",
       "<!-- data1_ArgR -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data1_ArgR</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArgR</text>\n",
       "</g>\n",
       "<!-- data1_ArgR&#45;&gt;data2_run2_ArgR -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_ArgR&#45;&gt;data2_run2_ArgR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"5\" d=\"M43.5,-57.9664C43.5,-52.6707 43.5,-46.8028 43.5,-41.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"5\" points=\"45.6876,-41.2222 43.5,-36.2222 41.3126,-41.2223 45.6876,-41.2222\"/>\n",
       "</g>\n",
       "<!-- data1_AtoC -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data1_AtoC</title>\n",
       "<text text-anchor=\"middle\" x=\"135.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AtoC</text>\n",
       "</g>\n",
       "<!-- data1_AtoC&#45;&gt;data2_run2_AtoC -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_AtoC&#45;&gt;data2_run2_AtoC</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"5\" d=\"M135.5,-57.9664C135.5,-52.6707 135.5,-46.8028 135.5,-41.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"5\" points=\"137.6876,-41.2222 135.5,-36.2222 133.3126,-41.2223 137.6876,-41.2222\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f4b7af8a4f0>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison across organisms\n",
    "`compare_ica` can also be used to compare ICA components between different\n",
    "organisms. To do so, you will need to provide an addition input, which is a\n",
    "string path to a Bidirectional Best Hit (BBH) orthology between your two organisms of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>subject</th>\n",
       "      <th>PID</th>\n",
       "      <th>alnLength</th>\n",
       "      <th>mismatchCount</th>\n",
       "      <th>gapOpenCount</th>\n",
       "      <th>queryStart</th>\n",
       "      <th>queryEnd</th>\n",
       "      <th>subjectStart</th>\n",
       "      <th>subjectEnd</th>\n",
       "      <th>eVal</th>\n",
       "      <th>bitScore</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>COV</th>\n",
       "      <th>BBH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0003</td>\n",
       "      <td>USA300HOU_RS06760</td>\n",
       "      <td>26.897</td>\n",
       "      <td>290</td>\n",
       "      <td>183</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>279</td>\n",
       "      <td>4</td>\n",
       "      <td>275</td>\n",
       "      <td>2.060000e-25</td>\n",
       "      <td>99.8</td>\n",
       "      <td>310</td>\n",
       "      <td>0.935484</td>\n",
       "      <td>&lt;=&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>b0007</td>\n",
       "      <td>USA300HOU_RS06905</td>\n",
       "      <td>40.638</td>\n",
       "      <td>470</td>\n",
       "      <td>257</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>463</td>\n",
       "      <td>3.650000e-114</td>\n",
       "      <td>342.0</td>\n",
       "      <td>476</td>\n",
       "      <td>0.987395</td>\n",
       "      <td>&lt;=&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>b0014</td>\n",
       "      <td>USA300HOU_RS08425</td>\n",
       "      <td>55.818</td>\n",
       "      <td>636</td>\n",
       "      <td>251</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>636</td>\n",
       "      <td>1</td>\n",
       "      <td>606</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>678.0</td>\n",
       "      <td>638</td>\n",
       "      <td>0.996865</td>\n",
       "      <td>&lt;=&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>b0015</td>\n",
       "      <td>USA300HOU_RS08420</td>\n",
       "      <td>46.982</td>\n",
       "      <td>381</td>\n",
       "      <td>187</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>371</td>\n",
       "      <td>1</td>\n",
       "      <td>376</td>\n",
       "      <td>8.040000e-110</td>\n",
       "      <td>323.0</td>\n",
       "      <td>376</td>\n",
       "      <td>1.013298</td>\n",
       "      <td>&lt;=&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>b0023</td>\n",
       "      <td>USA300HOU_RS08455</td>\n",
       "      <td>42.353</td>\n",
       "      <td>85</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>79</td>\n",
       "      <td>1</td>\n",
       "      <td>78</td>\n",
       "      <td>1.530000e-09</td>\n",
       "      <td>46.6</td>\n",
       "      <td>87</td>\n",
       "      <td>0.977011</td>\n",
       "      <td>&lt;=&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene            subject     PID  alnLength  mismatchCount  gapOpenCount  \\\n",
       "3   b0003  USA300HOU_RS06760  26.897        290            183             8   \n",
       "6   b0007  USA300HOU_RS06905  40.638        470            257             6   \n",
       "9   b0014  USA300HOU_RS08425  55.818        636            251             4   \n",
       "10  b0015  USA300HOU_RS08420  46.982        381            187             5   \n",
       "15  b0023  USA300HOU_RS08455  42.353         85             36             3   \n",
       "\n",
       "    queryStart  queryEnd  subjectStart  subjectEnd           eVal  bitScore  \\\n",
       "3            1       279             4         275   2.060000e-25      99.8   \n",
       "6            1       455             1         463  3.650000e-114     342.0   \n",
       "9            1       636             1         606   0.000000e+00     678.0   \n",
       "10           1       371             1         376  8.040000e-110     323.0   \n",
       "15           1        79             1          78   1.530000e-09      46.6   \n",
       "\n",
       "    gene_length       COV  BBH  \n",
       "3           310  0.935484  <=>  \n",
       "6           476  0.987395  <=>  \n",
       "9           638  0.996865  <=>  \n",
       "10          376  1.013298  <=>  \n",
       "15           87  0.977011  <=>  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_bbh = example_data.load_example_bbh()\n",
    "example_bbh.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also provide a path to your own orthology file, as long as it contains the\n",
    "columns ``gene`` and ``target``\n",
    "\n",
    "It is important that the order of the organism matches the order of the orthology\n",
    "file. Specifically, the organism for the ``M1`` file must be the organism that\n",
    "fills the ``gene`` column in the orthology file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "staph_data = example_data.load_staph_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"631pt\" height=\"312pt\"\n",
       " viewBox=\"0.00 0.00 631.00 312.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 308)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-308 627,-308 627,4 -4,4\"/>\n",
       "<!-- data2_Autolysin -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_Autolysin</title>\n",
       "<text text-anchor=\"middle\" x=\"54\" y=\"-224.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Autolysin</text>\n",
       "</g>\n",
       "<!-- data2_CcpA&#45;1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_CcpA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"181\" y=\"-224.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CcpA&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_CodY -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_CodY</title>\n",
       "<text text-anchor=\"middle\" x=\"369\" y=\"-224.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CodY</text>\n",
       "</g>\n",
       "<!-- data2_CymR -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_CymR</title>\n",
       "<text text-anchor=\"middle\" x=\"562\" y=\"-224.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CymR</text>\n",
       "</g>\n",
       "<!-- data2_Fur -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_Fur</title>\n",
       "<text text-anchor=\"middle\" x=\"54\" y=\"-121.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur</text>\n",
       "</g>\n",
       "<!-- data2_LacR -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data2_LacR</title>\n",
       "<text text-anchor=\"middle\" x=\"182\" y=\"-121.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">LacR</text>\n",
       "</g>\n",
       "<!-- data2_NreC/Rex -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data2_NreC/Rex</title>\n",
       "<text text-anchor=\"middle\" x=\"364\" y=\"-120.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NreC/Rex</text>\n",
       "</g>\n",
       "<!-- data2_PerR -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data2_PerR</title>\n",
       "<text text-anchor=\"middle\" x=\"556\" y=\"-121.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PerR</text>\n",
       "</g>\n",
       "<!-- data2_PurR -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data2_PurR</title>\n",
       "<text text-anchor=\"middle\" x=\"54\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR</text>\n",
       "</g>\n",
       "<!-- data2_PyrR -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data2_PyrR</title>\n",
       "<text text-anchor=\"middle\" x=\"185\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PyrR</text>\n",
       "</g>\n",
       "<!-- data2_Translation -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>data2_Translation</title>\n",
       "<text text-anchor=\"middle\" x=\"369\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"369\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data2_Vim&#45;3 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>data2_Vim&#45;3</title>\n",
       "<text text-anchor=\"middle\" x=\"556\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Vim&#45;3</text>\n",
       "</g>\n",
       "<!-- data1_AllR/AraC/FucR -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>data1_AllR/AraC/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"152\" y=\"-80.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AllR/Ar&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"152\" y=\"-65.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">aC/FucR</text>\n",
       "</g>\n",
       "<!-- data1_AllR/AraC/FucR&#45;&gt;data2_PyrR -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_AllR/AraC/FucR&#45;&gt;data2_PyrR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.31\" d=\"M162.8054,-57.6812C165.8012,-52.325 169.0825,-46.4585 172.1813,-40.9183\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.31\" points=\"173.8175,-41.5778 174.731,-36.3597 170.7629,-39.8692 173.8175,-41.5778\"/>\n",
       "</g>\n",
       "<!-- data1_Cbl+CysB -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>data1_Cbl+CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"529\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Cbl+CysB</text>\n",
       "</g>\n",
       "<!-- data1_Cbl+CysB&#45;&gt;data2_CymR -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_Cbl+CysB&#45;&gt;data2_CymR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.35\" d=\"M539.2605,-267.9664C542.3351,-262.5626 545.7486,-256.5631 548.9762,-250.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.35\" points=\"550.6805,-251.4335 551.6322,-246.2222 547.6385,-249.7026 550.6805,-251.4335\"/>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>data1_Crp&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"154\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Crp&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1&#45;&gt;data2_CcpA&#45;1 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_Crp&#45;1&#45;&gt;data2_CcpA&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.29\" d=\"M162.395,-267.9664C164.9105,-262.5626 167.7034,-256.5631 170.3442,-250.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.29\" points=\"171.9936,-251.4937 172.5172,-246.2222 168.8205,-250.0166 171.9936,-251.4937\"/>\n",
       "</g>\n",
       "<!-- data1_CsqR -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>data1_CsqR</title>\n",
       "<text text-anchor=\"middle\" x=\"182\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CsqR</text>\n",
       "</g>\n",
       "<!-- data1_CsqR&#45;&gt;data2_LacR -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_CsqR&#45;&gt;data2_LacR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.09\" d=\"M182,-164.9664C182,-159.6707 182,-153.8028 182,-148.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.09\" points=\"183.7501,-148.2222 182,-143.2222 180.2501,-148.2223 183.7501,-148.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CysB -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>data1_CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"596\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CysB</text>\n",
       "</g>\n",
       "<!-- data1_CysB&#45;&gt;data2_CymR -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_CysB&#45;&gt;data2_CymR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.55\" d=\"M585.4286,-267.9664C582.1975,-262.4545 578.6031,-256.3229 575.2191,-250.5503\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.55\" points=\"576.7204,-249.6507 572.682,-246.2222 573.7009,-251.4208 576.7204,-249.6507\"/>\n",
       "</g>\n",
       "<!-- data1_EvgA -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>data1_EvgA</title>\n",
       "<text text-anchor=\"middle\" x=\"282\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">EvgA</text>\n",
       "</g>\n",
       "<!-- data1_EvgA&#45;&gt;data2_NreC/Rex -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>data1_EvgA&#45;&gt;data2_NreC/Rex</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.95\" d=\"M307.0483,-164.9775C315.7101,-158.7452 325.4972,-151.7032 334.4837,-145.2373\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.95\" points=\"335.786,-146.4562 338.8226,-142.1155 333.7418,-143.6152 335.786,-146.4562\"/>\n",
       "</g>\n",
       "<!-- data1_Fnr -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>data1_Fnr</title>\n",
       "<text text-anchor=\"middle\" x=\"337\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fnr</text>\n",
       "</g>\n",
       "<!-- data1_Fnr&#45;&gt;data2_NreC/Rex -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>data1_Fnr&#45;&gt;data2_NreC/Rex</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.42\" d=\"M345.2476,-164.9775C347.8161,-159.3647 350.6852,-153.0952 353.3916,-147.1813\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.42\" points=\"355.1272,-147.5941 355.6166,-142.3193 351.9446,-146.1377 355.1272,-147.5941\"/>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>data1_Fur&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"27\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2&#45;&gt;data2_Fur -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>data1_Fur&#45;2&#45;&gt;data2_Fur</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.39\" d=\"M35.395,-164.9664C37.9105,-159.5626 40.7034,-153.5631 43.3442,-147.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.39\" points=\"44.9936,-148.4937 45.5172,-143.2222 41.8205,-147.0166 44.9936,-148.4937\"/>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>data1_Leu/Ile</title>\n",
       "<text text-anchor=\"middle\" x=\"369\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Leu/Ile</text>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile&#45;&gt;data2_CodY -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>data1_Leu/Ile&#45;&gt;data2_CodY</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.24\" d=\"M369,-267.9664C369,-262.6707 369,-256.8028 369,-251.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.24\" points=\"370.7501,-251.2222 369,-246.2222 367.2501,-251.2223 370.7501,-251.2222\"/>\n",
       "</g>\n",
       "<!-- data1_MalT -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>data1_MalT</title>\n",
       "<text text-anchor=\"middle\" x=\"209\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">MalT</text>\n",
       "</g>\n",
       "<!-- data1_MalT&#45;&gt;data2_CcpA&#45;1 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>data1_MalT&#45;&gt;data2_CcpA&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.27\" d=\"M200.2941,-267.9664C197.6854,-262.5626 194.7891,-256.5631 192.0505,-250.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.27\" points=\"193.5467,-249.9642 189.7969,-246.2222 190.3948,-251.4858 193.5467,-249.9642\"/>\n",
       "</g>\n",
       "<!-- data1_NarL -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>data1_NarL</title>\n",
       "<text text-anchor=\"middle\" x=\"392\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NarL</text>\n",
       "</g>\n",
       "<!-- data1_NarL&#45;&gt;data2_NreC/Rex -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>data1_NarL&#45;&gt;data2_NreC/Rex</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.74\" d=\"M383.4469,-164.9775C380.7833,-159.3647 377.8079,-153.0952 375.0013,-147.1813\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.74\" points=\"376.4187,-146.0862 372.6939,-142.3193 373.2567,-147.5868 376.4187,-146.0862\"/>\n",
       "</g>\n",
       "<!-- data1_NikR -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>data1_NikR</title>\n",
       "<text text-anchor=\"middle\" x=\"82\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NikR</text>\n",
       "</g>\n",
       "<!-- data1_NikR&#45;&gt;data2_Fur -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>data1_NikR&#45;&gt;data2_Fur</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.63\" d=\"M73.2941,-164.9664C70.6854,-159.5626 67.7891,-153.5631 65.0505,-147.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.63\" points=\"66.5467,-146.9642 62.7969,-143.2222 63.3948,-148.4858 66.5467,-146.9642\"/>\n",
       "</g>\n",
       "<!-- data1_OxyR -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>data1_OxyR</title>\n",
       "<text text-anchor=\"middle\" x=\"556\" y=\"-179.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">OxyR</text>\n",
       "</g>\n",
       "<!-- data1_OxyR&#45;&gt;data2_PerR -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>data1_OxyR&#45;&gt;data2_PerR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.09\" d=\"M556,-164.9664C556,-159.6707 556,-153.8028 556,-148.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.09\" points=\"557.7501,-148.2222 556,-143.2222 554.2501,-148.2223 557.7501,-148.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1 -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>data1_PurR&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"54\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1&#45;&gt;data2_PurR -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>data1_PurR&#45;1&#45;&gt;data2_PurR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"3.76\" d=\"M54,-58.9664C54,-53.6707 54,-47.8028 54,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"3.76\" points=\"55.7501,-42.2222 54,-37.2222 52.2501,-42.2223 55.7501,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;2 -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>data1_PurR&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"218\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;2&#45;&gt;data2_PyrR -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>data1_PurR&#45;2&#45;&gt;data2_PyrR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.93\" d=\"M207.9196,-58.9775C204.7175,-53.2525 201.1332,-46.8442 197.7676,-40.8269\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.93\" points=\"199.2146,-39.8289 195.2464,-36.3193 196.1599,-41.5374 199.2146,-39.8289\"/>\n",
       "</g>\n",
       "<!-- data1_RpoH -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>data1_RpoH</title>\n",
       "<text text-anchor=\"middle\" x=\"54\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RpoH</text>\n",
       "</g>\n",
       "<!-- data1_RpoH&#45;&gt;data2_Autolysin -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>data1_RpoH&#45;&gt;data2_Autolysin</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.67\" d=\"M54,-267.9664C54,-262.6707 54,-256.8028 54,-251.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.67\" points=\"55.7501,-251.2222 54,-246.2222 52.2501,-251.2223 55.7501,-251.2222\"/>\n",
       "</g>\n",
       "<!-- data1_nitrate&#45;related -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>data1_nitrate&#45;related</title>\n",
       "<text text-anchor=\"middle\" x=\"451\" y=\"-186.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">nitrate&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"451\" y=\"-171.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">&#45;related</text>\n",
       "</g>\n",
       "<!-- data1_nitrate&#45;related&#45;&gt;data2_NreC/Rex -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>data1_nitrate&#45;related&#45;&gt;data2_NreC/Rex</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.44\" d=\"M422.5129,-163.6812C413.7399,-157.7317 404.0383,-151.1524 395.0991,-145.0902\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.44\" points=\"395.8998,-143.5187 390.7794,-142.1607 393.9353,-146.4155 395.8998,-143.5187\"/>\n",
       "</g>\n",
       "<!-- data1_proVWX -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>data1_proVWX</title>\n",
       "<text text-anchor=\"middle\" x=\"556\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">proVWX</text>\n",
       "</g>\n",
       "<!-- data1_proVWX&#45;&gt;data2_Vim&#45;3 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>data1_proVWX&#45;&gt;data2_Vim&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.52\" d=\"M556,-58.9664C556,-53.6707 556,-47.8028 556,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.52\" points=\"557.7501,-42.2222 556,-37.2222 554.2501,-42.2223 557.7501,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_translation -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>data1_translation</title>\n",
       "<text text-anchor=\"middle\" x=\"369\" y=\"-82.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"369\" y=\"-67.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data1_translation&#45;&gt;data2_Translation -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>data1_translation&#45;&gt;data2_Translation</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"3.21\" d=\"M369,-59.6854C369,-54.5467 369,-48.9269 369,-43.549\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"3.21\" points=\"370.7501,-43.2784 369,-38.2785 367.2501,-43.2785 370.7501,-43.2784\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f4b78709160>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links,dots = compare_ica(ecoli_data.M,staph_data.M,\n",
    "                         ortho_file = example_bbh)\n",
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here it is much more clear that some iModulons are more similar between species, whereas others are loosely connected. Arrows can also indicate if an iModulon is composed of multiple iModulons from another dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('AllR/AraC/FucR', 'PyrR', 0.26151978205478876),\n",
       " ('Cbl+CysB', 'CymR', 0.2694541420808297),\n",
       " ('Crp-1', 'CcpA-1', 0.257653219615526),\n",
       " ('CsqR', 'LacR', 0.41725112391151326),\n",
       " ('CysB', 'CymR', 0.31057785046755065)]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, you can adjust the metric type and cutoff values for correlations\n",
    "by utilizing the ``method`` and ``cutoff`` parameters. The default for ``method`` is\n",
    " ``\"pearson\"``, but can be changed to ``\"spearman\"`.\n",
    "\n",
    "``cutoff`` takes a float value between 0 and 1 and will only\n",
    "show correlations greater than the ``cutoff`` value (default: 0.3)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"132pt\" height=\"106pt\"\n",
       " viewBox=\"0.00 0.00 131.50 106.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 102)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-102 127.5,-102 127.5,4 -4,4\"/>\n",
       "<!-- data2_PurR -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_PurR</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR</text>\n",
       "</g>\n",
       "<!-- data2_Translation -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_Translation</title>\n",
       "<text text-anchor=\"middle\" x=\"95.5\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"95.5\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data1_PurR&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1&#45;&gt;data2_PurR -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_PurR&#45;1&#45;&gt;data2_PurR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"3.76\" d=\"M29.5,-58.9664C29.5,-53.6707 29.5,-47.8028 29.5,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"3.76\" points=\"31.2501,-42.2222 29.5,-37.2222 27.7501,-42.2223 31.2501,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_translation -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data1_translation</title>\n",
       "<text text-anchor=\"middle\" x=\"95.5\" y=\"-82.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"95.5\" y=\"-67.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data1_translation&#45;&gt;data2_Translation -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_translation&#45;&gt;data2_Translation</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"3.21\" d=\"M95.5,-59.6854C95.5,-54.5467 95.5,-48.9269 95.5,-43.549\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"3.21\" points=\"97.2501,-43.2784 95.5,-38.2785 93.7501,-43.2785 97.2501,-43.2784\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f4b4ad9be50>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links,dots = compare_ica(ecoli_data.M,staph_data.M,\n",
    "                         ortho_file = example_bbh,\n",
    "                         cutoff = 0.6)\n",
    "dots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"230pt\" height=\"208pt\"\n",
       " viewBox=\"0.00 0.00 230.50 208.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 204)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-204 226.5,-204 226.5,4 -4,4\"/>\n",
       "<!-- data2_CcpA&#45;2 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_CcpA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"38.5\" y=\"-117.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CcpA&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_CodY -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_CodY</title>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CodY</text>\n",
       "</g>\n",
       "<!-- data2_NreC/Rex -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_NreC/Rex</title>\n",
       "<text text-anchor=\"middle\" x=\"66.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NreC/Rex</text>\n",
       "</g>\n",
       "<!-- data2_Translation -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_Translation</title>\n",
       "<text text-anchor=\"middle\" x=\"101.5\" y=\"-124.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"101.5\" y=\"-109.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data2_Unc&#45;1 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_Unc&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Unc&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data1_Crp&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Crp&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1&#45;&gt;data2_CodY -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_Crp&#45;1&#45;&gt;data2_CodY</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\".83\" d=\"M181.5,-161.9664C181.5,-156.6707 181.5,-150.8028 181.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\".83\" points=\"183.2501,-145.2222 181.5,-140.2222 179.7501,-145.2223 183.2501,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GntR/TyrR -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data1_GntR/TyrR</title>\n",
       "<text text-anchor=\"middle\" x=\"181.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GntR/TyrR</text>\n",
       "</g>\n",
       "<!-- data1_GntR/TyrR&#45;&gt;data2_Unc&#45;1 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_GntR/TyrR&#45;&gt;data2_Unc&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\".8\" d=\"M181.5,-57.9664C181.5,-52.6707 181.5,-46.8028 181.5,-41.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\".8\" points=\"183.2501,-41.2222 181.5,-36.2222 179.7501,-41.2223 183.2501,-41.2222\"/>\n",
       "</g>\n",
       "<!-- data1_His&#45;tRNA -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data1_His&#45;tRNA</title>\n",
       "<text text-anchor=\"middle\" x=\"66.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">His&#45;tRNA</text>\n",
       "</g>\n",
       "<!-- data1_His&#45;tRNA&#45;&gt;data2_NreC/Rex -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_His&#45;tRNA&#45;&gt;data2_NreC/Rex</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\".78\" d=\"M66.5,-57.9664C66.5,-52.6707 66.5,-46.8028 66.5,-41.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\".78\" points=\"68.2501,-41.2222 66.5,-36.2222 64.7501,-41.2223 68.2501,-41.2222\"/>\n",
       "</g>\n",
       "<!-- data1_translation -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data1_translation</title>\n",
       "<text text-anchor=\"middle\" x=\"105.5\" y=\"-184.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"105.5\" y=\"-169.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data1_translation&#45;&gt;data2_Translation -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_translation&#45;&gt;data2_Translation</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.46\" d=\"M104.2124,-161.6854C103.8698,-156.5467 103.4951,-150.9269 103.1366,-145.549\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.46\" points=\"104.864,-145.1509 102.7852,-140.2785 101.3718,-145.3838 104.864,-145.1509\"/>\n",
       "</g>\n",
       "<!-- data1_uncharacterized&#45;5 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data1_uncharacterized&#45;5</title>\n",
       "<text text-anchor=\"middle\" x=\"38.5\" y=\"-184.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"38.5\" y=\"-169.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;5</text>\n",
       "</g>\n",
       "<!-- data1_uncharacterized&#45;5&#45;&gt;data2_CcpA&#45;2 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>data1_uncharacterized&#45;5&#45;&gt;data2_CcpA&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\".78\" d=\"M38.5,-161.6854C38.5,-156.1249 38.5,-150.0011 38.5,-144.23\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\".78\" points=\"40.2501,-144.0553 38.5,-139.0554 36.7501,-144.0554 40.2501,-144.0553\"/>\n",
       "</g>\n",
       "<!-- data1_uncharacterized&#45;5&#45;&gt;data2_Translation -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_uncharacterized&#45;5&#45;&gt;data2_Translation</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\".95\" d=\"M58.7804,-161.6854C64.738,-156.0114 71.3115,-149.7509 77.4791,-143.8771\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\".95\" points=\"78.8438,-144.994 81.2576,-140.2785 76.43,-142.4595 78.8438,-144.994\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f4b4ad8cc10>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "links,dots = compare_ica(ecoli_data.M,staph_data.M,\n",
    "                         ortho_file = example_bbh,\n",
    "                         cutoff = 0.15, method = 'spearman')\n",
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a BBH orthology file\n",
    "To create your own BBH orthology file, pymodulon.compare contains a set of functions that will allow you to create the desired bbh.csv file for you specific comparison. This only needs to be done once per comparison.\n",
    "\n",
    "In order to make the BBH orthology file, you need to obtain the Genbank Full Genome files for your organisms. These can be found at https://www.ncbi.nlm.nih.gov/genome/. Search for both of your organisms/strains and download the \"Genbank (full)\" file. Once saved into a known location, you can use the ``make_prots`` function to convert the Genbank files into protein FASTA files.\n",
    "\n",
    "``make_prots`` recieves two parameters:\n",
    "\n",
    "* ``gbk``: Path to the genbank file (one file per function call)\n",
    "* ``out_path``: Path to the protein FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ecoli_gb= \"../pymodulon/test/data/ecoli.gb\"\n",
    "# mtb_gb = \"../pymodulon/test/data/mtb.gb\"\n",
    "# ecoli_out = \"../pymodulon/test/data/ecoli_prot.fasta\"\n",
    "# mtb_out = \"../pymodulon/test/data/mtb_prot.fasta\"\n",
    "\n",
    "# make_prots(ecoli_gb,ecoli_out)\n",
    "# make_prots(mtb_gb,mtb_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Genbank database files\n",
    "In order to run Bidirectional Best Hits between two organisms, it is required to build the necessary Genbank Database files from the newly created protein FASTA files. In order to do so, use the `make_prots_db` function. You will do this for both organims.\n",
    "\n",
    "`make_prots_db` recieves one parameter:\n",
    "\n",
    "* `fasta_file`: String path to protein FASTA file\n",
    "\n",
    "If there is an error, the function will print out error message. The function will also not execut if the necessary Genbank DB files already exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_prot_db(ecoli_out)\n",
    "# make_prot_db(mtb_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create BBH CSV\n",
    "Once the previous steps have been completed, you can now create the BBH CSV file. In order to do so, you can use the `get_bbh` function. \n",
    "\n",
    "`get_bbh` has the following parameters:\n",
    "\n",
    "* `db1`: String path to protein FASTA file (output of make_prots function) for organism 1\n",
    "* `db2`: String path to protein FASTA file (output of make_prots function) for organism 2\n",
    "* `outdir`: String path to output directory, default is \"bbh\" and will create the directory if it does not exist\n",
    "* `outname`: Default db1_vs_db2_parsed.csv where db[1-2] are the passed arguments name of the csv file where that will save the results\n",
    "* `mincov`: Minimum coverage to call hits in BLAST, must be between 0 and 1\n",
    "* `evalue`: evalue thershold for BLAST hits, Default .001\n",
    "* `threads`: Number of threads to run BLAST, Default 1\n",
    "* `force`: Whether to overwrite existing files or not\n",
    "* `savefiles`: Whether to save files to outdir\n",
    "\n",
    "The function will return a Pandas DataFrame containing all the BLAST hits between the two genes. The function will also save the DataFrame to a csv file, which you can then use in `compare_ica` between the two organisms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_bbh(ecoli_out,mtb_out, outdir = \"../pymodulon/test/data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
