{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iModulonDB Tutorial\n",
    "\n",
    "This notebook will help you add information and generate \n",
    "\n",
    "Follow the advice [here](https://github.com/SBRG/modulytics/wiki/Adding-a-Private-Project) to get started, and learn to use pymodulon with the initial Tutorial if needed.\n",
    "\n",
    "Please contact krychel@eng.ucsd.edu with any questions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.io import *\n",
    "from pymodulon.imodulondb import *\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Load Your Data\n",
    "\n",
    "More details on this can be found in the main tutorial.\n",
    "\n",
    "All that is absolutely required is the following:\n",
    "\n",
    "- M matrix (part of initial ICA output)\n",
    "- A matrix (part of initial ICA output)\n",
    "- X matrix (log_tpm.csv) -- initial data, preferably not normalized to the base condition.\n",
    "- Sample_table containing:\n",
    "    - Either a column labeled 'sample' or an index (leftmost column) that serves the purpose of sample labels. These should be unique to each sample (no duplicates in the entire table), and it must match the columns of A and X.\n",
    "    - A column labeled 'project' -- this is the largest grouping of samples; they may be downloaded from the same original paper or contain a set of samples/experiments with a common theme. The contents of this column will appear on the X axis when zoomed all the way out in the activity bar graphs on iModulonDB.\n",
    "    - A column labeled 'condition' -- this groups samples such that all biological replicates have the same condition name. Samples in the same project with matching conditions will be grouped into a single bar in the activity bar graphs. If you'd like to re-use condition names across different projects (for example, giving each project a 'control' condition), that is okay.\n",
    "    \n",
    "Additional information will only improve your version of the site!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = load_json_model('../example_data/example.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Fill out some metadata\n",
    "\n",
    "When you load an ica_data model, it fills out default metadata for you. Therefore you can technically skip this step.\n",
    "\n",
    "### 2.a: Splash Table (Dictionary)\n",
    "\n",
    "This table provides the basic elements shown on [the splash page](http://imodulondb.org) and the names of the directories in which all data will be stored.\n",
    "\n",
    "First, look at the default one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'large_title': 'New Dataset',\n",
       " 'subtitle': 'Unpublished study',\n",
       " 'author': 'Pymodulon User',\n",
       " 'organism_folder': 'new_org',\n",
       " 'dataset_folder': 'new_dataset'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.splash_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first three entries are shown on the splash page directly. The last two are important folder names. Typical organism folder names are in the form 'g_species'. Choose a name for your dataset that will be unique; it is likely the only one for your organism, so you could choose 'modulome' if that is the project you are working on.\n",
    "\n",
    "Let's update this table with your information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'large_title': '<i>M. favoritespeciesus</i>',\n",
       " 'subtitle': 'Modulome Dataset',\n",
       " 'author': 'Rychel, et al.',\n",
       " 'organism_folder': 'm_favoritespeciesus',\n",
       " 'dataset_folder': 'modulome'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FILL IN THESE DETAILS\n",
    "\n",
    "# <i> tags italicize the species name\n",
    "# So far on the site, large titles are all simply species names.\n",
    "ica_data.splash_table['large_title'] = '<i>M. favoritespeciesus</i>'\n",
    "\n",
    "# subtitle provides additional information about the dataset's origins\n",
    "ica_data.splash_table['subtitle'] = 'Modulome Dataset'\n",
    "ica_data.splash_table['author'] = 'Rychel, et al.'\n",
    "\n",
    "# these will correspond to actual folders, which will be generated for you.\n",
    "ica_data.splash_table['organism_folder'] = 'm_favoritespeciesus'\n",
    "ica_data.splash_table['dataset_folder'] = 'modulome'\n",
    "\n",
    "# make sure the output is desired\n",
    "ica_data.splash_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.b: Dataset Table (Dictionary)\n",
    "\n",
    "This table provides the basic elements shown in the [dataset page](https://imodulondb.org/dataset.html?organism=e_coli&dataset=precise1). The numerical elements are computed for you in the default.\n",
    "\n",
    "If desired, you can add your own entries to this with whatever names you'd like. They can even include links if you obey the [rules for making HTML links](https://www.w3schools.com/html/html_links.asp). For example, the published datasets include a 'Publication' entry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Title                             New Dataset\n",
       "Organism                         New Organism\n",
       "Strain                         Unknown Strain\n",
       "Number of Samples                         278\n",
       "Number of Unique Conditions               163\n",
       "Number of Genes                          3923\n",
       "Number of iModulons                        92\n",
       "dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.dataset_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Title                            M. favoritespeciesus Modulome\n",
       "Organism                           <i>M. favoritespeciesus</i>\n",
       "Strain                                                IMDB1000\n",
       "Number of Samples                                          278\n",
       "Number of Unique Conditions                                163\n",
       "Number of Genes                                           3923\n",
       "Number of iModulons                                         92\n",
       "Useful Additional Link         <a href=\"google.com\">Google</a>\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FILL IN THESE DETAILS, TOO\n",
    "\n",
    "# This title can combine some information from the splash table\n",
    "ica_data.dataset_table['Title'] = 'M. favoritespeciesus Modulome'\n",
    "ica_data.dataset_table['Organism'] = '<i>M. favoritespeciesus</i>'\n",
    "ica_data.dataset_table['Strain'] = 'IMDB1000'\n",
    "\n",
    "# additional rows if desired\n",
    "ica_data.dataset_table['Useful Additional Link'] = '<a href=\"google.com\">Google</a>'\n",
    "\n",
    "# check output\n",
    "ica_data.dataset_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Add Links (Optional)\n",
    "\n",
    "If your organism has a database like EcoCyc, SubtiWiki, or AureoWiki, you may be interested in adding links to it.\n",
    "\n",
    "iModulonDB has two kinds of links:\n",
    "\n",
    "- gene_links: \n",
    "    - Associate gene **loci** from the **index of ica_data.M** to websites\n",
    "    - On the [gene pages](https://imodulondb.org/gene.html?organism=e_coli&dataset=precise1&gene_id=b0002), these links have the name of the database that you are linking to. The name of the database needs to be saved as a string in ica_data.link_database.\n",
    "    - For programming reasons, the default value of gene_links is all genes as keys, each pointing to the value np.nan.\n",
    "    \n",
    "- tf_links:\n",
    "    - Associate **TF names** from **ica_data.TRN.regulator and ica_data.imodulon_table.regulator** to websites\n",
    "    - On the [iModulon pages](https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1&k=43), the name of the regulator becomes a link. You never have to state the name of this database, and it doesn't have to be the same one used for gene_links (i.e. ica_data.link_database does NOT have to be the one you use).\n",
    "    - For programming reasons, the default value is an empty dictionary.\n",
    "    \n",
    "Note that gene_links uses gene loci since not every gene has a readable name, but tf_links uses readable names since not all regulators have associated gene loci. This could create some confusion between the two types of links; for example, the gene b2741 in E. coli encodes the TF rpoS:\n",
    "\n",
    "- ica_data.gene_links\\['b2741'\\] = 'https://ecocyc.org/gene?orgid=ECOLI&id=EG10510 '\n",
    "- ica_data.tf_links\\['rpoS'\\] = 'http://regulondb.ccg.unam.mx/sigmulon?term=ECK125110294&organism=ECK12&format=jsp&type=sigmulon '\n",
    "\n",
    "In this example, ica_data.link_database = 'EcoCyc' since that is where the gene links point to.\n",
    "\n",
    "You are free to skip this step, or you can use any of the following strategies to generate your links:\n",
    "\n",
    "#### Strategy 1: Look at the URLs in your database.\n",
    "\n",
    "Search for genes in your database. You might notice that it always uses the same URL, but enters your search term in part of that URL. You could then take the pieces of that URL and use those to generate all needed URLs. Below is an example from psuedomonas aeruginosa:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'https://www.pseudomonas.com/primarySequenceFeature/list?strain_ids=107&term=Pseudomonas+aeruginosa+PAO1+%28Reference%29&c1=name&v1='\n",
    "suffix = '&e1=1&assembly=complete'\n",
    "\n",
    "temp_gene_links = dict()\n",
    "for gene in ica_data.M.index:\n",
    "    temp_gene_links[gene] = prefix + gene + suffix\n",
    "    \n",
    "temp_tf_links = dict()\n",
    "for gene in ica_data.trn.regulator.unique():\n",
    "    temp_tf_links[gene] = prefix + gene + suffix\n",
    "    \n",
    "# Then, test a few of the links and uncomment these lines\n",
    "\n",
    "#ica_data.gene_links = temp_gene_links\n",
    "#ica_data.tf_links = temp_tf_links"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may need to add a suffix in addition to a prefix. You may also have to deal with databases that use different loci abbreviations than the ones you have - this can be remedied by finding a mapping file between the two naming systems.\n",
    "\n",
    "#### Strategy 2: Make a file\n",
    "\n",
    "If you are more proficient in excel than python, have found a gene <-> link mapping file elsewhere, or already used strategy 1 to generate a file, you can make sure it has only two columns (gene and link) and no headers, then simply read it in as shown below.\n",
    "\n",
    "This may be especially useful if you have no database, but want to keep a file of gene_links or tf_links with specific papers instead of db entries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data.gene_links = 'gene_links.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy 3: Individual modifications\n",
    "\n",
    "You can also just tell the model the links one at a time. This can be useful if you generate all links with another strategy, but want to overide it in a few cases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data.gene_links['b2741'] = 'https://ecocyc.org/gene?orgid=ECOLI&id=EG10510'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### If using gene links, don't forget to update ica_data.link_database\n",
    "\n",
    "If you don't, the default value is 'External Database'. If you have no links, nothing will appear so this is irrelevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'External Database'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ica_data.link_database = 'EcoCyc'\n",
    "\n",
    "ica_data.link_database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4: Check for compatibility\n",
    "\n",
    "Though the code cannot check for all edge cases and concerns, you can use the function 'imodulondb_compatibility()' to see what information you are missing, mostly through a check of the columns of your DataFrames. \n",
    "\n",
    "#### 4.a: Modify column names\n",
    "\n",
    "First, let's see what we are missing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iModulon Table is missing a name column\n",
      "iModulon Table is missing a TF column\n",
      "Note that TF is used for coding purposes, so it uses '+' and '/' and must match the trn.\n",
      "iModulon Table is missing a Regulator column\n",
      "Note that Regulator is displayed in the dataset page so it can have spelled out 'and' and 'or' operators.If you have a TF column, a Regulator column will begenerated for you (but not vice versa).\n",
      "iModulon Table is missing a n_genes column\n",
      "\n",
      "Gene Table is missing the gene_product column\n",
      "Gene Table is missing the optional COG column\n",
      "\n",
      "Sample Table is missing the optional sample column\n",
      "The 0th column will be used to define the names of samples in the activity bar graph unless you add this column.\n"
     ]
    }
   ],
   "source": [
    "imodulondb_compatibility(ica_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's see if we actually DO have the information that the function believes is missing -- it is likely that some of the names of the columns are simply not what iModulonDB is expecting. If so, we simply rename it to what is expected.\n",
    "\n",
    "If you actually DO NOT have some of the information, that is okay -- after this step, we'll fill in the blanks with default values as needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FecI</th>\n",
       "      <td>fecI</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>Ferric citrate transport</td>\n",
       "      <td>Metal Homeostasis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iron-related</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>Related to iron metabolism</td>\n",
       "      <td>Biological Enrichment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RbsR</th>\n",
       "      <td>rbsR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>D-ribose catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgrT</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>Contains single dominating gene: sgrT</td>\n",
       "      <td>Biological Enrichment</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insertion</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>IS2 insertion element after laboratory evolution</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             regulator  pvalue  qvalue  precision    recall  f1score   TP  \\\n",
       "YieP              None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "GlpR              glpR     0.0     NaN   1.000000  1.000000      4.0  9.0   \n",
       "YgbI              None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "efeU-repair       None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "ydcI-KO           None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "...                ...     ...     ...        ...       ...      ...  ...   \n",
       "FecI              fecI     0.0     NaN   0.500000  1.000000      6.0  5.0   \n",
       "iron-related      None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "RbsR              rbsR     0.0     NaN   0.833333  0.555556      6.0  5.0   \n",
       "sgrT              None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "insertion         None     NaN     NaN        NaN       NaN      NaN  NaN   \n",
       "\n",
       "              regulon_size  n_regs  \\\n",
       "YieP                    11       1   \n",
       "GlpR                     9       1   \n",
       "YgbI                     9       1   \n",
       "efeU-repair              8       0   \n",
       "ydcI-KO                  3       0   \n",
       "...                    ...     ...   \n",
       "FecI                    10       1   \n",
       "iron-related            20       0   \n",
       "RbsR                     6       1   \n",
       "sgrT                     4       0   \n",
       "insertion                9       0   \n",
       "\n",
       "                                                       Function  \\\n",
       "YieP                                                    Unknown   \n",
       "GlpR                                        Glycerol catabolism   \n",
       "YgbI                                           Unknown Function   \n",
       "efeU-repair   Accounts for repair and expression of efeU operon   \n",
       "ydcI-KO                             Accounts for ydcI knock-out   \n",
       "...                                                         ...   \n",
       "FecI                                   Ferric citrate transport   \n",
       "iron-related                         Related to iron metabolism   \n",
       "RbsR                                        D-ribose catabolism   \n",
       "sgrT                      Contains single dominating gene: sgrT   \n",
       "insertion      IS2 insertion element after laboratory evolution   \n",
       "\n",
       "                               Category  single_gene  \n",
       "YieP                Regulator Discovery          NaN  \n",
       "GlpR          Carbon Source Utilization          NaN  \n",
       "YgbI                Regulator Discovery          NaN  \n",
       "efeU-repair         Genomic Alterations          NaN  \n",
       "ydcI-KO             Genomic Alterations          1.0  \n",
       "...                                 ...          ...  \n",
       "FecI                  Metal Homeostasis          NaN  \n",
       "iron-related      Biological Enrichment          NaN  \n",
       "RbsR          Carbon Source Utilization          NaN  \n",
       "sgrT              Biological Enrichment          1.0  \n",
       "insertion           Genomic Alterations          NaN  \n",
       "\n",
       "[92 rows x 12 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We do have regulators. Rename that column to TFs\n",
    "# We also have n_genes. Rename that column to n_genes\n",
    "rename_dict = {'regulator': 'TF', # this is more important than the 'Regulator' column\n",
    "               'regulon_size': 'n_genes'} \n",
    "ica_data.imodulon_table = ica_data.imodulon_table.rename(rename_dict, axis = 1)\n",
    "\n",
    "# We also have names. They were in the index, but the code wants them in the main dataframe as well.\n",
    "# If you would rather not add this column, the code will do so automatically.\n",
    "ica_data.imodulon_table['name'] = ica_data.imodulon_table.index\n",
    "\n",
    "# We don't have a separate Regulator column. Don't worry about that!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>cog</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>336</td>\n",
       "      <td>2799</td>\n",
       "      <td>+</td>\n",
       "      <td>thrA</td>\n",
       "      <td>2463</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fnr,ile-tRNA,lrp,rpoD,thr-tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>2800</td>\n",
       "      <td>3733</td>\n",
       "      <td>+</td>\n",
       "      <td>thrB</td>\n",
       "      <td>933</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fnr,ile-tRNA,lrp,rpoD,thr-tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>3733</td>\n",
       "      <td>5020</td>\n",
       "      <td>+</td>\n",
       "      <td>thrC</td>\n",
       "      <td>1287</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fnr,ile-tRNA,lrp,rpoD,thr-tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>5233</td>\n",
       "      <td>5530</td>\n",
       "      <td>+</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>297</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>5682</td>\n",
       "      <td>6459</td>\n",
       "      <td>-</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>777</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4688</th>\n",
       "      <td>290428</td>\n",
       "      <td>290633</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgS</td>\n",
       "      <td>205</td>\n",
       "      <td>ykgS</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4693</th>\n",
       "      <td>1204403</td>\n",
       "      <td>1205537</td>\n",
       "      <td>+</td>\n",
       "      <td>aaaE</td>\n",
       "      <td>1134</td>\n",
       "      <td>ymfTLMNR-beeE-jayE-ymfQ-ycfK-tfaP</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_1</th>\n",
       "      <td>1594126</td>\n",
       "      <td>1597987</td>\n",
       "      <td>-</td>\n",
       "      <td>yneO_1</td>\n",
       "      <td>3861</td>\n",
       "      <td>ydeK</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4696_2</th>\n",
       "      <td>1592664</td>\n",
       "      <td>1594125</td>\n",
       "      <td>-</td>\n",
       "      <td>yneO_2</td>\n",
       "      <td>1461</td>\n",
       "      <td>ydeK</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4705</th>\n",
       "      <td>852868</td>\n",
       "      <td>852997</td>\n",
       "      <td>-</td>\n",
       "      <td>mntS</td>\n",
       "      <td>129</td>\n",
       "      <td>rybA-mntS</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>mntR,nac,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3923 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           start     stop strand gene_name  length  \\\n",
       "b0002        336     2799      +      thrA    2463   \n",
       "b0003       2800     3733      +      thrB     933   \n",
       "b0004       3733     5020      +      thrC    1287   \n",
       "b0005       5233     5530      +      yaaX     297   \n",
       "b0006       5682     6459      -      yaaA     777   \n",
       "...          ...      ...    ...       ...     ...   \n",
       "b4688     290428   290633      +      ykgS     205   \n",
       "b4693    1204403  1205537      +      aaaE    1134   \n",
       "b4696_1  1594126  1597987      -    yneO_1    3861   \n",
       "b4696_2  1592664  1594125      -    yneO_2    1461   \n",
       "b4705     852868   852997      -      mntS     129   \n",
       "\n",
       "                                    operon  \\\n",
       "b0002                              thrLABC   \n",
       "b0003                              thrLABC   \n",
       "b0004                              thrLABC   \n",
       "b0005                                 yaaX   \n",
       "b0006                                 yaaA   \n",
       "...                                    ...   \n",
       "b4688                                 ykgS   \n",
       "b4693    ymfTLMNR-beeE-jayE-ymfQ-ycfK-tfaP   \n",
       "b4696_1                               ydeK   \n",
       "b4696_2                               ydeK   \n",
       "b4705                            rybA-mntS   \n",
       "\n",
       "                                         cog  \\\n",
       "b0002    Amino acid transport and metabolism   \n",
       "b0003    Amino acid transport and metabolism   \n",
       "b0004    Amino acid transport and metabolism   \n",
       "b0005                       Function unknown   \n",
       "b0006                       Function unknown   \n",
       "...                                      ...   \n",
       "b4688                      No COG Annotation   \n",
       "b4693                      No COG Annotation   \n",
       "b4696_1                    No COG Annotation   \n",
       "b4696_2                    No COG Annotation   \n",
       "b4705                      No COG Annotation   \n",
       "\n",
       "                                       regulator  \n",
       "b0002    arcA,crp,fnr,ile-tRNA,lrp,rpoD,thr-tRNA  \n",
       "b0003    arcA,crp,fnr,ile-tRNA,lrp,rpoD,thr-tRNA  \n",
       "b0004    arcA,crp,fnr,ile-tRNA,lrp,rpoD,thr-tRNA  \n",
       "b0005                                        NaN  \n",
       "b0006                                        NaN  \n",
       "...                                          ...  \n",
       "b4688                                        NaN  \n",
       "b4693                                        NaN  \n",
       "b4696_1                                      NaN  \n",
       "b4696_2                                      NaN  \n",
       "b4705                              mntR,nac,rpoD  \n",
       "\n",
       "[3923 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we have the \"missing\" cog column:\n",
    "rename_dict = {'cog': 'COG'}\n",
    "ica_data.gene_table = ica_data.gene_table.rename(rename_dict, axis = 1)\n",
    "\n",
    "# we don't have the gene_product column. This is a nice-to-have, but you can ignore it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "      <th>rep_id</th>\n",
       "      <th>contact</th>\n",
       "      <th>library creator</th>\n",
       "      <th>run_date</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>BAM</th>\n",
       "      <th>alignment</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "      <th>Public</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>control__wt_glc__1</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>04/17/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>control/MG1655-M9-glucose-1_S1_L001_R1_001.fas...</td>\n",
       "      <td>control/MG1655-M9-glucose-1_S1_L001_R2_001.fas...</td>\n",
       "      <td>control__wt_glc__1.bam</td>\n",
       "      <td>94.33</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control__wt_glc__2</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>2</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>04/17/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>control/MG1655-M9-glucose-2_S2_L001_R1_001.fas...</td>\n",
       "      <td>control/MG1655-M9-glucose-2_S2_L001_R2_001.fas...</td>\n",
       "      <td>control__wt_glc__2.bam</td>\n",
       "      <td>94.24</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>10/02/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WTDPD1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WTDPD1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_dpd__1.bam</td>\n",
       "      <td>98.04</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>2</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>10/02/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WTDPD2_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WTDPD2_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_dpd__2.bam</td>\n",
       "      <td>98.30</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>09/05/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WT-Fe2-1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WT-Fe2-1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_fe__1.bam</td>\n",
       "      <td>93.35</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU__menFentCubiC_ale36__2</th>\n",
       "      <td>Pseudogene Repair</td>\n",
       "      <td>efeU</td>\n",
       "      <td>menFentCubiC_ale36</td>\n",
       "      <td>2</td>\n",
       "      <td>Amitesh Anand</td>\n",
       "      <td>Richard Szubin</td>\n",
       "      <td>05/21/18</td>\n",
       "      <td>Escherichia coli EEP_menF_entC_ubiC.36.86.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>HiSeq 4000</td>\n",
       "      <td>Also used in quinone study</td>\n",
       "      <td>2</td>\n",
       "      <td>efeU/SRR8218074_1.fastq.gz</td>\n",
       "      <td>efeU/SRR8218074_2.fastq.gz</td>\n",
       "      <td>efeU__menFentCubiC_ale36__2.bam</td>\n",
       "      <td>93.44</td>\n",
       "      <td>doi.org/10.1038/s41564-018-0340-2</td>\n",
       "      <td>GSE122779</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU__menFentCubiC_ale37__1</th>\n",
       "      <td>Pseudogene Repair</td>\n",
       "      <td>efeU</td>\n",
       "      <td>menFentCubiC_ale37</td>\n",
       "      <td>1</td>\n",
       "      <td>Amitesh Anand</td>\n",
       "      <td>Richard Szubin</td>\n",
       "      <td>05/21/18</td>\n",
       "      <td>Escherichia coli EEP_menF_entC_ubiC.37.25.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>HiSeq 4000</td>\n",
       "      <td>Also used in quinone study</td>\n",
       "      <td>2</td>\n",
       "      <td>efeU/SRR8218075_1.fastq.gz</td>\n",
       "      <td>efeU/SRR8218075_2.fastq.gz</td>\n",
       "      <td>efeU__menFentCubiC_ale37__1.bam</td>\n",
       "      <td>94.89</td>\n",
       "      <td>doi.org/10.1038/s41564-018-0340-2</td>\n",
       "      <td>GSE122779</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU__menFentCubiC_ale37__2</th>\n",
       "      <td>Pseudogene Repair</td>\n",
       "      <td>efeU</td>\n",
       "      <td>menFentCubiC_ale37</td>\n",
       "      <td>2</td>\n",
       "      <td>Amitesh Anand</td>\n",
       "      <td>Richard Szubin</td>\n",
       "      <td>05/21/18</td>\n",
       "      <td>Escherichia coli EEP_menF_entC_ubiC.37.25.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>HiSeq 4000</td>\n",
       "      <td>Also used in quinone study</td>\n",
       "      <td>2</td>\n",
       "      <td>efeU/SRR8218076_1.fastq.gz</td>\n",
       "      <td>efeU/SRR8218076_2.fastq.gz</td>\n",
       "      <td>efeU__menFentCubiC_ale37__2.bam</td>\n",
       "      <td>94.62</td>\n",
       "      <td>doi.org/10.1038/s41564-018-0340-2</td>\n",
       "      <td>GSE122779</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU__menFentCubiC_ale38__1</th>\n",
       "      <td>Pseudogene Repair</td>\n",
       "      <td>efeU</td>\n",
       "      <td>menFentCubiC_ale38</td>\n",
       "      <td>1</td>\n",
       "      <td>Amitesh Anand</td>\n",
       "      <td>Richard Szubin</td>\n",
       "      <td>05/21/18</td>\n",
       "      <td>Escherichia coli EEP_menF_entC_ubiC.38.28.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>HiSeq 4000</td>\n",
       "      <td>Also used in quinone study</td>\n",
       "      <td>2</td>\n",
       "      <td>efeU/SRR8218077_1.fastq.gz</td>\n",
       "      <td>efeU/SRR8218077_2.fastq.gz</td>\n",
       "      <td>efeU__menFentCubiC_ale38__1.bam</td>\n",
       "      <td>95.65</td>\n",
       "      <td>doi.org/10.1038/s41564-018-0340-2</td>\n",
       "      <td>GSE122779</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU__menFentCubiC_ale38__2</th>\n",
       "      <td>Pseudogene Repair</td>\n",
       "      <td>efeU</td>\n",
       "      <td>menFentCubiC_ale38</td>\n",
       "      <td>2</td>\n",
       "      <td>Amitesh Anand</td>\n",
       "      <td>Richard Szubin</td>\n",
       "      <td>05/21/18</td>\n",
       "      <td>Escherichia coli EEP_menF_entC_ubiC.38.28.1</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>HiSeq 4000</td>\n",
       "      <td>Also used in quinone study</td>\n",
       "      <td>2</td>\n",
       "      <td>efeU/SRR8218078_1.fastq.gz</td>\n",
       "      <td>efeU/SRR8218078_2.fastq.gz</td>\n",
       "      <td>efeU__menFentCubiC_ale38__2.bam</td>\n",
       "      <td>94.94</td>\n",
       "      <td>doi.org/10.1038/s41564-018-0340-2</td>\n",
       "      <td>GSE122779</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>278 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         study  project           condition  \\\n",
       "control__wt_glc__1                     Control  control              wt_glc   \n",
       "control__wt_glc__2                     Control  control              wt_glc   \n",
       "fur__wt_dpd__1                             Fur      fur              wt_dpd   \n",
       "fur__wt_dpd__2                             Fur      fur              wt_dpd   \n",
       "fur__wt_fe__1                              Fur      fur               wt_fe   \n",
       "...                                        ...      ...                 ...   \n",
       "efeU__menFentCubiC_ale36__2  Pseudogene Repair     efeU  menFentCubiC_ale36   \n",
       "efeU__menFentCubiC_ale37__1  Pseudogene Repair     efeU  menFentCubiC_ale37   \n",
       "efeU__menFentCubiC_ale37__2  Pseudogene Repair     efeU  menFentCubiC_ale37   \n",
       "efeU__menFentCubiC_ale38__1  Pseudogene Repair     efeU  menFentCubiC_ale38   \n",
       "efeU__menFentCubiC_ale38__2  Pseudogene Repair     efeU  menFentCubiC_ale38   \n",
       "\n",
       "                             rep_id        contact library creator  run_date  \\\n",
       "control__wt_glc__1                1    Sangwoo Seo     Sangwoo Seo  04/17/13   \n",
       "control__wt_glc__2                2    Sangwoo Seo     Sangwoo Seo  04/17/13   \n",
       "fur__wt_dpd__1                    1    Sangwoo Seo     Sangwoo Seo  10/02/13   \n",
       "fur__wt_dpd__2                    2    Sangwoo Seo     Sangwoo Seo  10/02/13   \n",
       "fur__wt_fe__1                     1    Sangwoo Seo     Sangwoo Seo  09/05/13   \n",
       "...                             ...            ...             ...       ...   \n",
       "efeU__menFentCubiC_ale36__2       2  Amitesh Anand  Richard Szubin  05/21/18   \n",
       "efeU__menFentCubiC_ale37__1       1  Amitesh Anand  Richard Szubin  05/21/18   \n",
       "efeU__menFentCubiC_ale37__2       2  Amitesh Anand  Richard Szubin  05/21/18   \n",
       "efeU__menFentCubiC_ale38__1       1  Amitesh Anand  Richard Szubin  05/21/18   \n",
       "efeU__menFentCubiC_ale38__2       2  Amitesh Anand  Richard Szubin  05/21/18   \n",
       "\n",
       "                                                      Strain Description  \\\n",
       "control__wt_glc__1                          Escherichia coli K-12 MG1655   \n",
       "control__wt_glc__2                          Escherichia coli K-12 MG1655   \n",
       "fur__wt_dpd__1                              Escherichia coli K-12 MG1655   \n",
       "fur__wt_dpd__2                              Escherichia coli K-12 MG1655   \n",
       "fur__wt_fe__1                               Escherichia coli K-12 MG1655   \n",
       "...                                                                  ...   \n",
       "efeU__menFentCubiC_ale36__2  Escherichia coli EEP_menF_entC_ubiC.36.86.1   \n",
       "efeU__menFentCubiC_ale37__1  Escherichia coli EEP_menF_entC_ubiC.37.25.1   \n",
       "efeU__menFentCubiC_ale37__2  Escherichia coli EEP_menF_entC_ubiC.37.25.1   \n",
       "efeU__menFentCubiC_ale38__1  Escherichia coli EEP_menF_entC_ubiC.38.28.1   \n",
       "efeU__menFentCubiC_ale38__2  Escherichia coli EEP_menF_entC_ubiC.38.28.1   \n",
       "\n",
       "                             Strain Base Media  ... Sequencing Machine  \\\n",
       "control__wt_glc__1           MG1655         M9  ...              MiSeq   \n",
       "control__wt_glc__2           MG1655         M9  ...              MiSeq   \n",
       "fur__wt_dpd__1               MG1655         M9  ...              MiSeq   \n",
       "fur__wt_dpd__2               MG1655         M9  ...              MiSeq   \n",
       "fur__wt_fe__1                MG1655         M9  ...              MiSeq   \n",
       "...                             ...        ...  ...                ...   \n",
       "efeU__menFentCubiC_ale36__2  MG1655         M9  ...         HiSeq 4000   \n",
       "efeU__menFentCubiC_ale37__1  MG1655         M9  ...         HiSeq 4000   \n",
       "efeU__menFentCubiC_ale37__2  MG1655         M9  ...         HiSeq 4000   \n",
       "efeU__menFentCubiC_ale38__1  MG1655         M9  ...         HiSeq 4000   \n",
       "efeU__menFentCubiC_ale38__2  MG1655         M9  ...         HiSeq 4000   \n",
       "\n",
       "                                     Additional Details Biological Replicates  \\\n",
       "control__wt_glc__1                                 None                     2   \n",
       "control__wt_glc__2                                 None                     2   \n",
       "fur__wt_dpd__1                                     None                     2   \n",
       "fur__wt_dpd__2                                     None                     2   \n",
       "fur__wt_fe__1                                      None                     2   \n",
       "...                                                 ...                   ...   \n",
       "efeU__menFentCubiC_ale36__2  Also used in quinone study                     2   \n",
       "efeU__menFentCubiC_ale37__1  Also used in quinone study                     2   \n",
       "efeU__menFentCubiC_ale37__2  Also used in quinone study                     2   \n",
       "efeU__menFentCubiC_ale38__1  Also used in quinone study                     2   \n",
       "efeU__menFentCubiC_ale38__2  Also used in quinone study                     2   \n",
       "\n",
       "                                                                            R1  \\\n",
       "control__wt_glc__1           control/MG1655-M9-glucose-1_S1_L001_R1_001.fas...   \n",
       "control__wt_glc__2           control/MG1655-M9-glucose-2_S2_L001_R1_001.fas...   \n",
       "fur__wt_dpd__1                              fur/WTDPD1_S1_L001_R1_001.fastq.gz   \n",
       "fur__wt_dpd__2                              fur/WTDPD2_S1_L001_R1_001.fastq.gz   \n",
       "fur__wt_fe__1                             fur/WT-Fe2-1_S1_L001_R1_001.fastq.gz   \n",
       "...                                                                        ...   \n",
       "efeU__menFentCubiC_ale36__2                         efeU/SRR8218074_1.fastq.gz   \n",
       "efeU__menFentCubiC_ale37__1                         efeU/SRR8218075_1.fastq.gz   \n",
       "efeU__menFentCubiC_ale37__2                         efeU/SRR8218076_1.fastq.gz   \n",
       "efeU__menFentCubiC_ale38__1                         efeU/SRR8218077_1.fastq.gz   \n",
       "efeU__menFentCubiC_ale38__2                         efeU/SRR8218078_1.fastq.gz   \n",
       "\n",
       "                                                                            R2  \\\n",
       "control__wt_glc__1           control/MG1655-M9-glucose-1_S1_L001_R2_001.fas...   \n",
       "control__wt_glc__2           control/MG1655-M9-glucose-2_S2_L001_R2_001.fas...   \n",
       "fur__wt_dpd__1                              fur/WTDPD1_S1_L001_R2_001.fastq.gz   \n",
       "fur__wt_dpd__2                              fur/WTDPD2_S1_L001_R2_001.fastq.gz   \n",
       "fur__wt_fe__1                             fur/WT-Fe2-1_S1_L001_R2_001.fastq.gz   \n",
       "...                                                                        ...   \n",
       "efeU__menFentCubiC_ale36__2                         efeU/SRR8218074_2.fastq.gz   \n",
       "efeU__menFentCubiC_ale37__1                         efeU/SRR8218075_2.fastq.gz   \n",
       "efeU__menFentCubiC_ale37__2                         efeU/SRR8218076_2.fastq.gz   \n",
       "efeU__menFentCubiC_ale38__1                         efeU/SRR8218077_2.fastq.gz   \n",
       "efeU__menFentCubiC_ale38__2                         efeU/SRR8218078_2.fastq.gz   \n",
       "\n",
       "                                                         BAM  alignment  \\\n",
       "control__wt_glc__1                    control__wt_glc__1.bam      94.33   \n",
       "control__wt_glc__2                    control__wt_glc__2.bam      94.24   \n",
       "fur__wt_dpd__1                            fur__wt_dpd__1.bam      98.04   \n",
       "fur__wt_dpd__2                            fur__wt_dpd__2.bam      98.30   \n",
       "fur__wt_fe__1                              fur__wt_fe__1.bam      93.35   \n",
       "...                                                      ...        ...   \n",
       "efeU__menFentCubiC_ale36__2  efeU__menFentCubiC_ale36__2.bam      93.44   \n",
       "efeU__menFentCubiC_ale37__1  efeU__menFentCubiC_ale37__1.bam      94.89   \n",
       "efeU__menFentCubiC_ale37__2  efeU__menFentCubiC_ale37__2.bam      94.62   \n",
       "efeU__menFentCubiC_ale38__1  efeU__menFentCubiC_ale38__1.bam      95.65   \n",
       "efeU__menFentCubiC_ale38__2  efeU__menFentCubiC_ale38__2.bam      94.94   \n",
       "\n",
       "                                                           DOI        GEO  \\\n",
       "control__wt_glc__1                      doi.org/10.1101/080929   GSE65643   \n",
       "control__wt_glc__2                      doi.org/10.1101/080929   GSE65643   \n",
       "fur__wt_dpd__1                      doi.org/10.1038/ncomms5910   GSE54900   \n",
       "fur__wt_dpd__2                      doi.org/10.1038/ncomms5910   GSE54900   \n",
       "fur__wt_fe__1                       doi.org/10.1038/ncomms5910   GSE54900   \n",
       "...                                                        ...        ...   \n",
       "efeU__menFentCubiC_ale36__2  doi.org/10.1038/s41564-018-0340-2  GSE122779   \n",
       "efeU__menFentCubiC_ale37__1  doi.org/10.1038/s41564-018-0340-2  GSE122779   \n",
       "efeU__menFentCubiC_ale37__2  doi.org/10.1038/s41564-018-0340-2  GSE122779   \n",
       "efeU__menFentCubiC_ale38__1  doi.org/10.1038/s41564-018-0340-2  GSE122779   \n",
       "efeU__menFentCubiC_ale38__2  doi.org/10.1038/s41564-018-0340-2  GSE122779   \n",
       "\n",
       "                             Public  \n",
       "control__wt_glc__1                1  \n",
       "control__wt_glc__2                1  \n",
       "fur__wt_dpd__1                    1  \n",
       "fur__wt_dpd__2                    1  \n",
       "fur__wt_fe__1                     1  \n",
       "...                             ...  \n",
       "efeU__menFentCubiC_ale36__2       1  \n",
       "efeU__menFentCubiC_ale37__1       1  \n",
       "efeU__menFentCubiC_ale37__2       1  \n",
       "efeU__menFentCubiC_ale38__1       1  \n",
       "efeU__menFentCubiC_ale38__2       1  \n",
       "\n",
       "[278 rows x 32 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.sample_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nothing to change for the sample table -- the 0th column is the correct one to use.\n",
    "# However, you can optionally suppress a warning about this from later on if you name the index:\n",
    "ica_data.sample_table.index.name = 'sample'\n",
    "\n",
    "# Also, there is an error in the existing 'Biological Replicates' column.\n",
    "# This caused a \"cannot set a row with mismatched columns\" error in export (step 5).\n",
    "# The easiest way to fix it is to have the code re-compute this column in step 4b.\n",
    "# to make that happen, delete the existing one now.\n",
    "ica_data.sample_table = ica_data.sample_table.drop('Biological Replicates', axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.b: Fill in the unknowns (Optional)\n",
    "\n",
    "This step is already integrated in step 5 -- feel free to skip it. However, it might be a useful demonstration.\n",
    "\n",
    "Even if you only have the minimum (M, A, X, and sample_table) data, calling the compatibility function with write=True will fill in enough blanks to generate the dataset. In this example, we need the code to fix the things that we ignored during step 4a.\n",
    "\n",
    "In general, you may want to deepcopy the data before doing this. While the default columns shouldn't be an issue, the function requires that all iModulon names be numbers, which might not be desired for you. Under the hood in step 5, we automatically deep copy your data before changing anything. If you want access to the default values and columns, however, you may be interested in doing this without the deepcopy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iModulon Table is missing a Regulator column\n",
      "Note that Regulator is displayed in the dataset page so it can have spelled out 'and' and 'or' operators.If you have a TF column, a Regulator column will begenerated for you (but not vice versa).\n",
      "\n",
      "Gene Table is missing the gene_product column\n",
      "\n",
      "Sample Table is missing a Biological Replicates column\n"
     ]
    }
   ],
   "source": [
    "from copy import deepcopy\n",
    "\n",
    "new_data = deepcopy(ica_data)\n",
    "\n",
    "imodulondb_compatibility(new_data, write = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's call it again just to see what the output is now that the new elements are written:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "imodulondb_compatibility(new_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A very minimal output is good!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5: Export\n",
    "\n",
    "You are now ready to export! This process may take some time (up to around 1 hr for PRECISE2, but significantly shorter depending on the number of genes your organism has).\n",
    "\n",
    "The export function needs a path to export to. This should be path to where you cloned the iModulonDB repository.\n",
    "\n",
    "The function will first deepcopy your files and then call the compatibility function from above with write = True. Then it will generate all files that don't require iteration, then all iModulon files, then all gene files.\n",
    "\n",
    "As it goes, it will print anything you may want to be aware of. This is where issues with the TRN may become apparent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iModulon Table is missing a Regulator column\n",
      "Note that Regulator is displayed in the dataset page so it can have spelled out 'and' and 'or' operators.If you have a TF column, a Regulator column will begenerated for you (but not vice versa).\n",
      "\n",
      "Gene Table is missing the gene_product column\n",
      "\n",
      "Sample Table is missing a Biological Replicates column\n",
      "Writing main site files...\n",
      "Two progress bars will appear below. The second will take significantly longer than the first.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "26de27fbc33245128a71dccc9612b815",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value=''), FloatProgress(value=0.0, max=92.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TF has no associated expression profile: glpR\n",
      "If glpR is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: cecR\n",
      "If cecR is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: btsR\n",
      "If btsR is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: gutM\n",
      "If gutM is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: ile-tRNA\n",
      "If ile-tRNA is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: leu-tRNA\n",
      "If leu-tRNA is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: tPP\n",
      "If tPP is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: his-tRNA\n",
      "If his-tRNA is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: l-tryptophan\n",
      "If l-tryptophan is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "TF has no associated expression profile: trp-tRNA\n",
      "If trp-tRNA is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e7f95c50b3434725bdbed21bcf050457",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value=''), FloatProgress(value=0.0, max=3923.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "imodulondb_export(ica_data, '../../iModulonDB/Personal')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6: Update TRN, other gene info if desired\n",
    "\n",
    "In the output above, you will notice snippets such as this one:\n",
    "\n",
    "> TF has no associated expression profile: cecR\n",
    "\n",
    ">If cecR is not a gene, this behavior is expected.\n",
    "\n",
    ">If it is a gene, use consistent naming between the TRN and gene_table.\n",
    "\n",
    "If desired, you could look up each of these to find alternative names that may be used, then rename either the TRN (all instances of the name in the regulator column) or the gene_table (the name column in the row with the appropriate locus) so that they all match. This will not need to be done for every case; for example, 'ile-tRNA' is a leucine-bound tRNA molecule, which does not have an associated gene.\n",
    "\n",
    "After doing this, you would need to rerun the export function (step 5)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "start                829971\n",
       "stop                 830643\n",
       "strand                    -\n",
       "gene_name              ybiH\n",
       "length                  672\n",
       "operon         cecR-ybhGFSR\n",
       "COG           Transcription\n",
       "regulator    cecR,rpoD,ybiH\n",
       "Name: b0796, dtype: object"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# according to EcoCyc, the b-number for cecR is b0796 and its alternative name is 'ybiH'\n",
    "\n",
    "ica_data.gene_table.loc['b0796']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To remedy this matchup, I will replace 'ybiH' with 'cecR' in the gene_table.\n",
    "\n",
    "ica_data.gene_table.loc['b0796', 'gene_name'] = 'cecR'\n",
    "\n",
    "# repeat this process for all unmatched genes.\n",
    "\n",
    "# re-running step 5 would then correct this error."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7: Paste HTML\n",
    "\n",
    "This step only needs to occur once! Time to add the link to the splash page.\n",
    "\n",
    "1. Navigate to your iModulonDB/organisms/M_favoritespeciesus/modulome folder.\n",
    "2. Open the file 'html_for_splash.html' in a text editor\n",
    "3. Copy the entire contents to your clipboard\n",
    "4. Navigate back to the main repository folder\n",
    "5. Open 'index.html' in a text editor\n",
    "6. Scroll down to the comment near line 250 about \"NEW DATASETS\"\n",
    "7. Paste in the appropriate location. You may want to adjust the indentations.\n",
    "8. Save the file\n",
    "9. Enjoy your new website!\n",
    "\n",
    "Traveling to localhost:8000 in your browser should show you the site with a shiny new button for your dataset. Click it to see your new iModulons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Moving Forward\n",
    "\n",
    "Any time you make a change that you want reflected on your site, simply repeat step 5! If changes don't immediately appear on the browser, try a force refresh (Ctrl+Shift+R in Google Chrome)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Don't forget to save your file so that step 5 is all that is required in the future!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save_to_json(ica_data, '../example_data/example.json')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
