{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Initialize-your-IcaData-object\" data-toc-modified-id=\"Initialize-your-IcaData-object-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Initialize your IcaData object</a></span><ul class=\"toc-item\"><li><span><a href=\"#Minimum-requirements\" data-toc-modified-id=\"Minimum-requirements-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Minimum requirements</a></span></li><li><span><a href=\"#Viewing-matrices\" data-toc-modified-id=\"Viewing-matrices-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Viewing matrices</a></span></li><li><span><a href=\"#Sample,-gene,-and-iModulon-names\" data-toc-modified-id=\"Sample,-gene,-and-iModulon-names-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Sample, gene, and iModulon names</a></span></li><li><span><a href=\"#Adding-the-X-matrix\" data-toc-modified-id=\"Adding-the-X-matrix-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>Adding the X matrix</a></span></li></ul></li><li><span><a href=\"#Sample,-gene,-and-iModulon-tables\" data-toc-modified-id=\"Sample,-gene,-and-iModulon-tables-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Sample, gene, and iModulon tables</a></span><ul class=\"toc-item\"><li><span><a href=\"#Loading-tables\" data-toc-modified-id=\"Loading-tables-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Loading tables</a></span></li><li><span><a href=\"#Viewing-tables\" data-toc-modified-id=\"Viewing-tables-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Viewing tables</a></span></li></ul></li><li><span><a href=\"#TRN-Table\" data-toc-modified-id=\"TRN-Table-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>TRN Table</a></span><ul class=\"toc-item\"><li><span><a href=\"#Load-the-TRN\" data-toc-modified-id=\"Load-the-TRN-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Load the TRN</a></span></li></ul></li><li><span><a href=\"#iModulon-properties\" data-toc-modified-id=\"iModulon-properties-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>iModulon properties</a></span><ul class=\"toc-item\"><li><span><a href=\"#Inspecting-iModulons\" data-toc-modified-id=\"Inspecting-iModulons-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Inspecting iModulons</a></span></li><li><span><a href=\"#Renaming-iModulons\" data-toc-modified-id=\"Renaming-iModulons-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Renaming iModulons</a></span></li><li><span><a href=\"#Identifying-&quot;single-gene&quot;-iModulons\" data-toc-modified-id=\"Identifying-&quot;single-gene&quot;-iModulons-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Identifying \"single-gene\" iModulons</a></span></li></ul></li><li><span><a href=\"#Saving-and-Loading-IcaData-Objects\" data-toc-modified-id=\"Saving-and-Loading-IcaData-Objects-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Saving and Loading IcaData Objects</a></span></li><li><span><a href=\"#Explained-variance\" data-toc-modified-id=\"Explained-variance-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Explained variance</a></span></li><li><span><a href=\"#Gene-enrichment-analysis\" data-toc-modified-id=\"Gene-enrichment-analysis-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Gene enrichment analysis</a></span><ul class=\"toc-item\"><li><span><a href=\"#General-functions\" data-toc-modified-id=\"General-functions-7.1\"><span class=\"toc-item-num\">7.1&nbsp;&nbsp;</span>General functions</a></span></li><li><span><a href=\"#Using-the-IcaData-object\" data-toc-modified-id=\"Using-the-IcaData-object-7.2\"><span class=\"toc-item-num\">7.2&nbsp;&nbsp;</span>Using the IcaData object</a></span><ul class=\"toc-item\"><li><span><a href=\"#Save-to-sample-table\" data-toc-modified-id=\"Save-to-sample-table-7.2.1\"><span class=\"toc-item-num\">7.2.1&nbsp;&nbsp;</span>Save to sample table</a></span></li></ul></li></ul></li><li><span><a href=\"#iModulon-Thresholds\" data-toc-modified-id=\"iModulon-Thresholds-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>iModulon Thresholds</a></span><ul class=\"toc-item\"><li><span><a href=\"#Fitting-the-threshold\" data-toc-modified-id=\"Fitting-the-threshold-8.1\"><span class=\"toc-item-num\">8.1&nbsp;&nbsp;</span>Fitting the threshold</a></span></li><li><span><a href=\"#Changing-the-D'agostino-statistic\" data-toc-modified-id=\"Changing-the-D'agostino-statistic-8.2\"><span class=\"toc-item-num\">8.2&nbsp;&nbsp;</span>Changing the D'agostino statistic</a></span></li><li><span><a href=\"#Manually-changing-thresholds\" data-toc-modified-id=\"Manually-changing-thresholds-8.3\"><span class=\"toc-item-num\">8.3&nbsp;&nbsp;</span>Manually changing thresholds</a></span></li></ul></li><li><span><a href=\"#Inferring-iModulon-activities-of-new-data\" data-toc-modified-id=\"Inferring-iModulon-activities-of-new-data-9\"><span class=\"toc-item-num\">9&nbsp;&nbsp;</span>Inferring iModulon activities of new data</a></span></li><li><span><a href=\"#Comparing-ICA-Runs\" data-toc-modified-id=\"Comparing-ICA-Runs-10\"><span class=\"toc-item-num\">10&nbsp;&nbsp;</span>Comparing ICA Runs</a></span><ul class=\"toc-item\"><li><span><a href=\"#Comparsion-within-the-same-organism\" data-toc-modified-id=\"Comparsion-within-the-same-organism-10.1\"><span class=\"toc-item-num\">10.1&nbsp;&nbsp;</span>Comparsion within the same organism</a></span></li><li><span><a href=\"#Comparison-across-organisms\" data-toc-modified-id=\"Comparison-across-organisms-10.2\"><span class=\"toc-item-num\">10.2&nbsp;&nbsp;</span>Comparison across organisms</a></span><ul class=\"toc-item\"><li><span><a href=\"#Create-a-BBH-CSV-file\" data-toc-modified-id=\"Create-a-BBH-CSV-file-10.2.1\"><span class=\"toc-item-num\">10.2.1&nbsp;&nbsp;</span>Create a BBH CSV file</a></span></li><li><span><a href=\"#Create-Genbank-database-files\" data-toc-modified-id=\"Create-Genbank-database-files-10.2.2\"><span class=\"toc-item-num\">10.2.2&nbsp;&nbsp;</span>Create Genbank database files</a></span></li><li><span><a href=\"#Create-BBH-CSV\" data-toc-modified-id=\"Create-BBH-CSV-10.2.3\"><span class=\"toc-item-num\">10.2.3&nbsp;&nbsp;</span>Create BBH CSV</a></span></li></ul></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from os.path import abspath, join\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = abspath(join('..','pymodulon', 'test', 'precise_data'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Initialize your IcaData object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Minimum requirements\n",
    "The IcaData object only requires two matrices, the **M** and the **A** matrices. However, many functions (such as TRN enrichment) will be limited without including additional data.  \n",
    "These matrices can be added as either a filename or a pandas DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "M = join(DATA_DIR, 'M.csv')\n",
    "A = join(DATA_DIR, 'A.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = IcaData(M,A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.visualization import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viewing matrices\n",
    "You can directly view the M and A matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control__wt_glc__1</th>\n",
       "      <th>control__wt_glc__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_dpd__1</th>\n",
       "      <th>fur__delfur_dpd__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>...</th>\n",
       "      <th>efeU__menFentC_ale29__1</th>\n",
       "      <th>efeU__menFentC_ale29__2</th>\n",
       "      <th>efeU__menFentC_ale30__1</th>\n",
       "      <th>efeU__menFentC_ale30__2</th>\n",
       "      <th>efeU__menFentCubiC_ale36__1</th>\n",
       "      <th>efeU__menFentCubiC_ale36__2</th>\n",
       "      <th>efeU__menFentCubiC_ale37__1</th>\n",
       "      <th>efeU__menFentCubiC_ale37__2</th>\n",
       "      <th>efeU__menFentCubiC_ale38__1</th>\n",
       "      <th>efeU__menFentCubiC_ale38__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.076035</td>\n",
       "      <td>-0.076035</td>\n",
       "      <td>1.483992</td>\n",
       "      <td>1.719034</td>\n",
       "      <td>-1.184220</td>\n",
       "      <td>-1.350203</td>\n",
       "      <td>1.451882</td>\n",
       "      <td>1.329577</td>\n",
       "      <td>0.265718</td>\n",
       "      <td>0.024864</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.892372</td>\n",
       "      <td>-1.022146</td>\n",
       "      <td>0.253722</td>\n",
       "      <td>-0.880352</td>\n",
       "      <td>0.526509</td>\n",
       "      <td>1.261740</td>\n",
       "      <td>1.727782</td>\n",
       "      <td>1.512944</td>\n",
       "      <td>2.584012</td>\n",
       "      <td>2.787686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.408568</td>\n",
       "      <td>0.408568</td>\n",
       "      <td>1.760680</td>\n",
       "      <td>1.357274</td>\n",
       "      <td>-0.446482</td>\n",
       "      <td>-0.931725</td>\n",
       "      <td>0.517458</td>\n",
       "      <td>0.427993</td>\n",
       "      <td>-0.499529</td>\n",
       "      <td>-0.581335</td>\n",
       "      <td>...</td>\n",
       "      <td>1.781224</td>\n",
       "      <td>2.785099</td>\n",
       "      <td>0.439133</td>\n",
       "      <td>-0.015471</td>\n",
       "      <td>0.366544</td>\n",
       "      <td>0.369589</td>\n",
       "      <td>0.325470</td>\n",
       "      <td>1.027508</td>\n",
       "      <td>1.629269</td>\n",
       "      <td>2.154717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.103912</td>\n",
       "      <td>-0.103912</td>\n",
       "      <td>0.847276</td>\n",
       "      <td>1.530875</td>\n",
       "      <td>0.258492</td>\n",
       "      <td>0.188903</td>\n",
       "      <td>1.232846</td>\n",
       "      <td>0.883215</td>\n",
       "      <td>0.505356</td>\n",
       "      <td>0.132219</td>\n",
       "      <td>...</td>\n",
       "      <td>1.422466</td>\n",
       "      <td>2.396554</td>\n",
       "      <td>2.130935</td>\n",
       "      <td>1.656316</td>\n",
       "      <td>0.864338</td>\n",
       "      <td>0.093698</td>\n",
       "      <td>0.641705</td>\n",
       "      <td>1.306455</td>\n",
       "      <td>1.366912</td>\n",
       "      <td>2.452061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.456843</td>\n",
       "      <td>0.456843</td>\n",
       "      <td>5.890468</td>\n",
       "      <td>6.051415</td>\n",
       "      <td>-0.914670</td>\n",
       "      <td>-1.087245</td>\n",
       "      <td>5.210846</td>\n",
       "      <td>5.028641</td>\n",
       "      <td>3.820172</td>\n",
       "      <td>3.961980</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.343743</td>\n",
       "      <td>-1.991048</td>\n",
       "      <td>3.520794</td>\n",
       "      <td>3.105551</td>\n",
       "      <td>6.620301</td>\n",
       "      <td>5.400225</td>\n",
       "      <td>5.545880</td>\n",
       "      <td>5.555621</td>\n",
       "      <td>9.672308</td>\n",
       "      <td>10.109228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.269057</td>\n",
       "      <td>-0.269057</td>\n",
       "      <td>-0.254271</td>\n",
       "      <td>0.184537</td>\n",
       "      <td>0.651527</td>\n",
       "      <td>0.691914</td>\n",
       "      <td>0.016636</td>\n",
       "      <td>-0.465781</td>\n",
       "      <td>1.212097</td>\n",
       "      <td>1.122801</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.646464</td>\n",
       "      <td>-3.606055</td>\n",
       "      <td>-0.665810</td>\n",
       "      <td>-0.830586</td>\n",
       "      <td>-1.044574</td>\n",
       "      <td>0.119876</td>\n",
       "      <td>-0.045670</td>\n",
       "      <td>-0.826608</td>\n",
       "      <td>-1.393160</td>\n",
       "      <td>-1.587978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 278 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   control__wt_glc__1  control__wt_glc__2  fur__wt_dpd__1  fur__wt_dpd__2  \\\n",
       "0            0.076035           -0.076035        1.483992        1.719034   \n",
       "1           -0.408568            0.408568        1.760680        1.357274   \n",
       "2            0.103912           -0.103912        0.847276        1.530875   \n",
       "3           -0.456843            0.456843        5.890468        6.051415   \n",
       "4            0.269057           -0.269057       -0.254271        0.184537   \n",
       "\n",
       "   fur__wt_fe__1  fur__wt_fe__2  fur__delfur_dpd__1  fur__delfur_dpd__2  \\\n",
       "0      -1.184220      -1.350203            1.451882            1.329577   \n",
       "1      -0.446482      -0.931725            0.517458            0.427993   \n",
       "2       0.258492       0.188903            1.232846            0.883215   \n",
       "3      -0.914670      -1.087245            5.210846            5.028641   \n",
       "4       0.651527       0.691914            0.016636           -0.465781   \n",
       "\n",
       "   fur__delfur_fe2__1  fur__delfur_fe2__2  ...  efeU__menFentC_ale29__1  \\\n",
       "0            0.265718            0.024864  ...                -0.892372   \n",
       "1           -0.499529           -0.581335  ...                 1.781224   \n",
       "2            0.505356            0.132219  ...                 1.422466   \n",
       "3            3.820172            3.961980  ...                -2.343743   \n",
       "4            1.212097            1.122801  ...                -3.646464   \n",
       "\n",
       "   efeU__menFentC_ale29__2  efeU__menFentC_ale30__1  efeU__menFentC_ale30__2  \\\n",
       "0                -1.022146                 0.253722                -0.880352   \n",
       "1                 2.785099                 0.439133                -0.015471   \n",
       "2                 2.396554                 2.130935                 1.656316   \n",
       "3                -1.991048                 3.520794                 3.105551   \n",
       "4                -3.606055                -0.665810                -0.830586   \n",
       "\n",
       "   efeU__menFentCubiC_ale36__1  efeU__menFentCubiC_ale36__2  \\\n",
       "0                     0.526509                     1.261740   \n",
       "1                     0.366544                     0.369589   \n",
       "2                     0.864338                     0.093698   \n",
       "3                     6.620301                     5.400225   \n",
       "4                    -1.044574                     0.119876   \n",
       "\n",
       "   efeU__menFentCubiC_ale37__1  efeU__menFentCubiC_ale37__2  \\\n",
       "0                     1.727782                     1.512944   \n",
       "1                     0.325470                     1.027508   \n",
       "2                     0.641705                     1.306455   \n",
       "3                     5.545880                     5.555621   \n",
       "4                    -0.045670                    -0.826608   \n",
       "\n",
       "   efeU__menFentCubiC_ale38__1  efeU__menFentCubiC_ale38__2  \n",
       "0                     2.584012                     2.787686  \n",
       "1                     1.629269                     2.154717  \n",
       "2                     1.366912                     2.452061  \n",
       "3                     9.672308                    10.109228  \n",
       "4                    -1.393160                    -1.587978  \n",
       "\n",
       "[5 rows x 278 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.A.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample, gene, and iModulon names\n",
    "If the M and A matrices have sample/gene/iModulon names, these will be saved as the official sample/gene/iModulon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene names: ['b0002', 'b0003', 'b0004', 'b0005', 'b0006']\n",
      "Sample names: ['control__wt_glc__1', 'control__wt_glc__2', 'fur__wt_dpd__1', 'fur__wt_dpd__2', 'fur__wt_fe__1']\n",
      "iModulon names: [0, 1, 2, 3, 4]\n"
     ]
    }
   ],
   "source": [
    "print('Gene names:',ica_data.gene_names[:5])\n",
    "print('Sample names:',ica_data.sample_names[:5])\n",
    "print('iModulon names:',ica_data.imodulon_names[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding the X matrix\n",
    "The **X** matrix contains expression data and is mainly used for plotting functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = join(DATA_DIR, 'X.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.X = X"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample, gene, and iModulon tables\n",
    "You may load in additional data tables with information about your samples, genes, or iModulons. If you do not load any data, these tables are originally empty but can be altered like any pandas table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [b0002, b0003, b0004, b0005, b0006]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.gene_table['gene_name'] = 'example'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>example</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "b0002   example\n",
       "b0003   example\n",
       "b0004   example\n",
       "b0005   example\n",
       "b0006   example"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading tables\n",
    "Sample/gene/iModulon tables contain one sample/gene/iModulon per row, and information about the respective item in columns.  \n",
    "**The sample/gene/iModulon names must match the names in your M and A matrix.** However, the order of the names does not matter.  \n",
    "The tables can be passed in using either a filename or a pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene file: /home/anand/PycharmProjects/pymodulon/pymodulon/test/precise_data/gene_table.csv\n"
     ]
    }
   ],
   "source": [
    "gene_file = join(DATA_DIR, 'gene_table.csv')\n",
    "sample_file = join(DATA_DIR, 'sample_table.csv')\n",
    "imodulon_file = join(DATA_DIR, 'imodulon_table.csv')\n",
    "print('Gene file:',gene_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "ica_data.gene_table = gene_file\n",
    "ica_data.sample_table = sample_file\n",
    "ica_data.imodulon_table = imodulon_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viewing tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "      <th>rep_id</th>\n",
       "      <th>contact</th>\n",
       "      <th>library creator</th>\n",
       "      <th>run_date</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>BAM</th>\n",
       "      <th>alignment</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "      <th>Public</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>control__wt_glc__1</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>04/17/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>control/MG1655-M9-glucose-1_S1_L001_R1_001.fas...</td>\n",
       "      <td>control/MG1655-M9-glucose-1_S1_L001_R2_001.fas...</td>\n",
       "      <td>control__wt_glc__1.bam</td>\n",
       "      <td>94.33</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>control__wt_glc__2</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "      <td>wt_glc</td>\n",
       "      <td>2</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>04/17/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>control/MG1655-M9-glucose-2_S2_L001_R1_001.fas...</td>\n",
       "      <td>control/MG1655-M9-glucose-2_S2_L001_R2_001.fas...</td>\n",
       "      <td>control__wt_glc__2.bam</td>\n",
       "      <td>94.24</td>\n",
       "      <td>doi.org/10.1101/080929</td>\n",
       "      <td>GSE65643</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>10/02/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WTDPD1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WTDPD1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_dpd__1.bam</td>\n",
       "      <td>98.04</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_dpd</td>\n",
       "      <td>2</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>10/02/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WTDPD2_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WTDPD2_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_dpd__2.bam</td>\n",
       "      <td>98.30</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>1</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>Sangwoo Seo</td>\n",
       "      <td>09/05/13</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>fur/WT-Fe2-1_S1_L001_R1_001.fastq.gz</td>\n",
       "      <td>fur/WT-Fe2-1_S1_L001_R2_001.fastq.gz</td>\n",
       "      <td>fur__wt_fe__1.bam</td>\n",
       "      <td>93.35</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      study  project condition  rep_id      contact  \\\n",
       "control__wt_glc__1  Control  control    wt_glc       1  Sangwoo Seo   \n",
       "control__wt_glc__2  Control  control    wt_glc       2  Sangwoo Seo   \n",
       "fur__wt_dpd__1          Fur      fur    wt_dpd       1  Sangwoo Seo   \n",
       "fur__wt_dpd__2          Fur      fur    wt_dpd       2  Sangwoo Seo   \n",
       "fur__wt_fe__1           Fur      fur     wt_fe       1  Sangwoo Seo   \n",
       "\n",
       "                   library creator  run_date            Strain Description  \\\n",
       "control__wt_glc__1     Sangwoo Seo  04/17/13  Escherichia coli K-12 MG1655   \n",
       "control__wt_glc__2     Sangwoo Seo  04/17/13  Escherichia coli K-12 MG1655   \n",
       "fur__wt_dpd__1         Sangwoo Seo  10/02/13  Escherichia coli K-12 MG1655   \n",
       "fur__wt_dpd__2         Sangwoo Seo  10/02/13  Escherichia coli K-12 MG1655   \n",
       "fur__wt_fe__1          Sangwoo Seo  09/05/13  Escherichia coli K-12 MG1655   \n",
       "\n",
       "                    Strain Base Media  ... Sequencing Machine  \\\n",
       "control__wt_glc__1  MG1655         M9  ...              MiSeq   \n",
       "control__wt_glc__2  MG1655         M9  ...              MiSeq   \n",
       "fur__wt_dpd__1      MG1655         M9  ...              MiSeq   \n",
       "fur__wt_dpd__2      MG1655         M9  ...              MiSeq   \n",
       "fur__wt_fe__1       MG1655         M9  ...              MiSeq   \n",
       "\n",
       "                   Additional Details Biological Replicates  \\\n",
       "control__wt_glc__1                NaN                     2   \n",
       "control__wt_glc__2                NaN                     2   \n",
       "fur__wt_dpd__1                    NaN                     2   \n",
       "fur__wt_dpd__2                    NaN                     2   \n",
       "fur__wt_fe__1                     NaN                     2   \n",
       "\n",
       "                                                                   R1  \\\n",
       "control__wt_glc__1  control/MG1655-M9-glucose-1_S1_L001_R1_001.fas...   \n",
       "control__wt_glc__2  control/MG1655-M9-glucose-2_S2_L001_R1_001.fas...   \n",
       "fur__wt_dpd__1                     fur/WTDPD1_S1_L001_R1_001.fastq.gz   \n",
       "fur__wt_dpd__2                     fur/WTDPD2_S1_L001_R1_001.fastq.gz   \n",
       "fur__wt_fe__1                    fur/WT-Fe2-1_S1_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                                                   R2  \\\n",
       "control__wt_glc__1  control/MG1655-M9-glucose-1_S1_L001_R2_001.fas...   \n",
       "control__wt_glc__2  control/MG1655-M9-glucose-2_S2_L001_R2_001.fas...   \n",
       "fur__wt_dpd__1                     fur/WTDPD1_S1_L001_R2_001.fastq.gz   \n",
       "fur__wt_dpd__2                     fur/WTDPD2_S1_L001_R2_001.fastq.gz   \n",
       "fur__wt_fe__1                    fur/WT-Fe2-1_S1_L001_R2_001.fastq.gz   \n",
       "\n",
       "                                       BAM  alignment  \\\n",
       "control__wt_glc__1  control__wt_glc__1.bam      94.33   \n",
       "control__wt_glc__2  control__wt_glc__2.bam      94.24   \n",
       "fur__wt_dpd__1          fur__wt_dpd__1.bam      98.04   \n",
       "fur__wt_dpd__2          fur__wt_dpd__2.bam      98.30   \n",
       "fur__wt_fe__1            fur__wt_fe__1.bam      93.35   \n",
       "\n",
       "                                           DOI       GEO  Public  \n",
       "control__wt_glc__1      doi.org/10.1101/080929  GSE65643       1  \n",
       "control__wt_glc__2      doi.org/10.1101/080929  GSE65643       1  \n",
       "fur__wt_dpd__1      doi.org/10.1038/ncomms5910  GSE54900       1  \n",
       "fur__wt_dpd__2      doi.org/10.1038/ncomms5910  GSE54900       1  \n",
       "fur__wt_fe__1       doi.org/10.1038/ncomms5910  GSE54900       1  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.sample_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>COG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>336</td>\n",
       "      <td>2799</td>\n",
       "      <td>+</td>\n",
       "      <td>thrA</td>\n",
       "      <td>2463</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>2800</td>\n",
       "      <td>3733</td>\n",
       "      <td>+</td>\n",
       "      <td>thrB</td>\n",
       "      <td>933</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>3733</td>\n",
       "      <td>5020</td>\n",
       "      <td>+</td>\n",
       "      <td>thrC</td>\n",
       "      <td>1287</td>\n",
       "      <td>thrLABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>5233</td>\n",
       "      <td>5530</td>\n",
       "      <td>+</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>297</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>5682</td>\n",
       "      <td>6459</td>\n",
       "      <td>-</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>777</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start  stop strand gene_name  length   operon  \\\n",
       "b0002    336  2799      +      thrA    2463  thrLABC   \n",
       "b0003   2800  3733      +      thrB     933  thrLABC   \n",
       "b0004   3733  5020      +      thrC    1287  thrLABC   \n",
       "b0005   5233  5530      +      yaaX     297     yaaX   \n",
       "b0006   5682  6459      -      yaaA     777     yaaA   \n",
       "\n",
       "                                       COG  \n",
       "b0002  Amino acid transport and metabolism  \n",
       "b0003  Amino acid transport and metabolism  \n",
       "b0004  Amino acid transport and metabolism  \n",
       "b0005                     Function unknown  \n",
       "b0006                     Function unknown  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YieP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YgbI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>efeU-repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ydcI-KO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          name regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "0         YieP       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "1         GlpR      glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "2         YgbI       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "3  efeU-repair       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "4      ydcI-KO       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "   regulon_size  n_regs                                           Function  \\\n",
       "0            11       1                                            Unknown   \n",
       "1             9       1                                Glycerol catabolism   \n",
       "2             9       1                                   Unknown Function   \n",
       "3             8       0  Accounts for repair and expression of efeU operon   \n",
       "4             3       0                        Accounts for ydcI knock-out   \n",
       "\n",
       "                    Category  \n",
       "0        Regulator Discovery  \n",
       "1  Carbon Source Utilization  \n",
       "2        Regulator Discovery  \n",
       "3        Genomic Alterations  \n",
       "4        Genomic Alterations  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRN Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the TRN\n",
    "Again, this table can be passed in using either a filename or a pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/anand/PycharmProjects/pymodulon/pymodulon/core.py:271: UserWarning: The following genes are in the TRN but not in your M matrix: {'b4341', 'b0616', 'b4416', 'b1314', 'b0672', 'b4194', 'b3138', 'b0553', 'b1576', 'b1691', 'b2483', 'b4697', 'b0510', 'b1573', 'b2334', 'b4439', 'b1715', 'b3077', 'b3423', 'b2337', 'b2824', 'b1229', 'b2488', 'b3046_1', 'b1317', 'b3141', 'b0246', 'b4685', 'b0319', 'b3427', 'b0240_1', 'b4598', 'b2734', 'b2485', 'b3119', 'b0668', 'b4347', 'b4699', 'b0001', 'b0617', 'b3027', 'b1578', 'b4438', 'b4432', 'b4415', 'b1577', 'b3257', 'b3707', 'b3044', 'b4687', 'b1575', 'b0044', 'b2018', 'b1555', 'b0551', 'b1160', 'b4613', 'b4101', 'b4533', 'b2407', 'b3121', 'b1396', 'b4414', 'b2759', 'b4513', 'b0362', 'G0-10544', 'b4618', 'b4673', 'b4365', 'b4100', 'b1553', 'b1373', 'b1532', 'b0664', 'b0247', 'b1510', 'b4502', 'b4369', 'b4671', 'b2849', 'b0939', 'b1041', 'b3137', 'b0248', 'b0554', 'b2139', 'b3392', 'b3045', 'b4599', 'b2092', 'b4075', 'b2189', 'b0218', 'b3268_1', 'b3864', 'b4449', 'b0240', 'b4504', 'b0691', 'b1862', 'b2934', 'b2198', 'b1169', 'b3171', 'b1312', 'b1567', 'b4368', 'b3123', 'b2815', 'b4653', 'b2228', 'b4532', 'b4104_1', 'b2083', 'unannotated', 'b2490', 'b0256', 'b4458', 'b0301', 'b2891_1', 'b4678', 'b1307', 'b0971', 'b4490_1', 'b3537', 'b0769', 'b4103', 'b4176', 'b0249', 'b0673', 'b0883', 'b4436', 'b0245', 'b1408', 'b2336', 'b0035', 'b2850', 'b3330', 'b4451', 'b3378', 'b0140', 'b2767', 'b4408', 'b3120', 'b0670', 'b1572', 'b4524', 'b3818', 'b2624', 'b2338', 'b3890', 'b0303', 'b4520', 'b1500', 'b3394', 'b1569', 'b2335', 'b4097', 'b3242', 'b2339', 'b1556', 'b3768', 'b2598', 'b1016', 'b1554', 'b4702', 'b2814', 'b4402', 'b1265', 'b2621', 'b0613', 'b0614', 'b1568', 'b1409', 'b3331', 'b1017', 'b2123', 'b0108', 'b4663', 'b2481', 'b4637', 'b0665', 'b2940', 'b3393', 'b1311', 'b3268', 'b3767', 'b1977', 'b4443', 'b4525', 'b4670', 'b0075', 'b2641_1', 'b4104_2', 'b4418', 'b3334', 'b2462', 'b2816', 'b4669', 'b4370', 'b3329', 'b0556', 'b0216', 'b1371', 'b1043', 'b3174', 'b1571', 'b2333', 'b0666', 'b2649', 'b1995', 'b2258', 'b1401', 'b2448', 'b4566', 'b1445'}\n",
      "  warnings.warn('The following genes are in the TRN but not in '\n"
     ]
    }
   ],
   "source": [
    "trn = pd.read_csv(join(DATA_DIR, 'trn.csv'))\n",
    "ica_data.trn = trn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>regulator_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>effect</th>\n",
       "      <th>ev_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FMN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b3041</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-tryptophan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b3708</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-tryptophan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b3709</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TPP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b0066</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TPP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>b0067</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      regulator  regulator_id  gene_name gene_id effect  ev_level\n",
       "0           FMN           NaN        NaN   b3041      -       NaN\n",
       "1  L-tryptophan           NaN        NaN   b3708      +       NaN\n",
       "2  L-tryptophan           NaN        NaN   b3709      +       NaN\n",
       "3           TPP           NaN        NaN   b0066      -       NaN\n",
       "4           TPP           NaN        NaN   b0067      -       NaN"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iModulon properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspecting iModulons\n",
    "The `view_imodulon` shows the information about each gene in the iModulon. Most information is stored in the `gene_table`, but the `regulator` column comes from the TRN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0306</th>\n",
       "      <td>0.237575</td>\n",
       "      <td>321607</td>\n",
       "      <td>322327</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgE</td>\n",
       "      <td>720</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>lrp,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0307</th>\n",
       "      <td>0.157379</td>\n",
       "      <td>322337</td>\n",
       "      <td>323765</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgF</td>\n",
       "      <td>1428</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>lrp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0308</th>\n",
       "      <td>0.157886</td>\n",
       "      <td>323757</td>\n",
       "      <td>324453</td>\n",
       "      <td>+</td>\n",
       "      <td>ykgG</td>\n",
       "      <td>696</td>\n",
       "      <td>ykgEFG</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>lrp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1826</th>\n",
       "      <td>-0.124414</td>\n",
       "      <td>1908622</td>\n",
       "      <td>1908766</td>\n",
       "      <td>-</td>\n",
       "      <td>mgrB</td>\n",
       "      <td>144</td>\n",
       "      <td>mgrB</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>phoP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2141</th>\n",
       "      <td>0.264099</td>\n",
       "      <td>2230623</td>\n",
       "      <td>2231022</td>\n",
       "      <td>+</td>\n",
       "      <td>yohJ</td>\n",
       "      <td>399</td>\n",
       "      <td>yohJK</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2142</th>\n",
       "      <td>0.202429</td>\n",
       "      <td>2231018</td>\n",
       "      <td>2231714</td>\n",
       "      <td>+</td>\n",
       "      <td>yohK</td>\n",
       "      <td>696</td>\n",
       "      <td>yohJK</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3755</th>\n",
       "      <td>-0.348577</td>\n",
       "      <td>3940634</td>\n",
       "      <td>3941327</td>\n",
       "      <td>-</td>\n",
       "      <td>yieP</td>\n",
       "      <td>693</td>\n",
       "      <td>yieP-hsrA</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4116</th>\n",
       "      <td>0.137769</td>\n",
       "      <td>4337167</td>\n",
       "      <td>4337929</td>\n",
       "      <td>-</td>\n",
       "      <td>adiY</td>\n",
       "      <td>762</td>\n",
       "      <td>adiY</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>h-NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4354</th>\n",
       "      <td>0.136132</td>\n",
       "      <td>4589128</td>\n",
       "      <td>4591279</td>\n",
       "      <td>-</td>\n",
       "      <td>yjiY</td>\n",
       "      <td>2151</td>\n",
       "      <td>btsT</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>btsR,crp,rpoS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length     operon  \\\n",
       "b0306     0.237575   321607   322327      +      ykgE     720     ykgEFG   \n",
       "b0307     0.157379   322337   323765      +      ykgF    1428     ykgEFG   \n",
       "b0308     0.157886   323757   324453      +      ykgG     696     ykgEFG   \n",
       "b1826    -0.124414  1908622  1908766      -      mgrB     144       mgrB   \n",
       "b2141     0.264099  2230623  2231022      +      yohJ     399      yohJK   \n",
       "b2142     0.202429  2231018  2231714      +      yohK     696      yohJK   \n",
       "b3755    -0.348577  3940634  3941327      -      yieP     693  yieP-hsrA   \n",
       "b4116     0.137769  4337167  4337929      -      adiY     762       adiY   \n",
       "b4354     0.136132  4589128  4591279      -      yjiY    2151       btsT   \n",
       "\n",
       "                                          COG      regulator  \n",
       "b0306        Energy production and conversion       lrp,yieP  \n",
       "b0307        Energy production and conversion            lrp  \n",
       "b0308                        Function unknown            lrp  \n",
       "b1826                        Function unknown           phoP  \n",
       "b2141                        Function unknown           yieP  \n",
       "b2142  Cell wall/membrane/envelope biogenesis            NaN  \n",
       "b3755                           Transcription            NaN  \n",
       "b4116                           Transcription           h-NS  \n",
       "b4354          Signal transduction mechanisms  btsR,crp,rpoS  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Renaming iModulons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>YieP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YgbI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>efeU-repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ydcI-KO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             name regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "YieP         YieP       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "1            GlpR      glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "2            YgbI       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "3     efeU-repair       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "4         ydcI-KO       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "      regulon_size  n_regs                                           Function  \\\n",
       "YieP            11       1                                            Unknown   \n",
       "1                9       1                                Glycerol catabolism   \n",
       "2                9       1                                   Unknown Function   \n",
       "3                8       0  Accounts for repair and expression of efeU operon   \n",
       "4                3       0                        Accounts for ydcI knock-out   \n",
       "\n",
       "                       Category  \n",
       "YieP        Regulator Discovery  \n",
       "1     Carbon Source Utilization  \n",
       "2           Regulator Discovery  \n",
       "3           Genomic Alterations  \n",
       "4           Genomic Alterations  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename iModulons using a dict\n",
    "ica_data.rename_imodulons({0:'YieP'})\n",
    "\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "name                                                                     \n",
       "YieP              NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "GlpR             glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "YgbI              NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "efeU-repair       NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "ydcI-KO           NaN     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "             regulon_size  n_regs  \\\n",
       "name                                \n",
       "YieP                   11       1   \n",
       "GlpR                    9       1   \n",
       "YgbI                    9       1   \n",
       "efeU-repair             8       0   \n",
       "ydcI-KO                 3       0   \n",
       "\n",
       "                                                      Function  \\\n",
       "name                                                             \n",
       "YieP                                                   Unknown   \n",
       "GlpR                                       Glycerol catabolism   \n",
       "YgbI                                          Unknown Function   \n",
       "efeU-repair  Accounts for repair and expression of efeU operon   \n",
       "ydcI-KO                            Accounts for ydcI knock-out   \n",
       "\n",
       "                              Category  \n",
       "name                                    \n",
       "YieP               Regulator Discovery  \n",
       "GlpR         Carbon Source Utilization  \n",
       "YgbI               Regulator Discovery  \n",
       "efeU-repair        Genomic Alterations  \n",
       "ydcI-KO            Genomic Alterations  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename iModulons using the iModulon table\n",
    "ica_data.rename_imodulons(column='name')\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>name</th>\n",
       "      <th>YieP</th>\n",
       "      <th>GlpR</th>\n",
       "      <th>YgbI</th>\n",
       "      <th>efeU-repair</th>\n",
       "      <th>ydcI-KO</th>\n",
       "      <th>BW25113</th>\n",
       "      <th>proVWX</th>\n",
       "      <th>DhaR/Mlc</th>\n",
       "      <th>ArgR</th>\n",
       "      <th>YneJ</th>\n",
       "      <th>...</th>\n",
       "      <th>RpoS</th>\n",
       "      <th>membrane</th>\n",
       "      <th>duplication-1</th>\n",
       "      <th>uncharacterized-3</th>\n",
       "      <th>ExuR/FucR</th>\n",
       "      <th>FecI</th>\n",
       "      <th>iron-related</th>\n",
       "      <th>RbsR</th>\n",
       "      <th>sgrT</th>\n",
       "      <th>insertion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>0.072693</td>\n",
       "      <td>-0.006320</td>\n",
       "      <td>-0.038374</td>\n",
       "      <td>-0.010045</td>\n",
       "      <td>-0.004345</td>\n",
       "      <td>-0.005599</td>\n",
       "      <td>0.009043</td>\n",
       "      <td>-0.005072</td>\n",
       "      <td>-0.012186</td>\n",
       "      <td>-0.013543</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.010152</td>\n",
       "      <td>-0.012124</td>\n",
       "      <td>-0.012027</td>\n",
       "      <td>-0.002123</td>\n",
       "      <td>-0.000036</td>\n",
       "      <td>-0.013667</td>\n",
       "      <td>-0.018883</td>\n",
       "      <td>0.004408</td>\n",
       "      <td>-0.006315</td>\n",
       "      <td>-0.001780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>0.015975</td>\n",
       "      <td>-0.009969</td>\n",
       "      <td>-0.024571</td>\n",
       "      <td>-0.006775</td>\n",
       "      <td>-0.003583</td>\n",
       "      <td>0.006680</td>\n",
       "      <td>-0.010841</td>\n",
       "      <td>-0.012620</td>\n",
       "      <td>-0.003685</td>\n",
       "      <td>0.003512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004702</td>\n",
       "      <td>0.004402</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>-0.004415</td>\n",
       "      <td>-0.001514</td>\n",
       "      <td>-0.005052</td>\n",
       "      <td>0.015502</td>\n",
       "      <td>0.000860</td>\n",
       "      <td>-0.022438</td>\n",
       "      <td>-0.020087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>0.015319</td>\n",
       "      <td>-0.003808</td>\n",
       "      <td>-0.027018</td>\n",
       "      <td>-0.004455</td>\n",
       "      <td>-0.008418</td>\n",
       "      <td>0.003245</td>\n",
       "      <td>-0.016494</td>\n",
       "      <td>-0.013818</td>\n",
       "      <td>-0.008104</td>\n",
       "      <td>0.002144</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003507</td>\n",
       "      <td>0.005342</td>\n",
       "      <td>0.003759</td>\n",
       "      <td>0.000700</td>\n",
       "      <td>0.001859</td>\n",
       "      <td>-0.002344</td>\n",
       "      <td>0.016263</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>-0.016565</td>\n",
       "      <td>-0.014142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>0.008229</td>\n",
       "      <td>-0.000685</td>\n",
       "      <td>-0.005215</td>\n",
       "      <td>-0.002029</td>\n",
       "      <td>-0.028069</td>\n",
       "      <td>0.014748</td>\n",
       "      <td>-0.019240</td>\n",
       "      <td>-0.002403</td>\n",
       "      <td>-0.008507</td>\n",
       "      <td>-0.019092</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.006856</td>\n",
       "      <td>0.010676</td>\n",
       "      <td>0.008874</td>\n",
       "      <td>0.018497</td>\n",
       "      <td>0.000700</td>\n",
       "      <td>-0.013267</td>\n",
       "      <td>0.015265</td>\n",
       "      <td>0.016886</td>\n",
       "      <td>-0.015319</td>\n",
       "      <td>-0.020727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>-0.004941</td>\n",
       "      <td>-0.001849</td>\n",
       "      <td>-0.002326</td>\n",
       "      <td>0.006075</td>\n",
       "      <td>0.001607</td>\n",
       "      <td>0.009678</td>\n",
       "      <td>0.001661</td>\n",
       "      <td>0.003754</td>\n",
       "      <td>-0.000152</td>\n",
       "      <td>-0.001013</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014761</td>\n",
       "      <td>0.005071</td>\n",
       "      <td>0.001496</td>\n",
       "      <td>-0.002506</td>\n",
       "      <td>-0.002819</td>\n",
       "      <td>0.005800</td>\n",
       "      <td>0.003975</td>\n",
       "      <td>-0.010976</td>\n",
       "      <td>-0.008756</td>\n",
       "      <td>0.006169</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "name       YieP      GlpR      YgbI  efeU-repair   ydcI-KO   BW25113  \\\n",
       "b0002  0.072693 -0.006320 -0.038374    -0.010045 -0.004345 -0.005599   \n",
       "b0003  0.015975 -0.009969 -0.024571    -0.006775 -0.003583  0.006680   \n",
       "b0004  0.015319 -0.003808 -0.027018    -0.004455 -0.008418  0.003245   \n",
       "b0005  0.008229 -0.000685 -0.005215    -0.002029 -0.028069  0.014748   \n",
       "b0006 -0.004941 -0.001849 -0.002326     0.006075  0.001607  0.009678   \n",
       "\n",
       "name     proVWX  DhaR/Mlc      ArgR      YneJ  ...      RpoS  membrane  \\\n",
       "b0002  0.009043 -0.005072 -0.012186 -0.013543  ... -0.010152 -0.012124   \n",
       "b0003 -0.010841 -0.012620 -0.003685  0.003512  ...  0.004702  0.004402   \n",
       "b0004 -0.016494 -0.013818 -0.008104  0.002144  ...  0.003507  0.005342   \n",
       "b0005 -0.019240 -0.002403 -0.008507 -0.019092  ... -0.006856  0.010676   \n",
       "b0006  0.001661  0.003754 -0.000152 -0.001013  ... -0.014761  0.005071   \n",
       "\n",
       "name   duplication-1  uncharacterized-3  ExuR/FucR      FecI  iron-related  \\\n",
       "b0002      -0.012027          -0.002123  -0.000036 -0.013667     -0.018883   \n",
       "b0003       0.001182          -0.004415  -0.001514 -0.005052      0.015502   \n",
       "b0004       0.003759           0.000700   0.001859 -0.002344      0.016263   \n",
       "b0005       0.008874           0.018497   0.000700 -0.013267      0.015265   \n",
       "b0006       0.001496          -0.002506  -0.002819  0.005800      0.003975   \n",
       "\n",
       "name       RbsR      sgrT  insertion  \n",
       "b0002  0.004408 -0.006315  -0.001780  \n",
       "b0003  0.000860 -0.022438  -0.020087  \n",
       "b0004  0.002486 -0.016565  -0.014142  \n",
       "b0005  0.016886 -0.015319  -0.020727  \n",
       "b0006 -0.010976 -0.008756   0.006169  \n",
       "\n",
       "[5 rows x 92 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This also renames the M and A matrix\n",
    "ica_data.M.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control__wt_glc__1</th>\n",
       "      <th>control__wt_glc__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_dpd__1</th>\n",
       "      <th>fur__delfur_dpd__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>...</th>\n",
       "      <th>efeU__menFentC_ale29__1</th>\n",
       "      <th>efeU__menFentC_ale29__2</th>\n",
       "      <th>efeU__menFentC_ale30__1</th>\n",
       "      <th>efeU__menFentC_ale30__2</th>\n",
       "      <th>efeU__menFentCubiC_ale36__1</th>\n",
       "      <th>efeU__menFentCubiC_ale36__2</th>\n",
       "      <th>efeU__menFentCubiC_ale37__1</th>\n",
       "      <th>efeU__menFentCubiC_ale37__2</th>\n",
       "      <th>efeU__menFentCubiC_ale38__1</th>\n",
       "      <th>efeU__menFentCubiC_ale38__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>0.076035</td>\n",
       "      <td>-0.076035</td>\n",
       "      <td>1.483992</td>\n",
       "      <td>1.719034</td>\n",
       "      <td>-1.184220</td>\n",
       "      <td>-1.350203</td>\n",
       "      <td>1.451882</td>\n",
       "      <td>1.329577</td>\n",
       "      <td>0.265718</td>\n",
       "      <td>0.024864</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.892372</td>\n",
       "      <td>-1.022146</td>\n",
       "      <td>0.253722</td>\n",
       "      <td>-0.880352</td>\n",
       "      <td>0.526509</td>\n",
       "      <td>1.261740</td>\n",
       "      <td>1.727782</td>\n",
       "      <td>1.512944</td>\n",
       "      <td>2.584012</td>\n",
       "      <td>2.787686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>-0.408568</td>\n",
       "      <td>0.408568</td>\n",
       "      <td>1.760680</td>\n",
       "      <td>1.357274</td>\n",
       "      <td>-0.446482</td>\n",
       "      <td>-0.931725</td>\n",
       "      <td>0.517458</td>\n",
       "      <td>0.427993</td>\n",
       "      <td>-0.499529</td>\n",
       "      <td>-0.581335</td>\n",
       "      <td>...</td>\n",
       "      <td>1.781224</td>\n",
       "      <td>2.785099</td>\n",
       "      <td>0.439133</td>\n",
       "      <td>-0.015471</td>\n",
       "      <td>0.366544</td>\n",
       "      <td>0.369589</td>\n",
       "      <td>0.325470</td>\n",
       "      <td>1.027508</td>\n",
       "      <td>1.629269</td>\n",
       "      <td>2.154717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>0.103912</td>\n",
       "      <td>-0.103912</td>\n",
       "      <td>0.847276</td>\n",
       "      <td>1.530875</td>\n",
       "      <td>0.258492</td>\n",
       "      <td>0.188903</td>\n",
       "      <td>1.232846</td>\n",
       "      <td>0.883215</td>\n",
       "      <td>0.505356</td>\n",
       "      <td>0.132219</td>\n",
       "      <td>...</td>\n",
       "      <td>1.422466</td>\n",
       "      <td>2.396554</td>\n",
       "      <td>2.130935</td>\n",
       "      <td>1.656316</td>\n",
       "      <td>0.864338</td>\n",
       "      <td>0.093698</td>\n",
       "      <td>0.641705</td>\n",
       "      <td>1.306455</td>\n",
       "      <td>1.366912</td>\n",
       "      <td>2.452061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>-0.456843</td>\n",
       "      <td>0.456843</td>\n",
       "      <td>5.890468</td>\n",
       "      <td>6.051415</td>\n",
       "      <td>-0.914670</td>\n",
       "      <td>-1.087245</td>\n",
       "      <td>5.210846</td>\n",
       "      <td>5.028641</td>\n",
       "      <td>3.820172</td>\n",
       "      <td>3.961980</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.343743</td>\n",
       "      <td>-1.991048</td>\n",
       "      <td>3.520794</td>\n",
       "      <td>3.105551</td>\n",
       "      <td>6.620301</td>\n",
       "      <td>5.400225</td>\n",
       "      <td>5.545880</td>\n",
       "      <td>5.555621</td>\n",
       "      <td>9.672308</td>\n",
       "      <td>10.109228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>0.269057</td>\n",
       "      <td>-0.269057</td>\n",
       "      <td>-0.254271</td>\n",
       "      <td>0.184537</td>\n",
       "      <td>0.651527</td>\n",
       "      <td>0.691914</td>\n",
       "      <td>0.016636</td>\n",
       "      <td>-0.465781</td>\n",
       "      <td>1.212097</td>\n",
       "      <td>1.122801</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.646464</td>\n",
       "      <td>-3.606055</td>\n",
       "      <td>-0.665810</td>\n",
       "      <td>-0.830586</td>\n",
       "      <td>-1.044574</td>\n",
       "      <td>0.119876</td>\n",
       "      <td>-0.045670</td>\n",
       "      <td>-0.826608</td>\n",
       "      <td>-1.393160</td>\n",
       "      <td>-1.587978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 278 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             control__wt_glc__1  control__wt_glc__2  fur__wt_dpd__1  \\\n",
       "name                                                                  \n",
       "YieP                   0.076035           -0.076035        1.483992   \n",
       "GlpR                  -0.408568            0.408568        1.760680   \n",
       "YgbI                   0.103912           -0.103912        0.847276   \n",
       "efeU-repair           -0.456843            0.456843        5.890468   \n",
       "ydcI-KO                0.269057           -0.269057       -0.254271   \n",
       "\n",
       "             fur__wt_dpd__2  fur__wt_fe__1  fur__wt_fe__2  fur__delfur_dpd__1  \\\n",
       "name                                                                            \n",
       "YieP               1.719034      -1.184220      -1.350203            1.451882   \n",
       "GlpR               1.357274      -0.446482      -0.931725            0.517458   \n",
       "YgbI               1.530875       0.258492       0.188903            1.232846   \n",
       "efeU-repair        6.051415      -0.914670      -1.087245            5.210846   \n",
       "ydcI-KO            0.184537       0.651527       0.691914            0.016636   \n",
       "\n",
       "             fur__delfur_dpd__2  fur__delfur_fe2__1  fur__delfur_fe2__2  ...  \\\n",
       "name                                                                     ...   \n",
       "YieP                   1.329577            0.265718            0.024864  ...   \n",
       "GlpR                   0.427993           -0.499529           -0.581335  ...   \n",
       "YgbI                   0.883215            0.505356            0.132219  ...   \n",
       "efeU-repair            5.028641            3.820172            3.961980  ...   \n",
       "ydcI-KO               -0.465781            1.212097            1.122801  ...   \n",
       "\n",
       "             efeU__menFentC_ale29__1  efeU__menFentC_ale29__2  \\\n",
       "name                                                            \n",
       "YieP                       -0.892372                -1.022146   \n",
       "GlpR                        1.781224                 2.785099   \n",
       "YgbI                        1.422466                 2.396554   \n",
       "efeU-repair                -2.343743                -1.991048   \n",
       "ydcI-KO                    -3.646464                -3.606055   \n",
       "\n",
       "             efeU__menFentC_ale30__1  efeU__menFentC_ale30__2  \\\n",
       "name                                                            \n",
       "YieP                        0.253722                -0.880352   \n",
       "GlpR                        0.439133                -0.015471   \n",
       "YgbI                        2.130935                 1.656316   \n",
       "efeU-repair                 3.520794                 3.105551   \n",
       "ydcI-KO                    -0.665810                -0.830586   \n",
       "\n",
       "             efeU__menFentCubiC_ale36__1  efeU__menFentCubiC_ale36__2  \\\n",
       "name                                                                    \n",
       "YieP                            0.526509                     1.261740   \n",
       "GlpR                            0.366544                     0.369589   \n",
       "YgbI                            0.864338                     0.093698   \n",
       "efeU-repair                     6.620301                     5.400225   \n",
       "ydcI-KO                        -1.044574                     0.119876   \n",
       "\n",
       "             efeU__menFentCubiC_ale37__1  efeU__menFentCubiC_ale37__2  \\\n",
       "name                                                                    \n",
       "YieP                            1.727782                     1.512944   \n",
       "GlpR                            0.325470                     1.027508   \n",
       "YgbI                            0.641705                     1.306455   \n",
       "efeU-repair                     5.545880                     5.555621   \n",
       "ydcI-KO                        -0.045670                    -0.826608   \n",
       "\n",
       "             efeU__menFentCubiC_ale38__1  efeU__menFentCubiC_ale38__2  \n",
       "name                                                                   \n",
       "YieP                            2.584012                     2.787686  \n",
       "GlpR                            1.629269                     2.154717  \n",
       "YgbI                            1.366912                     2.452061  \n",
       "efeU-repair                     9.672308                    10.109228  \n",
       "ydcI-KO                        -1.393160                    -1.587978  \n",
       "\n",
       "[5 rows x 278 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.A.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying \"single-gene\" iModulons\n",
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "2. Artificial knock-out of single genes\n",
    "3. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The `find_single_gene_imodulons` function identifies iModulons that are likely dominated by a single gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ydcI-KO', 'fur-KO', 'Pyruvate', 'thrA-KO', 'sgrT']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.find_single_gene_imodulons(save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>fur-KO</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for fur knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       regulator  pvalue  qvalue  precision  recall  f1score  TP  \\\n",
       "name                                                               \n",
       "fur-KO       NaN     NaN     NaN        NaN     NaN      NaN NaN   \n",
       "\n",
       "        regulon_size  n_regs                    Function             Category  \\\n",
       "name                                                                            \n",
       "fur-KO            13       0  Accounts for fur knock-out  Genomic Alterations   \n",
       "\n",
       "       single_gene  \n",
       "name                \n",
       "fur-KO        True  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.loc[['fur-KO']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saving and Loading IcaData Objects\n",
    "To facilitate data sharing, you can save IcaData objects as json files that can be easily re-loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data,'../example_data/example.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = load_json_model('../example_data/example.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            regulator  pvalue  qvalue  precision  recall  f1score   TP  \\\n",
       "YieP             None     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "GlpR             glpR     0.0     NaN        1.0     1.0      4.0  9.0   \n",
       "YgbI             None     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "efeU-repair      None     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "ydcI-KO          None     NaN     NaN        NaN     NaN      NaN  NaN   \n",
       "\n",
       "             regulon_size  n_regs  \\\n",
       "YieP                   11       1   \n",
       "GlpR                    9       1   \n",
       "YgbI                    9       1   \n",
       "efeU-repair             8       0   \n",
       "ydcI-KO                 3       0   \n",
       "\n",
       "                                                      Function  \\\n",
       "YieP                                                   Unknown   \n",
       "GlpR                                       Glycerol catabolism   \n",
       "YgbI                                          Unknown Function   \n",
       "efeU-repair  Accounts for repair and expression of efeU operon   \n",
       "ydcI-KO                            Accounts for ydcI knock-out   \n",
       "\n",
       "                              Category  single_gene  \n",
       "YieP               Regulator Discovery          NaN  \n",
       "GlpR         Carbon Source Utilization          NaN  \n",
       "YgbI               Regulator Discovery          NaN  \n",
       "efeU-repair        Genomic Alterations          NaN  \n",
       "ydcI-KO            Genomic Alterations          1.0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explained variance\n",
    "To compute the total amount of expression variation explained by iModulons, use the `explained_variance` function. You can also limit by genes, samples, and iModulons selected to compute partial explained variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69.46207831718387"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "explained_variance(ica_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "92.783126613826"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "explained_variance(ica_data,genes='thrA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code shows you how to identify iModulons that explain the largest fraction of variance in the dataset. These iModulons tend to represent global regulators."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "rec_var = {}\n",
    "for k in ica_data.imodulon_names:\n",
    "    rec_var[k] = explained_variance(ica_data,imodulons=k)\n",
    "df_rec_var = pd.Series(rec_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RpoS                 4.324150\n",
       "FlhDC                4.217993\n",
       "Fnr                  3.485507\n",
       "FliA                 3.425892\n",
       "uncharacterized-5    3.285922\n",
       "Fur-1                2.054392\n",
       "Crp-1                1.861757\n",
       "ArcA-1               1.790581\n",
       "PurR-2               1.755976\n",
       "NtrC+RpoN            1.678451\n",
       "dtype: float64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rec_var.sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.enrichment import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform a basic enrichment test between two samples, use the `compute_enrichment` function.\n",
    "Optional arguments:\n",
    "* `label`: Label for your target gene set (e.g. regulator name or GO term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set = ['b0002','b0003','b0004'] # e.g. an iModulon or other gene set\n",
    "target_genes = ['b0002','b0005','b0007'] # e.g. a regulon or genes in a COG category\n",
    "all_genes = ica_data.M.index # List of all genes in your expression dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pvalue             0.002293\n",
       "precision          0.333333\n",
       "recall             0.333333\n",
       "f1score            0.333333\n",
       "TP                 1.000000\n",
       "target_set_size    3.000000\n",
       "gene_set_size      3.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_enrichment(gene_set,target_genes,all_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform an enrichment test against a regulon in a TRN table, use the `compute_regulon` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/anand/PycharmProjects/pymodulon/pymodulon/enrichment.py:149: UserWarning: Some genes are in the regulon but not in all_genes. These genes are removed before enrichment analysis.\n",
      "  warnings.warn('Some genes are in the regulon but not in all_genes. '\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "pvalue             0.000131\n",
       "precision          0.015000\n",
       "recall             1.000000\n",
       "f1score            0.029557\n",
       "TP                 3.000000\n",
       "regulon_size     200.000000\n",
       "gene_set_size      3.000000\n",
       "n_regs             1.000000\n",
       "Name: lrp, dtype: float64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_regulon_enrichment(gene_set,'lrp',all_genes,trn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To perform an enrichment against all regulons in a TRN table, use the `compute_trn_enrichment` function. Optional arguments:\n",
    "* `max_regs`: Maximum number of regulators to include in complex regulon (default: 1) \n",
    "* `fdr`: False detection rate for Benjamini Hochberg correction\n",
    "* `method`: How to combine complex regulons. `'or'` computes enrichment against union of regulons, `'and'` computes enrichment against intersection of regulons, and `'both'` performs both tests (default: `'both'`) \n",
    "* `force`: Allows computation of >2 regulators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>gene_set_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>qvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thr-tRNA</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ile-tRNA</th>\n",
       "      <td>8.354256e-09</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pvalue  precision  recall  f1score   TP  regulon_size  \\\n",
       "thr-tRNA  0.000000e+00   1.000000     1.0      1.0  3.0           3.0   \n",
       "ile-tRNA  8.354256e-09   0.333333     1.0      0.5  3.0           9.0   \n",
       "\n",
       "          gene_set_size  n_regs    qvalue  \n",
       "thr-tRNA            3.0     1.0  0.000000  \n",
       "ile-tRNA            3.0     1.0  0.000002  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_trn_enrichment(gene_set,all_genes,trn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Using >2 maximum regulators may take time to compute. To perform analysis, use force=True",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-41-48a7f67043f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcompute_trn_enrichment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_set\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mall_genes\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtrn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmax_regs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/PycharmProjects/pymodulon/pymodulon/enrichment.py\u001b[0m in \u001b[0;36mcompute_trn_enrichment\u001b[0;34m(gene_set, all_genes, trn, max_regs, fdr, method, force)\u001b[0m\n\u001b[1;32m    183\u001b[0m     \u001b[0;31m# Warning if max_regs is too high\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mmax_regs\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mforce\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 185\u001b[0;31m         raise ValueError(\n\u001b[0m\u001b[1;32m    186\u001b[0m             \u001b[0;34m'Using >2 maximum regulators may take time to compute. '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    187\u001b[0m             'To perform analysis, use force=True')\n",
      "\u001b[0;31mValueError\u001b[0m: Using >2 maximum regulators may take time to compute. To perform analysis, use force=True"
     ]
    }
   ],
   "source": [
    "compute_trn_enrichment(gene_set,all_genes,trn,max_regs=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/anand/PycharmProjects/pymodulon/pymodulon/enrichment.py:149: UserWarning: Some genes are in the regulon but not in all_genes. These genes are removed before enrichment analysis.\n",
      "  warnings.warn('Some genes are in the regulon but not in all_genes. '\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>gene_set_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>qvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>arcA+ile-tRNA+rpoD</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arcA+lrp+thr-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arcA+rpoD+thr-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>crp+fnr+ile-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>crp+fnr+thr-tRNA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    pvalue  precision  recall  f1score   TP  regulon_size  \\\n",
       "arcA+ile-tRNA+rpoD     0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "arcA+lrp+thr-tRNA      0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "arcA+rpoD+thr-tRNA     0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "crp+fnr+ile-tRNA       0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "crp+fnr+thr-tRNA       0.0        1.0     1.0      1.0  3.0           3.0   \n",
       "\n",
       "                    gene_set_size  n_regs  qvalue  \n",
       "arcA+ile-tRNA+rpoD            3.0     3.0     0.0  \n",
       "arcA+lrp+thr-tRNA             3.0     3.0     0.0  \n",
       "arcA+rpoD+thr-tRNA            3.0     3.0     0.0  \n",
       "crp+fnr+ile-tRNA              3.0     3.0     0.0  \n",
       "crp+fnr+thr-tRNA              3.0     3.0     0.0  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_trn_enrichment(gene_set,all_genes,trn,max_regs=3,force=True).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the IcaData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pvalue           6.515058e-24\n",
       "precision        8.888889e-01\n",
       "recall           1.000000e+00\n",
       "f1score          9.411765e-01\n",
       "TP               8.000000e+00\n",
       "regulon_size     9.000000e+00\n",
       "imodulon_size    8.000000e+00\n",
       "n_regs           1.000000e+00\n",
       "Name: glpR, dtype: float64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compare a single iModulon to a simple regulon\n",
    "ica_data.compute_regulon_enrichment('GlpR',regulator='glpR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pvalue           3.933564e-10\n",
       "precision        2.105263e-01\n",
       "recall           1.000000e+00\n",
       "f1score          3.478261e-01\n",
       "TP               4.000000e+00\n",
       "regulon_size     1.900000e+01\n",
       "imodulon_size    4.000000e+00\n",
       "n_regs           2.000000e+00\n",
       "Name: cusR/cueR, dtype: float64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compare a single iModulon to a complex regulon\n",
    "ica_data.compute_regulon_enrichment('Copper','cusR/cueR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GlpR</td>\n",
       "      <td>glpR</td>\n",
       "      <td>6.515058e-24</td>\n",
       "      <td>3.075107e-21</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>efeU-repair</td>\n",
       "      <td>fur</td>\n",
       "      <td>1.043681e-08</td>\n",
       "      <td>4.926172e-06</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.076087</td>\n",
       "      <td>7.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BW25113</td>\n",
       "      <td>mhpR</td>\n",
       "      <td>1.084432e-09</td>\n",
       "      <td>5.118520e-07</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.421053</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>proVWX</td>\n",
       "      <td>betI</td>\n",
       "      <td>9.274727e-09</td>\n",
       "      <td>4.377671e-06</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DhaR/Mlc</td>\n",
       "      <td>mlc</td>\n",
       "      <td>4.163677e-11</td>\n",
       "      <td>1.965255e-08</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>iron-related</td>\n",
       "      <td>fur</td>\n",
       "      <td>2.433376e-09</td>\n",
       "      <td>5.373956e-07</td>\n",
       "      <td>0.097143</td>\n",
       "      <td>0.253731</td>\n",
       "      <td>0.140496</td>\n",
       "      <td>17.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>iron-related</td>\n",
       "      <td>ihf</td>\n",
       "      <td>1.353532e-09</td>\n",
       "      <td>5.373956e-07</td>\n",
       "      <td>0.087558</td>\n",
       "      <td>0.283582</td>\n",
       "      <td>0.133803</td>\n",
       "      <td>19.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>iron-related</td>\n",
       "      <td>arcA</td>\n",
       "      <td>6.417929e-09</td>\n",
       "      <td>6.058525e-07</td>\n",
       "      <td>0.057269</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.099808</td>\n",
       "      <td>26.0</td>\n",
       "      <td>454.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>iron-related</td>\n",
       "      <td>narL</td>\n",
       "      <td>5.852193e-09</td>\n",
       "      <td>6.058525e-07</td>\n",
       "      <td>0.118644</td>\n",
       "      <td>0.208955</td>\n",
       "      <td>0.151351</td>\n",
       "      <td>14.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>RbsR</td>\n",
       "      <td>rbsR</td>\n",
       "      <td>3.488522e-15</td>\n",
       "      <td>1.646582e-12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.631579</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0            GlpR      glpR  6.515058e-24  3.075107e-21   0.888889  1.000000   \n",
       "1     efeU-repair       fur  1.043681e-08  4.926172e-06   0.040000  0.777778   \n",
       "2         BW25113      mhpR  1.084432e-09  5.118520e-07   0.666667  0.307692   \n",
       "3          proVWX      betI  9.274727e-09  4.377671e-06   1.000000  0.100000   \n",
       "4        DhaR/Mlc       mlc  4.163677e-11  1.965255e-08   0.500000  0.384615   \n",
       "..            ...       ...           ...           ...        ...       ...   \n",
       "151  iron-related       fur  2.433376e-09  5.373956e-07   0.097143  0.253731   \n",
       "152  iron-related       ihf  1.353532e-09  5.373956e-07   0.087558  0.283582   \n",
       "153  iron-related      arcA  6.417929e-09  6.058525e-07   0.057269  0.388060   \n",
       "154  iron-related      narL  5.852193e-09  6.058525e-07   0.118644  0.208955   \n",
       "155          RbsR      rbsR  3.488522e-15  1.646582e-12   0.666667  0.600000   \n",
       "\n",
       "      f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0    0.941176   8.0           9.0            8.0     1.0  \n",
       "1    0.076087   7.0         175.0            9.0     1.0  \n",
       "2    0.421053   4.0           6.0           13.0     1.0  \n",
       "3    0.181818   4.0           4.0           40.0     1.0  \n",
       "4    0.434783   5.0          10.0           13.0     1.0  \n",
       "..        ...   ...           ...            ...     ...  \n",
       "151  0.140496  17.0         175.0           67.0     1.0  \n",
       "152  0.133803  19.0         217.0           67.0     1.0  \n",
       "153  0.099808  26.0         454.0           67.0     1.0  \n",
       "154  0.151351  14.0         118.0           67.0     1.0  \n",
       "155  0.631579   6.0           9.0           10.0     1.0  \n",
       "\n",
       "[156 rows x 11 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all significant enrichments for all iModulons. \n",
    "# Optional parameters are same as the global compute_trn_enrichment function\n",
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save to sample table\n",
    "You can directly save iModulon enrichments to the sample table using either the `ica_data.compute_regulon_enrichment` function or `ica_data.compute_trn_enrichment` function. Columns unrelated to regulon enrichments (e.g. `Function` and `Category`) are left alone."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator  pvalue  qvalue  precision  recall  f1score   TP  regulon_size  \\\n",
       "GlpR      glpR     0.0     NaN        1.0     1.0      4.0  9.0             9   \n",
       "\n",
       "      n_regs             Function                   Category  single_gene  \n",
       "GlpR       1  Glycerol catabolism  Carbon Source Utilization          NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.loc[['GlpR']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>single_gene</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>crp</td>\n",
       "      <td>2.412433e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.013629</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.026891</td>\n",
       "      <td>8.0</td>\n",
       "      <td>587.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator        pvalue  qvalue  precision  recall   f1score   TP  \\\n",
       "GlpR       crp  2.412433e-07     NaN   0.013629     1.0  0.026891  8.0   \n",
       "\n",
       "      regulon_size  n_regs             Function                   Category  \\\n",
       "GlpR         587.0     1.0  Glycerol catabolism  Carbon Source Utilization   \n",
       "\n",
       "      single_gene  imodulon_size  \n",
       "GlpR          NaN            8.0  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.compute_regulon_enrichment('GlpR',regulator='crp',save=True)\n",
    "ica_data.imodulon_table.loc[['GlpR']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>Function</th>\n",
       "      <th>single_gene</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>glpR</td>\n",
       "      <td>6.515058e-24</td>\n",
       "      <td>3.075107e-21</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Glycerol catabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Unknown Function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Regulator Discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>fur</td>\n",
       "      <td>1.043681e-08</td>\n",
       "      <td>4.926172e-06</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.076087</td>\n",
       "      <td>7.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Accounts for repair and expression of efeU operon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Accounts for ydcI knock-out</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FecI</th>\n",
       "      <td>fecI</td>\n",
       "      <td>1.631427e-14</td>\n",
       "      <td>7.700337e-12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Ferric citrate transport</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metal Homeostasis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iron-related</th>\n",
       "      <td>fnr</td>\n",
       "      <td>3.415650e-09</td>\n",
       "      <td>5.373956e-07</td>\n",
       "      <td>0.058957</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.102362</td>\n",
       "      <td>26.0</td>\n",
       "      <td>441.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Related to iron metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biological Enrichment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RbsR</th>\n",
       "      <td>rbsR</td>\n",
       "      <td>3.488522e-15</td>\n",
       "      <td>1.646582e-12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.631579</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>D-ribose catabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Carbon Source Utilization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgrT</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Contains single dominating gene: sgrT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Biological Enrichment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insertion</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>IS2 insertion element after laboratory evolution</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genomic Alterations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             regulator        pvalue        qvalue  precision    recall  \\\n",
       "YieP              None           NaN           NaN        NaN       NaN   \n",
       "GlpR              glpR  6.515058e-24  3.075107e-21   0.888889  1.000000   \n",
       "YgbI              None           NaN           NaN        NaN       NaN   \n",
       "efeU-repair        fur  1.043681e-08  4.926172e-06   0.040000  0.777778   \n",
       "ydcI-KO           None           NaN           NaN        NaN       NaN   \n",
       "...                ...           ...           ...        ...       ...   \n",
       "FecI              fecI  1.631427e-14  7.700337e-12   1.000000  0.555556   \n",
       "iron-related       fnr  3.415650e-09  5.373956e-07   0.058957  0.388060   \n",
       "RbsR              rbsR  3.488522e-15  1.646582e-12   0.666667  0.600000   \n",
       "sgrT              None           NaN           NaN        NaN       NaN   \n",
       "insertion         None           NaN           NaN        NaN       NaN   \n",
       "\n",
       "               f1score    TP  regulon_size  imodulon_size  n_regs  \\\n",
       "YieP               NaN   NaN          11.0            NaN     1.0   \n",
       "GlpR          0.941176   8.0           9.0            8.0     1.0   \n",
       "YgbI               NaN   NaN           9.0            NaN     1.0   \n",
       "efeU-repair   0.076087   7.0         175.0            9.0     1.0   \n",
       "ydcI-KO            NaN   NaN           3.0            NaN     0.0   \n",
       "...                ...   ...           ...            ...     ...   \n",
       "FecI          0.714286   5.0           5.0            9.0     1.0   \n",
       "iron-related  0.102362  26.0         441.0           67.0     1.0   \n",
       "RbsR          0.631579   6.0           9.0           10.0     1.0   \n",
       "sgrT               NaN   NaN           4.0            NaN     0.0   \n",
       "insertion          NaN   NaN           9.0            NaN     0.0   \n",
       "\n",
       "                                                       Function  single_gene  \\\n",
       "YieP                                                    Unknown          NaN   \n",
       "GlpR                                        Glycerol catabolism          NaN   \n",
       "YgbI                                           Unknown Function          NaN   \n",
       "efeU-repair   Accounts for repair and expression of efeU operon          NaN   \n",
       "ydcI-KO                             Accounts for ydcI knock-out          1.0   \n",
       "...                                                         ...          ...   \n",
       "FecI                                   Ferric citrate transport          NaN   \n",
       "iron-related                         Related to iron metabolism          NaN   \n",
       "RbsR                                        D-ribose catabolism          NaN   \n",
       "sgrT                      Contains single dominating gene: sgrT          1.0   \n",
       "insertion      IS2 insertion element after laboratory evolution          NaN   \n",
       "\n",
       "                               Category  \n",
       "YieP                Regulator Discovery  \n",
       "GlpR          Carbon Source Utilization  \n",
       "YgbI                Regulator Discovery  \n",
       "efeU-repair         Genomic Alterations  \n",
       "ydcI-KO             Genomic Alterations  \n",
       "...                                 ...  \n",
       "FecI                  Metal Homeostasis  \n",
       "iron-related      Biological Enrichment  \n",
       "RbsR          Carbon Source Utilization  \n",
       "sgrT              Biological Enrichment  \n",
       "insertion           Genomic Alterations  \n",
       "\n",
       "[92 rows x 13 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The enrichment with the lowest qvalue is saved as the official enrichment\n",
    "ica_data.compute_trn_enrichment(save=True)\n",
    "ica_data.imodulon_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iModulon Thresholds\n",
    "iModulon thresholds are computed using an outlier test (See Methods section \"*Determination of the gene coefficient threshold*\" in [Sastry et al 2019](https://www.nature.com/articles/s41467-019-13483-w#Sec11)). This outlier test relies on the D'agostino test statistic.\n",
    "* If you have a relatively complete TRN, you may fit this test statistic using the `reoptimize_thresholds` function.\n",
    "* To manually adjust this statistic, use the `recompute_thresholds` function.\n",
    "* To change a single iModulon threshold, use the `change_threshold` function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fitting the threshold\n",
    "The `reoptimize_threshold` function may take a few minutes to complete. Fortunately, this only needs to be performed once. You can turn off the progress bar or plot using the `progress` and `plot` arguments, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data.reoptimize_thresholds()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Changing the D'agostino statistic\n",
    "To test out different D'agostino test statistics, use the `recompute_thresholds` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2239</th>\n",
       "      <td>0.211384</td>\n",
       "      <td>2349934</td>\n",
       "      <td>2351011</td>\n",
       "      <td>-</td>\n",
       "      <td>glpQ</td>\n",
       "      <td>1077</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2240</th>\n",
       "      <td>0.306134</td>\n",
       "      <td>2351015</td>\n",
       "      <td>2352374</td>\n",
       "      <td>-</td>\n",
       "      <td>glpT</td>\n",
       "      <td>1359</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2241</th>\n",
       "      <td>0.375662</td>\n",
       "      <td>2352646</td>\n",
       "      <td>2354275</td>\n",
       "      <td>+</td>\n",
       "      <td>glpA</td>\n",
       "      <td>1629</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2242</th>\n",
       "      <td>0.328961</td>\n",
       "      <td>2354264</td>\n",
       "      <td>2355524</td>\n",
       "      <td>+</td>\n",
       "      <td>glpB</td>\n",
       "      <td>1260</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2243</th>\n",
       "      <td>0.315752</td>\n",
       "      <td>2355520</td>\n",
       "      <td>2356711</td>\n",
       "      <td>+</td>\n",
       "      <td>glpC</td>\n",
       "      <td>1191</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3426</th>\n",
       "      <td>0.350034</td>\n",
       "      <td>3562012</td>\n",
       "      <td>3563518</td>\n",
       "      <td>+</td>\n",
       "      <td>glpD</td>\n",
       "      <td>1506</td>\n",
       "      <td>glpD</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,glpR,rpoD,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3926</th>\n",
       "      <td>0.290235</td>\n",
       "      <td>4115713</td>\n",
       "      <td>4117222</td>\n",
       "      <td>-</td>\n",
       "      <td>glpK</td>\n",
       "      <td>1509</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3927</th>\n",
       "      <td>0.312307</td>\n",
       "      <td>4117244</td>\n",
       "      <td>4118090</td>\n",
       "      <td>-</td>\n",
       "      <td>glpF</td>\n",
       "      <td>846</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length  operon  \\\n",
       "b2239     0.211384  2349934  2351011      -      glpQ    1077   glpTQ   \n",
       "b2240     0.306134  2351015  2352374      -      glpT    1359   glpTQ   \n",
       "b2241     0.375662  2352646  2354275      +      glpA    1629  glpABC   \n",
       "b2242     0.328961  2354264  2355524      +      glpB    1260  glpABC   \n",
       "b2243     0.315752  2355520  2356711      +      glpC    1191  glpABC   \n",
       "b3426     0.350034  3562012  3563518      +      glpD    1506    glpD   \n",
       "b3926     0.290235  4115713  4117222      -      glpK    1509  glpFKX   \n",
       "b3927     0.312307  4117244  4118090      -      glpF     846  glpFKX   \n",
       "\n",
       "                                         COG  \\\n",
       "b2239       Energy production and conversion   \n",
       "b2240  Carbohydrate transport and metabolism   \n",
       "b2241       Energy production and conversion   \n",
       "b2242    Amino acid transport and metabolism   \n",
       "b2243       Energy production and conversion   \n",
       "b3426       Energy production and conversion   \n",
       "b3926       Energy production and conversion   \n",
       "b3927  Carbohydrate transport and metabolism   \n",
       "\n",
       "                                  regulator  \n",
       "b2239         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2240         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2241  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2242  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2243  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b3426               arcA,crp,glpR,rpoD,yieP  \n",
       "b3926                         crp,glpR,rpoD  \n",
       "b3927                         crp,glpR,rpoD  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('GlpR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.recompute_thresholds(300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b1199</th>\n",
       "      <td>0.045686</td>\n",
       "      <td>1249124</td>\n",
       "      <td>1249757</td>\n",
       "      <td>-</td>\n",
       "      <td>dhaL</td>\n",
       "      <td>633</td>\n",
       "      <td>dhaKLM</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>arcA,dhaR,fnr,lrp,rpoS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1200</th>\n",
       "      <td>0.045491</td>\n",
       "      <td>1249767</td>\n",
       "      <td>1250838</td>\n",
       "      <td>-</td>\n",
       "      <td>dhaK</td>\n",
       "      <td>1071</td>\n",
       "      <td>dhaKLM</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>arcA,dhaR,fnr,lrp,rpoS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2239</th>\n",
       "      <td>0.211384</td>\n",
       "      <td>2349934</td>\n",
       "      <td>2351011</td>\n",
       "      <td>-</td>\n",
       "      <td>glpQ</td>\n",
       "      <td>1077</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2240</th>\n",
       "      <td>0.306134</td>\n",
       "      <td>2351015</td>\n",
       "      <td>2352374</td>\n",
       "      <td>-</td>\n",
       "      <td>glpT</td>\n",
       "      <td>1359</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2241</th>\n",
       "      <td>0.375662</td>\n",
       "      <td>2352646</td>\n",
       "      <td>2354275</td>\n",
       "      <td>+</td>\n",
       "      <td>glpA</td>\n",
       "      <td>1629</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2242</th>\n",
       "      <td>0.328961</td>\n",
       "      <td>2354264</td>\n",
       "      <td>2355524</td>\n",
       "      <td>+</td>\n",
       "      <td>glpB</td>\n",
       "      <td>1260</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2243</th>\n",
       "      <td>0.315752</td>\n",
       "      <td>2355520</td>\n",
       "      <td>2356711</td>\n",
       "      <td>+</td>\n",
       "      <td>glpC</td>\n",
       "      <td>1191</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3011</th>\n",
       "      <td>0.049324</td>\n",
       "      <td>3155354</td>\n",
       "      <td>3156518</td>\n",
       "      <td>+</td>\n",
       "      <td>yqhD</td>\n",
       "      <td>1164</td>\n",
       "      <td>yqhD-dkgA</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>rpoD,yqhC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3426</th>\n",
       "      <td>0.350034</td>\n",
       "      <td>3562012</td>\n",
       "      <td>3563518</td>\n",
       "      <td>+</td>\n",
       "      <td>glpD</td>\n",
       "      <td>1506</td>\n",
       "      <td>glpD</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,glpR,rpoD,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3925</th>\n",
       "      <td>0.084278</td>\n",
       "      <td>4114568</td>\n",
       "      <td>4115579</td>\n",
       "      <td>-</td>\n",
       "      <td>glpX</td>\n",
       "      <td>1011</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3926</th>\n",
       "      <td>0.290235</td>\n",
       "      <td>4115713</td>\n",
       "      <td>4117222</td>\n",
       "      <td>-</td>\n",
       "      <td>glpK</td>\n",
       "      <td>1509</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3927</th>\n",
       "      <td>0.312307</td>\n",
       "      <td>4117244</td>\n",
       "      <td>4118090</td>\n",
       "      <td>-</td>\n",
       "      <td>glpF</td>\n",
       "      <td>846</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length     operon  \\\n",
       "b1199     0.045686  1249124  1249757      -      dhaL     633     dhaKLM   \n",
       "b1200     0.045491  1249767  1250838      -      dhaK    1071     dhaKLM   \n",
       "b2239     0.211384  2349934  2351011      -      glpQ    1077      glpTQ   \n",
       "b2240     0.306134  2351015  2352374      -      glpT    1359      glpTQ   \n",
       "b2241     0.375662  2352646  2354275      +      glpA    1629     glpABC   \n",
       "b2242     0.328961  2354264  2355524      +      glpB    1260     glpABC   \n",
       "b2243     0.315752  2355520  2356711      +      glpC    1191     glpABC   \n",
       "b3011     0.049324  3155354  3156518      +      yqhD    1164  yqhD-dkgA   \n",
       "b3426     0.350034  3562012  3563518      +      glpD    1506       glpD   \n",
       "b3925     0.084278  4114568  4115579      -      glpX    1011     glpFKX   \n",
       "b3926     0.290235  4115713  4117222      -      glpK    1509     glpFKX   \n",
       "b3927     0.312307  4117244  4118090      -      glpF     846     glpFKX   \n",
       "\n",
       "                                         COG  \\\n",
       "b1199  Carbohydrate transport and metabolism   \n",
       "b1200  Carbohydrate transport and metabolism   \n",
       "b2239       Energy production and conversion   \n",
       "b2240  Carbohydrate transport and metabolism   \n",
       "b2241       Energy production and conversion   \n",
       "b2242    Amino acid transport and metabolism   \n",
       "b2243       Energy production and conversion   \n",
       "b3011       Energy production and conversion   \n",
       "b3426       Energy production and conversion   \n",
       "b3925  Carbohydrate transport and metabolism   \n",
       "b3926       Energy production and conversion   \n",
       "b3927  Carbohydrate transport and metabolism   \n",
       "\n",
       "                                  regulator  \n",
       "b1199                arcA,dhaR,fnr,lrp,rpoS  \n",
       "b1200                arcA,dhaR,fnr,lrp,rpoS  \n",
       "b2239         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2240         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2241  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2242  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2243  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b3011                             rpoD,yqhC  \n",
       "b3426               arcA,crp,glpR,rpoD,yieP  \n",
       "b3925                         crp,glpR,rpoD  \n",
       "b3926                         crp,glpR,rpoD  \n",
       "b3927                         crp,glpR,rpoD  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('GlpR')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manually changing thresholds\n",
    "To manually change a threshold, use the `change_threshold` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.change_threshold('GlpR',0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>COG</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2239</th>\n",
       "      <td>0.211384</td>\n",
       "      <td>2349934</td>\n",
       "      <td>2351011</td>\n",
       "      <td>-</td>\n",
       "      <td>glpQ</td>\n",
       "      <td>1077</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2240</th>\n",
       "      <td>0.306134</td>\n",
       "      <td>2351015</td>\n",
       "      <td>2352374</td>\n",
       "      <td>-</td>\n",
       "      <td>glpT</td>\n",
       "      <td>1359</td>\n",
       "      <td>glpTQ</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,fis,fnr,glpR,ihf,nac,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2241</th>\n",
       "      <td>0.375662</td>\n",
       "      <td>2352646</td>\n",
       "      <td>2354275</td>\n",
       "      <td>+</td>\n",
       "      <td>glpA</td>\n",
       "      <td>1629</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2242</th>\n",
       "      <td>0.328961</td>\n",
       "      <td>2354264</td>\n",
       "      <td>2355524</td>\n",
       "      <td>+</td>\n",
       "      <td>glpB</td>\n",
       "      <td>1260</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2243</th>\n",
       "      <td>0.315752</td>\n",
       "      <td>2355520</td>\n",
       "      <td>2356711</td>\n",
       "      <td>+</td>\n",
       "      <td>glpC</td>\n",
       "      <td>1191</td>\n",
       "      <td>glpABC</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3426</th>\n",
       "      <td>0.350034</td>\n",
       "      <td>3562012</td>\n",
       "      <td>3563518</td>\n",
       "      <td>+</td>\n",
       "      <td>glpD</td>\n",
       "      <td>1506</td>\n",
       "      <td>glpD</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>arcA,crp,glpR,rpoD,yieP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3926</th>\n",
       "      <td>0.290235</td>\n",
       "      <td>4115713</td>\n",
       "      <td>4117222</td>\n",
       "      <td>-</td>\n",
       "      <td>glpK</td>\n",
       "      <td>1509</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3927</th>\n",
       "      <td>0.312307</td>\n",
       "      <td>4117244</td>\n",
       "      <td>4118090</td>\n",
       "      <td>-</td>\n",
       "      <td>glpF</td>\n",
       "      <td>846</td>\n",
       "      <td>glpFKX</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>crp,glpR,rpoD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_weight    start     stop strand gene_name  length  operon  \\\n",
       "b2239     0.211384  2349934  2351011      -      glpQ    1077   glpTQ   \n",
       "b2240     0.306134  2351015  2352374      -      glpT    1359   glpTQ   \n",
       "b2241     0.375662  2352646  2354275      +      glpA    1629  glpABC   \n",
       "b2242     0.328961  2354264  2355524      +      glpB    1260  glpABC   \n",
       "b2243     0.315752  2355520  2356711      +      glpC    1191  glpABC   \n",
       "b3426     0.350034  3562012  3563518      +      glpD    1506    glpD   \n",
       "b3926     0.290235  4115713  4117222      -      glpK    1509  glpFKX   \n",
       "b3927     0.312307  4117244  4118090      -      glpF     846  glpFKX   \n",
       "\n",
       "                                         COG  \\\n",
       "b2239       Energy production and conversion   \n",
       "b2240  Carbohydrate transport and metabolism   \n",
       "b2241       Energy production and conversion   \n",
       "b2242    Amino acid transport and metabolism   \n",
       "b2243       Energy production and conversion   \n",
       "b3426       Energy production and conversion   \n",
       "b3926       Energy production and conversion   \n",
       "b3927  Carbohydrate transport and metabolism   \n",
       "\n",
       "                                  regulator  \n",
       "b2239         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2240         crp,fis,fnr,glpR,ihf,nac,rpoD  \n",
       "b2241  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2242  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b2243  arcA,crp,fis,flhD;flhC,fnr,glpR,rpoD  \n",
       "b3426               arcA,crp,glpR,rpoD,yieP  \n",
       "b3926                         crp,glpR,rpoD  \n",
       "b3927                         crp,glpR,rpoD  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('GlpR')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inferring iModulon activities of new data\n",
    "If you have a new dataset that is external to your project, or of a different strain, you can infer the iModulon activities. You'll need to confirm two things:\n",
    "1. Your dataset must use the same, or similar, gene identifiers as your ICA data object\n",
    "1. Data must be centered to a reference condition\n",
    "\n",
    "Only genes shared between your ICA data object and the new expression profiles will be used to project your data (i.e. 'not_a_gene' will be essentially ignored by the function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>-0.003615</td>\n",
       "      <td>-0.289353</td>\n",
       "      <td>0.636527</td>\n",
       "      <td>0.819793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>0.307588</td>\n",
       "      <td>0.128414</td>\n",
       "      <td>0.954439</td>\n",
       "      <td>1.334385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>-0.264995</td>\n",
       "      <td>-0.546017</td>\n",
       "      <td>-0.202697</td>\n",
       "      <td>0.119195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>0.428769</td>\n",
       "      <td>0.123564</td>\n",
       "      <td>-0.865171</td>\n",
       "      <td>-0.951179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>-0.144870</td>\n",
       "      <td>-0.090152</td>\n",
       "      <td>-0.131039</td>\n",
       "      <td>-0.124079</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fur__wt_fe__1  fur__wt_fe__2  fur__wt_dpd__1  fur__wt_dpd__2\n",
       "b0002      -0.003615      -0.289353        0.636527        0.819793\n",
       "b0003       0.307588       0.128414        0.954439        1.334385\n",
       "b0004      -0.264995      -0.546017       -0.202697        0.119195\n",
       "b0005       0.428769       0.123564       -0.865171       -0.951179\n",
       "b0006      -0.144870      -0.090152       -0.131039       -0.124079"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data = ica_data.X[['fur__wt_fe__1','fur__wt_fe__2','fur__wt_dpd__1','fur__wt_dpd__2']]\n",
    "test_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>0.142869</td>\n",
       "      <td>-0.142869</td>\n",
       "      <td>0.783011</td>\n",
       "      <td>0.966277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>0.089587</td>\n",
       "      <td>-0.089587</td>\n",
       "      <td>0.736438</td>\n",
       "      <td>1.116384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>0.140511</td>\n",
       "      <td>-0.140511</td>\n",
       "      <td>0.202810</td>\n",
       "      <td>0.524701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>0.152603</td>\n",
       "      <td>-0.152603</td>\n",
       "      <td>-1.141338</td>\n",
       "      <td>-1.227346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>-0.027359</td>\n",
       "      <td>0.027359</td>\n",
       "      <td>-0.013528</td>\n",
       "      <td>-0.006568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fur__wt_fe__1  fur__wt_fe__2  fur__wt_dpd__1  fur__wt_dpd__2\n",
       "b0002       0.142869      -0.142869        0.783011        0.966277\n",
       "b0003       0.089587      -0.089587        0.736438        1.116384\n",
       "b0004       0.140511      -0.140511        0.202810        0.524701\n",
       "b0005       0.152603      -0.152603       -1.141338       -1.227346\n",
       "b0006      -0.027359       0.027359       -0.013528       -0.006568"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Center your data to a reference condition (or replicates of the condition, in this case)\n",
    "centered_test_data = test_data.sub(test_data[['fur__wt_fe__1','fur__wt_fe__2']].mean(axis=1),axis=0)\n",
    "centered_test_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YieP</th>\n",
       "      <td>0.083389</td>\n",
       "      <td>-0.083389</td>\n",
       "      <td>2.569008</td>\n",
       "      <td>2.788578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GlpR</th>\n",
       "      <td>0.244025</td>\n",
       "      <td>-0.244025</td>\n",
       "      <td>2.517578</td>\n",
       "      <td>2.110451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YgbI</th>\n",
       "      <td>0.034920</td>\n",
       "      <td>-0.034920</td>\n",
       "      <td>0.589780</td>\n",
       "      <td>1.294429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeU-repair</th>\n",
       "      <td>0.104035</td>\n",
       "      <td>-0.104035</td>\n",
       "      <td>7.608159</td>\n",
       "      <td>7.924332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ydcI-KO</th>\n",
       "      <td>-0.012846</td>\n",
       "      <td>0.012846</td>\n",
       "      <td>-0.531643</td>\n",
       "      <td>-0.017053</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             fur__wt_fe__1  fur__wt_fe__2  fur__wt_dpd__1  fur__wt_dpd__2\n",
       "YieP              0.083389      -0.083389        2.569008        2.788578\n",
       "GlpR              0.244025      -0.244025        2.517578        2.110451\n",
       "YgbI              0.034920      -0.034920        0.589780        1.294429\n",
       "efeU-repair       0.104035      -0.104035        7.608159        7.924332\n",
       "ydcI-KO          -0.012846       0.012846       -0.531643       -0.017053"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "activities = infer_activities(ica_data,centered_test_data)\n",
    "activities.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_dpd__1</th>\n",
       "      <th>fur__wt_dpd__2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>proVWX</th>\n",
       "      <td>10.860168</td>\n",
       "      <td>11.736302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MetJ</th>\n",
       "      <td>11.066912</td>\n",
       "      <td>8.789003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nitrate-related</th>\n",
       "      <td>11.414305</td>\n",
       "      <td>12.855044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fur-1</th>\n",
       "      <td>22.464512</td>\n",
       "      <td>20.664499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RpoS</th>\n",
       "      <td>24.199762</td>\n",
       "      <td>25.250345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 fur__wt_dpd__1  fur__wt_dpd__2\n",
       "proVWX                10.860168       11.736302\n",
       "MetJ                  11.066912        8.789003\n",
       "nitrate-related       11.414305       12.855044\n",
       "Fur-1                 22.464512       20.664499\n",
       "RpoS                  24.199762       25.250345"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "activities[['fur__wt_dpd__1','fur__wt_dpd__2']].sort_values('fur__wt_dpd__1').tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparing ICA Runs\n",
    "\n",
    "`compare_ica` is a function specifically built to compare the components between different ICA runs. The function will return a Digraph plot which shows connections between ICA components of different runs, as well as Dictionary List of the one-to-one connections.\n",
    "\n",
    "The function requires a minimum of two inputs if running a comparison beween two ICA runs for the same organism:\n",
    "\n",
    "* S1: Pandas Dataframe of M matrix 1\n",
    "* S2: Pandas Dataframe of M Matrix 2\n",
    "\n",
    "`compare_ica` can be found in **pymodulon.compare**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.compare import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparsion within the same organism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load comparison M matrix\n",
    "ica_run_diff = pd.read_csv(\"../pymodulon/test/data/ecoli_M_mod.csv\",index_col=0)\n",
    "ica_run_diff.columns = ['run2_'+str(i) for i in ica_run_diff.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 92/92 [00:01<00:00, 74.99it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"720pt\" height=\"595pt\"\n",
       " viewBox=\"0.00 0.00 720.00 594.89\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(.6383 .6383) rotate(0) translate(4 928)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-928 1124,-928 1124,4 -4,4\"/>\n",
       "<!-- data2_run2_0 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_run2_0</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_0</text>\n",
       "</g>\n",
       "<!-- data2_run2_1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_run2_1</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_1</text>\n",
       "</g>\n",
       "<!-- data2_run2_10 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_run2_10</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_10</text>\n",
       "</g>\n",
       "<!-- data2_run2_11 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_run2_11</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_11</text>\n",
       "</g>\n",
       "<!-- data2_run2_12 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_run2_12</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-842.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_12</text>\n",
       "</g>\n",
       "<!-- data2_run2_13 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data2_run2_13</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_13</text>\n",
       "</g>\n",
       "<!-- data2_run2_14 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data2_run2_14</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_14</text>\n",
       "</g>\n",
       "<!-- data2_run2_15 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data2_run2_15</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-842.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_15</text>\n",
       "</g>\n",
       "<!-- data2_run2_16 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data2_run2_16</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_16</text>\n",
       "</g>\n",
       "<!-- data2_run2_17 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data2_run2_17</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_17</text>\n",
       "</g>\n",
       "<!-- data2_run2_18 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>data2_run2_18</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_18</text>\n",
       "</g>\n",
       "<!-- data2_run2_19 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>data2_run2_19</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_19</text>\n",
       "</g>\n",
       "<!-- data2_run2_2 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>data2_run2_2</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-738.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_2</text>\n",
       "</g>\n",
       "<!-- data2_run2_20 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>data2_run2_20</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_20</text>\n",
       "</g>\n",
       "<!-- data2_run2_21 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>data2_run2_21</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_21</text>\n",
       "</g>\n",
       "<!-- data2_run2_22 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>data2_run2_22</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_22</text>\n",
       "</g>\n",
       "<!-- data2_run2_23 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>data2_run2_23</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_23</text>\n",
       "</g>\n",
       "<!-- data2_run2_24 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>data2_run2_24</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_24</text>\n",
       "</g>\n",
       "<!-- data2_run2_25 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>data2_run2_25</title>\n",
       "<text text-anchor=\"middle\" x=\"862.5\" y=\"-738.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_25</text>\n",
       "</g>\n",
       "<!-- data2_run2_26 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>data2_run2_26</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-738.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_26</text>\n",
       "</g>\n",
       "<!-- data2_run2_27 -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>data2_run2_27</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_27</text>\n",
       "</g>\n",
       "<!-- data2_run2_28 -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>data2_run2_28</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_28</text>\n",
       "</g>\n",
       "<!-- data2_run2_29 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>data2_run2_29</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_29</text>\n",
       "</g>\n",
       "<!-- data2_run2_3 -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>data2_run2_3</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_3</text>\n",
       "</g>\n",
       "<!-- data2_run2_30 -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>data2_run2_30</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-634.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_30</text>\n",
       "</g>\n",
       "<!-- data2_run2_31 -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>data2_run2_31</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_31</text>\n",
       "</g>\n",
       "<!-- data2_run2_32 -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>data2_run2_32</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_32</text>\n",
       "</g>\n",
       "<!-- data2_run2_33 -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>data2_run2_33</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_33</text>\n",
       "</g>\n",
       "<!-- data2_run2_34 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>data2_run2_34</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_34</text>\n",
       "</g>\n",
       "<!-- data2_run2_35 -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>data2_run2_35</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-635.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_35</text>\n",
       "</g>\n",
       "<!-- data2_run2_36 -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>data2_run2_36</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_36</text>\n",
       "</g>\n",
       "<!-- data2_run2_37 -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>data2_run2_37</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_37</text>\n",
       "</g>\n",
       "<!-- data2_run2_38 -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>data2_run2_38</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_38</text>\n",
       "</g>\n",
       "<!-- data2_run2_39 -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>data2_run2_39</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_39</text>\n",
       "</g>\n",
       "<!-- data2_run2_4 -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>data2_run2_4</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_4</text>\n",
       "</g>\n",
       "<!-- data2_run2_40 -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>data2_run2_40</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_40</text>\n",
       "</g>\n",
       "<!-- data2_run2_41 -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>data2_run2_41</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_41</text>\n",
       "</g>\n",
       "<!-- data2_run2_42 -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>data2_run2_42</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_42</text>\n",
       "</g>\n",
       "<!-- data2_run2_43 -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>data2_run2_43</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_43</text>\n",
       "</g>\n",
       "<!-- data2_run2_44 -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>data2_run2_44</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-532.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_44</text>\n",
       "</g>\n",
       "<!-- data2_run2_45 -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>data2_run2_45</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_45</text>\n",
       "</g>\n",
       "<!-- data2_run2_46 -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>data2_run2_46</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_46</text>\n",
       "</g>\n",
       "<!-- data2_run2_47 -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>data2_run2_47</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_47</text>\n",
       "</g>\n",
       "<!-- data2_run2_48 -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>data2_run2_48</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_48</text>\n",
       "</g>\n",
       "<!-- data2_run2_49 -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>data2_run2_49</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_49</text>\n",
       "</g>\n",
       "<!-- data2_run2_5 -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>data2_run2_5</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_5</text>\n",
       "</g>\n",
       "<!-- data2_run2_50 -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>data2_run2_50</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-428.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_50</text>\n",
       "</g>\n",
       "<!-- data2_run2_51 -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>data2_run2_51</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_51</text>\n",
       "</g>\n",
       "<!-- data2_run2_52 -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>data2_run2_52</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_52</text>\n",
       "</g>\n",
       "<!-- data2_run2_53 -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>data2_run2_53</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_53</text>\n",
       "</g>\n",
       "<!-- data2_run2_54 -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>data2_run2_54</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_54</text>\n",
       "</g>\n",
       "<!-- data2_run2_55 -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>data2_run2_55</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_55</text>\n",
       "</g>\n",
       "<!-- data2_run2_56 -->\n",
       "<g id=\"node53\" class=\"node\">\n",
       "<title>data2_run2_56</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-324.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_56</text>\n",
       "</g>\n",
       "<!-- data2_run2_57 -->\n",
       "<g id=\"node54\" class=\"node\">\n",
       "<title>data2_run2_57</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-324.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_57</text>\n",
       "</g>\n",
       "<!-- data2_run2_58 -->\n",
       "<g id=\"node55\" class=\"node\">\n",
       "<title>data2_run2_58</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_58</text>\n",
       "</g>\n",
       "<!-- data2_run2_59 -->\n",
       "<g id=\"node56\" class=\"node\">\n",
       "<title>data2_run2_59</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_59</text>\n",
       "</g>\n",
       "<!-- data2_run2_6 -->\n",
       "<g id=\"node57\" class=\"node\">\n",
       "<title>data2_run2_6</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_6</text>\n",
       "</g>\n",
       "<!-- data2_run2_60 -->\n",
       "<g id=\"node58\" class=\"node\">\n",
       "<title>data2_run2_60</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-324.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_60</text>\n",
       "</g>\n",
       "<!-- data2_run2_61 -->\n",
       "<g id=\"node59\" class=\"node\">\n",
       "<title>data2_run2_61</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-324.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_61</text>\n",
       "</g>\n",
       "<!-- data2_run2_62 -->\n",
       "<g id=\"node60\" class=\"node\">\n",
       "<title>data2_run2_62</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_62</text>\n",
       "</g>\n",
       "<!-- data2_run2_63 -->\n",
       "<g id=\"node61\" class=\"node\">\n",
       "<title>data2_run2_63</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_63</text>\n",
       "</g>\n",
       "<!-- data2_run2_64 -->\n",
       "<g id=\"node62\" class=\"node\">\n",
       "<title>data2_run2_64</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_64</text>\n",
       "</g>\n",
       "<!-- data2_run2_65 -->\n",
       "<g id=\"node63\" class=\"node\">\n",
       "<title>data2_run2_65</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_65</text>\n",
       "</g>\n",
       "<!-- data2_run2_66 -->\n",
       "<g id=\"node64\" class=\"node\">\n",
       "<title>data2_run2_66</title>\n",
       "<text text-anchor=\"middle\" x=\"314.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_66</text>\n",
       "</g>\n",
       "<!-- data2_run2_67 -->\n",
       "<g id=\"node65\" class=\"node\">\n",
       "<title>data2_run2_67</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_67</text>\n",
       "</g>\n",
       "<!-- data2_run2_68 -->\n",
       "<g id=\"node66\" class=\"node\">\n",
       "<title>data2_run2_68</title>\n",
       "<text text-anchor=\"middle\" x=\"929.5\" y=\"-738.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_68</text>\n",
       "</g>\n",
       "<!-- data2_run2_69 -->\n",
       "<g id=\"node67\" class=\"node\">\n",
       "<title>data2_run2_69</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_69</text>\n",
       "</g>\n",
       "<!-- data2_run2_7 -->\n",
       "<g id=\"node68\" class=\"node\">\n",
       "<title>data2_run2_7</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_7</text>\n",
       "</g>\n",
       "<!-- data2_run2_70 -->\n",
       "<g id=\"node69\" class=\"node\">\n",
       "<title>data2_run2_70</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_70</text>\n",
       "</g>\n",
       "<!-- data2_run2_72 -->\n",
       "<g id=\"node70\" class=\"node\">\n",
       "<title>data2_run2_72</title>\n",
       "<text text-anchor=\"middle\" x=\"996.5\" y=\"-738.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_72</text>\n",
       "</g>\n",
       "<!-- data2_run2_73 -->\n",
       "<g id=\"node71\" class=\"node\">\n",
       "<title>data2_run2_73</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-220.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_73</text>\n",
       "</g>\n",
       "<!-- data2_run2_74 -->\n",
       "<g id=\"node72\" class=\"node\">\n",
       "<title>data2_run2_74</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_74</text>\n",
       "</g>\n",
       "<!-- data2_run2_75 -->\n",
       "<g id=\"node73\" class=\"node\">\n",
       "<title>data2_run2_75</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_75</text>\n",
       "</g>\n",
       "<!-- data2_run2_76 -->\n",
       "<g id=\"node74\" class=\"node\">\n",
       "<title>data2_run2_76</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_76</text>\n",
       "</g>\n",
       "<!-- data2_run2_77 -->\n",
       "<g id=\"node75\" class=\"node\">\n",
       "<title>data2_run2_77</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_77</text>\n",
       "</g>\n",
       "<!-- data2_run2_78 -->\n",
       "<g id=\"node76\" class=\"node\">\n",
       "<title>data2_run2_78</title>\n",
       "<text text-anchor=\"middle\" x=\"793.5\" y=\"-738.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_78</text>\n",
       "</g>\n",
       "<!-- data2_run2_79 -->\n",
       "<g id=\"node77\" class=\"node\">\n",
       "<title>data2_run2_79</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_79</text>\n",
       "</g>\n",
       "<!-- data2_run2_80 -->\n",
       "<g id=\"node78\" class=\"node\">\n",
       "<title>data2_run2_80</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_80</text>\n",
       "</g>\n",
       "<!-- data2_run2_81 -->\n",
       "<g id=\"node79\" class=\"node\">\n",
       "<title>data2_run2_81</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_81</text>\n",
       "</g>\n",
       "<!-- data2_run2_82 -->\n",
       "<g id=\"node80\" class=\"node\">\n",
       "<title>data2_run2_82</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_82</text>\n",
       "</g>\n",
       "<!-- data2_run2_83 -->\n",
       "<g id=\"node81\" class=\"node\">\n",
       "<title>data2_run2_83</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_83</text>\n",
       "</g>\n",
       "<!-- data2_run2_84 -->\n",
       "<g id=\"node82\" class=\"node\">\n",
       "<title>data2_run2_84</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_84</text>\n",
       "</g>\n",
       "<!-- data2_run2_85 -->\n",
       "<g id=\"node83\" class=\"node\">\n",
       "<title>data2_run2_85</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_85</text>\n",
       "</g>\n",
       "<!-- data2_run2_86 -->\n",
       "<g id=\"node84\" class=\"node\">\n",
       "<title>data2_run2_86</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_86</text>\n",
       "</g>\n",
       "<!-- data2_run2_87 -->\n",
       "<g id=\"node85\" class=\"node\">\n",
       "<title>data2_run2_87</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_87</text>\n",
       "</g>\n",
       "<!-- data2_run2_88 -->\n",
       "<g id=\"node86\" class=\"node\">\n",
       "<title>data2_run2_88</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_88</text>\n",
       "</g>\n",
       "<!-- data2_run2_89 -->\n",
       "<g id=\"node87\" class=\"node\">\n",
       "<title>data2_run2_89</title>\n",
       "<text text-anchor=\"middle\" x=\"381.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_89</text>\n",
       "</g>\n",
       "<!-- data2_run2_9 -->\n",
       "<g id=\"node88\" class=\"node\">\n",
       "<title>data2_run2_9</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_9</text>\n",
       "</g>\n",
       "<!-- data2_run2_90 -->\n",
       "<g id=\"node89\" class=\"node\">\n",
       "<title>data2_run2_90</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_90</text>\n",
       "</g>\n",
       "<!-- data2_run2_91 -->\n",
       "<g id=\"node90\" class=\"node\">\n",
       "<title>data2_run2_91</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_91</text>\n",
       "</g>\n",
       "<!-- data2_run2_93 -->\n",
       "<g id=\"node91\" class=\"node\">\n",
       "<title>data2_run2_93</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_93</text>\n",
       "</g>\n",
       "<!-- data1_YieP -->\n",
       "<g id=\"node92\" class=\"node\">\n",
       "<title>data1_YieP</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YieP</text>\n",
       "</g>\n",
       "<!-- data1_YieP&#45;&gt;data2_run2_44 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_YieP&#45;&gt;data2_run2_44</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M1078.5,-575.9664C1078.5,-570.6707 1078.5,-564.8028 1078.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"1080.6351,-559.2222 1078.5,-554.2222 1076.3651,-559.2223 1080.6351,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GlpR -->\n",
       "<g id=\"node93\" class=\"node\">\n",
       "<title>data1_GlpR</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GlpR</text>\n",
       "</g>\n",
       "<!-- data1_GlpR&#45;&gt;data2_run2_24 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_GlpR&#45;&gt;data2_run2_24</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M714.5,-782.9664C714.5,-777.6707 714.5,-771.8028 714.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"716.6745,-766.2222 714.5,-761.2222 712.3257,-766.2223 716.6745,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_YgbI -->\n",
       "<g id=\"node94\" class=\"node\">\n",
       "<title>data1_YgbI</title>\n",
       "<text text-anchor=\"middle\" x=\"793.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YgbI</text>\n",
       "</g>\n",
       "<!-- data1_YgbI&#45;&gt;data2_run2_25 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_YgbI&#45;&gt;data2_run2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.05\" d=\"M814.5772,-782.9775C821.6664,-776.9157 829.6514,-770.0879 837.0411,-763.7692\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.05\" points=\"838.4128,-764.8989 841.0757,-760.3193 836.1382,-762.2387 838.4128,-764.8989\"/>\n",
       "</g>\n",
       "<!-- data1_YgbI&#45;&gt;data2_run2_78 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_YgbI&#45;&gt;data2_run2_78</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.78\" d=\"M793.5,-782.9775C793.5,-777.477 793.5,-771.3458 793.5,-765.5365\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.78\" points=\"795.5913,-765.3193 793.5,-760.3193 791.4088,-765.3194 795.5913,-765.3193\"/>\n",
       "</g>\n",
       "<!-- data1_efeU&#45;&#45;\n",
       "repair -->\n",
       "<g id=\"node95\" class=\"node\">\n",
       "<title>data1_efeU&#45;&#45;\n",
       "repair</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-804.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">efeU&#45;&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-789.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">repair</text>\n",
       "</g>\n",
       "<!-- data1_efeU&#45;&#45;\n",
       "repair&#45;&gt;data2_run2_2 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_efeU&#45;&#45;\n",
       "repair&#45;&gt;data2_run2_2</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M231.5,-781.6812C231.5,-776.5393 231.5,-770.927 231.5,-765.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"233.6438,-765.3596 231.5,-760.3597 229.3563,-765.3597 233.6438,-765.3596\"/>\n",
       "</g>\n",
       "<!-- data1_ydcI&#45;KO -->\n",
       "<g id=\"node96\" class=\"node\">\n",
       "<title>data1_ydcI&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ydcI&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_ydcI&#45;KO&#45;&gt;data2_run2_79 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>data1_ydcI&#45;KO&#45;&gt;data2_run2_79</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M347.5,-161.9664C347.5,-156.6707 347.5,-150.8028 347.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"349.6351,-145.2222 347.5,-140.2222 345.3651,-145.2223 349.6351,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_BW25113 -->\n",
       "<g id=\"node97\" class=\"node\">\n",
       "<title>data1_BW25113</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">BW25113</text>\n",
       "</g>\n",
       "<!-- data1_BW25113&#45;&gt;data2_run2_7 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>data1_BW25113&#45;&gt;data2_run2_7</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.84\" d=\"M630.5,-264.9664C630.5,-259.6707 630.5,-253.8028 630.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.84\" points=\"632.6176,-248.2222 630.5,-243.2222 628.3826,-248.2223 632.6176,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_proVWX -->\n",
       "<g id=\"node98\" class=\"node\">\n",
       "<title>data1_proVWX</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">proVWX</text>\n",
       "</g>\n",
       "<!-- data1_proVWX&#45;&gt;data2_run2_11 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>data1_proVWX&#45;&gt;data2_run2_11</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.68\" d=\"M347.5,-886.9664C347.5,-881.6707 347.5,-875.8028 347.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.68\" points=\"349.5476,-870.2222 347.5,-865.2222 345.4526,-870.2223 349.5476,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_DhaR/Mlc -->\n",
       "<g id=\"node99\" class=\"node\">\n",
       "<title>data1_DhaR/Mlc</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">DhaR/Mlc</text>\n",
       "</g>\n",
       "<!-- data1_DhaR/Mlc&#45;&gt;data2_run2_32 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>data1_DhaR/Mlc&#45;&gt;data2_run2_32</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.84\" d=\"M630.5,-678.9664C630.5,-673.6707 630.5,-667.8028 630.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.84\" points=\"632.6176,-662.2222 630.5,-657.2222 628.3826,-662.2223 632.6176,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_ArgR -->\n",
       "<g id=\"node100\" class=\"node\">\n",
       "<title>data1_ArgR</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArgR</text>\n",
       "</g>\n",
       "<!-- data1_ArgR&#45;&gt;data2_run2_9 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>data1_ArgR&#45;&gt;data2_run2_9</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M347.5,-58.9664C347.5,-53.6707 347.5,-47.8028 347.5,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"349.6657,-42.2222 347.5,-37.2222 345.3345,-42.2223 349.6657,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_YneJ -->\n",
       "<g id=\"node101\" class=\"node\">\n",
       "<title>data1_YneJ</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YneJ</text>\n",
       "</g>\n",
       "<!-- data1_YneJ&#45;&gt;data2_run2_82 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>data1_YneJ&#45;&gt;data2_run2_82</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.84\" d=\"M630.5,-161.9664C630.5,-156.6707 630.5,-150.8028 630.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.84\" points=\"632.6176,-145.2222 630.5,-140.2222 628.3826,-145.2223 632.6176,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NarL -->\n",
       "<g id=\"node102\" class=\"node\">\n",
       "<title>data1_NarL</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NarL</text>\n",
       "</g>\n",
       "<!-- data1_NarL&#45;&gt;data2_run2_83 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>data1_NarL&#45;&gt;data2_run2_83</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M714.5,-161.9664C714.5,-156.6707 714.5,-150.8028 714.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"716.6526,-145.2222 714.5,-140.2222 712.3476,-145.2223 716.6526,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_e14&#45;de&#45;\n",
       "letion -->\n",
       "<g id=\"node103\" class=\"node\">\n",
       "<title>data1_e14&#45;de&#45;\n",
       "letion</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-390.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">e14&#45;de&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-375.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">letion</text>\n",
       "</g>\n",
       "<!-- data1_e14&#45;de&#45;\n",
       "letion&#45;&gt;data2_run2_61 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>data1_e14&#45;de&#45;\n",
       "letion&#45;&gt;data2_run2_61</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M894.5,-367.6812C894.5,-362.5393 894.5,-356.927 894.5,-351.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"896.6526,-351.3596 894.5,-346.3597 892.3476,-351.3597 896.6526,-351.3596\"/>\n",
       "</g>\n",
       "<!-- data1_MalT -->\n",
       "<g id=\"node104\" class=\"node\">\n",
       "<title>data1_MalT</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">MalT</text>\n",
       "</g>\n",
       "<!-- data1_MalT&#45;&gt;data2_run2_5 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>data1_MalT&#45;&gt;data2_run2_5</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M545.5,-472.9664C545.5,-467.6707 545.5,-461.8028 545.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"547.6613,-456.2222 545.5,-451.2222 543.3388,-456.2223 547.6613,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_OxyR -->\n",
       "<g id=\"node105\" class=\"node\">\n",
       "<title>data1_OxyR</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">OxyR</text>\n",
       "</g>\n",
       "<!-- data1_OxyR&#45;&gt;data2_run2_0 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>data1_OxyR&#45;&gt;data2_run2_0</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M43.5,-886.9664C43.5,-881.6707 43.5,-875.8028 43.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"45.6613,-870.2222 43.5,-865.2222 41.3388,-870.2223 45.6613,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;2 -->\n",
       "<g id=\"node106\" class=\"node\">\n",
       "<title>data1_Crp&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Crp&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;2&#45;&gt;data2_run2_37 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>data1_Crp&#45;2&#45;&gt;data2_run2_37</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.96\" d=\"M138.5,-575.9664C138.5,-570.6707 138.5,-564.8028 138.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.96\" points=\"140.6701,-559.2222 138.5,-554.2222 136.3301,-559.2223 140.6701,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_AtoC -->\n",
       "<g id=\"node107\" class=\"node\">\n",
       "<title>data1_AtoC</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AtoC</text>\n",
       "</g>\n",
       "<!-- data1_AtoC&#45;&gt;data2_run2_6 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>data1_AtoC&#45;&gt;data2_run2_6</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.81\" d=\"M630.5,-368.9664C630.5,-363.6707 630.5,-357.8028 630.5,-352.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.81\" points=\"632.6045,-352.2222 630.5,-347.2222 628.3957,-352.2223 632.6045,-352.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1 -->\n",
       "<g id=\"node108\" class=\"node\">\n",
       "<title>data1_PurR&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1&#45;&gt;data2_run2_64 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>data1_PurR&#45;1&#45;&gt;data2_run2_64</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M138.5,-264.9664C138.5,-259.6707 138.5,-253.8028 138.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"140.6613,-248.2222 138.5,-243.2222 136.3388,-248.2223 140.6613,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_yheO&#45;KO -->\n",
       "<g id=\"node109\" class=\"node\">\n",
       "<title>data1_yheO&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">yheO&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_yheO&#45;KO&#45;&gt;data2_run2_46 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>data1_yheO&#45;KO&#45;&gt;data2_run2_46</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M138.5,-472.9664C138.5,-467.6707 138.5,-461.8028 138.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"140.6307,-456.2222 138.5,-451.2222 136.3695,-456.2223 140.6307,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CsqR -->\n",
       "<g id=\"node110\" class=\"node\">\n",
       "<title>data1_CsqR</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CsqR</text>\n",
       "</g>\n",
       "<!-- data1_CsqR&#45;&gt;data2_run2_14 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>data1_CsqR&#45;&gt;data2_run2_14</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M630.5,-886.9664C630.5,-881.6707 630.5,-875.8028 630.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"632.6657,-870.2222 630.5,-865.2222 628.3345,-870.2223 632.6657,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_MetJ -->\n",
       "<g id=\"node111\" class=\"node\">\n",
       "<title>data1_MetJ</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">MetJ</text>\n",
       "</g>\n",
       "<!-- data1_MetJ&#45;&gt;data2_run2_1 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>data1_MetJ&#45;&gt;data2_run2_1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M138.5,-886.9664C138.5,-881.6707 138.5,-875.8028 138.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"140.6657,-870.2222 138.5,-865.2222 136.3345,-870.2223 140.6657,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_entC&#45;m&#45;\n",
       "enF&#45;KO -->\n",
       "<g id=\"node112\" class=\"node\">\n",
       "<title>data1_entC&#45;m&#45;\n",
       "enF&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-908.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">entC&#45;m&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-893.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">enF&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_entC&#45;m&#45;\n",
       "enF&#45;KO&#45;&gt;data2_run2_15 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>data1_entC&#45;m&#45;\n",
       "enF&#45;KO&#45;&gt;data2_run2_15</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.93\" d=\"M714.5,-885.6812C714.5,-880.5393 714.5,-874.927 714.5,-869.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.93\" points=\"716.657,-869.3596 714.5,-864.3597 712.3432,-869.3597 716.657,-869.3596\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;6 -->\n",
       "<g id=\"node113\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;6</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-80.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-65.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;6</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;6&#45;&gt;data2_run2_86 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;6&#45;&gt;data2_run2_86</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.85\" d=\"M43.5,-57.6812C43.5,-52.5393 43.5,-46.927 43.5,-41.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.85\" points=\"45.622,-41.3596 43.5,-36.3597 41.3782,-41.3597 45.622,-41.3596\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;2 -->\n",
       "<g id=\"node114\" class=\"node\">\n",
       "<title>data1_PurR&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;2&#45;&gt;data2_run2_81 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>data1_PurR&#45;2&#45;&gt;data2_run2_81</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M545.5,-161.9664C545.5,-156.6707 545.5,-150.8028 545.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"547.6613,-145.2222 545.5,-140.2222 543.3388,-145.2223 547.6613,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Cbl+CysB -->\n",
       "<g id=\"node115\" class=\"node\">\n",
       "<title>data1_Cbl+CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Cbl+CysB</text>\n",
       "</g>\n",
       "<!-- data1_Cbl+CysB&#45;&gt;data2_run2_28 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>data1_Cbl+CysB&#45;&gt;data2_run2_28</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M138.5,-678.9664C138.5,-673.6707 138.5,-667.8028 138.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"140.6438,-662.2222 138.5,-657.2222 136.3563,-662.2223 140.6438,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GlcC -->\n",
       "<g id=\"node116\" class=\"node\">\n",
       "<title>data1_GlcC</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GlcC</text>\n",
       "</g>\n",
       "<!-- data1_GlcC&#45;&gt;data2_run2_54 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>data1_GlcC&#45;&gt;data2_run2_54</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M43.5,-368.9664C43.5,-363.6707 43.5,-357.8028 43.5,-352.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"45.6395,-352.2222 43.5,-347.2222 41.3607,-352.2223 45.6395,-352.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1 -->\n",
       "<g id=\"node117\" class=\"node\">\n",
       "<title>data1_Crp&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Crp&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1&#45;&gt;data2_run2_93 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>data1_Crp&#45;1&#45;&gt;data2_run2_93</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M630.5,-58.9664C630.5,-53.6707 630.5,-47.8028 630.5,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"632.6307,-42.2222 630.5,-37.2222 628.3695,-42.2223 632.6307,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_YiaJ -->\n",
       "<g id=\"node118\" class=\"node\">\n",
       "<title>data1_YiaJ</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YiaJ</text>\n",
       "</g>\n",
       "<!-- data1_YiaJ&#45;&gt;data2_run2_55 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>data1_YiaJ&#45;&gt;data2_run2_55</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M138.5,-368.9664C138.5,-363.6707 138.5,-357.8028 138.5,-352.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"140.6745,-352.2222 138.5,-347.2222 136.3257,-352.2223 140.6745,-352.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GcvA -->\n",
       "<g id=\"node119\" class=\"node\">\n",
       "<title>data1_GcvA</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GcvA</text>\n",
       "</g>\n",
       "<!-- data1_GcvA&#45;&gt;data2_run2_20 -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>data1_GcvA&#45;&gt;data2_run2_20</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M347.5,-782.9664C347.5,-777.6707 347.5,-771.8028 347.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"349.6438,-766.2222 347.5,-761.2222 345.3563,-766.2223 349.6438,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_XylR -->\n",
       "<g id=\"node120\" class=\"node\">\n",
       "<title>data1_XylR</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">XylR</text>\n",
       "</g>\n",
       "<!-- data1_XylR&#45;&gt;data2_run2_40 -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>data1_XylR&#45;&gt;data2_run2_40</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M545.5,-575.9664C545.5,-570.6707 545.5,-564.8028 545.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"547.6657,-559.2222 545.5,-554.2222 543.3345,-559.2223 547.6657,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_fur&#45;KO -->\n",
       "<g id=\"node121\" class=\"node\">\n",
       "<title>data1_fur&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">fur&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_fur&#45;KO&#45;&gt;data2_run2_74 -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>data1_fur&#45;KO&#45;&gt;data2_run2_74</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M1078.5,-264.9664C1078.5,-259.6707 1078.5,-253.8028 1078.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"1080.6307,-248.2222 1078.5,-243.2222 1076.3695,-248.2223 1080.6307,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RpoH -->\n",
       "<g id=\"node122\" class=\"node\">\n",
       "<title>data1_RpoH</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RpoH</text>\n",
       "</g>\n",
       "<!-- data1_RpoH&#45;&gt;data2_run2_39 -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>data1_RpoH&#45;&gt;data2_run2_39</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M347.5,-575.9664C347.5,-570.6707 347.5,-564.8028 347.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"349.6307,-559.2222 347.5,-554.2222 345.3695,-559.2223 349.6307,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GadEWX -->\n",
       "<g id=\"node123\" class=\"node\">\n",
       "<title>data1_GadEWX</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GadEWX</text>\n",
       "</g>\n",
       "<!-- data1_GadEWX&#45;&gt;data2_run2_87 -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>data1_GadEWX&#45;&gt;data2_run2_87</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.98\" d=\"M138.5,-58.9664C138.5,-53.6707 138.5,-47.8028 138.5,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.98\" points=\"140.6788,-42.2222 138.5,-37.2222 136.3213,-42.2223 140.6788,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_trans&#45;\n",
       "lation -->\n",
       "<g id=\"node124\" class=\"node\">\n",
       "<title>data1_trans&#45;\n",
       "lation</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-804.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-789.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data1_trans&#45;\n",
       "lation&#45;&gt;data2_run2_26 -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>data1_trans&#45;\n",
       "lation&#45;&gt;data2_run2_26</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M1078.5,-781.6812C1078.5,-776.5393 1078.5,-770.927 1078.5,-765.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"1080.6088,-765.3596 1078.5,-760.3597 1076.3913,-765.3597 1080.6088,-765.3596\"/>\n",
       "</g>\n",
       "<!-- data1_PrpR -->\n",
       "<g id=\"node125\" class=\"node\">\n",
       "<title>data1_PrpR</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PrpR</text>\n",
       "</g>\n",
       "<!-- data1_PrpR&#45;&gt;data2_run2_84 -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>data1_PrpR&#45;&gt;data2_run2_84</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M894.5,-161.9664C894.5,-156.6707 894.5,-150.8028 894.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"896.6395,-145.2222 894.5,-140.2222 892.3607,-145.2223 896.6395,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GadWX -->\n",
       "<g id=\"node126\" class=\"node\">\n",
       "<title>data1_GadWX</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GadWX</text>\n",
       "</g>\n",
       "<!-- data1_GadWX&#45;&gt;data2_run2_62 -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>data1_GadWX&#45;&gt;data2_run2_62</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.81\" d=\"M1078.5,-368.9664C1078.5,-363.6707 1078.5,-357.8028 1078.5,-352.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.81\" points=\"1080.6045,-352.2222 1078.5,-347.2222 1076.3957,-352.2223 1080.6045,-352.2222\"/>\n",
       "</g>\n",
       "<!-- data1_fimbriae -->\n",
       "<g id=\"node127\" class=\"node\">\n",
       "<title>data1_fimbriae</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">fimbriae</text>\n",
       "</g>\n",
       "<!-- data1_fimbriae&#45;&gt;data2_run2_52 -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>data1_fimbriae&#45;&gt;data2_run2_52</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.93\" d=\"M894.5,-472.9664C894.5,-467.6707 894.5,-461.8028 894.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.93\" points=\"896.657,-456.2222 894.5,-451.2222 892.3432,-456.2223 896.657,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CecR -->\n",
       "<g id=\"node128\" class=\"node\">\n",
       "<title>data1_CecR</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CecR</text>\n",
       "</g>\n",
       "<!-- data1_CecR&#45;&gt;data2_run2_65 -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>data1_CecR&#45;&gt;data2_run2_65</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.86\" d=\"M231.5,-264.9664C231.5,-259.6707 231.5,-253.8028 231.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.86\" points=\"233.6263,-248.2222 231.5,-243.2222 229.3738,-248.2223 233.6263,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;4 -->\n",
       "<g id=\"node129\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;4</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-700.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-685.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;4</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;4&#45;&gt;data2_run2_30 -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;4&#45;&gt;data2_run2_30</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.8\" d=\"M460.5,-677.6812C460.5,-672.5393 460.5,-666.927 460.5,-661.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.8\" points=\"462.6001,-661.3596 460.5,-656.3597 458.4001,-661.3597 462.6001,-661.3596\"/>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;1 -->\n",
       "<g id=\"node130\" class=\"node\">\n",
       "<title>data1_ArcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;1&#45;&gt;data2_run2_49 -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>data1_ArcA&#45;1&#45;&gt;data2_run2_49</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M460.5,-472.9664C460.5,-467.6707 460.5,-461.8028 460.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"462.6307,-456.2222 460.5,-451.2222 458.3695,-456.2223 462.6307,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_nitrate&#45;\n",
       "&#45;related -->\n",
       "<g id=\"node131\" class=\"node\">\n",
       "<title>data1_nitrate&#45;\n",
       "&#45;related</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-494.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">nitrate&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-479.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">&#45;related</text>\n",
       "</g>\n",
       "<!-- data1_nitrate&#45;\n",
       "&#45;related&#45;&gt;data2_run2_50 -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>data1_nitrate&#45;\n",
       "&#45;related&#45;&gt;data2_run2_50</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M630.5,-471.6812C630.5,-466.5393 630.5,-460.927 630.5,-455.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"632.6395,-455.3596 630.5,-450.3597 628.3607,-455.3597 632.6395,-455.3596\"/>\n",
       "</g>\n",
       "<!-- data1_Fnr -->\n",
       "<g id=\"node132\" class=\"node\">\n",
       "<title>data1_Fnr</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fnr</text>\n",
       "</g>\n",
       "<!-- data1_Fnr&#45;&gt;data2_run2_17 -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>data1_Fnr&#45;&gt;data2_run2_17</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M1078.5,-886.9664C1078.5,-881.6707 1078.5,-875.8028 1078.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"1080.6526,-870.2222 1078.5,-865.2222 1076.3476,-870.2223 1080.6526,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CdaR -->\n",
       "<g id=\"node133\" class=\"node\">\n",
       "<title>data1_CdaR</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CdaR</text>\n",
       "</g>\n",
       "<!-- data1_CdaR&#45;&gt;data2_run2_38 -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>data1_CdaR&#45;&gt;data2_run2_38</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M231.5,-575.9664C231.5,-570.6707 231.5,-564.8028 231.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"233.6482,-559.2222 231.5,-554.2222 229.352,-559.2223 233.6482,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Pyruvate -->\n",
       "<g id=\"node134\" class=\"node\">\n",
       "<title>data1_Pyruvate</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Pyruvate</text>\n",
       "</g>\n",
       "<!-- data1_Pyruvate&#45;&gt;data2_run2_67 -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>data1_Pyruvate&#45;&gt;data2_run2_67</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M460.5,-264.9664C460.5,-259.6707 460.5,-253.8028 460.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"462.6351,-248.2222 460.5,-243.2222 458.3651,-248.2223 462.6351,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_SrlR+GutM -->\n",
       "<g id=\"node135\" class=\"node\">\n",
       "<title>data1_SrlR+GutM</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">SrlR+GutM</text>\n",
       "</g>\n",
       "<!-- data1_SrlR+GutM&#45;&gt;data2_run2_76 -->\n",
       "<g id=\"edge45\" class=\"edge\">\n",
       "<title>data1_SrlR+GutM&#45;&gt;data2_run2_76</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.8\" d=\"M138.5,-161.9664C138.5,-156.6707 138.5,-150.8028 138.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.8\" points=\"140.6001,-145.2222 138.5,-140.2222 136.4001,-145.2223 140.6001,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NagC/TyrR -->\n",
       "<g id=\"node136\" class=\"node\">\n",
       "<title>data1_NagC/TyrR</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NagC/TyrR</text>\n",
       "</g>\n",
       "<!-- data1_NagC/TyrR&#45;&gt;data2_run2_10 -->\n",
       "<g id=\"edge46\" class=\"edge\">\n",
       "<title>data1_NagC/TyrR&#45;&gt;data2_run2_10</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M231.5,-886.9664C231.5,-881.6707 231.5,-875.8028 231.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"233.6657,-870.2222 231.5,-865.2222 229.3345,-870.2223 233.6657,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CpxR -->\n",
       "<g id=\"node137\" class=\"node\">\n",
       "<title>data1_CpxR</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CpxR</text>\n",
       "</g>\n",
       "<!-- data1_CpxR&#45;&gt;data2_run2_23 -->\n",
       "<g id=\"edge47\" class=\"edge\">\n",
       "<title>data1_CpxR&#45;&gt;data2_run2_23</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.81\" d=\"M630.5,-782.9664C630.5,-777.6707 630.5,-771.8028 630.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.81\" points=\"632.6045,-766.2222 630.5,-761.2222 628.3957,-766.2223 632.6045,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_thrA&#45;KO -->\n",
       "<g id=\"node138\" class=\"node\">\n",
       "<title>data1_thrA&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">thrA&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_thrA&#45;KO&#45;&gt;data2_run2_36 -->\n",
       "<g id=\"edge48\" class=\"edge\">\n",
       "<title>data1_thrA&#45;KO&#45;&gt;data2_run2_36</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.64\" d=\"M43.5,-575.9664C43.5,-570.6707 43.5,-564.8028 43.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.64\" points=\"45.5301,-559.2222 43.5,-554.2222 41.4701,-559.2223 45.5301,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Lrp -->\n",
       "<g id=\"node139\" class=\"node\">\n",
       "<title>data1_Lrp</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Lrp</text>\n",
       "</g>\n",
       "<!-- data1_Lrp&#45;&gt;data2_run2_48 -->\n",
       "<g id=\"edge49\" class=\"edge\">\n",
       "<title>data1_Lrp&#45;&gt;data2_run2_48</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M347.5,-472.9664C347.5,-467.6707 347.5,-461.8028 347.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"349.6351,-456.2222 347.5,-451.2222 345.3651,-456.2223 349.6351,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;1 -->\n",
       "<g id=\"node140\" class=\"node\">\n",
       "<title>data1_deletion&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"138.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">deletion&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;1&#45;&gt;data2_run2_19 -->\n",
       "<g id=\"edge50\" class=\"edge\">\n",
       "<title>data1_deletion&#45;1&#45;&gt;data2_run2_19</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M138.5,-782.9664C138.5,-777.6707 138.5,-771.8028 138.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"140.6657,-766.2222 138.5,-761.2222 136.3345,-766.2223 140.6657,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile -->\n",
       "<g id=\"node141\" class=\"node\">\n",
       "<title>data1_Leu/Ile</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Leu/Ile</text>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile&#45;&gt;data2_run2_66 -->\n",
       "<g id=\"edge51\" class=\"edge\">\n",
       "<title>data1_Leu/Ile&#45;&gt;data2_run2_66</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.5\" d=\"M337.2395,-264.9664C334.1649,-259.5626 330.7514,-253.5631 327.5238,-247.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.5\" points=\"329.0517,-246.5945 324.8678,-243.2222 325.6293,-248.5417 329.0517,-246.5945\"/>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile&#45;&gt;data2_run2_89 -->\n",
       "<g id=\"edge52\" class=\"edge\">\n",
       "<title>data1_Leu/Ile&#45;&gt;data2_run2_89</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.56\" d=\"M358.0714,-264.9664C361.3025,-259.4545 364.8969,-253.3229 368.2809,-247.5503\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.56\" points=\"369.7991,-248.4208 370.818,-243.2222 366.7796,-246.6507 369.7991,-248.4208\"/>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;2 -->\n",
       "<g id=\"node142\" class=\"node\">\n",
       "<title>data1_deletion&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">deletion&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;2&#45;&gt;data2_run2_29 -->\n",
       "<g id=\"edge53\" class=\"edge\">\n",
       "<title>data1_deletion&#45;2&#45;&gt;data2_run2_29</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M231.5,-678.9664C231.5,-673.6707 231.5,-667.8028 231.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"233.6745,-662.2222 231.5,-657.2222 229.3257,-662.2223 233.6745,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;1 -->\n",
       "<g id=\"node143\" class=\"node\">\n",
       "<title>data1_Fur&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;1&#45;&gt;data2_run2_34 -->\n",
       "<g id=\"edge54\" class=\"edge\">\n",
       "<title>data1_Fur&#45;1&#45;&gt;data2_run2_34</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M894.5,-678.9664C894.5,-673.6707 894.5,-667.8028 894.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"896.6657,-662.2222 894.5,-657.2222 892.3345,-662.2223 896.6657,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;5 -->\n",
       "<g id=\"node144\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;5</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-286.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-271.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;5</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;5&#45;&gt;data2_run2_73 -->\n",
       "<g id=\"edge55\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;5&#45;&gt;data2_run2_73</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M894.5,-263.6812C894.5,-258.5393 894.5,-252.927 894.5,-247.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"896.6088,-247.3596 894.5,-242.3597 892.3913,-247.3597 896.6088,-247.3596\"/>\n",
       "</g>\n",
       "<!-- data1_Thiamine -->\n",
       "<g id=\"node145\" class=\"node\">\n",
       "<title>data1_Thiamine</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Thiamine</text>\n",
       "</g>\n",
       "<!-- data1_Thiamine&#45;&gt;data2_run2_80 -->\n",
       "<g id=\"edge56\" class=\"edge\">\n",
       "<title>data1_Thiamine&#45;&gt;data2_run2_80</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M460.5,-161.9664C460.5,-156.6707 460.5,-150.8028 460.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"462.6526,-145.2222 460.5,-140.2222 458.3476,-145.2223 462.6526,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CysB -->\n",
       "<g id=\"node146\" class=\"node\">\n",
       "<title>data1_CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CysB</text>\n",
       "</g>\n",
       "<!-- data1_CysB&#45;&gt;data2_run2_47 -->\n",
       "<g id=\"edge57\" class=\"edge\">\n",
       "<title>data1_CysB&#45;&gt;data2_run2_47</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M231.5,-472.9664C231.5,-467.6707 231.5,-461.8028 231.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"233.6745,-456.2222 231.5,-451.2222 229.3257,-456.2223 233.6745,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_FliA -->\n",
       "<g id=\"node147\" class=\"node\">\n",
       "<title>data1_FliA</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FliA</text>\n",
       "</g>\n",
       "<!-- data1_FliA&#45;&gt;data2_run2_21 -->\n",
       "<g id=\"edge58\" class=\"edge\">\n",
       "<title>data1_FliA&#45;&gt;data2_run2_21</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M460.5,-782.9664C460.5,-777.6707 460.5,-771.8028 460.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"462.6745,-766.2222 460.5,-761.2222 458.3257,-766.2223 462.6745,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PuuR -->\n",
       "<g id=\"node148\" class=\"node\">\n",
       "<title>data1_PuuR</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PuuR</text>\n",
       "</g>\n",
       "<!-- data1_PuuR&#45;&gt;data2_run2_63 -->\n",
       "<g id=\"edge59\" class=\"edge\">\n",
       "<title>data1_PuuR&#45;&gt;data2_run2_63</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M43.5,-264.9664C43.5,-259.6707 43.5,-253.8028 43.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"45.6351,-248.2222 43.5,-243.2222 41.3651,-248.2223 45.6351,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NikR -->\n",
       "<g id=\"node149\" class=\"node\">\n",
       "<title>data1_NikR</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NikR</text>\n",
       "</g>\n",
       "<!-- data1_NikR&#45;&gt;data2_run2_59 -->\n",
       "<g id=\"edge60\" class=\"edge\">\n",
       "<title>data1_NikR&#45;&gt;data2_run2_59</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M545.5,-368.9664C545.5,-363.6707 545.5,-357.8028 545.5,-352.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"547.6307,-352.2222 545.5,-347.2222 543.3695,-352.2223 547.6307,-352.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NtrC+RpoN -->\n",
       "<g id=\"node150\" class=\"node\">\n",
       "<title>data1_NtrC+RpoN</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NtrC+RpoN</text>\n",
       "</g>\n",
       "<!-- data1_NtrC+RpoN&#45;&gt;data2_run2_18 -->\n",
       "<g id=\"edge61\" class=\"edge\">\n",
       "<title>data1_NtrC+RpoN&#45;&gt;data2_run2_18</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.98\" d=\"M43.5,-782.9664C43.5,-777.6707 43.5,-771.8028 43.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.98\" points=\"45.6788,-766.2222 43.5,-761.2222 41.3213,-766.2223 45.6788,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RcsAB -->\n",
       "<g id=\"node151\" class=\"node\">\n",
       "<title>data1_RcsAB</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RcsAB</text>\n",
       "</g>\n",
       "<!-- data1_RcsAB&#45;&gt;data2_run2_85 -->\n",
       "<g id=\"edge62\" class=\"edge\">\n",
       "<title>data1_RcsAB&#45;&gt;data2_run2_85</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M1078.5,-161.9664C1078.5,-156.6707 1078.5,-150.8028 1078.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"1080.6745,-145.2222 1078.5,-140.2222 1076.3257,-145.2223 1080.6745,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_His&#45;tRNA -->\n",
       "<g id=\"node152\" class=\"node\">\n",
       "<title>data1_His&#45;tRNA</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">His&#45;tRNA</text>\n",
       "</g>\n",
       "<!-- data1_His&#45;tRNA&#45;&gt;data2_run2_75 -->\n",
       "<g id=\"edge63\" class=\"edge\">\n",
       "<title>data1_His&#45;tRNA&#45;&gt;data2_run2_75</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M43.5,-161.9664C43.5,-156.6707 43.5,-150.8028 43.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"45.6482,-145.2222 43.5,-140.2222 41.352,-145.2223 45.6482,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2 -->\n",
       "<g id=\"node153\" class=\"node\">\n",
       "<title>data1_Fur&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2&#45;&gt;data2_run2_53 -->\n",
       "<g id=\"edge64\" class=\"edge\">\n",
       "<title>data1_Fur&#45;2&#45;&gt;data2_run2_53</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.86\" d=\"M1078.5,-472.9664C1078.5,-467.6707 1078.5,-461.8028 1078.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.86\" points=\"1080.6263,-456.2222 1078.5,-451.2222 1076.3738,-456.2223 1080.6263,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Copper -->\n",
       "<g id=\"node154\" class=\"node\">\n",
       "<title>data1_Copper</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Copper</text>\n",
       "</g>\n",
       "<!-- data1_Copper&#45;&gt;data2_run2_42 -->\n",
       "<g id=\"edge65\" class=\"edge\">\n",
       "<title>data1_Copper&#45;&gt;data2_run2_42</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M714.5,-575.9664C714.5,-570.6707 714.5,-564.8028 714.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"716.6745,-559.2222 714.5,-554.2222 712.3257,-559.2223 716.6745,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR -->\n",
       "<g id=\"node155\" class=\"node\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"943.5\" y=\"-804.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AllR/Ar&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"943.5\" y=\"-789.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">aC/FucR</text>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_25 -->\n",
       "<g id=\"edge66\" class=\"edge\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.32\" d=\"M916.9776,-781.6812C908.9581,-775.8398 900.1052,-769.3914 891.909,-763.4213\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.32\" points=\"892.5044,-761.6901 887.4325,-760.1607 890.4437,-764.5191 892.5044,-761.6901\"/>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_68 -->\n",
       "<g id=\"edge67\" class=\"edge\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_68</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.09\" d=\"M938.9159,-781.6812C937.6703,-776.4322 936.3085,-770.693 935.0172,-765.2511\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.09\" points=\"936.7137,-764.8205 933.8565,-760.3597 933.3083,-765.6287 936.7137,-764.8205\"/>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_72 -->\n",
       "<g id=\"edge68\" class=\"edge\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_72</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.23\" d=\"M960.8542,-781.6812C965.8581,-776.1108 971.3579,-769.9884 976.5081,-764.2551\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.23\" points=\"978.0428,-765.316 980.0074,-760.3597 975.2893,-762.8426 978.0428,-765.316\"/>\n",
       "</g>\n",
       "<!-- data1_Cra -->\n",
       "<g id=\"node156\" class=\"node\">\n",
       "<title>data1_Cra</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Cra</text>\n",
       "</g>\n",
       "<!-- data1_Cra&#45;&gt;data2_run2_4 -->\n",
       "<g id=\"edge69\" class=\"edge\">\n",
       "<title>data1_Cra&#45;&gt;data2_run2_4</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.96\" d=\"M460.5,-575.9664C460.5,-570.6707 460.5,-564.8028 460.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.96\" points=\"462.6701,-559.2222 460.5,-554.2222 458.3301,-559.2223 462.6701,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;2 -->\n",
       "<g id=\"node157\" class=\"node\">\n",
       "<title>data1_ArcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;2&#45;&gt;data2_run2_3 -->\n",
       "<g id=\"edge70\" class=\"edge\">\n",
       "<title>data1_ArcA&#45;2&#45;&gt;data2_run2_3</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M347.5,-678.9664C347.5,-673.6707 347.5,-667.8028 347.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"349.6088,-662.2222 347.5,-657.2222 345.3913,-662.2223 349.6088,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_EvgA -->\n",
       "<g id=\"node158\" class=\"node\">\n",
       "<title>data1_EvgA</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">EvgA</text>\n",
       "</g>\n",
       "<!-- data1_EvgA&#45;&gt;data2_run2_33 -->\n",
       "<g id=\"edge71\" class=\"edge\">\n",
       "<title>data1_EvgA&#45;&gt;data2_run2_33</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M714.5,-678.9664C714.5,-673.6707 714.5,-667.8028 714.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"716.6745,-662.2222 714.5,-657.2222 712.3257,-662.2223 716.6745,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;2 -->\n",
       "<g id=\"node159\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-908.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-893.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;2&#45;&gt;data2_run2_12 -->\n",
       "<g id=\"edge72\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;2&#45;&gt;data2_run2_12</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M460.5,-885.6812C460.5,-880.5393 460.5,-874.927 460.5,-869.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"462.6307,-869.3596 460.5,-864.3597 458.3695,-869.3597 462.6307,-869.3596\"/>\n",
       "</g>\n",
       "<!-- data1_Zinc -->\n",
       "<g id=\"node160\" class=\"node\">\n",
       "<title>data1_Zinc</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Zinc</text>\n",
       "</g>\n",
       "<!-- data1_Zinc&#45;&gt;data2_run2_43 -->\n",
       "<g id=\"edge73\" class=\"edge\">\n",
       "<title>data1_Zinc&#45;&gt;data2_run2_43</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.99\" d=\"M894.5,-575.9664C894.5,-570.6707 894.5,-564.8028 894.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.99\" points=\"896.6832,-559.2222 894.5,-554.2222 892.317,-559.2223 896.6832,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_SoxS -->\n",
       "<g id=\"node161\" class=\"node\">\n",
       "<title>data1_SoxS</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">SoxS</text>\n",
       "</g>\n",
       "<!-- data1_SoxS&#45;&gt;data2_run2_22 -->\n",
       "<g id=\"edge74\" class=\"edge\">\n",
       "<title>data1_SoxS&#45;&gt;data2_run2_22</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M545.5,-782.9664C545.5,-777.6707 545.5,-771.8028 545.5,-766.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"547.6745,-766.2222 545.5,-761.2222 543.3257,-766.2223 547.6745,-766.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Nac -->\n",
       "<g id=\"node162\" class=\"node\">\n",
       "<title>data1_Nac</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Nac</text>\n",
       "</g>\n",
       "<!-- data1_Nac&#45;&gt;data2_run2_31 -->\n",
       "<g id=\"edge75\" class=\"edge\">\n",
       "<title>data1_Nac&#45;&gt;data2_run2_31</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M545.5,-678.9664C545.5,-673.6707 545.5,-667.8028 545.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"547.6438,-662.2222 545.5,-657.2222 543.3563,-662.2223 547.6438,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_curli -->\n",
       "<g id=\"node163\" class=\"node\">\n",
       "<title>data1_curli</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">curli</text>\n",
       "</g>\n",
       "<!-- data1_curli&#45;&gt;data2_run2_88 -->\n",
       "<g id=\"edge76\" class=\"edge\">\n",
       "<title>data1_curli&#45;&gt;data2_run2_88</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.85\" d=\"M231.5,-58.9664C231.5,-53.6707 231.5,-47.8028 231.5,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.85\" points=\"233.622,-42.2222 231.5,-37.2222 229.3782,-42.2223 233.622,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_gadWX&#45;KO -->\n",
       "<g id=\"node164\" class=\"node\">\n",
       "<title>data1_gadWX&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">gadWX&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_gadWX&#45;KO&#45;&gt;data2_run2_45 -->\n",
       "<g id=\"edge77\" class=\"edge\">\n",
       "<title>data1_gadWX&#45;KO&#45;&gt;data2_run2_45</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M43.5,-472.9664C43.5,-467.6707 43.5,-461.8028 43.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"45.6482,-456.2222 43.5,-451.2222 41.352,-456.2223 45.6482,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_flu&#45;yeeRS -->\n",
       "<g id=\"node165\" class=\"node\">\n",
       "<title>data1_flu&#45;yeeRS</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">flu&#45;yeeRS</text>\n",
       "</g>\n",
       "<!-- data1_flu&#45;yeeRS&#45;&gt;data2_run2_77 -->\n",
       "<g id=\"edge78\" class=\"edge\">\n",
       "<title>data1_flu&#45;yeeRS&#45;&gt;data2_run2_77</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M231.5,-161.9664C231.5,-156.6707 231.5,-150.8028 231.5,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"233.6613,-145.2222 231.5,-140.2222 229.3388,-145.2223 233.6613,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;1 -->\n",
       "<g id=\"node166\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-390.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"347.5\" y=\"-375.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;1&#45;&gt;data2_run2_57 -->\n",
       "<g id=\"edge79\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;1&#45;&gt;data2_run2_57</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.85\" d=\"M347.5,-367.6812C347.5,-362.5393 347.5,-356.927 347.5,-351.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.85\" points=\"349.622,-351.3596 347.5,-346.3597 345.3782,-351.3597 349.622,-351.3596\"/>\n",
       "</g>\n",
       "<!-- data1_FadR/IclR -->\n",
       "<g id=\"node167\" class=\"node\">\n",
       "<title>data1_FadR/IclR</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FadR/IclR</text>\n",
       "</g>\n",
       "<!-- data1_FadR/IclR&#45;&gt;data2_run2_58 -->\n",
       "<g id=\"edge80\" class=\"edge\">\n",
       "<title>data1_FadR/IclR&#45;&gt;data2_run2_58</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M460.5,-368.9664C460.5,-363.6707 460.5,-357.8028 460.5,-352.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"462.6482,-352.2222 460.5,-347.2222 458.352,-352.2223 462.6482,-352.2222\"/>\n",
       "</g>\n",
       "<!-- data1_lipopolys&#45;\n",
       "accharide -->\n",
       "<g id=\"node168\" class=\"node\">\n",
       "<title>data1_lipopolys&#45;\n",
       "accharide</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-390.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lipopolys&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-375.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">accharide</text>\n",
       "</g>\n",
       "<!-- data1_lipopolys&#45;\n",
       "accharide&#45;&gt;data2_run2_60 -->\n",
       "<g id=\"edge81\" class=\"edge\">\n",
       "<title>data1_lipopolys&#45;\n",
       "accharide&#45;&gt;data2_run2_60</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.86\" d=\"M714.5,-367.6812C714.5,-362.5393 714.5,-356.927 714.5,-351.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.86\" points=\"716.6263,-351.3596 714.5,-346.3597 712.3738,-351.3597 716.6263,-351.3596\"/>\n",
       "</g>\n",
       "<!-- data1_FlhDC -->\n",
       "<g id=\"node169\" class=\"node\">\n",
       "<title>data1_FlhDC</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FlhDC</text>\n",
       "</g>\n",
       "<!-- data1_FlhDC&#45;&gt;data2_run2_70 -->\n",
       "<g id=\"edge82\" class=\"edge\">\n",
       "<title>data1_FlhDC&#45;&gt;data2_run2_70</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.98\" d=\"M714.5,-264.9664C714.5,-259.6707 714.5,-253.8028 714.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.98\" points=\"716.6788,-248.2222 714.5,-243.2222 712.3213,-248.2223 716.6788,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_crp&#45;KO -->\n",
       "<g id=\"node170\" class=\"node\">\n",
       "<title>data1_crp&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"714.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">crp&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_crp&#45;KO&#45;&gt;data2_run2_51 -->\n",
       "<g id=\"edge83\" class=\"edge\">\n",
       "<title>data1_crp&#45;KO&#45;&gt;data2_run2_51</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M714.5,-472.9664C714.5,-467.6707 714.5,-461.8028 714.5,-456.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"716.6395,-456.2222 714.5,-451.2222 712.3607,-456.2223 716.6395,-456.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GntR/TyrR -->\n",
       "<g id=\"node171\" class=\"node\">\n",
       "<title>data1_GntR/TyrR</title>\n",
       "<text text-anchor=\"middle\" x=\"894.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GntR/TyrR</text>\n",
       "</g>\n",
       "<!-- data1_GntR/TyrR&#45;&gt;data2_run2_16 -->\n",
       "<g id=\"edge84\" class=\"edge\">\n",
       "<title>data1_GntR/TyrR&#45;&gt;data2_run2_16</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M894.5,-886.9664C894.5,-881.6707 894.5,-875.8028 894.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"896.6657,-870.2222 894.5,-865.2222 892.3345,-870.2223 896.6657,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_purR&#45;KO -->\n",
       "<g id=\"node172\" class=\"node\">\n",
       "<title>data1_purR&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"630.5\" y=\"-590.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">purR&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_purR&#45;KO&#45;&gt;data2_run2_41 -->\n",
       "<g id=\"edge85\" class=\"edge\">\n",
       "<title>data1_purR&#45;KO&#45;&gt;data2_run2_41</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.78\" d=\"M630.5,-575.9664C630.5,-570.6707 630.5,-564.8028 630.5,-559.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.78\" points=\"632.5913,-559.2222 630.5,-554.2222 628.4088,-559.2223 632.5913,-559.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RpoS -->\n",
       "<g id=\"node173\" class=\"node\">\n",
       "<title>data1_RpoS</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-901.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RpoS</text>\n",
       "</g>\n",
       "<!-- data1_RpoS&#45;&gt;data2_run2_13 -->\n",
       "<g id=\"edge86\" class=\"edge\">\n",
       "<title>data1_RpoS&#45;&gt;data2_run2_13</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M545.5,-886.9664C545.5,-881.6707 545.5,-875.8028 545.5,-870.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"547.6745,-870.2222 545.5,-865.2222 543.3257,-870.2223 547.6745,-870.2222\"/>\n",
       "</g>\n",
       "<!-- data1_membrane -->\n",
       "<g id=\"node174\" class=\"node\">\n",
       "<title>data1_membrane</title>\n",
       "<text text-anchor=\"middle\" x=\"1078.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">membrane</text>\n",
       "</g>\n",
       "<!-- data1_membrane&#45;&gt;data2_run2_35 -->\n",
       "<g id=\"edge87\" class=\"edge\">\n",
       "<title>data1_membrane&#45;&gt;data2_run2_35</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M1078.5,-678.9664C1078.5,-673.6707 1078.5,-667.8028 1078.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"1080.6526,-662.2222 1078.5,-657.2222 1076.3476,-662.2223 1080.6526,-662.2222\"/>\n",
       "</g>\n",
       "<!-- data1_duplic&#45;\n",
       "ation&#45;1 -->\n",
       "<g id=\"node175\" class=\"node\">\n",
       "<title>data1_duplic&#45;\n",
       "ation&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-390.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">duplic&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"231.5\" y=\"-375.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ation&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_duplic&#45;\n",
       "ation&#45;1&#45;&gt;data2_run2_56 -->\n",
       "<g id=\"edge88\" class=\"edge\">\n",
       "<title>data1_duplic&#45;\n",
       "ation&#45;1&#45;&gt;data2_run2_56</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.96\" d=\"M231.5,-367.6812C231.5,-362.5393 231.5,-356.927 231.5,-351.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.96\" points=\"233.6701,-351.3596 231.5,-346.3597 229.3301,-351.3597 233.6701,-351.3596\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;3 -->\n",
       "<g id=\"node176\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;3</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-80.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-65.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;3</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;3&#45;&gt;data2_run2_91 -->\n",
       "<g id=\"edge89\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;3&#45;&gt;data2_run2_91</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M545.5,-57.6812C545.5,-52.5393 545.5,-46.927 545.5,-41.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"547.6088,-41.3596 545.5,-36.3597 543.3913,-41.3597 547.6088,-41.3596\"/>\n",
       "</g>\n",
       "<!-- data1_ExuR/FucR -->\n",
       "<g id=\"node177\" class=\"node\">\n",
       "<title>data1_ExuR/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"864.5\" y=\"-797.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ExuR/FucR</text>\n",
       "</g>\n",
       "<!-- data1_ExuR/FucR&#45;&gt;data2_run2_25 -->\n",
       "<g id=\"edge90\" class=\"edge\">\n",
       "<title>data1_ExuR/FucR&#45;&gt;data2_run2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.11\" d=\"M863.8891,-782.9775C863.7026,-777.477 863.4948,-771.3458 863.2978,-765.5365\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.11\" points=\"865.0395,-765.2572 863.121,-760.3193 861.5415,-765.3758 865.0395,-765.2572\"/>\n",
       "</g>\n",
       "<!-- data1_ExuR/FucR&#45;&gt;data2_run2_68 -->\n",
       "<g id=\"edge91\" class=\"edge\">\n",
       "<title>data1_ExuR/FucR&#45;&gt;data2_run2_68</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.75\" d=\"M884.3553,-782.9775C891.0336,-776.9157 898.5557,-770.0879 905.5169,-763.7692\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.75\" points=\"907.0121,-765.2187 909.3177,-760.3193 904.2186,-762.1411 907.0121,-765.2187\"/>\n",
       "</g>\n",
       "<!-- data1_FecI -->\n",
       "<g id=\"node178\" class=\"node\">\n",
       "<title>data1_FecI</title>\n",
       "<text text-anchor=\"middle\" x=\"545.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FecI</text>\n",
       "</g>\n",
       "<!-- data1_FecI&#45;&gt;data2_run2_69 -->\n",
       "<g id=\"edge92\" class=\"edge\">\n",
       "<title>data1_FecI&#45;&gt;data2_run2_69</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M545.5,-264.9664C545.5,-259.6707 545.5,-253.8028 545.5,-248.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"547.6351,-248.2222 545.5,-243.2222 543.3651,-248.2223 547.6351,-248.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RbsR -->\n",
       "<g id=\"node179\" class=\"node\">\n",
       "<title>data1_RbsR</title>\n",
       "<text text-anchor=\"middle\" x=\"460.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RbsR</text>\n",
       "</g>\n",
       "<!-- data1_RbsR&#45;&gt;data2_run2_90 -->\n",
       "<g id=\"edge93\" class=\"edge\">\n",
       "<title>data1_RbsR&#45;&gt;data2_run2_90</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.63\" d=\"M460.5,-58.9664C460.5,-53.6707 460.5,-47.8028 460.5,-42.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.63\" points=\"462.5257,-42.2222 460.5,-37.2222 458.4745,-42.2223 462.5257,-42.2222\"/>\n",
       "</g>\n",
       "<!-- data1_sgrT -->\n",
       "<g id=\"node180\" class=\"node\">\n",
       "<title>data1_sgrT</title>\n",
       "<text text-anchor=\"middle\" x=\"43.5\" y=\"-693.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">sgrT</text>\n",
       "</g>\n",
       "<!-- data1_sgrT&#45;&gt;data2_run2_27 -->\n",
       "<g id=\"edge94\" class=\"edge\">\n",
       "<title>data1_sgrT&#45;&gt;data2_run2_27</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.88\" d=\"M43.5,-678.9664C43.5,-673.6707 43.5,-667.8028 43.5,-662.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.88\" points=\"45.2501,-662.2222 43.5,-657.2222 41.7501,-662.2223 45.2501,-662.2222\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7fe01a2139a0>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots,links = compare_ica(ica_data.M,ica_run_diff)\n",
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For comparisons of ICA components between the same organism,\n",
    " you find that components will be very closely correlated with each other.\n",
    "  The thickness of the arrow will indicate the correlation coefficient between\n",
    "  components, and arrows can also indicate if components split/merge.\n",
    "\n",
    "You can also see components which do not have any correlations by using the\n",
    "keyword `show_all`. The default for `show_all` is **False**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 92/92 [00:01<00:00, 78.21it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"720pt\" height=\"631pt\"\n",
       " viewBox=\"0.00 0.00 720.00 630.97\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(.6452 .6452) rotate(0) translate(4 974)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-974 1112,-974 1112,4 -4,4\"/>\n",
       "<!-- data2_run2_0 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_run2_0</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_0</text>\n",
       "</g>\n",
       "<!-- data2_run2_1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_run2_1</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_1</text>\n",
       "</g>\n",
       "<!-- data2_run2_10 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_run2_10</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_10</text>\n",
       "</g>\n",
       "<!-- data2_run2_11 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_run2_11</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_11</text>\n",
       "</g>\n",
       "<!-- data2_run2_12 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_run2_12</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_12</text>\n",
       "</g>\n",
       "<!-- data2_run2_13 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data2_run2_13</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_13</text>\n",
       "</g>\n",
       "<!-- data2_run2_14 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data2_run2_14</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_14</text>\n",
       "</g>\n",
       "<!-- data2_run2_15 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data2_run2_15</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-888.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_15</text>\n",
       "</g>\n",
       "<!-- data2_run2_16 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data2_run2_16</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_16</text>\n",
       "</g>\n",
       "<!-- data2_run2_17 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data2_run2_17</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-889.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_17</text>\n",
       "</g>\n",
       "<!-- data2_run2_18 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>data2_run2_18</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_18</text>\n",
       "</g>\n",
       "<!-- data2_run2_19 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>data2_run2_19</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_19</text>\n",
       "</g>\n",
       "<!-- data2_run2_2 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>data2_run2_2</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-784.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_2</text>\n",
       "</g>\n",
       "<!-- data2_run2_20 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>data2_run2_20</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_20</text>\n",
       "</g>\n",
       "<!-- data2_run2_21 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>data2_run2_21</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_21</text>\n",
       "</g>\n",
       "<!-- data2_run2_22 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>data2_run2_22</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_22</text>\n",
       "</g>\n",
       "<!-- data2_run2_23 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>data2_run2_23</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_23</text>\n",
       "</g>\n",
       "<!-- data2_run2_24 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>data2_run2_24</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-785.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_24</text>\n",
       "</g>\n",
       "<!-- data2_run2_25 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>data2_run2_25</title>\n",
       "<text text-anchor=\"middle\" x=\"850.5\" y=\"-784.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_25</text>\n",
       "</g>\n",
       "<!-- data2_run2_26 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>data2_run2_26</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-784.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_26</text>\n",
       "</g>\n",
       "<!-- data2_run2_27 -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>data2_run2_27</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_27</text>\n",
       "</g>\n",
       "<!-- data2_run2_28 -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>data2_run2_28</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_28</text>\n",
       "</g>\n",
       "<!-- data2_run2_29 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>data2_run2_29</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_29</text>\n",
       "</g>\n",
       "<!-- data2_run2_3 -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>data2_run2_3</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_3</text>\n",
       "</g>\n",
       "<!-- data2_run2_30 -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>data2_run2_30</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-680.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_30</text>\n",
       "</g>\n",
       "<!-- data2_run2_31 -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>data2_run2_31</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_31</text>\n",
       "</g>\n",
       "<!-- data2_run2_32 -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>data2_run2_32</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_32</text>\n",
       "</g>\n",
       "<!-- data2_run2_33 -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>data2_run2_33</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_33</text>\n",
       "</g>\n",
       "<!-- data2_run2_34 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>data2_run2_34</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_34</text>\n",
       "</g>\n",
       "<!-- data2_run2_35 -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>data2_run2_35</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-681.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_35</text>\n",
       "</g>\n",
       "<!-- data2_run2_36 -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>data2_run2_36</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_36</text>\n",
       "</g>\n",
       "<!-- data2_run2_37 -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>data2_run2_37</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_37</text>\n",
       "</g>\n",
       "<!-- data2_run2_38 -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>data2_run2_38</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_38</text>\n",
       "</g>\n",
       "<!-- data2_run2_39 -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>data2_run2_39</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_39</text>\n",
       "</g>\n",
       "<!-- data2_run2_4 -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>data2_run2_4</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_4</text>\n",
       "</g>\n",
       "<!-- data2_run2_40 -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>data2_run2_40</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_40</text>\n",
       "</g>\n",
       "<!-- data2_run2_41 -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>data2_run2_41</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_41</text>\n",
       "</g>\n",
       "<!-- data2_run2_42 -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>data2_run2_42</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_42</text>\n",
       "</g>\n",
       "<!-- data2_run2_43 -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>data2_run2_43</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_43</text>\n",
       "</g>\n",
       "<!-- data2_run2_44 -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>data2_run2_44</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-578.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_44</text>\n",
       "</g>\n",
       "<!-- data2_run2_45 -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>data2_run2_45</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_45</text>\n",
       "</g>\n",
       "<!-- data2_run2_46 -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>data2_run2_46</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_46</text>\n",
       "</g>\n",
       "<!-- data2_run2_47 -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>data2_run2_47</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_47</text>\n",
       "</g>\n",
       "<!-- data2_run2_48 -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>data2_run2_48</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_48</text>\n",
       "</g>\n",
       "<!-- data2_run2_49 -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>data2_run2_49</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_49</text>\n",
       "</g>\n",
       "<!-- data2_run2_5 -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>data2_run2_5</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_5</text>\n",
       "</g>\n",
       "<!-- data2_run2_50 -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>data2_run2_50</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-474.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_50</text>\n",
       "</g>\n",
       "<!-- data2_run2_51 -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>data2_run2_51</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_51</text>\n",
       "</g>\n",
       "<!-- data2_run2_52 -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>data2_run2_52</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_52</text>\n",
       "</g>\n",
       "<!-- data2_run2_53 -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>data2_run2_53</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-475.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_53</text>\n",
       "</g>\n",
       "<!-- data2_run2_54 -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>data2_run2_54</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-371.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_54</text>\n",
       "</g>\n",
       "<!-- data2_run2_55 -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>data2_run2_55</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-371.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_55</text>\n",
       "</g>\n",
       "<!-- data2_run2_56 -->\n",
       "<g id=\"node53\" class=\"node\">\n",
       "<title>data2_run2_56</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-370.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_56</text>\n",
       "</g>\n",
       "<!-- data2_run2_57 -->\n",
       "<g id=\"node54\" class=\"node\">\n",
       "<title>data2_run2_57</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-370.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_57</text>\n",
       "</g>\n",
       "<!-- data2_run2_58 -->\n",
       "<g id=\"node55\" class=\"node\">\n",
       "<title>data2_run2_58</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-371.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_58</text>\n",
       "</g>\n",
       "<!-- data2_run2_59 -->\n",
       "<g id=\"node56\" class=\"node\">\n",
       "<title>data2_run2_59</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-371.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_59</text>\n",
       "</g>\n",
       "<!-- data2_run2_6 -->\n",
       "<g id=\"node57\" class=\"node\">\n",
       "<title>data2_run2_6</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-371.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_6</text>\n",
       "</g>\n",
       "<!-- data2_run2_60 -->\n",
       "<g id=\"node58\" class=\"node\">\n",
       "<title>data2_run2_60</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-370.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_60</text>\n",
       "</g>\n",
       "<!-- data2_run2_61 -->\n",
       "<g id=\"node59\" class=\"node\">\n",
       "<title>data2_run2_61</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-370.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_61</text>\n",
       "</g>\n",
       "<!-- data2_run2_62 -->\n",
       "<g id=\"node60\" class=\"node\">\n",
       "<title>data2_run2_62</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-371.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_62</text>\n",
       "</g>\n",
       "<!-- data2_run2_63 -->\n",
       "<g id=\"node61\" class=\"node\">\n",
       "<title>data2_run2_63</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_63</text>\n",
       "</g>\n",
       "<!-- data2_run2_64 -->\n",
       "<g id=\"node62\" class=\"node\">\n",
       "<title>data2_run2_64</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_64</text>\n",
       "</g>\n",
       "<!-- data2_run2_65 -->\n",
       "<g id=\"node63\" class=\"node\">\n",
       "<title>data2_run2_65</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_65</text>\n",
       "</g>\n",
       "<!-- data2_run2_66 -->\n",
       "<g id=\"node64\" class=\"node\">\n",
       "<title>data2_run2_66</title>\n",
       "<text text-anchor=\"middle\" x=\"311.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_66</text>\n",
       "</g>\n",
       "<!-- data2_run2_67 -->\n",
       "<g id=\"node65\" class=\"node\">\n",
       "<title>data2_run2_67</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_67</text>\n",
       "</g>\n",
       "<!-- data2_run2_68 -->\n",
       "<g id=\"node66\" class=\"node\">\n",
       "<title>data2_run2_68</title>\n",
       "<text text-anchor=\"middle\" x=\"917.5\" y=\"-784.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_68</text>\n",
       "</g>\n",
       "<!-- data2_run2_69 -->\n",
       "<g id=\"node67\" class=\"node\">\n",
       "<title>data2_run2_69</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_69</text>\n",
       "</g>\n",
       "<!-- data2_run2_7 -->\n",
       "<g id=\"node68\" class=\"node\">\n",
       "<title>data2_run2_7</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_7</text>\n",
       "</g>\n",
       "<!-- data2_run2_70 -->\n",
       "<g id=\"node69\" class=\"node\">\n",
       "<title>data2_run2_70</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_70</text>\n",
       "</g>\n",
       "<!-- data2_run2_71 -->\n",
       "<g id=\"node70\" class=\"node\">\n",
       "<title>data2_run2_71</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-296.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">run2_71</text>\n",
       "</g>\n",
       "<!-- data2_run2_72 -->\n",
       "<g id=\"node71\" class=\"node\">\n",
       "<title>data2_run2_72</title>\n",
       "<text text-anchor=\"middle\" x=\"984.5\" y=\"-784.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_72</text>\n",
       "</g>\n",
       "<!-- data2_run2_73 -->\n",
       "<g id=\"node72\" class=\"node\">\n",
       "<title>data2_run2_73</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-266.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_73</text>\n",
       "</g>\n",
       "<!-- data2_run2_74 -->\n",
       "<g id=\"node73\" class=\"node\">\n",
       "<title>data2_run2_74</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_74</text>\n",
       "</g>\n",
       "<!-- data2_run2_75 -->\n",
       "<g id=\"node74\" class=\"node\">\n",
       "<title>data2_run2_75</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_75</text>\n",
       "</g>\n",
       "<!-- data2_run2_76 -->\n",
       "<g id=\"node75\" class=\"node\">\n",
       "<title>data2_run2_76</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_76</text>\n",
       "</g>\n",
       "<!-- data2_run2_77 -->\n",
       "<g id=\"node76\" class=\"node\">\n",
       "<title>data2_run2_77</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_77</text>\n",
       "</g>\n",
       "<!-- data2_run2_78 -->\n",
       "<g id=\"node77\" class=\"node\">\n",
       "<title>data2_run2_78</title>\n",
       "<text text-anchor=\"middle\" x=\"781.5\" y=\"-784.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_78</text>\n",
       "</g>\n",
       "<!-- data2_run2_79 -->\n",
       "<g id=\"node78\" class=\"node\">\n",
       "<title>data2_run2_79</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_79</text>\n",
       "</g>\n",
       "<!-- data2_run2_8 -->\n",
       "<g id=\"node79\" class=\"node\">\n",
       "<title>data2_run2_8</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-193.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">run2_8</text>\n",
       "</g>\n",
       "<!-- data2_run2_80 -->\n",
       "<g id=\"node80\" class=\"node\">\n",
       "<title>data2_run2_80</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_80</text>\n",
       "</g>\n",
       "<!-- data2_run2_81 -->\n",
       "<g id=\"node81\" class=\"node\">\n",
       "<title>data2_run2_81</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_81</text>\n",
       "</g>\n",
       "<!-- data2_run2_82 -->\n",
       "<g id=\"node82\" class=\"node\">\n",
       "<title>data2_run2_82</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_82</text>\n",
       "</g>\n",
       "<!-- data2_run2_83 -->\n",
       "<g id=\"node83\" class=\"node\">\n",
       "<title>data2_run2_83</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-164.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_83</text>\n",
       "</g>\n",
       "<!-- data2_run2_84 -->\n",
       "<g id=\"node84\" class=\"node\">\n",
       "<title>data2_run2_84</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_84</text>\n",
       "</g>\n",
       "<!-- data2_run2_85 -->\n",
       "<g id=\"node85\" class=\"node\">\n",
       "<title>data2_run2_85</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_85</text>\n",
       "</g>\n",
       "<!-- data2_run2_86 -->\n",
       "<g id=\"node86\" class=\"node\">\n",
       "<title>data2_run2_86</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-60.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_86</text>\n",
       "</g>\n",
       "<!-- data2_run2_87 -->\n",
       "<g id=\"node87\" class=\"node\">\n",
       "<title>data2_run2_87</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_87</text>\n",
       "</g>\n",
       "<!-- data2_run2_88 -->\n",
       "<g id=\"node88\" class=\"node\">\n",
       "<title>data2_run2_88</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_88</text>\n",
       "</g>\n",
       "<!-- data2_run2_89 -->\n",
       "<g id=\"node89\" class=\"node\">\n",
       "<title>data2_run2_89</title>\n",
       "<text text-anchor=\"middle\" x=\"378.5\" y=\"-267.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_89</text>\n",
       "</g>\n",
       "<!-- data2_run2_9 -->\n",
       "<g id=\"node90\" class=\"node\">\n",
       "<title>data2_run2_9</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_9</text>\n",
       "</g>\n",
       "<!-- data2_run2_90 -->\n",
       "<g id=\"node91\" class=\"node\">\n",
       "<title>data2_run2_90</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_90</text>\n",
       "</g>\n",
       "<!-- data2_run2_91 -->\n",
       "<g id=\"node92\" class=\"node\">\n",
       "<title>data2_run2_91</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-60.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_91</text>\n",
       "</g>\n",
       "<!-- data2_run2_92 -->\n",
       "<g id=\"node93\" class=\"node\">\n",
       "<title>data2_run2_92</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-90.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">run2_92</text>\n",
       "</g>\n",
       "<!-- data2_run2_93 -->\n",
       "<g id=\"node94\" class=\"node\">\n",
       "<title>data2_run2_93</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-61.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_93</text>\n",
       "</g>\n",
       "<!-- data1_YieP -->\n",
       "<g id=\"node95\" class=\"node\">\n",
       "<title>data1_YieP</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YieP</text>\n",
       "</g>\n",
       "<!-- data1_YieP&#45;&gt;data2_run2_44 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_YieP&#45;&gt;data2_run2_44</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M1066.5,-621.9664C1066.5,-616.6707 1066.5,-610.8028 1066.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"1068.6351,-605.2222 1066.5,-600.2222 1064.3651,-605.2223 1068.6351,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GlpR -->\n",
       "<g id=\"node96\" class=\"node\">\n",
       "<title>data1_GlpR</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GlpR</text>\n",
       "</g>\n",
       "<!-- data1_GlpR&#45;&gt;data2_run2_24 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_GlpR&#45;&gt;data2_run2_24</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M701.5,-828.9664C701.5,-823.6707 701.5,-817.8028 701.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"703.6745,-812.2222 701.5,-807.2222 699.3257,-812.2223 703.6745,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_YgbI -->\n",
       "<g id=\"node97\" class=\"node\">\n",
       "<title>data1_YgbI</title>\n",
       "<text text-anchor=\"middle\" x=\"781.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YgbI</text>\n",
       "</g>\n",
       "<!-- data1_YgbI&#45;&gt;data2_run2_25 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_YgbI&#45;&gt;data2_run2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.05\" d=\"M802.5772,-828.9775C809.6664,-822.9157 817.6514,-816.0879 825.0411,-809.7692\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.05\" points=\"826.4128,-810.8989 829.0757,-806.3193 824.1382,-808.2387 826.4128,-810.8989\"/>\n",
       "</g>\n",
       "<!-- data1_YgbI&#45;&gt;data2_run2_78 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_YgbI&#45;&gt;data2_run2_78</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.78\" d=\"M781.5,-828.9775C781.5,-823.477 781.5,-817.3458 781.5,-811.5365\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.78\" points=\"783.5913,-811.3193 781.5,-806.3193 779.4088,-811.3194 783.5913,-811.3193\"/>\n",
       "</g>\n",
       "<!-- data1_efeU&#45;&#45;\n",
       "repair -->\n",
       "<g id=\"node98\" class=\"node\">\n",
       "<title>data1_efeU&#45;&#45;\n",
       "repair</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-850.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">efeU&#45;&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-835.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">repair</text>\n",
       "</g>\n",
       "<!-- data1_efeU&#45;&#45;\n",
       "repair&#45;&gt;data2_run2_2 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_efeU&#45;&#45;\n",
       "repair&#45;&gt;data2_run2_2</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M227.5,-827.6812C227.5,-822.5393 227.5,-816.927 227.5,-811.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"229.6438,-811.3596 227.5,-806.3597 225.3563,-811.3597 229.6438,-811.3596\"/>\n",
       "</g>\n",
       "<!-- data1_ydcI&#45;KO -->\n",
       "<g id=\"node99\" class=\"node\">\n",
       "<title>data1_ydcI&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ydcI&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_ydcI&#45;KO&#45;&gt;data2_run2_79 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>data1_ydcI&#45;KO&#45;&gt;data2_run2_79</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M457.5,-207.9664C457.5,-202.6707 457.5,-196.8028 457.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"459.6351,-191.2222 457.5,-186.2222 455.3651,-191.2223 459.6351,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_BW25113 -->\n",
       "<g id=\"node100\" class=\"node\">\n",
       "<title>data1_BW25113</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">BW25113</text>\n",
       "</g>\n",
       "<!-- data1_BW25113&#45;&gt;data2_run2_7 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>data1_BW25113&#45;&gt;data2_run2_7</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.84\" d=\"M616.5,-310.9664C616.5,-305.6707 616.5,-299.8028 616.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.84\" points=\"618.6176,-294.2222 616.5,-289.2222 614.3826,-294.2223 618.6176,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_proVWX -->\n",
       "<g id=\"node101\" class=\"node\">\n",
       "<title>data1_proVWX</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">proVWX</text>\n",
       "</g>\n",
       "<!-- data1_proVWX&#45;&gt;data2_run2_11 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>data1_proVWX&#45;&gt;data2_run2_11</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.68\" d=\"M344.5,-932.9664C344.5,-927.6707 344.5,-921.8028 344.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.68\" points=\"346.5476,-916.2222 344.5,-911.2222 342.4526,-916.2223 346.5476,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_DhaR/Mlc -->\n",
       "<g id=\"node102\" class=\"node\">\n",
       "<title>data1_DhaR/Mlc</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">DhaR/Mlc</text>\n",
       "</g>\n",
       "<!-- data1_DhaR/Mlc&#45;&gt;data2_run2_32 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>data1_DhaR/Mlc&#45;&gt;data2_run2_32</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.84\" d=\"M616.5,-724.9664C616.5,-719.6707 616.5,-713.8028 616.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.84\" points=\"618.6176,-708.2222 616.5,-703.2222 614.3826,-708.2223 618.6176,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_ArgR -->\n",
       "<g id=\"node103\" class=\"node\">\n",
       "<title>data1_ArgR</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArgR</text>\n",
       "</g>\n",
       "<!-- data1_ArgR&#45;&gt;data2_run2_9 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>data1_ArgR&#45;&gt;data2_run2_9</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M537.5,-104.9664C537.5,-99.6707 537.5,-93.8028 537.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"539.6657,-88.2222 537.5,-83.2222 535.3345,-88.2223 539.6657,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_YneJ -->\n",
       "<g id=\"node104\" class=\"node\">\n",
       "<title>data1_YneJ</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YneJ</text>\n",
       "</g>\n",
       "<!-- data1_YneJ&#45;&gt;data2_run2_82 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>data1_YneJ&#45;&gt;data2_run2_82</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.84\" d=\"M882.5,-207.9664C882.5,-202.6707 882.5,-196.8028 882.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.84\" points=\"884.6176,-191.2222 882.5,-186.2222 880.3826,-191.2223 884.6176,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NarL -->\n",
       "<g id=\"node105\" class=\"node\">\n",
       "<title>data1_NarL</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NarL</text>\n",
       "</g>\n",
       "<!-- data1_NarL&#45;&gt;data2_run2_83 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>data1_NarL&#45;&gt;data2_run2_83</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M1066.5,-207.9664C1066.5,-202.6707 1066.5,-196.8028 1066.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"1068.6526,-191.2222 1066.5,-186.2222 1064.3476,-191.2223 1068.6526,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_e14&#45;de&#45;\n",
       "letion -->\n",
       "<g id=\"node106\" class=\"node\">\n",
       "<title>data1_e14&#45;de&#45;\n",
       "letion</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-436.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">e14&#45;de&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-421.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">letion</text>\n",
       "</g>\n",
       "<!-- data1_e14&#45;de&#45;\n",
       "letion&#45;&gt;data2_run2_61 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>data1_e14&#45;de&#45;\n",
       "letion&#45;&gt;data2_run2_61</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M882.5,-413.6812C882.5,-408.5393 882.5,-402.927 882.5,-397.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"884.6526,-397.3596 882.5,-392.3597 880.3476,-397.3597 884.6526,-397.3596\"/>\n",
       "</g>\n",
       "<!-- data1_MalT -->\n",
       "<g id=\"node107\" class=\"node\">\n",
       "<title>data1_MalT</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">MalT</text>\n",
       "</g>\n",
       "<!-- data1_MalT&#45;&gt;data2_run2_5 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>data1_MalT&#45;&gt;data2_run2_5</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M537.5,-518.9664C537.5,-513.6707 537.5,-507.8028 537.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"539.6613,-502.2222 537.5,-497.2222 535.3388,-502.2223 539.6613,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_OxyR -->\n",
       "<g id=\"node108\" class=\"node\">\n",
       "<title>data1_OxyR</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">OxyR</text>\n",
       "</g>\n",
       "<!-- data1_OxyR&#45;&gt;data2_run2_0 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>data1_OxyR&#45;&gt;data2_run2_0</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M45.5,-932.9664C45.5,-927.6707 45.5,-921.8028 45.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"47.6613,-916.2222 45.5,-911.2222 43.3388,-916.2223 47.6613,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;2 -->\n",
       "<g id=\"node109\" class=\"node\">\n",
       "<title>data1_Crp&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Crp&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;2&#45;&gt;data2_run2_37 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>data1_Crp&#45;2&#45;&gt;data2_run2_37</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.96\" d=\"M137.5,-621.9664C137.5,-616.6707 137.5,-610.8028 137.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.96\" points=\"139.6701,-605.2222 137.5,-600.2222 135.3301,-605.2223 139.6701,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_AtoC -->\n",
       "<g id=\"node110\" class=\"node\">\n",
       "<title>data1_AtoC</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AtoC</text>\n",
       "</g>\n",
       "<!-- data1_AtoC&#45;&gt;data2_run2_6 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>data1_AtoC&#45;&gt;data2_run2_6</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.81\" d=\"M616.5,-414.9664C616.5,-409.6707 616.5,-403.8028 616.5,-398.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.81\" points=\"618.6045,-398.2222 616.5,-393.2222 614.3957,-398.2223 618.6045,-398.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1 -->\n",
       "<g id=\"node111\" class=\"node\">\n",
       "<title>data1_PurR&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1&#45;&gt;data2_run2_64 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>data1_PurR&#45;1&#45;&gt;data2_run2_64</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M137.5,-310.9664C137.5,-305.6707 137.5,-299.8028 137.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"139.6613,-294.2222 137.5,-289.2222 135.3388,-294.2223 139.6613,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_yheO&#45;KO -->\n",
       "<g id=\"node112\" class=\"node\">\n",
       "<title>data1_yheO&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">yheO&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_yheO&#45;KO&#45;&gt;data2_run2_46 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>data1_yheO&#45;KO&#45;&gt;data2_run2_46</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M137.5,-518.9664C137.5,-513.6707 137.5,-507.8028 137.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"139.6307,-502.2222 137.5,-497.2222 135.3695,-502.2223 139.6307,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CsqR -->\n",
       "<g id=\"node113\" class=\"node\">\n",
       "<title>data1_CsqR</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CsqR</text>\n",
       "</g>\n",
       "<!-- data1_CsqR&#45;&gt;data2_run2_14 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>data1_CsqR&#45;&gt;data2_run2_14</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M616.5,-932.9664C616.5,-927.6707 616.5,-921.8028 616.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"618.6657,-916.2222 616.5,-911.2222 614.3345,-916.2223 618.6657,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_MetJ -->\n",
       "<g id=\"node114\" class=\"node\">\n",
       "<title>data1_MetJ</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">MetJ</text>\n",
       "</g>\n",
       "<!-- data1_MetJ&#45;&gt;data2_run2_1 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>data1_MetJ&#45;&gt;data2_run2_1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M137.5,-932.9664C137.5,-927.6707 137.5,-921.8028 137.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"139.6657,-916.2222 137.5,-911.2222 135.3345,-916.2223 139.6657,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_entC&#45;m&#45;\n",
       "enF&#45;KO -->\n",
       "<g id=\"node115\" class=\"node\">\n",
       "<title>data1_entC&#45;m&#45;\n",
       "enF&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-954.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">entC&#45;m&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-939.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">enF&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_entC&#45;m&#45;\n",
       "enF&#45;KO&#45;&gt;data2_run2_15 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>data1_entC&#45;m&#45;\n",
       "enF&#45;KO&#45;&gt;data2_run2_15</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.93\" d=\"M701.5,-931.6812C701.5,-926.5393 701.5,-920.927 701.5,-915.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.93\" points=\"703.657,-915.3596 701.5,-910.3597 699.3432,-915.3597 703.657,-915.3596\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;6 -->\n",
       "<g id=\"node116\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;6</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-126.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-111.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;6</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;6&#45;&gt;data2_run2_86 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;6&#45;&gt;data2_run2_86</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.85\" d=\"M227.5,-103.6812C227.5,-98.5393 227.5,-92.927 227.5,-87.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.85\" points=\"229.622,-87.3596 227.5,-82.3597 225.3782,-87.3597 229.622,-87.3596\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;2 -->\n",
       "<g id=\"node117\" class=\"node\">\n",
       "<title>data1_PurR&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;2&#45;&gt;data2_run2_81 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>data1_PurR&#45;2&#45;&gt;data2_run2_81</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M701.5,-207.9664C701.5,-202.6707 701.5,-196.8028 701.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"703.6613,-191.2222 701.5,-186.2222 699.3388,-191.2223 703.6613,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Cbl+CysB -->\n",
       "<g id=\"node118\" class=\"node\">\n",
       "<title>data1_Cbl+CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Cbl+CysB</text>\n",
       "</g>\n",
       "<!-- data1_Cbl+CysB&#45;&gt;data2_run2_28 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>data1_Cbl+CysB&#45;&gt;data2_run2_28</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M137.5,-724.9664C137.5,-719.6707 137.5,-713.8028 137.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"139.6438,-708.2222 137.5,-703.2222 135.3563,-708.2223 139.6438,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GlcC -->\n",
       "<g id=\"node119\" class=\"node\">\n",
       "<title>data1_GlcC</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GlcC</text>\n",
       "</g>\n",
       "<!-- data1_GlcC&#45;&gt;data2_run2_54 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>data1_GlcC&#45;&gt;data2_run2_54</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M45.5,-414.9664C45.5,-409.6707 45.5,-403.8028 45.5,-398.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"47.6395,-398.2222 45.5,-393.2222 43.3607,-398.2223 47.6395,-398.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1 -->\n",
       "<g id=\"node120\" class=\"node\">\n",
       "<title>data1_Crp&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Crp&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Crp&#45;1&#45;&gt;data2_run2_93 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>data1_Crp&#45;1&#45;&gt;data2_run2_93</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M1066.5,-104.9664C1066.5,-99.6707 1066.5,-93.8028 1066.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"1068.6307,-88.2222 1066.5,-83.2222 1064.3695,-88.2223 1068.6307,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_YiaJ -->\n",
       "<g id=\"node121\" class=\"node\">\n",
       "<title>data1_YiaJ</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">YiaJ</text>\n",
       "</g>\n",
       "<!-- data1_YiaJ&#45;&gt;data2_run2_55 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>data1_YiaJ&#45;&gt;data2_run2_55</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M137.5,-414.9664C137.5,-409.6707 137.5,-403.8028 137.5,-398.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"139.6745,-398.2222 137.5,-393.2222 135.3257,-398.2223 139.6745,-398.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GcvA -->\n",
       "<g id=\"node122\" class=\"node\">\n",
       "<title>data1_GcvA</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GcvA</text>\n",
       "</g>\n",
       "<!-- data1_GcvA&#45;&gt;data2_run2_20 -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>data1_GcvA&#45;&gt;data2_run2_20</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M344.5,-828.9664C344.5,-823.6707 344.5,-817.8028 344.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"346.6438,-812.2222 344.5,-807.2222 342.3563,-812.2223 346.6438,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_XylR -->\n",
       "<g id=\"node123\" class=\"node\">\n",
       "<title>data1_XylR</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">XylR</text>\n",
       "</g>\n",
       "<!-- data1_XylR&#45;&gt;data2_run2_40 -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>data1_XylR&#45;&gt;data2_run2_40</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M537.5,-621.9664C537.5,-616.6707 537.5,-610.8028 537.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"539.6657,-605.2222 537.5,-600.2222 535.3345,-605.2223 539.6657,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_fur&#45;KO -->\n",
       "<g id=\"node124\" class=\"node\">\n",
       "<title>data1_fur&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">fur&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_fur&#45;KO&#45;&gt;data2_run2_74 -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>data1_fur&#45;KO&#45;&gt;data2_run2_74</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M45.5,-207.9664C45.5,-202.6707 45.5,-196.8028 45.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"47.6307,-191.2222 45.5,-186.2222 43.3695,-191.2223 47.6307,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RpoH -->\n",
       "<g id=\"node125\" class=\"node\">\n",
       "<title>data1_RpoH</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RpoH</text>\n",
       "</g>\n",
       "<!-- data1_RpoH&#45;&gt;data2_run2_39 -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>data1_RpoH&#45;&gt;data2_run2_39</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M344.5,-621.9664C344.5,-616.6707 344.5,-610.8028 344.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"346.6307,-605.2222 344.5,-600.2222 342.3695,-605.2223 346.6307,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GadEWX -->\n",
       "<g id=\"node126\" class=\"node\">\n",
       "<title>data1_GadEWX</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GadEWX</text>\n",
       "</g>\n",
       "<!-- data1_GadEWX&#45;&gt;data2_run2_87 -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>data1_GadEWX&#45;&gt;data2_run2_87</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.98\" d=\"M344.5,-104.9664C344.5,-99.6707 344.5,-93.8028 344.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.98\" points=\"346.6788,-88.2222 344.5,-83.2222 342.3213,-88.2223 346.6788,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_trans&#45;\n",
       "lation -->\n",
       "<g id=\"node127\" class=\"node\">\n",
       "<title>data1_trans&#45;\n",
       "lation</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-850.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-835.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lation</text>\n",
       "</g>\n",
       "<!-- data1_trans&#45;\n",
       "lation&#45;&gt;data2_run2_26 -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>data1_trans&#45;\n",
       "lation&#45;&gt;data2_run2_26</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M1066.5,-827.6812C1066.5,-822.5393 1066.5,-816.927 1066.5,-811.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"1068.6088,-811.3596 1066.5,-806.3597 1064.3913,-811.3597 1068.6088,-811.3596\"/>\n",
       "</g>\n",
       "<!-- data1_PrpR -->\n",
       "<g id=\"node128\" class=\"node\">\n",
       "<title>data1_PrpR</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PrpR</text>\n",
       "</g>\n",
       "<!-- data1_PrpR&#45;&gt;data2_run2_84 -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>data1_PrpR&#45;&gt;data2_run2_84</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M45.5,-104.9664C45.5,-99.6707 45.5,-93.8028 45.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"47.6395,-88.2222 45.5,-83.2222 43.3607,-88.2223 47.6395,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GadWX -->\n",
       "<g id=\"node129\" class=\"node\">\n",
       "<title>data1_GadWX</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GadWX</text>\n",
       "</g>\n",
       "<!-- data1_GadWX&#45;&gt;data2_run2_62 -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>data1_GadWX&#45;&gt;data2_run2_62</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.81\" d=\"M1066.5,-414.9664C1066.5,-409.6707 1066.5,-403.8028 1066.5,-398.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.81\" points=\"1068.6045,-398.2222 1066.5,-393.2222 1064.3957,-398.2223 1068.6045,-398.2222\"/>\n",
       "</g>\n",
       "<!-- data1_fimbriae -->\n",
       "<g id=\"node130\" class=\"node\">\n",
       "<title>data1_fimbriae</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">fimbriae</text>\n",
       "</g>\n",
       "<!-- data1_fimbriae&#45;&gt;data2_run2_52 -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>data1_fimbriae&#45;&gt;data2_run2_52</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.93\" d=\"M882.5,-518.9664C882.5,-513.6707 882.5,-507.8028 882.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.93\" points=\"884.657,-502.2222 882.5,-497.2222 880.3432,-502.2223 884.657,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CecR -->\n",
       "<g id=\"node131\" class=\"node\">\n",
       "<title>data1_CecR</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CecR</text>\n",
       "</g>\n",
       "<!-- data1_CecR&#45;&gt;data2_run2_65 -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>data1_CecR&#45;&gt;data2_run2_65</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.86\" d=\"M227.5,-310.9664C227.5,-305.6707 227.5,-299.8028 227.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.86\" points=\"229.6263,-294.2222 227.5,-289.2222 225.3738,-294.2223 229.6263,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;4 -->\n",
       "<g id=\"node132\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;4</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-746.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-731.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;4</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;4&#45;&gt;data2_run2_30 -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;4&#45;&gt;data2_run2_30</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.8\" d=\"M457.5,-723.6812C457.5,-718.5393 457.5,-712.927 457.5,-707.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.8\" points=\"459.6001,-707.3596 457.5,-702.3597 455.4001,-707.3597 459.6001,-707.3596\"/>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;1 -->\n",
       "<g id=\"node133\" class=\"node\">\n",
       "<title>data1_ArcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;1&#45;&gt;data2_run2_49 -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>data1_ArcA&#45;1&#45;&gt;data2_run2_49</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M457.5,-518.9664C457.5,-513.6707 457.5,-507.8028 457.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"459.6307,-502.2222 457.5,-497.2222 455.3695,-502.2223 459.6307,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_nitrate&#45;\n",
       "&#45;related -->\n",
       "<g id=\"node134\" class=\"node\">\n",
       "<title>data1_nitrate&#45;\n",
       "&#45;related</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-540.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">nitrate&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-525.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">&#45;related</text>\n",
       "</g>\n",
       "<!-- data1_nitrate&#45;\n",
       "&#45;related&#45;&gt;data2_run2_50 -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>data1_nitrate&#45;\n",
       "&#45;related&#45;&gt;data2_run2_50</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M616.5,-517.6812C616.5,-512.5393 616.5,-506.927 616.5,-501.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"618.6395,-501.3596 616.5,-496.3597 614.3607,-501.3597 618.6395,-501.3596\"/>\n",
       "</g>\n",
       "<!-- data1_Fnr -->\n",
       "<g id=\"node135\" class=\"node\">\n",
       "<title>data1_Fnr</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fnr</text>\n",
       "</g>\n",
       "<!-- data1_Fnr&#45;&gt;data2_run2_17 -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>data1_Fnr&#45;&gt;data2_run2_17</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M1066.5,-932.9664C1066.5,-927.6707 1066.5,-921.8028 1066.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"1068.6526,-916.2222 1066.5,-911.2222 1064.3476,-916.2223 1068.6526,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CdaR -->\n",
       "<g id=\"node136\" class=\"node\">\n",
       "<title>data1_CdaR</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CdaR</text>\n",
       "</g>\n",
       "<!-- data1_CdaR&#45;&gt;data2_run2_38 -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>data1_CdaR&#45;&gt;data2_run2_38</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M227.5,-621.9664C227.5,-616.6707 227.5,-610.8028 227.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"229.6482,-605.2222 227.5,-600.2222 225.352,-605.2223 229.6482,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Pyruvate -->\n",
       "<g id=\"node137\" class=\"node\">\n",
       "<title>data1_Pyruvate</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Pyruvate</text>\n",
       "</g>\n",
       "<!-- data1_Pyruvate&#45;&gt;data2_run2_67 -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>data1_Pyruvate&#45;&gt;data2_run2_67</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M457.5,-310.9664C457.5,-305.6707 457.5,-299.8028 457.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"459.6351,-294.2222 457.5,-289.2222 455.3651,-294.2223 459.6351,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_SrlR+GutM -->\n",
       "<g id=\"node138\" class=\"node\">\n",
       "<title>data1_SrlR+GutM</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">SrlR+GutM</text>\n",
       "</g>\n",
       "<!-- data1_SrlR+GutM&#45;&gt;data2_run2_76 -->\n",
       "<g id=\"edge45\" class=\"edge\">\n",
       "<title>data1_SrlR+GutM&#45;&gt;data2_run2_76</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.8\" d=\"M227.5,-207.9664C227.5,-202.6707 227.5,-196.8028 227.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.8\" points=\"229.6001,-191.2222 227.5,-186.2222 225.4001,-191.2223 229.6001,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NagC/TyrR -->\n",
       "<g id=\"node139\" class=\"node\">\n",
       "<title>data1_NagC/TyrR</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NagC/TyrR</text>\n",
       "</g>\n",
       "<!-- data1_NagC/TyrR&#45;&gt;data2_run2_10 -->\n",
       "<g id=\"edge46\" class=\"edge\">\n",
       "<title>data1_NagC/TyrR&#45;&gt;data2_run2_10</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M227.5,-932.9664C227.5,-927.6707 227.5,-921.8028 227.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"229.6657,-916.2222 227.5,-911.2222 225.3345,-916.2223 229.6657,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CpxR -->\n",
       "<g id=\"node140\" class=\"node\">\n",
       "<title>data1_CpxR</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CpxR</text>\n",
       "</g>\n",
       "<!-- data1_CpxR&#45;&gt;data2_run2_23 -->\n",
       "<g id=\"edge47\" class=\"edge\">\n",
       "<title>data1_CpxR&#45;&gt;data2_run2_23</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.81\" d=\"M616.5,-828.9664C616.5,-823.6707 616.5,-817.8028 616.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.81\" points=\"618.6045,-812.2222 616.5,-807.2222 614.3957,-812.2223 618.6045,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_thrA&#45;KO -->\n",
       "<g id=\"node141\" class=\"node\">\n",
       "<title>data1_thrA&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">thrA&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_thrA&#45;KO&#45;&gt;data2_run2_36 -->\n",
       "<g id=\"edge48\" class=\"edge\">\n",
       "<title>data1_thrA&#45;KO&#45;&gt;data2_run2_36</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.64\" d=\"M45.5,-621.9664C45.5,-616.6707 45.5,-610.8028 45.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.64\" points=\"47.5301,-605.2222 45.5,-600.2222 43.4701,-605.2223 47.5301,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Lrp -->\n",
       "<g id=\"node142\" class=\"node\">\n",
       "<title>data1_Lrp</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Lrp</text>\n",
       "</g>\n",
       "<!-- data1_Lrp&#45;&gt;data2_run2_48 -->\n",
       "<g id=\"edge49\" class=\"edge\">\n",
       "<title>data1_Lrp&#45;&gt;data2_run2_48</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M344.5,-518.9664C344.5,-513.6707 344.5,-507.8028 344.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"346.6351,-502.2222 344.5,-497.2222 342.3651,-502.2223 346.6351,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;1 -->\n",
       "<g id=\"node143\" class=\"node\">\n",
       "<title>data1_deletion&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">deletion&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;1&#45;&gt;data2_run2_19 -->\n",
       "<g id=\"edge50\" class=\"edge\">\n",
       "<title>data1_deletion&#45;1&#45;&gt;data2_run2_19</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M137.5,-828.9664C137.5,-823.6707 137.5,-817.8028 137.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"139.6657,-812.2222 137.5,-807.2222 135.3345,-812.2223 139.6657,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile -->\n",
       "<g id=\"node144\" class=\"node\">\n",
       "<title>data1_Leu/Ile</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Leu/Ile</text>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile&#45;&gt;data2_run2_66 -->\n",
       "<g id=\"edge51\" class=\"edge\">\n",
       "<title>data1_Leu/Ile&#45;&gt;data2_run2_66</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.5\" d=\"M334.2395,-310.9664C331.1649,-305.5626 327.7514,-299.5631 324.5238,-293.8903\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.5\" points=\"326.0517,-292.5945 321.8678,-289.2222 322.6293,-294.5417 326.0517,-292.5945\"/>\n",
       "</g>\n",
       "<!-- data1_Leu/Ile&#45;&gt;data2_run2_89 -->\n",
       "<g id=\"edge52\" class=\"edge\">\n",
       "<title>data1_Leu/Ile&#45;&gt;data2_run2_89</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.56\" d=\"M355.0714,-310.9664C358.3025,-305.4545 361.8969,-299.3229 365.2809,-293.5503\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.56\" points=\"366.7991,-294.4208 367.818,-289.2222 363.7796,-292.6507 366.7991,-294.4208\"/>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;2 -->\n",
       "<g id=\"node145\" class=\"node\">\n",
       "<title>data1_deletion&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">deletion&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_deletion&#45;2&#45;&gt;data2_run2_29 -->\n",
       "<g id=\"edge53\" class=\"edge\">\n",
       "<title>data1_deletion&#45;2&#45;&gt;data2_run2_29</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M227.5,-724.9664C227.5,-719.6707 227.5,-713.8028 227.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"229.6745,-708.2222 227.5,-703.2222 225.3257,-708.2223 229.6745,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;1 -->\n",
       "<g id=\"node146\" class=\"node\">\n",
       "<title>data1_Fur&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;1&#45;&gt;data2_run2_34 -->\n",
       "<g id=\"edge54\" class=\"edge\">\n",
       "<title>data1_Fur&#45;1&#45;&gt;data2_run2_34</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M882.5,-724.9664C882.5,-719.6707 882.5,-713.8028 882.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"884.6657,-708.2222 882.5,-703.2222 880.3345,-708.2223 884.6657,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;5 -->\n",
       "<g id=\"node147\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;5</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-332.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-317.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;5</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;5&#45;&gt;data2_run2_73 -->\n",
       "<g id=\"edge55\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;5&#45;&gt;data2_run2_73</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M1066.5,-309.6812C1066.5,-304.5393 1066.5,-298.927 1066.5,-293.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"1068.6088,-293.3596 1066.5,-288.3597 1064.3913,-293.3597 1068.6088,-293.3596\"/>\n",
       "</g>\n",
       "<!-- data1_Thiamine -->\n",
       "<g id=\"node148\" class=\"node\">\n",
       "<title>data1_Thiamine</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Thiamine</text>\n",
       "</g>\n",
       "<!-- data1_Thiamine&#45;&gt;data2_run2_80 -->\n",
       "<g id=\"edge56\" class=\"edge\">\n",
       "<title>data1_Thiamine&#45;&gt;data2_run2_80</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M616.5,-207.9664C616.5,-202.6707 616.5,-196.8028 616.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"618.6526,-191.2222 616.5,-186.2222 614.3476,-191.2223 618.6526,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CysB -->\n",
       "<g id=\"node149\" class=\"node\">\n",
       "<title>data1_CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CysB</text>\n",
       "</g>\n",
       "<!-- data1_CysB&#45;&gt;data2_run2_47 -->\n",
       "<g id=\"edge57\" class=\"edge\">\n",
       "<title>data1_CysB&#45;&gt;data2_run2_47</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M227.5,-518.9664C227.5,-513.6707 227.5,-507.8028 227.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"229.6745,-502.2222 227.5,-497.2222 225.3257,-502.2223 229.6745,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_FliA -->\n",
       "<g id=\"node150\" class=\"node\">\n",
       "<title>data1_FliA</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FliA</text>\n",
       "</g>\n",
       "<!-- data1_FliA&#45;&gt;data2_run2_21 -->\n",
       "<g id=\"edge58\" class=\"edge\">\n",
       "<title>data1_FliA&#45;&gt;data2_run2_21</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M457.5,-828.9664C457.5,-823.6707 457.5,-817.8028 457.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"459.6745,-812.2222 457.5,-807.2222 455.3257,-812.2223 459.6745,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PuuR -->\n",
       "<g id=\"node151\" class=\"node\">\n",
       "<title>data1_PuuR</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PuuR</text>\n",
       "</g>\n",
       "<!-- data1_PuuR&#45;&gt;data2_run2_63 -->\n",
       "<g id=\"edge59\" class=\"edge\">\n",
       "<title>data1_PuuR&#45;&gt;data2_run2_63</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M45.5,-310.9664C45.5,-305.6707 45.5,-299.8028 45.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"47.6351,-294.2222 45.5,-289.2222 43.3651,-294.2223 47.6351,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NikR -->\n",
       "<g id=\"node152\" class=\"node\">\n",
       "<title>data1_NikR</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NikR</text>\n",
       "</g>\n",
       "<!-- data1_NikR&#45;&gt;data2_run2_59 -->\n",
       "<g id=\"edge60\" class=\"edge\">\n",
       "<title>data1_NikR&#45;&gt;data2_run2_59</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M537.5,-414.9664C537.5,-409.6707 537.5,-403.8028 537.5,-398.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"539.6307,-398.2222 537.5,-393.2222 535.3695,-398.2223 539.6307,-398.2222\"/>\n",
       "</g>\n",
       "<!-- data1_NtrC+RpoN -->\n",
       "<g id=\"node153\" class=\"node\">\n",
       "<title>data1_NtrC+RpoN</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">NtrC+RpoN</text>\n",
       "</g>\n",
       "<!-- data1_NtrC+RpoN&#45;&gt;data2_run2_18 -->\n",
       "<g id=\"edge61\" class=\"edge\">\n",
       "<title>data1_NtrC+RpoN&#45;&gt;data2_run2_18</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.98\" d=\"M45.5,-828.9664C45.5,-823.6707 45.5,-817.8028 45.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.98\" points=\"47.6788,-812.2222 45.5,-807.2222 43.3213,-812.2223 47.6788,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RcsAB -->\n",
       "<g id=\"node154\" class=\"node\">\n",
       "<title>data1_RcsAB</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RcsAB</text>\n",
       "</g>\n",
       "<!-- data1_RcsAB&#45;&gt;data2_run2_85 -->\n",
       "<g id=\"edge62\" class=\"edge\">\n",
       "<title>data1_RcsAB&#45;&gt;data2_run2_85</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M137.5,-104.9664C137.5,-99.6707 137.5,-93.8028 137.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"139.6745,-88.2222 137.5,-83.2222 135.3257,-88.2223 139.6745,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_His&#45;tRNA -->\n",
       "<g id=\"node155\" class=\"node\">\n",
       "<title>data1_His&#45;tRNA</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">His&#45;tRNA</text>\n",
       "</g>\n",
       "<!-- data1_His&#45;tRNA&#45;&gt;data2_run2_75 -->\n",
       "<g id=\"edge63\" class=\"edge\">\n",
       "<title>data1_His&#45;tRNA&#45;&gt;data2_run2_75</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M137.5,-207.9664C137.5,-202.6707 137.5,-196.8028 137.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"139.6482,-191.2222 137.5,-186.2222 135.352,-191.2223 139.6482,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2 -->\n",
       "<g id=\"node156\" class=\"node\">\n",
       "<title>data1_Fur&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2&#45;&gt;data2_run2_53 -->\n",
       "<g id=\"edge64\" class=\"edge\">\n",
       "<title>data1_Fur&#45;2&#45;&gt;data2_run2_53</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.86\" d=\"M1066.5,-518.9664C1066.5,-513.6707 1066.5,-507.8028 1066.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.86\" points=\"1068.6263,-502.2222 1066.5,-497.2222 1064.3738,-502.2223 1068.6263,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Copper -->\n",
       "<g id=\"node157\" class=\"node\">\n",
       "<title>data1_Copper</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Copper</text>\n",
       "</g>\n",
       "<!-- data1_Copper&#45;&gt;data2_run2_42 -->\n",
       "<g id=\"edge65\" class=\"edge\">\n",
       "<title>data1_Copper&#45;&gt;data2_run2_42</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M701.5,-621.9664C701.5,-616.6707 701.5,-610.8028 701.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"703.6745,-605.2222 701.5,-600.2222 699.3257,-605.2223 703.6745,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR -->\n",
       "<g id=\"node158\" class=\"node\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"931.5\" y=\"-850.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">AllR/Ar&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"931.5\" y=\"-835.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">aC/FucR</text>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_25 -->\n",
       "<g id=\"edge66\" class=\"edge\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.32\" d=\"M904.9776,-827.6812C896.9581,-821.8398 888.1052,-815.3914 879.909,-809.4213\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.32\" points=\"880.5044,-807.6901 875.4325,-806.1607 878.4437,-810.5191 880.5044,-807.6901\"/>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_68 -->\n",
       "<g id=\"edge67\" class=\"edge\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_68</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.09\" d=\"M926.9159,-827.6812C925.6703,-822.4322 924.3085,-816.693 923.0172,-811.2511\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.09\" points=\"924.7137,-810.8205 921.8565,-806.3597 921.3083,-811.6287 924.7137,-810.8205\"/>\n",
       "</g>\n",
       "<!-- data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_72 -->\n",
       "<g id=\"edge68\" class=\"edge\">\n",
       "<title>data1_AllR/Ar&#45;\n",
       "aC/FucR&#45;&gt;data2_run2_72</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.23\" d=\"M948.8542,-827.6812C953.8581,-822.1108 959.3579,-815.9884 964.5081,-810.2551\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.23\" points=\"966.0428,-811.316 968.0074,-806.3597 963.2893,-808.8426 966.0428,-811.316\"/>\n",
       "</g>\n",
       "<!-- data1_Cra -->\n",
       "<g id=\"node159\" class=\"node\">\n",
       "<title>data1_Cra</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Cra</text>\n",
       "</g>\n",
       "<!-- data1_Cra&#45;&gt;data2_run2_4 -->\n",
       "<g id=\"edge69\" class=\"edge\">\n",
       "<title>data1_Cra&#45;&gt;data2_run2_4</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.96\" d=\"M457.5,-621.9664C457.5,-616.6707 457.5,-610.8028 457.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.96\" points=\"459.6701,-605.2222 457.5,-600.2222 455.3301,-605.2223 459.6701,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;2 -->\n",
       "<g id=\"node160\" class=\"node\">\n",
       "<title>data1_ArcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ArcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_ArcA&#45;2&#45;&gt;data2_run2_3 -->\n",
       "<g id=\"edge70\" class=\"edge\">\n",
       "<title>data1_ArcA&#45;2&#45;&gt;data2_run2_3</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M344.5,-724.9664C344.5,-719.6707 344.5,-713.8028 344.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"346.6088,-708.2222 344.5,-703.2222 342.3913,-708.2223 346.6088,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_EvgA -->\n",
       "<g id=\"node161\" class=\"node\">\n",
       "<title>data1_EvgA</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">EvgA</text>\n",
       "</g>\n",
       "<!-- data1_EvgA&#45;&gt;data2_run2_33 -->\n",
       "<g id=\"edge71\" class=\"edge\">\n",
       "<title>data1_EvgA&#45;&gt;data2_run2_33</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M701.5,-724.9664C701.5,-719.6707 701.5,-713.8028 701.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"703.6745,-708.2222 701.5,-703.2222 699.3257,-708.2223 703.6745,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;2 -->\n",
       "<g id=\"node162\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-954.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-939.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;2&#45;&gt;data2_run2_12 -->\n",
       "<g id=\"edge72\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;2&#45;&gt;data2_run2_12</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.87\" d=\"M457.5,-931.6812C457.5,-926.5393 457.5,-920.927 457.5,-915.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.87\" points=\"459.6307,-915.3596 457.5,-910.3597 455.3695,-915.3597 459.6307,-915.3596\"/>\n",
       "</g>\n",
       "<!-- data1_Zinc -->\n",
       "<g id=\"node163\" class=\"node\">\n",
       "<title>data1_Zinc</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Zinc</text>\n",
       "</g>\n",
       "<!-- data1_Zinc&#45;&gt;data2_run2_43 -->\n",
       "<g id=\"edge73\" class=\"edge\">\n",
       "<title>data1_Zinc&#45;&gt;data2_run2_43</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.99\" d=\"M882.5,-621.9664C882.5,-616.6707 882.5,-610.8028 882.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.99\" points=\"884.6832,-605.2222 882.5,-600.2222 880.317,-605.2223 884.6832,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_SoxS -->\n",
       "<g id=\"node164\" class=\"node\">\n",
       "<title>data1_SoxS</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">SoxS</text>\n",
       "</g>\n",
       "<!-- data1_SoxS&#45;&gt;data2_run2_22 -->\n",
       "<g id=\"edge74\" class=\"edge\">\n",
       "<title>data1_SoxS&#45;&gt;data2_run2_22</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M537.5,-828.9664C537.5,-823.6707 537.5,-817.8028 537.5,-812.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"539.6745,-812.2222 537.5,-807.2222 535.3257,-812.2223 539.6745,-812.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Nac -->\n",
       "<g id=\"node165\" class=\"node\">\n",
       "<title>data1_Nac</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Nac</text>\n",
       "</g>\n",
       "<!-- data1_Nac&#45;&gt;data2_run2_31 -->\n",
       "<g id=\"edge75\" class=\"edge\">\n",
       "<title>data1_Nac&#45;&gt;data2_run2_31</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M537.5,-724.9664C537.5,-719.6707 537.5,-713.8028 537.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"539.6438,-708.2222 537.5,-703.2222 535.3563,-708.2223 539.6438,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Tryptophan -->\n",
       "<g id=\"node166\" class=\"node\">\n",
       "<title>data1_Tryptophan</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">Tryptophan</text>\n",
       "</g>\n",
       "<!-- data1_curli -->\n",
       "<g id=\"node167\" class=\"node\">\n",
       "<title>data1_curli</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">curli</text>\n",
       "</g>\n",
       "<!-- data1_curli&#45;&gt;data2_run2_88 -->\n",
       "<g id=\"edge76\" class=\"edge\">\n",
       "<title>data1_curli&#45;&gt;data2_run2_88</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.85\" d=\"M457.5,-104.9664C457.5,-99.6707 457.5,-93.8028 457.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.85\" points=\"459.622,-88.2222 457.5,-83.2222 455.3782,-88.2223 459.622,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_gadWX&#45;KO -->\n",
       "<g id=\"node168\" class=\"node\">\n",
       "<title>data1_gadWX&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">gadWX&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_gadWX&#45;KO&#45;&gt;data2_run2_45 -->\n",
       "<g id=\"edge77\" class=\"edge\">\n",
       "<title>data1_gadWX&#45;KO&#45;&gt;data2_run2_45</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M45.5,-518.9664C45.5,-513.6707 45.5,-507.8028 45.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"47.6482,-502.2222 45.5,-497.2222 43.352,-502.2223 47.6482,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_flu&#45;yeeRS -->\n",
       "<g id=\"node169\" class=\"node\">\n",
       "<title>data1_flu&#45;yeeRS</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-222.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">flu&#45;yeeRS</text>\n",
       "</g>\n",
       "<!-- data1_flu&#45;yeeRS&#45;&gt;data2_run2_77 -->\n",
       "<g id=\"edge78\" class=\"edge\">\n",
       "<title>data1_flu&#45;yeeRS&#45;&gt;data2_run2_77</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.94\" d=\"M344.5,-207.9664C344.5,-202.6707 344.5,-196.8028 344.5,-191.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.94\" points=\"346.6613,-191.2222 344.5,-186.2222 342.3388,-191.2223 346.6613,-191.2222\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;1 -->\n",
       "<g id=\"node170\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-436.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"344.5\" y=\"-421.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;1&#45;&gt;data2_run2_57 -->\n",
       "<g id=\"edge79\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;1&#45;&gt;data2_run2_57</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.85\" d=\"M344.5,-413.6812C344.5,-408.5393 344.5,-402.927 344.5,-397.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.85\" points=\"346.622,-397.3596 344.5,-392.3597 342.3782,-397.3597 346.622,-397.3596\"/>\n",
       "</g>\n",
       "<!-- data1_FadR/IclR -->\n",
       "<g id=\"node171\" class=\"node\">\n",
       "<title>data1_FadR/IclR</title>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FadR/IclR</text>\n",
       "</g>\n",
       "<!-- data1_FadR/IclR&#45;&gt;data2_run2_58 -->\n",
       "<g id=\"edge80\" class=\"edge\">\n",
       "<title>data1_FadR/IclR&#45;&gt;data2_run2_58</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M457.5,-414.9664C457.5,-409.6707 457.5,-403.8028 457.5,-398.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"459.6482,-398.2222 457.5,-393.2222 455.352,-398.2223 459.6482,-398.2222\"/>\n",
       "</g>\n",
       "<!-- data1_lipopolys&#45;\n",
       "accharide -->\n",
       "<g id=\"node172\" class=\"node\">\n",
       "<title>data1_lipopolys&#45;\n",
       "accharide</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-436.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">lipopolys&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-421.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">accharide</text>\n",
       "</g>\n",
       "<!-- data1_lipopolys&#45;\n",
       "accharide&#45;&gt;data2_run2_60 -->\n",
       "<g id=\"edge81\" class=\"edge\">\n",
       "<title>data1_lipopolys&#45;\n",
       "accharide&#45;&gt;data2_run2_60</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.86\" d=\"M701.5,-413.6812C701.5,-408.5393 701.5,-402.927 701.5,-397.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.86\" points=\"703.6263,-397.3596 701.5,-392.3597 699.3738,-397.3597 703.6263,-397.3596\"/>\n",
       "</g>\n",
       "<!-- data1_FlhDC -->\n",
       "<g id=\"node173\" class=\"node\">\n",
       "<title>data1_FlhDC</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FlhDC</text>\n",
       "</g>\n",
       "<!-- data1_FlhDC&#45;&gt;data2_run2_70 -->\n",
       "<g id=\"edge82\" class=\"edge\">\n",
       "<title>data1_FlhDC&#45;&gt;data2_run2_70</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.98\" d=\"M701.5,-310.9664C701.5,-305.6707 701.5,-299.8028 701.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.98\" points=\"703.6788,-294.2222 701.5,-289.2222 699.3213,-294.2223 703.6788,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_crp&#45;KO -->\n",
       "<g id=\"node174\" class=\"node\">\n",
       "<title>data1_crp&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">crp&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_crp&#45;KO&#45;&gt;data2_run2_51 -->\n",
       "<g id=\"edge83\" class=\"edge\">\n",
       "<title>data1_crp&#45;KO&#45;&gt;data2_run2_51</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.89\" d=\"M701.5,-518.9664C701.5,-513.6707 701.5,-507.8028 701.5,-502.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.89\" points=\"703.6395,-502.2222 701.5,-497.2222 699.3607,-502.2223 703.6395,-502.2222\"/>\n",
       "</g>\n",
       "<!-- data1_GntR/TyrR -->\n",
       "<g id=\"node175\" class=\"node\">\n",
       "<title>data1_GntR/TyrR</title>\n",
       "<text text-anchor=\"middle\" x=\"882.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GntR/TyrR</text>\n",
       "</g>\n",
       "<!-- data1_GntR/TyrR&#45;&gt;data2_run2_16 -->\n",
       "<g id=\"edge84\" class=\"edge\">\n",
       "<title>data1_GntR/TyrR&#45;&gt;data2_run2_16</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.95\" d=\"M882.5,-932.9664C882.5,-927.6707 882.5,-921.8028 882.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.95\" points=\"884.6657,-916.2222 882.5,-911.2222 880.3345,-916.2223 884.6657,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_purR&#45;KO -->\n",
       "<g id=\"node176\" class=\"node\">\n",
       "<title>data1_purR&#45;KO</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-636.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">purR&#45;KO</text>\n",
       "</g>\n",
       "<!-- data1_purR&#45;KO&#45;&gt;data2_run2_41 -->\n",
       "<g id=\"edge85\" class=\"edge\">\n",
       "<title>data1_purR&#45;KO&#45;&gt;data2_run2_41</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.78\" d=\"M616.5,-621.9664C616.5,-616.6707 616.5,-610.8028 616.5,-605.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.78\" points=\"618.5913,-605.2222 616.5,-600.2222 614.4088,-605.2223 618.5913,-605.2222\"/>\n",
       "</g>\n",
       "<!-- data1_RpoS -->\n",
       "<g id=\"node177\" class=\"node\">\n",
       "<title>data1_RpoS</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-947.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RpoS</text>\n",
       "</g>\n",
       "<!-- data1_RpoS&#45;&gt;data2_run2_13 -->\n",
       "<g id=\"edge86\" class=\"edge\">\n",
       "<title>data1_RpoS&#45;&gt;data2_run2_13</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.97\" d=\"M537.5,-932.9664C537.5,-927.6707 537.5,-921.8028 537.5,-916.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.97\" points=\"539.6745,-916.2222 537.5,-911.2222 535.3257,-916.2223 539.6745,-916.2222\"/>\n",
       "</g>\n",
       "<!-- data1_membrane -->\n",
       "<g id=\"node178\" class=\"node\">\n",
       "<title>data1_membrane</title>\n",
       "<text text-anchor=\"middle\" x=\"1066.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">membrane</text>\n",
       "</g>\n",
       "<!-- data1_membrane&#45;&gt;data2_run2_35 -->\n",
       "<g id=\"edge87\" class=\"edge\">\n",
       "<title>data1_membrane&#45;&gt;data2_run2_35</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.92\" d=\"M1066.5,-724.9664C1066.5,-719.6707 1066.5,-713.8028 1066.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.92\" points=\"1068.6526,-708.2222 1066.5,-703.2222 1064.3476,-708.2223 1068.6526,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_duplic&#45;\n",
       "ation&#45;1 -->\n",
       "<g id=\"node179\" class=\"node\">\n",
       "<title>data1_duplic&#45;\n",
       "ation&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-436.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">duplic&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-421.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ation&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_duplic&#45;\n",
       "ation&#45;1&#45;&gt;data2_run2_56 -->\n",
       "<g id=\"edge88\" class=\"edge\">\n",
       "<title>data1_duplic&#45;\n",
       "ation&#45;1&#45;&gt;data2_run2_56</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.96\" d=\"M227.5,-413.6812C227.5,-408.5393 227.5,-402.927 227.5,-397.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.96\" points=\"229.6701,-397.3596 227.5,-392.3597 225.3301,-397.3597 229.6701,-397.3596\"/>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;3 -->\n",
       "<g id=\"node180\" class=\"node\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;3</title>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-126.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">uncharac&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"701.5\" y=\"-111.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">terized&#45;3</text>\n",
       "</g>\n",
       "<!-- data1_uncharac&#45;\n",
       "terized&#45;3&#45;&gt;data2_run2_91 -->\n",
       "<g id=\"edge89\" class=\"edge\">\n",
       "<title>data1_uncharac&#45;\n",
       "terized&#45;3&#45;&gt;data2_run2_91</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.82\" d=\"M701.5,-103.6812C701.5,-98.5393 701.5,-92.927 701.5,-87.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.82\" points=\"703.6088,-87.3596 701.5,-82.3597 699.3913,-87.3597 703.6088,-87.3596\"/>\n",
       "</g>\n",
       "<!-- data1_ExuR/FucR -->\n",
       "<g id=\"node181\" class=\"node\">\n",
       "<title>data1_ExuR/FucR</title>\n",
       "<text text-anchor=\"middle\" x=\"852.5\" y=\"-843.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ExuR/FucR</text>\n",
       "</g>\n",
       "<!-- data1_ExuR/FucR&#45;&gt;data2_run2_25 -->\n",
       "<g id=\"edge90\" class=\"edge\">\n",
       "<title>data1_ExuR/FucR&#45;&gt;data2_run2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.11\" d=\"M851.8891,-828.9775C851.7026,-823.477 851.4948,-817.3458 851.2978,-811.5365\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.11\" points=\"853.0395,-811.2572 851.121,-806.3193 849.5415,-811.3758 853.0395,-811.2572\"/>\n",
       "</g>\n",
       "<!-- data1_ExuR/FucR&#45;&gt;data2_run2_68 -->\n",
       "<g id=\"edge91\" class=\"edge\">\n",
       "<title>data1_ExuR/FucR&#45;&gt;data2_run2_68</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.75\" d=\"M872.3553,-828.9775C879.0336,-822.9157 886.5557,-816.0879 893.5169,-809.7692\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.75\" points=\"895.0121,-811.2187 897.3177,-806.3193 892.2186,-808.1411 895.0121,-811.2187\"/>\n",
       "</g>\n",
       "<!-- data1_FecI -->\n",
       "<g id=\"node182\" class=\"node\">\n",
       "<title>data1_FecI</title>\n",
       "<text text-anchor=\"middle\" x=\"537.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FecI</text>\n",
       "</g>\n",
       "<!-- data1_FecI&#45;&gt;data2_run2_69 -->\n",
       "<g id=\"edge92\" class=\"edge\">\n",
       "<title>data1_FecI&#45;&gt;data2_run2_69</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.88\" d=\"M537.5,-310.9664C537.5,-305.6707 537.5,-299.8028 537.5,-294.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.88\" points=\"539.6351,-294.2222 537.5,-289.2222 535.3651,-294.2223 539.6351,-294.2222\"/>\n",
       "</g>\n",
       "<!-- data1_iron&#45;r&#45;\n",
       "elated -->\n",
       "<g id=\"node183\" class=\"node\">\n",
       "<title>data1_iron&#45;r&#45;\n",
       "elated</title>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">iron&#45;r&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"137.5\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">elated</text>\n",
       "</g>\n",
       "<!-- data1_RbsR -->\n",
       "<g id=\"node184\" class=\"node\">\n",
       "<title>data1_RbsR</title>\n",
       "<text text-anchor=\"middle\" x=\"616.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">RbsR</text>\n",
       "</g>\n",
       "<!-- data1_RbsR&#45;&gt;data2_run2_90 -->\n",
       "<g id=\"edge93\" class=\"edge\">\n",
       "<title>data1_RbsR&#45;&gt;data2_run2_90</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.63\" d=\"M616.5,-104.9664C616.5,-99.6707 616.5,-93.8028 616.5,-88.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.63\" points=\"618.5257,-88.2222 616.5,-83.2222 614.4745,-88.2223 618.5257,-88.2222\"/>\n",
       "</g>\n",
       "<!-- data1_sgrT -->\n",
       "<g id=\"node185\" class=\"node\">\n",
       "<title>data1_sgrT</title>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-739.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">sgrT</text>\n",
       "</g>\n",
       "<!-- data1_sgrT&#45;&gt;data2_run2_27 -->\n",
       "<g id=\"edge94\" class=\"edge\">\n",
       "<title>data1_sgrT&#45;&gt;data2_run2_27</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.88\" d=\"M45.5,-724.9664C45.5,-719.6707 45.5,-713.8028 45.5,-708.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.88\" points=\"47.2501,-708.2222 45.5,-703.2222 43.7501,-708.2223 47.2501,-708.2222\"/>\n",
       "</g>\n",
       "<!-- data1_insertion -->\n",
       "<g id=\"node186\" class=\"node\">\n",
       "<title>data1_insertion</title>\n",
       "<text text-anchor=\"middle\" x=\"227.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"#ff0000\">insertion</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7fe01a229d00>"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots,links = compare_ica(ica_data.M,ica_run_diff,show_all = True)\n",
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, you can adjust the metric type and cutoff values for correlations\n",
    "by utilizing the `metric` and `cutoff` parameters. The default for metric is\n",
    " \"*pearson*\", but can be changed to \"*spearman*\" by inputting the string.\n",
    "\n",
    "`cutoff` takes a float value between 0 and 1 and will change compare_ica to only\n",
    "show correlations greater than the `cutoff` value.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 92/92 [00:10<00:00,  8.92it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"148pt\" height=\"206pt\"\n",
       " viewBox=\"0.00 0.00 148.00 206.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 202)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-202 144,-202 144,4 -4,4\"/>\n",
       "<!-- data2_run2_22 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_run2_22</title>\n",
       "<text text-anchor=\"middle\" x=\"33\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_22</text>\n",
       "</g>\n",
       "<!-- data2_run2_47 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_run2_47</title>\n",
       "<text text-anchor=\"middle\" x=\"107\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_47</text>\n",
       "</g>\n",
       "<!-- data2_run2_56 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_run2_56</title>\n",
       "<text text-anchor=\"middle\" x=\"33\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">run2_56</text>\n",
       "</g>\n",
       "<!-- data1_CysB -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data1_CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"107\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CysB</text>\n",
       "</g>\n",
       "<!-- data1_CysB&#45;&gt;data2_run2_47 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_CysB&#45;&gt;data2_run2_47</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.91\" d=\"M107,-161.9664C107,-156.6707 107,-150.8028 107,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.91\" points=\"109.1482,-145.2222 107,-140.2222 104.852,-145.2223 109.1482,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_SoxS -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data1_SoxS</title>\n",
       "<text text-anchor=\"middle\" x=\"33\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">SoxS</text>\n",
       "</g>\n",
       "<!-- data1_SoxS&#45;&gt;data2_run2_22 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_SoxS&#45;&gt;data2_run2_22</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.93\" d=\"M33,-161.9664C33,-156.6707 33,-150.8028 33,-145.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.93\" points=\"35.157,-145.2222 33,-140.2222 30.8432,-145.2223 35.157,-145.2222\"/>\n",
       "</g>\n",
       "<!-- data1_duplic&#45;\n",
       "ation&#45;1 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data1_duplic&#45;\n",
       "ation&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"33\" y=\"-80.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">duplic&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"33\" y=\"-65.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">ation&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_duplic&#45;\n",
       "ation&#45;1&#45;&gt;data2_run2_56 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_duplic&#45;\n",
       "ation&#45;1&#45;&gt;data2_run2_56</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"4.9\" d=\"M33,-57.6812C33,-52.5393 33,-46.927 33,-41.5847\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"4.9\" points=\"35.1438,-41.3596 33,-36.3597 30.8563,-41.3597 35.1438,-41.3596\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7fe01a2126a0>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots,links = compare_ica(ica_data.M,ica_run_diff,metric=\"spearman\",cutoff=.98)\n",
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison across organisms\n",
    "`compare_ica` can also be used to compare ICA components between different\n",
    "organisms. To do so, you will need to provide an addition input, which is a\n",
    "string path to a Bidirectional Best Hit (BBH) between your two organisms of interest.\n",
    "\n",
    "A Github repository containing many BBH CSV's can be found at the following\n",
    "link: https://github.com/SBRG/modulome_compare_data/tree/master/bbh_csv\n",
    "\n",
    "You can also provide a path to your own BBH file, as long as it contains the\n",
    "columns `**gene**` and `**target**`\n",
    "\n",
    "It is important that the order of the organism matches the order of the BBH\n",
    "file. Specifically, the organism for the `M1` file must be the organism that\n",
    "fills the `gene` column in the bbh csv. Otherwise, errors will occur."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_org = load_json_model('../example_data/modulome_example/saci.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 92/92 [00:00<00:00, 251.48it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"215pt\" height=\"204pt\"\n",
       " viewBox=\"0.00 0.00 214.50 204.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 200)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-200 210.5,-200 210.5,4 -4,4\"/>\n",
       "<!-- data2_1 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_1</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">1</text>\n",
       "</g>\n",
       "<!-- data2_23 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_23</title>\n",
       "<text text-anchor=\"middle\" x=\"105.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">23</text>\n",
       "</g>\n",
       "<!-- data2_25 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_25</title>\n",
       "<text text-anchor=\"middle\" x=\"179.5\" y=\"-116.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">25</text>\n",
       "</g>\n",
       "<!-- data2_29 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_29</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">29</text>\n",
       "</g>\n",
       "<!-- data2_FadR -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_FadR</title>\n",
       "<text text-anchor=\"middle\" x=\"105.5\" y=\"-14.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FadR</text>\n",
       "</g>\n",
       "<!-- data1_GlpR -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data1_GlpR</title>\n",
       "<text text-anchor=\"middle\" x=\"105.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">GlpR</text>\n",
       "</g>\n",
       "<!-- data1_GlpR&#45;&gt;data2_23 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_GlpR&#45;&gt;data2_23</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.68\" d=\"M105.5,-159.9664C105.5,-154.6707 105.5,-148.8028 105.5,-143.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.68\" points=\"107.2501,-143.2222 105.5,-138.2222 103.7501,-143.2223 107.2501,-143.2222\"/>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data1_PurR&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">PurR&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_PurR&#45;1&#45;&gt;data2_29 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_PurR&#45;1&#45;&gt;data2_29</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"3.18\" d=\"M29.5,-57.9664C29.5,-52.6707 29.5,-46.8028 29.5,-41.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"3.18\" points=\"31.2501,-41.2222 29.5,-36.2222 27.7501,-41.2223 31.2501,-41.2222\"/>\n",
       "</g>\n",
       "<!-- data1_CysB -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data1_CysB</title>\n",
       "<text text-anchor=\"middle\" x=\"29.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">CysB</text>\n",
       "</g>\n",
       "<!-- data1_CysB&#45;&gt;data2_1 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_CysB&#45;&gt;data2_1</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"3.05\" d=\"M29.5,-159.9664C29.5,-154.6707 29.5,-148.8028 29.5,-143.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"3.05\" points=\"31.2501,-143.2222 29.5,-138.2222 27.7501,-143.2223 31.2501,-143.2222\"/>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data1_Fur&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"179.5\" y=\"-174.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">Fur&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_Fur&#45;2&#45;&gt;data2_25 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_Fur&#45;2&#45;&gt;data2_25</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"2.01\" d=\"M179.5,-159.9664C179.5,-154.6707 179.5,-148.8028 179.5,-143.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"2.01\" points=\"181.2501,-143.2222 179.5,-138.2222 177.7501,-143.2223 181.2501,-143.2222\"/>\n",
       "</g>\n",
       "<!-- data1_FadR/IclR -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data1_FadR/IclR</title>\n",
       "<text text-anchor=\"middle\" x=\"105.5\" y=\"-72.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\" fill=\"#000000\">FadR/IclR</text>\n",
       "</g>\n",
       "<!-- data1_FadR/IclR&#45;&gt;data2_FadR -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_FadR/IclR&#45;&gt;data2_FadR</title>\n",
       "<path fill=\"none\" stroke=\"#000000\" stroke-width=\"1.99\" d=\"M105.5,-57.9664C105.5,-52.6707 105.5,-46.8028 105.5,-41.231\"/>\n",
       "<polygon fill=\"#000000\" stroke=\"#000000\" stroke-width=\"1.99\" points=\"107.2501,-41.2222 105.5,-36.2222 103.7501,-41.2223 107.2501,-41.2222\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7fe01a229820>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots,links = compare_ica(ica_data.M,other_org.M, cutoff=0.3,\n",
    "                         ortho_file = '../example_data/example_bbh.csv')\n",
    "dots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a BBH CSV file\n",
    "If running a comparison between two organisms, you may find that the bbh.csv does not exist within the *modulome_compare_data* repository. As such, pymodulon.compare contains a set of functions that will allow you to create the desired bbh.csv file for you specific comparison. This only needs to be done once per combination.\n",
    "\n",
    "In order to make the bbh.csv, you need to obtain the Genbank Full Genome files for your organisms. These can be found at https://www.ncbi.nlm.nih.gov/genome/. Search for both your organisms/strains and download the \"Genbank (full)\" file. Once saved into a known location, you will use the `make_prots` function to convert the Genbank files into protein FASTA files.\n",
    "\n",
    "`make_prots` recieves two parameters:\n",
    "* `gbk`: Path to the genbank file (one file per function call)\n",
    "* `out_path`: Path to the protein FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_gb= \"../pymodulon/test/data/ecoli.gb\"\n",
    "mtb_gb = \"../pymodulon/test/data/mtb.gb\"\n",
    "ecoli_out = \"../pymodulon/test/data/ecoli_prot.fasta\"\n",
    "mtb_out = \"../pymodulon/test/data/mtb_prot.fasta\"\n",
    "\n",
    "make_prots(ecoli_gb,ecoli_out)\n",
    "make_prots(mtb_gb,mtb_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Genbank database files\n",
    "In order to run Bidirectional Best Hits between two organisms, it is required to build the necessary Genbank Database files from the newly created protein FASTA files. In order to do so, use the `make_prots_db` function. You will do this for both organims.\n",
    "\n",
    "`make_prots_db` recieves one parameter:\n",
    "* `fasta_file`: String path to protein FASTA file\n",
    "\n",
    "If there is an error, the function will print out error message. The function will also not execut if the necessary Genbank DB files already exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running makeblastdb with following command line...\n",
      "makeblastdb -in ../pymodulon/test/data/ecoli_prot.fasta -parse_seqids -dbtype prot\n",
      "\n",
      "makeblastdb run failed. Make sure makeblastdb is installed and working properly, and that the protein FASTA file contains no duplicate genes. View the output below to see what error occured:\n",
      "\n",
      "CompletedProcess(args=['makeblastdb', '-in', '../pymodulon/test/data/ecoli_prot.fasta', '-parse_seqids', '-dbtype', 'prot'], returncode=1, stdout=b'\\n\\nBuilding a new DB, current time: 10/15/2020 15:40:25\\nNew DB name:   /home/anand/PycharmProjects/pymodulon/pymodulon/test/data/ecoli_prot.fasta\\nNew DB title:  ../pymodulon/test/data/ecoli_prot.fasta\\nSequence type: Protein\\nDeleted existing Protein BLAST database named /home/anand/PycharmProjects/pymodulon/pymodulon/test/data/ecoli_prot.fasta\\nKeep MBits: T\\nMaximum file size: 1000000000B\\n\\n\\n', stderr=b'BLAST Database creation error: Error: Duplicate seq_ids are found: \\nLCL|B0470\\n\\n')\n",
      "running makeblastdb with following command line...\n",
      "makeblastdb -in ../pymodulon/test/data/mtb_prot.fasta -parse_seqids -dbtype prot\n",
      "Protein DB files created successfully\n"
     ]
    }
   ],
   "source": [
    "make_prot_db(ecoli_out)\n",
    "make_prot_db(mtb_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create BBH CSV\n",
    "Once the previous steps have been completed, you can now create the BBH CSV file. In order to do so, you can use the `get_bbh` function. \n",
    "\n",
    "`get_bbh` has the following parameters:\n",
    "* `db1`: String path to protein FASTA file (output of make_prots function) for organism 1\n",
    "* `db2`: String path to protein FASTA file (output of make_prots function) for organism 2\n",
    "* `outdir`: String path to output directory, default is \"bbh\" and will create the directory if it does not exist\n",
    "* `outname`: Default db1_vs_db2_parsed.csv where db[1-2] are the passed arguments name of the csv file where that will save the results\n",
    "* `mincov`: Minimum coverage to call hits in BLAST, must be between 0 and 1\n",
    "* `evalue`: evalue thershold for BLAST hits, Default .001\n",
    "* `threads`: Number of threads to run BLAST, Default 1\n",
    "* `force`: Whether to overwrite existing files or not\n",
    "* `savefiles`: Whether to save files to outdir\n",
    "\n",
    "The function will return a Pandas DataFrame containing all the BLAST hits between the two genes. The function will also save the DataFrame to a csv file, which you can then use in `compare_ica` between the two organisms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_bbh(ecoli_out,mtb_out, outdir = \"../pymodulon/test/data/\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
